// @solidxai/core-ui/dist/components/layout/Layout.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { usePathname } from "../../hooks/usePathname";
import { useSearchParams } from "../../hooks/useSearchParams";
import { PrimeReactContext } from 'primereact/api';
import { useEventListener, useUnmountEffect } from 'primereact/hooks';
import { classNames } from 'primereact/utils';
import React, { useContext, useEffect, useRef } from 'react';
import { useSelector } from 'react-redux';
import AppConfig from './AppConfig';
import { LayoutContext } from './context/layoutcontext';
import AppSidebar from './AppSidebar';
import SolidPopupContainer from '../common/SolidPopupContainer';
import { useSession } from "../../hooks/useSession";
import { getExtensionFunction } from '../../helpers/registry';
import { env } from "../../adapters/env";
export var Layout = function (_a) {
    var _b;
    var children = _a.children;
    var _c = useContext(LayoutContext), layoutConfig = _c.layoutConfig, layoutState = _c.layoutState, setLayoutState = _c.setLayoutState;
    var setRipple = useContext(PrimeReactContext).setRipple;
    // const topbarRef = useRef<AppTopbarRef>(null);
    var sidebarRef = useRef(null);
    var _d = useEventListener({
        type: 'click',
        listener: function (event) {
            var _a, _b;
            var isOutsideClicked = !(((_a = sidebarRef.current) === null || _a === void 0 ? void 0 : _a.isSameNode(event.target)) ||
                ((_b = sidebarRef.current) === null || _b === void 0 ? void 0 : _b.contains(event.target))
            // topbarRef.current?.menubutton?.isSameNode(event.target as Node) ||
            // topbarRef.current?.menubutton?.contains(event.target as Node)
            );
            if (isOutsideClicked) {
                hideMenu();
            }
        }
    }), bindMenuOutsideClickListener = _d[0], unbindMenuOutsideClickListener = _d[1];
    var pathname = usePathname();
    var searchParams = useSearchParams();
    useEffect(function () {
        hideMenu();
        hideProfileMenu();
    }, [pathname, searchParams]);
    // const [bindProfileMenuOutsideClickListener, unbindProfileMenuOutsideClickListener] = useEventListener({
    //     type: 'click',
    //     listener: (event) => {
    //         const isOutsideClicked = !(
    //             topbarRef.current?.topbarmenu?.isSameNode(event.target as Node) ||
    //             topbarRef.current?.topbarmenu?.contains(event.target as Node) ||
    //             topbarRef.current?.topbarmenubutton?.isSameNode(event.target as Node) ||
    //             topbarRef.current?.topbarmenubutton?.contains(event.target as Node)
    //         );
    //         if (isOutsideClicked) {
    //             hideProfileMenu();
    //         }
    //     }
    // });
    var hideMenu = function () {
        setLayoutState(function (prevLayoutState) { return (__assign(__assign({}, prevLayoutState), { overlayMenuActive: false, staticMenuMobileActive: false, menuHoverActive: false })); });
        unbindMenuOutsideClickListener();
        unblockBodyScroll();
    };
    var hideProfileMenu = function () {
        setLayoutState(function (prevLayoutState) { return (__assign(__assign({}, prevLayoutState), { profileSidebarVisible: false })); });
        // unbindProfileMenuOutsideClickListener();
    };
    var blockBodyScroll = function () {
        if (document.body.classList) {
            document.body.classList.add('blocked-scroll');
        }
        else {
            document.body.className += ' blocked-scroll';
        }
    };
    var unblockBodyScroll = function () {
        if (document.body.classList) {
            document.body.classList.remove('blocked-scroll');
        }
        else {
            document.body.className = document.body.className.replace(new RegExp('(^|\\b)' + 'blocked-scroll'.split(' ').join('|') + '(\\b|$)', 'gi'), ' ');
        }
    };
    useEffect(function () {
        if (layoutState.overlayMenuActive || layoutState.staticMenuMobileActive) {
            bindMenuOutsideClickListener();
        }
        layoutState.staticMenuMobileActive && blockBodyScroll();
    }, [layoutState.overlayMenuActive, layoutState.staticMenuMobileActive]);
    // useEffect(() => {
    //     if (layoutState.profileSidebarVisible) {
    //         bindProfileMenuOutsideClickListener();
    //     }
    // }, [layoutState.profileSidebarVisible]);
    useUnmountEffect(function () {
        unbindMenuOutsideClickListener();
        // unbindProfileMenuOutsideClickListener();
    });
    var containerClass = classNames('layout-wrapper', {
        'layout-overlay-active': layoutState.overlayMenuActive,
        'layout-mobile-active': layoutState.staticMenuMobileActive,
        'p-input-filled': layoutConfig.inputStyle === 'filled',
    });
    var visibleNavbar = useSelector(function (state) { return state.navbarState; }).visibleNavbar; // Get the visibility state of sidebar-two
    var session = useSession();
    var user = (_b = session === null || session === void 0 ? void 0 : session.data) === null || _b === void 0 ? void 0 : _b.user;
    var hasRunRef = useRef(false);
    useEffect(function () {
        if (hasRunRef.current)
            return;
        if (!session || !user)
            return;
        var navEntry = performance.getEntriesByType("navigation")[0];
        var isReload = (navEntry === null || navEntry === void 0 ? void 0 : navEntry.type) === "reload";
        // Detect first mount (after login redirect)
        var isFirstMount = !sessionStorage.getItem("app-mounted");
        // Allow execution only for:
        //  1. First mount
        //  2. Reload
        if (isFirstMount || isReload) {
            sessionStorage.setItem("app-mounted", "true");
        }
        else {
            return;
        }
        hasRunRef.current = true;
        var handleDynamicFunction = function () { return __awaiter(void 0, void 0, void 0, function () {
            var dynamicHeader, event, DynamicFunctionComponent;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        dynamicHeader = env("SOLIDX_ON_APPLICATION_MOUNT_HANDLER");
                        event = {
                            type: "onApplicationMount",
                            user: user,
                            session: session.data
                        };
                        if (!dynamicHeader) return [3 /*break*/, 2];
                        DynamicFunctionComponent = getExtensionFunction(dynamicHeader);
                        if (!DynamicFunctionComponent) return [3 /*break*/, 2];
                        return [4 /*yield*/, DynamicFunctionComponent(event)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        }); };
        handleDynamicFunction();
    }, [session, session === null || session === void 0 ? void 0 : session.data, user]);
    // const dispatch = useDispatch()
    // const { data: solidSettingsDataInitialData } = useGetSolidSettingsQuery("")
    // useEffect(() => {
    //     if (solidSettingsDataInitialData) {
    //         dispatch(setSolidSettings(solidSettingsDataInitialData?.data));
    //     }
    // }, [solidSettingsDataInitialData]);
    // const toggleBothSidebars = () => {
    //     if (visibleNavbar) {
    //         dispatch(toggleNavbar());   // close both
    //     } else {
    //         dispatch(showNavbar());     // open both
    //     }
    // };
    return (_jsx(React.Fragment, { children: _jsxs("div", { className: containerClass, children: [_jsx(AppSidebar, {}), _jsx(SolidPopupContainer, {}), _jsx("div", { className: "main-content ".concat(visibleNavbar ? "shifted" : ""), children: children }), _jsx(AppConfig, {}), _jsx("div", { className: "layout-mask" })] }) }));
};
