// @solidxai/core-ui/dist/components/core/list/SolidColumnSelector/SolidListColumnSelector.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useFormik } from 'formik';
import { Button } from 'primereact/button';
import { Checkbox } from 'primereact/checkbox';
import { useRef, useState } from 'react';
import { createSolidEntityApi } from '../../../../redux/api/solidEntityApi';
import { Toast } from "primereact/toast";
import { DragDropContext, Droppable, Draggable } from '@hello-pangea/dnd';
import styles from './SolidListColumnSelector.module.css';
import { ERROR_MESSAGES } from '../../../../constants/error-messages';
import showToast from "../../../../helpers/showToast";
export var SolidListColumnSelector = function (_a) {
    var _b, _c;
    var listViewMetaData = _a.listViewMetaData, customizeLayout = _a.customizeLayout;
    console.log("listViewMetaData column selector", listViewMetaData);
    var toast = useRef(null);
    var _d = useState(false), isDragging = _d[0], setIsDragging = _d[1];
    var entityApi = createSolidEntityApi('userViewMetadata');
    var useUpsertSolidEntityMutation = entityApi.useUpsertSolidEntityMutation;
    var _e = useUpsertSolidEntityMutation(), upsertUserView = _e[0], _f = _e[1], isLoading = _f.isLoading, viewCreateError = _f.error, isSuccess = _f.isSuccess, data = _f.data;
    if (!listViewMetaData) {
        return;
    }
    if (!listViewMetaData.data) {
        return;
    }
    var solidView = (_b = listViewMetaData === null || listViewMetaData === void 0 ? void 0 : listViewMetaData.data) === null || _b === void 0 ? void 0 : _b.solidView;
    // This is a key value map of field name vs field metadata.
    var solidFieldsMetadata = (_c = listViewMetaData === null || listViewMetaData === void 0 ? void 0 : listViewMetaData.data) === null || _c === void 0 ? void 0 : _c.solidFieldsMetadata;
    if (!solidView || !solidFieldsMetadata) {
        return;
    }
    var checkedFieldNames = new Set(solidView.layout.children.map(function (col) { return col.attrs.name; }));
    var solidListColumns = Object.entries(solidFieldsMetadata).map(function (_a) {
        var key = _a[0], field = _a[1];
        return ({
            name: field.displayName,
            key: key,
        });
    });
    var _g = useState(function () {
        var selectedOrder = solidView.layout.children.map(function (child) { return child.attrs.name; });
        var allColumns = Object.entries(solidFieldsMetadata).map(function (_a) {
            var key = _a[0], field = _a[1];
            return ({
                name: field.displayName,
                key: key,
            });
        });
        var selectedFields = [];
        var remainingFields = [];
        var usedKeys = new Set();
        var _loop_1 = function (key) {
            var column = allColumns.find(function (col) { return col.key === key; });
            if (column) {
                selectedFields.push(column);
                usedKeys.add(key);
            }
        };
        // First, add selected fields in the order of solidView
        for (var _i = 0, selectedOrder_1 = selectedOrder; _i < selectedOrder_1.length; _i++) {
            var key = selectedOrder_1[_i];
            _loop_1(key);
        }
        // Then, add remaining fields that are not selected
        for (var _a = 0, allColumns_1 = allColumns; _a < allColumns_1.length; _a++) {
            var col = allColumns_1[_a];
            if (!usedKeys.has(col.key)) {
                remainingFields.push(col);
            }
        }
        return __spreadArray(__spreadArray([], selectedFields, true), remainingFields, true);
    }), fields = _g[0], setFields = _g[1];
    var formik = useFormik({
        initialValues: {
            selectedColumns: solidListColumns.filter(function (col) { return checkedFieldNames.has(col.key); }),
        },
        onSubmit: function (values) { return __awaiter(void 0, void 0, void 0, function () {
            var selectedKeys, currentChildren, allFieldMeta, newChildren, updatedView, response, error_1;
            var _a, _b, _c, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        selectedKeys = values.selectedColumns.map(function (col) { return col.key; });
                        currentChildren = solidView.layout.children;
                        allFieldMeta = solidFieldsMetadata;
                        newChildren = fields
                            .filter(function (col) { return selectedKeys.includes(col.key); })
                            .map(function (_a) {
                            var _b, _c;
                            var key = _a.key;
                            var existingChild = currentChildren.find(function (child) { return child.attrs.name === key; });
                            if (existingChild)
                                return existingChild;
                            // @ts-ignore
                            var fieldType = (_b = allFieldMeta[key]) === null || _b === void 0 ? void 0 : _b.type;
                            var isTextType = fieldType === "shortText" || fieldType === "longText" || fieldType === "selectionStatic" || fieldType === "selectionDynamic";
                            return {
                                type: 'field',
                                attrs: __assign({ name: key, label: ((_c = allFieldMeta[key]) === null || _c === void 0 ? void 0 : _c.displayName) || key }, (isTextType ? { isSearchable: true } : {})),
                            };
                        });
                        updatedView = {
                            layout: __assign(__assign({}, solidView.layout), { children: newChildren })
                        };
                        _e.label = 1;
                    case 1:
                        _e.trys.push([1, 4, , 5]);
                        if (!((_b = (_a = listViewMetaData === null || listViewMetaData === void 0 ? void 0 : listViewMetaData.data) === null || _a === void 0 ? void 0 : _a.solidView) === null || _b === void 0 ? void 0 : _b.id)) return [3 /*break*/, 3];
                        return [4 /*yield*/, upsertUserView({
                                viewMetadataId: (_d = (_c = listViewMetaData === null || listViewMetaData === void 0 ? void 0 : listViewMetaData.data) === null || _c === void 0 ? void 0 : _c.solidView) === null || _d === void 0 ? void 0 : _d.id,
                                layout: JSON.stringify(updatedView.layout),
                            }).unwrap()];
                    case 2:
                        response = _e.sent();
                        if (response.statusCode === 200) {
                            showToast(toast, "success", ERROR_MESSAGES.LAYOUT, ERROR_MESSAGES.FORM_LAYOUT_UPDATE);
                            window.location.reload();
                        }
                        _e.label = 3;
                    case 3: return [3 /*break*/, 5];
                    case 4:
                        error_1 = _e.sent();
                        console.error(ERROR_MESSAGES.UPDATING_USER, error_1);
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        }); },
    });
    var onDragEnd = function (result) {
        setIsDragging(false);
        if (!result.destination)
            return;
        var reordered = __spreadArray([], fields, true);
        var moved = reordered.splice(result.source.index, 1)[0];
        reordered.splice(result.destination.index, 0, moved);
        setFields(reordered);
    };
    return (_jsxs(_Fragment, { children: [_jsx(Toast, { ref: toast }), _jsxs("form", { onSubmit: formik.handleSubmit, className: "flex flex-column gap-1 px-2", children: [_jsx(DragDropContext, { onDragEnd: onDragEnd, onDragStart: function () { return setIsDragging(true); }, children: _jsx(Droppable, { droppableId: "columns", children: function (provided) { return (_jsxs("div", __assign({}, provided.droppableProps, { ref: provided.innerRef, className: "flex flex-column cogwheel-column-filter px-1 ".concat(isDragging ? styles.SolidColumnDragContextActive : ''), style: { maxHeight: 400, overflowY: 'auto' }, children: [fields.map(function (column, index) { return (_jsx(Draggable, { draggableId: column.key, index: index, children: function (provided, snapshot) { return (_jsxs("div", __assign({ ref: provided.innerRef }, provided.draggableProps, provided.dragHandleProps, { className: "flex align-items-center justify-content-between gap-3 px-3 py-2 ".concat(snapshot.isDragging ? styles.SolidColumnDraggedActiveElement : ''), style: __assign({}, provided.draggableProps.style), children: [_jsxs("div", { className: 'flex align-items-center gap-1', children: [_jsx(Checkbox, { inputId: column.key, name: "selectedColumns", value: column, onChange: function () {
                                                                var isChecked = formik.values.selectedColumns.some(function (item) { return item.key === column.key; });
                                                                formik.setFieldValue("selectedColumns", isChecked
                                                                    ? formik.values.selectedColumns.filter(function (item) { return item.key !== column.key; })
                                                                    : __spreadArray(__spreadArray([], formik.values.selectedColumns, true), [column], false));
                                                            }, checked: formik.values.selectedColumns.some(function (item) { return item.key === column.key; }), className: "text-base flex align-items-center" }), _jsx("label", { htmlFor: column.key, className: "ml-2 text-base", children: column.name })] }), _jsx(DragActive, { active: snapshot.isDragging })] }))); } }, column.key)); }), provided.placeholder] }))); } }) }), _jsxs("div", { className: "p-3 flex gap-2", children: [_jsx(Button, { type: 'submit', label: "Apply", size: "small" }), _jsx(Button, { type: 'button', outlined: true, label: "Cancel", size: "small", 
                                // @ts-ignore
                                onClick: function (e) { return customizeLayout.current.hide(e); } })] })] })] }));
};
var DragActive = function (_a) {
    var active = _a.active;
    return (_jsxs("svg", { xmlns: "http://www.w3.org/2000/svg", width: "16", height: "16", viewBox: "0 0 16 16", fill: "none", children: [_jsx("path", { d: "M7.46354 11.7331C7.46354 12.0997 7.33299 12.4136 7.07187 12.6747C6.81076 12.9359 6.49687 13.0664 6.13021 13.0664C5.76354 13.0664 5.44965 12.9359 5.18854 12.6747C4.92743 12.4136 4.79688 12.0997 4.79688 11.7331C4.79688 11.3664 4.92743 11.0525 5.18854 10.7914C5.44965 10.5303 5.76354 10.3997 6.13021 10.3997C6.49687 10.3997 6.81076 10.5303 7.07187 10.7914C7.33299 11.0525 7.46354 11.3664 7.46354 11.7331ZM7.46354 7.73307C7.46354 8.09974 7.33299 8.41363 7.07187 8.67474C6.81076 8.93585 6.49687 9.06641 6.13021 9.06641C5.76354 9.06641 5.44965 8.93585 5.18854 8.67474C4.92743 8.41363 4.79688 8.09974 4.79688 7.73307C4.79688 7.36641 4.92743 7.05252 5.18854 6.79141C5.44965 6.5303 5.76354 6.39974 6.13021 6.39974C6.49687 6.39974 6.81076 6.5303 7.07187 6.79141C7.33299 7.05252 7.46354 7.36641 7.46354 7.73307ZM7.46354 3.73307C7.46354 4.09974 7.33299 4.41363 7.07187 4.67474C6.81076 4.93585 6.49687 5.06641 6.13021 5.06641C5.76354 5.06641 5.44965 4.93585 5.18854 4.67474C4.92743 4.41363 4.79688 4.09974 4.79688 3.73307C4.79688 3.36641 4.92743 3.05252 5.18854 2.79141C5.44965 2.5303 5.76354 2.39974 6.13021 2.39974C6.49687 2.39974 6.81076 2.5303 7.07187 2.79141C7.33299 3.05252 7.46354 3.36641 7.46354 3.73307Z", fill: "var(--icon-color)", "fill-opacity": active ? "0.75" : "0.25" }), _jsx("path", { d: "M11.737 11.7331C11.737 12.0997 11.6064 12.4136 11.3453 12.6747C11.0842 12.9359 10.7703 13.0664 10.4036 13.0664C10.037 13.0664 9.72309 12.9359 9.46198 12.6747C9.20087 12.4136 9.07031 12.0997 9.07031 11.7331C9.07031 11.3664 9.20087 11.0525 9.46198 10.7914C9.72309 10.5303 10.037 10.3997 10.4036 10.3997C10.7703 10.3997 11.0842 10.5303 11.3453 10.7914C11.6064 11.0525 11.737 11.3664 11.737 11.7331ZM11.737 7.73307C11.737 8.09974 11.6064 8.41363 11.3453 8.67474C11.0842 8.93585 10.7703 9.06641 10.4036 9.06641C10.037 9.06641 9.72309 8.93585 9.46198 8.67474C9.20087 8.41363 9.07031 8.09974 9.07031 7.73307C9.07031 7.36641 9.20087 7.05252 9.46198 6.79141C9.72309 6.5303 10.037 6.39974 10.4036 6.39974C10.7703 6.39974 11.0842 6.5303 11.3453 6.79141C11.6064 7.05252 11.737 7.36641 11.737 7.73307ZM11.737 3.73307C11.737 4.09974 11.6064 4.41363 11.3453 4.67474C11.0842 4.93585 10.7703 5.06641 10.4036 5.06641C10.037 5.06641 9.72309 4.93585 9.46198 4.67474C9.20087 4.41363 9.07031 4.09974 9.07031 3.73307C9.07031 3.36641 9.20087 3.05252 9.46198 2.79141C9.72309 2.5303 10.037 2.39974 10.4036 2.39974C10.7703 2.39974 11.0842 2.5303 11.3453 2.79141C11.6064 3.05252 11.737 3.36641 11.737 3.73307Z", fill: "var(--icon-color)", "fill-opacity": active ? "0.75" : "0.25" })] }));
};
