// @solidxai/core-ui/dist/components/core/kanban/KanbanBoard.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
// @ts-nocheck
import { useState } from "react";
import { DragDropContext } from "@hello-pangea/dnd";
import KanbanColumn from "./KanbanColumn";
export var KanbanBoard = function (_a) {
    var groupByFieldName = _a.groupByFieldName, groupedView = _a.groupedView, kanbanViewData = _a.kanbanViewData, maxSwimLanesCount = _a.maxSwimLanesCount, solidKanbanViewMetaData = _a.solidKanbanViewMetaData, setKanbanViewData = _a.setKanbanViewData, handleLoadMore = _a.handleLoadMore, onDragEnd = _a.onDragEnd, handleSwimLanePagination = _a.handleSwimLanePagination, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox, editButtonUrl = _a.editButtonUrl;
    var _b = useState(true), loading = _b[0], setLoading = _b[1];
    // State to manage the folded status of each column
    var _c = useState({}), foldedStates = _c[0], setFoldedStates = _c[1];
    // Toggle fold (not yet implemented)
    var toggleFold = function (status) {
        setFoldedStates(function (prevFoldedStates) {
            var _a;
            return (__assign(__assign({}, prevFoldedStates), (_a = {}, _a[status] = !prevFoldedStates[status], _a)));
        });
    };
    // Render the Kanban board
    return (
    //@ts-ignore
    _jsx("div", { className: "solid-kanban-board-wrapper", children: _jsx(DragDropContext, { onDragEnd: onDragEnd, children: _jsxs("div", { className: "flex gap-4 px-4 md:px-5 py-3 md:py-4 solid-kanban-board-scroll-context", children: [kanbanViewData.map(function (data) {
                        var _a;
                        // Find the displayName for the groupName from solidKanbanViewMetaData.solidFieldsMetadata
                        var label = data.groupName;
                        var fieldMeta = (_a = solidKanbanViewMetaData === null || solidKanbanViewMetaData === void 0 ? void 0 : solidKanbanViewMetaData.solidFieldsMetadata) === null || _a === void 0 ? void 0 : _a[groupByFieldName];
                        if (fieldMeta &&
                            fieldMeta.type === "selectionStatic" &&
                            Array.isArray(fieldMeta.selectionStaticValues)) {
                            var match = fieldMeta.selectionStaticValues.find(function (v) {
                                var _a = v.split(":"), value = _a[0], displayName = _a[1];
                                return value === data.groupName;
                            });
                            if (match) {
                                label = match.split(":")[1];
                            }
                        }
                        var group = {
                            label: label,
                            count: data.groupData.meta.totalRecords,
                            limit: data.groupData.meta.perPage,
                            currentPage: data.groupData.meta.currentPage,
                            folded: foldedStates[data.groupName] || false,
                        };
                        return (_jsx(KanbanColumn, { groupedView: groupedView, groupByField: data.groupName, group: group, solidKanbanViewMetaData: solidKanbanViewMetaData, groupData: data.groupData.records, toggleFold: toggleFold, handleLoadMore: handleLoadMore, setLightboxUrls: setLightboxUrls, setOpenLightbox: setOpenLightbox, editButtonUrl: editButtonUrl }, data.groupName));
                    }), groupedView !== false && kanbanViewData.length < maxSwimLanesCount &&
                        _jsx("div", { children: _jsxs("a", { size: "small", className: "kaban-swimlane-load-more", style: { textWrap: 'nowrap' }, text: true, onClick: handleSwimLanePagination, children: ["Load More...(", maxSwimLanesCount - kanbanViewData.length, ")"] }) })] }) }) }));
};
export default KanbanBoard;
