// @solidxai/core-ui/dist/adapters/auth/handleLogout.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { ERROR_MESSAGES } from "../../constants/error-messages";
import { getSession, signOut } from "./index";
import { solidPost } from "../../http/solidHttp";
export function handleLogout(_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k;
    var toast = _a.toast;
    return __awaiter(this, void 0, void 0, function () {
        var session, refreshToken, response, error_1, err, message;
        return __generator(this, function (_l) {
            switch (_l.label) {
                case 0: return [4 /*yield*/, getSession()];
                case 1:
                    session = _l.sent();
                    refreshToken = (_b = session === null || session === void 0 ? void 0 : session.user) === null || _b === void 0 ? void 0 : _b.refreshToken;
                    _l.label = 2;
                case 2:
                    _l.trys.push([2, 7, , 8]);
                    return [4 /*yield*/, solidPost("/iam/logout", {
                            refreshToken: refreshToken, // Pass refresh token in body
                        })];
                case 3:
                    response = _l.sent();
                    console.log("logout response", response);
                    if (!(((_c = response === null || response === void 0 ? void 0 : response.data) === null || _c === void 0 ? void 0 : _c.statusCode) === 200)) return [3 /*break*/, 5];
                    return [4 /*yield*/, signOut({ callbackUrl: '/auth/login' })];
                case 4:
                    _l.sent();
                    return [3 /*break*/, 6];
                case 5:
                    (_d = toast === null || toast === void 0 ? void 0 : toast.current) === null || _d === void 0 ? void 0 : _d.show({
                        severity: 'error',
                        summary: ERROR_MESSAGES.LOGOUT_FAILED,
                        detail: "".concat((_f = (_e = response === null || response === void 0 ? void 0 : response.data) === null || _e === void 0 ? void 0 : _e.data) === null || _f === void 0 ? void 0 : _f.status),
                        life: 3000,
                    });
                    _l.label = 6;
                case 6: return [3 /*break*/, 8];
                case 7:
                    error_1 = _l.sent();
                    err = error_1;
                    message = ((_j = (_h = (_g = err.response) === null || _g === void 0 ? void 0 : _g.data) === null || _h === void 0 ? void 0 : _h.data) === null || _j === void 0 ? void 0 : _j.message) || err.message || ERROR_MESSAGES.LOGOUT_FAILED;
                    (_k = toast === null || toast === void 0 ? void 0 : toast.current) === null || _k === void 0 ? void 0 : _k.show({
                        severity: 'error',
                        summary: ERROR_MESSAGES.LOGOUT_FAILED,
                        detail: message,
                        life: 3000,
                    });
                    return [3 /*break*/, 8];
                case 8: return [2 /*return*/];
            }
        });
    });
}
