// @solidxai/core-ui/dist/adapters/auth/signIn.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jwtDecode } from "jwt-decode";
import { env } from "../env";
import { saveSession } from "./storage";
import { eventBus, AppEvents } from "../../helpers/eventBus";
import { solidPost } from "../..//http/solidHttp";
export function signIn(provider, options) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
    if (options === void 0) { options = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var identifier, password, apiUrl, response, accessToken, refreshToken, user, decoded, accessTokenExpires, session, error_1, message;
        return __generator(this, function (_o) {
            switch (_o.label) {
                case 0:
                    if (provider !== "credentials") {
                        return [2 /*return*/, { ok: false, error: "Unsupported provider: ".concat(provider), status: 400, url: null }];
                    }
                    identifier = options.identifier, password = options.password;
                    apiUrl = env("API_URL");
                    if (!apiUrl) {
                        return [2 /*return*/, { ok: false, error: "API_URL is not configured", status: 500, url: null }];
                    }
                    _o.label = 1;
                case 1:
                    _o.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, solidPost("".concat(apiUrl, "/api/iam/authenticate"), {
                            username: identifier,
                            email: identifier.includes('@') ? identifier : null,
                            password: password,
                        }, {
                            headers: {
                                accept: "*/*",
                                "Content-Type": "application/json",
                            },
                        })];
                case 2:
                    response = _o.sent();
                    accessToken = (_b = (_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.accessToken;
                    refreshToken = (_d = (_c = response === null || response === void 0 ? void 0 : response.data) === null || _c === void 0 ? void 0 : _c.data) === null || _d === void 0 ? void 0 : _d.refreshToken;
                    user = ((_f = (_e = response === null || response === void 0 ? void 0 : response.data) === null || _e === void 0 ? void 0 : _e.data) === null || _f === void 0 ? void 0 : _f.user) || {};
                    if (!accessToken || !refreshToken) {
                        return [2 /*return*/, { ok: false, error: "Missing tokens in response", status: response.status || 500, url: null }];
                    }
                    decoded = jwtDecode(accessToken);
                    accessTokenExpires = decoded.exp ? decoded.exp * 1000 : undefined;
                    session = {
                        user: __assign(__assign({}, user), { accessToken: accessToken, refreshToken: refreshToken, accessTokenExpires: accessTokenExpires }),
                        error: null,
                    };
                    saveSession(session);
                    eventBus.emit(AppEvents.SessionUpdated, session);
                    return [2 /*return*/, { ok: true, error: null, status: response.status || 200, url: null }];
                case 3:
                    error_1 = _o.sent();
                    message = ((_h = (_g = error_1 === null || error_1 === void 0 ? void 0 : error_1.response) === null || _g === void 0 ? void 0 : _g.data) === null || _h === void 0 ? void 0 : _h.message) || ((_l = (_k = (_j = error_1 === null || error_1 === void 0 ? void 0 : error_1.response) === null || _j === void 0 ? void 0 : _j.data) === null || _k === void 0 ? void 0 : _k.data) === null || _l === void 0 ? void 0 : _l.message) || (error_1 === null || error_1 === void 0 ? void 0 : error_1.message) || "Login failed";
                    return [2 /*return*/, { ok: false, error: message, status: ((_m = error_1 === null || error_1 === void 0 ? void 0 : error_1.response) === null || _m === void 0 ? void 0 : _m.status) || 500, url: null }];
                case 4: return [2 /*return*/];
            }
        });
    });
}
