// @solidxai/core-ui/dist/components/core/form/fields/relations/SolidRelationManyToManyField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Message } from "primereact/message";
import { useEffect, useState } from "react";
import * as Yup from 'yup';
import { getExtensionComponent } from "../../../../../helpers/registry";
import { AutoComplete } from "primereact/autocomplete";
import { Button } from "primereact/button";
import { useRelationEntityHandler } from "./widgets/helpers/useRelationEntityHandler";
import { InlineRelationEntityDialog } from "./widgets/helpers/InlineRelationEntityDialog";
import { capitalize } from "lodash";
import { Checkbox } from "primereact/checkbox";
import { Panel } from "primereact/panel";
import { SolidFieldTooltip } from "../../../../../components/common/SolidFieldTooltip";
import qs from 'qs';
// import Handlebars from "handlebars/dist/handlebars";
import * as Handlebars from "handlebars";
import { ERROR_MESSAGES } from "../../../../../constants/error-messages";
var SolidRelationManyToManyField = /** @class */ (function () {
    function SolidRelationManyToManyField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidRelationManyToManyField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b;
            var manyToManyFieldData = this.fieldContext.data[this.fieldContext.field.attrs.name];
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var userKeyField = (_b = (_a = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _a === void 0 ? void 0 : _a.userKeyField) === null || _b === void 0 ? void 0 : _b.name;
            if (manyToManyFieldData) {
                return manyToManyFieldData.map(function (e) {
                    var manyToManyColVal = e[userKeyField] || '';
                    return {
                        label: manyToManyColVal,
                        value: (e === null || e === void 0 ? void 0 : e.id) || '',
                        original: e
                    };
                });
            }
            return [];
        }
    });
    Object.defineProperty(SolidRelationManyToManyField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            //if empty then clear the field
            if (value && value.length === 0) {
                formData.append("".concat(fieldLayoutInfo.attrs.name, "Command"), "clear");
            }
            if (value && value.length > 0) {
                var shouldUseOriginal_1 = value.every(function (item) { return item.original && item.original.id; });
                value.forEach(function (item, index) {
                    if (shouldUseOriginal_1) {
                        formData.append("".concat(fieldLayoutInfo.attrs.name, "Ids[").concat(index, "]"), item.value);
                    }
                    else {
                        formData.append("".concat(fieldLayoutInfo.attrs.name, "[").concat(index, "]"), JSON.stringify(item.original));
                    }
                });
                if (shouldUseOriginal_1) {
                    formData.append("".concat(fieldLayoutInfo.attrs.name, "Command"), "set");
                }
                else {
                    formData.append("".concat(fieldLayoutInfo.attrs.name, "Command"), "update");
                }
            }
        }
    });
    Object.defineProperty(SolidRelationManyToManyField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            var schema = Yup.array();
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // 1. required 
            if (fieldMetadata.required) {
                schema = schema
                    .min(1, ERROR_MESSAGES.SELECT_ATLEAST_ONE(fieldLabel))
                    .required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
            }
            return schema;
        }
    });
    Object.defineProperty(SolidRelationManyToManyField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a, _b;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultRelationManyToManyAutoCompleteFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultRelationOneToManyFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" && (_jsx(_Fragment, { children: editWidget &&
                                this.renderExtensionRenderMode(editWidget, formik) }))] }) }));
        }
    });
    Object.defineProperty(SolidRelationManyToManyField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidRelationManyToManyField;
}());
export { SolidRelationManyToManyField };
export var DefaultRelationManyToManyAutoCompleteFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var disabled = (_e = fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.disabled;
    var readOnly = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.readOnly;
    var whereClause = fieldLayoutInfo.attrs.whereClause;
    var _h = useState(false), visibleCreateDialog = _h[0], setVisibleCreateDialog = _h[1];
    var _j = useRelationEntityHandler({ fieldContext: fieldContext, formik: formik }), autoCompleteItems = _j.autoCompleteItems, fetchRelationEntities = _j.fetchRelationEntities, addNewRelation = _j.addNewRelation;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    // const onChange = (e: any) => {
    //     formik.setFieldValue(fieldContext.field.attrs.name, e.value);
    // };
    var autoCompleteSearch = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, fixedFilterToBeApplied, fixedFilterParsed, convertedFixedFilter, fixedFilterTemplate, renderedFilter, parsedFilter_1, isValid, autocompleteQs;
        var _a, _b;
        var _c, _d, _e, _f, _g, _h;
        return __generator(this, function (_j) {
            queryData = {
                offset: 0,
                limit: 1000,
                filters: {
                    $and: [
                        (_a = {},
                            _a[(_d = (_c = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _c === void 0 ? void 0 : _c.userKeyField) === null || _d === void 0 ? void 0 : _d.name] = (_b = {},
                                _b[((_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.autocompleteMatchMode) || '$containsi'] = event.query,
                                _b),
                            _a)
                    ]
                }
            };
            fixedFilterToBeApplied = false;
            fixedFilterParsed = false;
            if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter) || ((_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.whereClause)) {
                convertedFixedFilter = ((_g = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.whereClause) ? (_h = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.whereClause : fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter;
                fixedFilterToBeApplied = true;
                fixedFilterTemplate = Handlebars.compile(convertedFixedFilter);
                renderedFilter = fixedFilterTemplate(formik.values);
                try {
                    parsedFilter_1 = JSON.parse(renderedFilter);
                    isValid = function (obj) {
                        if (!obj || typeof obj !== 'object')
                            return false;
                        var hasValidValue = function (val) {
                            if (val === null || val === undefined || val === '')
                                return false;
                            if (typeof val === 'object') {
                                return Object.values(val).some(hasValidValue);
                            }
                            return true;
                        };
                        return hasValidValue(parsedFilter_1);
                    };
                    if (isValid(parsedFilter_1)) {
                        queryData.filters.$and.push(parsedFilter_1);
                        fixedFilterParsed = true;
                    }
                    else {
                        console.warn(ERROR_MESSAGES.SKIPPING_EMPTY_FIXED_FILTER, parsedFilter_1);
                    }
                }
                catch (e) {
                    console.error(ERROR_MESSAGES.INVALID_JSON_WHERECLAUSE, renderedFilter);
                    parsedFilter_1 = {};
                }
            }
            autocompleteQs = qs.stringify(queryData, {
                encodeValuesOnly: true,
            });
            // if (whereClause) {
            //     autocompleteQs = `${autocompleteQs}&${whereClause}`;
            // }
            if (fixedFilterToBeApplied && !fixedFilterParsed) {
                console.error(ERROR_MESSAGES.FIXED_FILTER_NOT_APPLIED);
            }
            else {
                //  const autocompleteQs = qs.stringify(queryData, {
                //     encodeValuesOnly: true,
                // });
                fetchRelationEntities(autocompleteQs);
            }
            return [2 /*return*/];
        });
    }); };
    return (_jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsxs("div", { className: "flex align-items-center gap-3", children: [_jsx(AutoComplete, __assign({ readOnly: readOnly || readOnlyPermission, disabled: disabled || readOnlyPermission, multiple: true }, formik.getFieldProps(fieldLayoutInfo.attrs.name), { id: fieldLayoutInfo.attrs.name, field: "label", value: formik.values[fieldLayoutInfo.attrs.name] || '', dropdown: !readOnlyPermission, suggestions: autoCompleteItems, completeMethod: autoCompleteSearch, onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, className: "solid-standard-autocomplete w-full" })), fieldContext.field.attrs.inlineCreate && (_jsxs(_Fragment, { children: [_jsx("div", { children: _jsx(Button, { icon: "pi pi-plus", rounded: true, outlined: true, "aria-label": "Filter", type: "button", size: "small", onClick: function () { return setVisibleCreateDialog(true); }, className: "custom-add-button" }) }), _jsx(InlineRelationEntityDialog, { visible: visibleCreateDialog, setVisible: setVisibleCreateDialog, fieldContext: fieldContext, onCreate: addNewRelation })] }))] })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_g = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _g === void 0 ? void 0 : _g.toString() }) }))] }));
};
export var DefaultRelationManyToManyCheckBoxFormEditWidget = function (_a) {
    var _b;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var showFieldLabel = (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var _c = useState(false), visibleCreateDialog = _c[0], setVisibleCreateDialog = _c[1];
    var _d = useRelationEntityHandler({ fieldContext: fieldContext, formik: formik }), autoCompleteItems = _d.autoCompleteItems, fetchRelationEntities = _d.fetchRelationEntities, addNewRelation = _d.addNewRelation;
    useEffect(function () {
        var _a, _b, _c;
        var fieldMetadata = fieldContext.fieldMetadata;
        var fieldLayoutInfo = fieldContext.field;
        var queryData = {
            offset: 0,
            limit: 1000,
            filters: {
                $and: []
            }
        };
        var fixedFilterToBeApplied = false;
        var fixedFilterParsed = false;
        if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter) || ((_a = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.whereClause)) {
            var convertedFixedFilter = ((_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.whereClause) ? (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.whereClause : fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter;
            fixedFilterToBeApplied = true;
            var fixedFilterTemplate = Handlebars.compile(convertedFixedFilter);
            var renderedFilter = fixedFilterTemplate(formik.values);
            var parsedFilter_2;
            try {
                parsedFilter_2 = JSON.parse(renderedFilter);
                var isValid = function (obj) {
                    if (!obj || typeof obj !== 'object')
                        return false;
                    var hasValidValue = function (val) {
                        if (val === null || val === undefined || val === '')
                            return false;
                        if (typeof val === 'object') {
                            return Object.values(val).some(hasValidValue);
                        }
                        return true;
                    };
                    return hasValidValue(parsedFilter_2);
                };
                if (isValid(parsedFilter_2)) {
                    queryData.filters.$and.push(parsedFilter_2);
                    fixedFilterParsed = true;
                }
                else {
                    console.warn(ERROR_MESSAGES.SKIPPING_EMPTY_FIXED_FILTER, parsedFilter_2);
                }
            }
            catch (e) {
                console.error(ERROR_MESSAGES.INVALID_JSON_WHERECLAUSE, renderedFilter);
                parsedFilter_2 = {};
            }
        }
        if (fixedFilterToBeApplied && !fixedFilterParsed) {
            console.error(ERROR_MESSAGES.FIXED_FILTER_NOT_APPLIED);
        }
        else {
            var autocompleteQs = qs.stringify(queryData, {
                encodeValuesOnly: true,
            });
            fetchRelationEntities(autocompleteQs);
        }
    }, [fieldContext, formik.values]);
    var handleCheckboxChange = function (e) {
        if (formik.values[fieldLayoutInfo.attrs.name].some(function (item) { return item.value === e.value; })) {
            formik.setFieldValue(fieldLayoutInfo.attrs.name, formik.values[fieldLayoutInfo.attrs.name].filter(function (s) { return s.value !== e.value; }));
        }
        else {
            formik.setFieldValue(fieldLayoutInfo.attrs.name, __spreadArray(__spreadArray([], formik.values[fieldLayoutInfo.attrs.name], true), [e], false));
        }
    };
    var headerTemplate = function (options) {
        var className = "".concat(options.className, " justify-content-space-between");
        return (_jsxs("div", { className: className, children: [_jsxs("div", { className: "flex align-items-center gap-3", children: [showFieldLabel != false &&
                            _jsxs("label", { className: "form-field-label", children: [capitalize(fieldLayoutInfo.attrs.name), fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), fieldContext.field.attrs.inlineCreate && (_jsxs(_Fragment, { children: [_jsx(Button, { icon: "pi pi-plus", rounded: true, outlined: true, "aria-label": "Filter", type: "button", size: "small", onClick: function () { return setVisibleCreateDialog(true); }, className: "custom-add-button" }), _jsx(InlineRelationEntityDialog, { visible: visibleCreateDialog, setVisible: setVisibleCreateDialog, fieldContext: fieldContext, onCreate: addNewRelation })] }))] }), _jsx("div", { children: options.togglerElement })] }));
    };
    return (_jsx("div", { children: _jsx(Panel, { toggleable: true, headerTemplate: headerTemplate, children: _jsx("div", { className: "formgrid grid", children: autoCompleteItems && autoCompleteItems.map(function (a, i) {
                    return (_jsxs("div", { className: "field col-6 flex gap-2 ".concat(i >= 2 ? 'mt-3' : ''), children: [_jsx(Checkbox, { readOnly: readOnlyPermission, inputId: a.label, checked: formik.values[fieldLayoutInfo.attrs.name].some(function (item) { return item.label === a.label; }), onChange: function () { return handleCheckboxChange(a); } }), _jsxs("label", { htmlFor: a.label, className: "form-field-label m-0", children: [" ", a.label] })] }, a.label));
                }) }) }) }));
};
