// @solidxai/core-ui/dist/components/core/form/fields/relations/SolidRelationOneToManyField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { SolidListView } from "../../../../../components/core/list/SolidListView";
import { camelCase, capitalize } from "lodash";
import { Dialog } from "primereact/dialog";
import { useEffect, useState } from "react";
import * as Yup from 'yup';
import SolidFormView from '../../../../../components/core/form/SolidFormView';
import { usePathname } from "../../../../../hooks/usePathname";
import { useRouter } from "../../../../../hooks/useRouter";
import { getExtensionComponent } from "../../../../../helpers/registry";
import { SolidFieldTooltip } from "../../../../../components/common/SolidFieldTooltip";
import { Button } from "primereact/button";
import { ERROR_MESSAGES } from "../../../../../constants/error-messages";
var SolidRelationOneToManyField = /** @class */ (function () {
    function SolidRelationOneToManyField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidRelationOneToManyField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            // const manyToOneFieldData = this.fieldContext.data[this.fieldContext.field.attrs.name];
            // const fieldMetadata = this.fieldContext.fieldMetadata;
            // const userKeyField = fieldMetadata.relationModel.userKeyField.name;
            // const manyToOneColVal = manyToOneFieldData ? manyToOneFieldData[userKeyField] : '';
            // return { label: manyToOneColVal || '', value: manyToOneFieldData?.id || '' };
            return '';
        }
    });
    Object.defineProperty(SolidRelationOneToManyField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            // if(!value || value.length === 0) {
            //     formData.append(`${fieldLayoutInfo.attrs.name}Command`, "clear");
            // }
            if (value) {
                formData.append("".concat(fieldLayoutInfo.attrs.name, "Id"), value.value);
            }
        }
    });
    Object.defineProperty(SolidRelationOneToManyField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            var schema = Yup.object();
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // 1. required 
            if (fieldMetadata.required) {
                schema = schema.required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
            }
            return schema;
        }
    });
    Object.defineProperty(SolidRelationOneToManyField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a, _b, _c, _d, _e;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
            var userKeyFieldName = (_d = (_c = fieldMetadata.relationModel) === null || _c === void 0 ? void 0 : _c.userKeyField) === null || _d === void 0 ? void 0 : _d.name;
            var DynamicWidget = getExtensionComponent("SolidFormFieldRelationViewModeWidget");
            var widgetProps = {
                label: fieldLabel,
                value: (((_e = this.fieldContext.data) === null || _e === void 0 ? void 0 : _e[fieldLayoutInfo.attrs.name]) || []).map(function (item) { var _a; return ({ label: (_a = item[userKeyFieldName]) !== null && _a !== void 0 ? _a : '' }); }),
                layout: fieldLayoutInfo
            };
            // DefaultRelationOneToManyFormEditWidget
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultRelationOneToManyFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultRelationOneToManyFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" && (_jsx(_Fragment, { children: editWidget &&
                                this.renderExtensionRenderMode(editWidget, formik) }))] }) }));
        }
    });
    Object.defineProperty(SolidRelationOneToManyField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidRelationOneToManyField;
}());
export { SolidRelationOneToManyField };
export var DefaultRelationOneToManyFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var router = useRouter();
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var _w = useState(false), visibleCreateRelationEntity = _w[0], setvisibleCreateRelationEntity = _w[1];
    var _x = useState(), listViewParams = _x[0], setListViewParams = _x[1];
    var _y = useState(), formViewParams = _y[0], setformViewParams = _y[1];
    var _z = useState(false), refreshList = _z[0], setRefreshList = _z[1]; // Added state for rerender
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var pathname = usePathname();
    var lastPathSegment = pathname.split('/').pop();
    var userKeyField = (_f = Object.entries(fieldContext.solidFormViewMetaData.data.solidFieldsMetadata).find(function (_a) {
        var _ = _a[0], value = _a[1];
        return value.isUserKey;
    })) === null || _f === void 0 ? void 0 : _f[0];
    var _0 = useState(false), showSaveParentEntityConfirmationPopup = _0[0], setShowSaveParentEntityConfirmationPopup = _0[1];
    useEffect(function () {
        var urlParams = new URLSearchParams(window.location.search);
        var childEntity = urlParams.get('childEntity');
        if (childEntity === fieldLayoutInfo.attrs.name && lastPathSegment !== "new") {
            handlePopupOpen('new');
        }
    }, []);
    var handlePopupOpen = function (id) {
        var _a;
        var _b, _c;
        if (lastPathSegment === "new") {
            setShowSaveParentEntityConfirmationPopup(true);
        }
        else {
            var formviewparams = {
                moduleName: fieldContext.fieldMetadata.relationModelModuleName,
                id: id,
                embeded: true,
                isCustomCreate: false,
                customLayout: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineCreateLayout,
                modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
                parentFieldName: fieldContext.fieldMetadata.relationCoModelFieldName,
                parentData: userKeyField ? (_a = {}, _a[userKeyField] = { solidManyToOneLabel: fieldContext.data[userKeyField], solidManyToOneValue: fieldContext.data['id'] }, _a) : {},
                onEmbeddedFormSave: fieldContext.onEmbeddedFormSave,
                inlineCreateAutoSave: (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.inlineCreateAutoSave,
            };
            setformViewParams(formviewparams);
            setvisibleCreateRelationEntity(true);
        }
    };
    var handlePopupClose = function () {
        var _a;
        var _b, _c;
        setvisibleCreateRelationEntity(false);
        var currentUrl = new URL(window.location.href);
        currentUrl.searchParams.delete('childEntity');
        router.push(currentUrl.toString());
        setRefreshList(function (prev) { return !prev; });
        var customFilter = fieldContext.fieldMetadata.relationCoModelFieldName ? fieldContext.fieldMetadata.relationCoModelFieldName : "".concat(fieldContext.modelName);
        var lisviewparams = {
            moduleName: fieldContext.fieldMetadata.relationModelModuleName,
            modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
            inlineCreate: readOnlyPermission === false ? true : false,
            customLayout: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineListLayout,
            embeded: true,
            id: fieldContext.data ? (_c = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.data) === null || _c === void 0 ? void 0 : _c.id : 'new',
            customFilter: (_a = {},
                _a[customFilter] = {
                    id: {
                        $eq: fieldContext.data && fieldContext.data.id !== undefined ? fieldContext.data.id : -1
                    }
                },
                _a)
        };
        setListViewParams(lisviewparams);
    };
    //Intial Params 
    useEffect(function () {
        var _a, _b;
        var _c, _d, _e, _f;
        var customFilter = fieldContext.fieldMetadata.relationCoModelFieldName ? fieldContext.fieldMetadata.relationCoModelFieldName : "".concat(fieldContext.modelName);
        var listviewparams = {
            moduleName: fieldContext.fieldMetadata.relationModelModuleName,
            modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
            inlineCreate: readOnlyPermission === false ? true : false,
            customLayout: (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.inlineListLayout,
            embeded: true,
            id: fieldContext.data ? (_d = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.data) === null || _d === void 0 ? void 0 : _d.id : 'new',
            customFilter: (_a = {},
                _a[customFilter] = {
                    id: {
                        $eq: fieldContext.data && fieldContext.data.id !== undefined ? fieldContext.data.id : -1
                    }
                },
                _a)
        };
        setListViewParams(listviewparams);
        var formviewparams = {
            moduleName: fieldContext.fieldMetadata.relationModelModuleName,
            modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
            parentFieldName: fieldContext.fieldMetadata.relationCoModelFieldName,
            id: "new",
            embeded: true,
            isCustomCreate: false,
            customLayout: (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.inlineCreateLayout,
            parentData: userKeyField ? (_b = {}, _b[userKeyField] = { solidManyToOneLabel: fieldContext.data[userKeyField], solidManyToOneValue: fieldContext.data['id'] }, _b) : {},
            onEmbeddedFormSave: fieldContext.onEmbeddedFormSave,
            inlineCreateAutoSave: (_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.inlineCreateAutoSave,
        };
        setformViewParams(formviewparams);
    }, [readOnlyPermission]);
    var fieldDisabled = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.disabled;
    var fieldReadonly = (_h = fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.readonly;
    var formDisabled = (_l = (_k = (_j = solidFormViewMetaData.data.solidView) === null || _j === void 0 ? void 0 : _j.layout) === null || _k === void 0 ? void 0 : _k.attrs) === null || _l === void 0 ? void 0 : _l.disabled;
    var formReadonly = (_p = (_o = (_m = solidFormViewMetaData.data.solidView) === null || _m === void 0 ? void 0 : _m.layout) === null || _o === void 0 ? void 0 : _o.attrs) === null || _p === void 0 ? void 0 : _p.readonly;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var saveParentEntity = function () { return __awaiter(void 0, void 0, void 0, function () {
        var currentPath, currentUrl, updatedPath, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    currentPath = window.location.pathname;
                    currentUrl = new URL(window.location.href);
                    currentUrl.searchParams.set('childEntity', fieldLayoutInfo.attrs.name);
                    currentUrl.searchParams.set('viewMode', 'edit');
                    updatedPath = currentUrl.toString();
                    _b.label = 1;
                case 1:
                    _b.trys.push([1, 3, , 4]);
                    console.log("updatedPath", updatedPath);
                    router.push(updatedPath);
                    return [4 /*yield*/, formik.handleSubmit()];
                case 2:
                    _b.sent();
                    return [3 /*break*/, 4];
                case 3:
                    _a = _b.sent();
                    return [3 /*break*/, 4];
                case 4:
                    setShowSaveParentEntityConfirmationPopup(false);
                    return [2 /*return*/];
            }
        });
    }); };
    return (_jsxs("div", { children: [showFieldLabel != false &&
                _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), listViewParams &&
                _jsx(SolidListView, __assign({}, listViewParams, { handlePopUpOpen: handlePopupOpen }), refreshList.toString()), readOnlyPermission !== true && formViewParams &&
                _jsx(RenderSolidFormEmbededView, { formik: formik, fieldContext: fieldContext, visibleCreateRelationEntity: visibleCreateRelationEntity, setvisibleCreateRelationEntity: setvisibleCreateRelationEntity, formViewParams: formViewParams, handlePopupClose: handlePopupClose }), _jsx(Dialog, { showHeader: false, headerClassName: "py-2", contentClassName: "px-0 pb-0", className: "solid-confirm-dialog", contentStyle: { borderRadius: 6 }, visible: showSaveParentEntityConfirmationPopup, style: { width: '20vw' }, onHide: function () { if (!showSaveParentEntityConfirmationPopup)
                    return; setShowSaveParentEntityConfirmationPopup(false); }, children: _jsxs("div", { className: "p-4", children: [_jsxs("p", { className: "m-0 solid-primary-title", style: { fontSize: 16 }, children: ["Before Creating ", fieldLabel, " you need to save ", ((_s = (_r = (_q = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _q === void 0 ? void 0 : _q.solidView) === null || _r === void 0 ? void 0 : _r.model) === null || _s === void 0 ? void 0 : _s.displayName) ? (_v = (_u = (_t = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _t === void 0 ? void 0 : _t.solidView) === null || _u === void 0 ? void 0 : _u.model) === null || _v === void 0 ? void 0 : _v.displayName : capitalize(fieldContext.modelName), ". Please click save if you wish to proceed ?"] }), _jsxs("div", { className: "flex align-items-center justify-content-start gap-2 mt-3", children: [_jsx(Button, { label: "Save", size: "small", onClick: saveParentEntity }), _jsx(Button, { label: "Cancel", size: "small", onClick: function () { return setShowSaveParentEntityConfirmationPopup(false); }, outlined: true, className: 'bg-primary-reverse' })] })] }) })] }));
};
export var DefaultRelationOneToManyFormViewWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var _q = useState(false), visibleCreateRelationEntity = _q[0], setvisibleCreateRelationEntity = _q[1];
    var _r = useState(), listViewParams = _r[0], setListViewParams = _r[1];
    var _s = useState(), formViewParams = _s[0], setformViewParams = _s[1];
    var _t = useState(false), refreshList = _t[0], setRefreshList = _t[1]; // Added state for rerender
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var pathname = usePathname();
    var lastPathSegment = pathname.split('/').pop();
    var userKeyField = (_f = Object.entries(fieldContext.solidFormViewMetaData.data.solidFieldsMetadata).find(function (_a) {
        var _ = _a[0], value = _a[1];
        return value.isUserKey;
    })) === null || _f === void 0 ? void 0 : _f[0];
    var handlePopupOpen = function (id) {
        var _a;
        var _b, _c;
        var formviewparams = {
            moduleName: fieldContext.fieldMetadata.relationModelModuleName,
            id: id,
            embeded: true,
            isCustomCreate: false,
            customLayout: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineCreateLayout,
            modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
            parentFieldName: fieldContext.fieldMetadata.relationCoModelFieldName,
            parentData: userKeyField ? (_a = {}, _a[userKeyField] = { solidManyToOneLabel: fieldContext.data[userKeyField], solidManyToOneValue: fieldContext.data['id'] }, _a) : {},
            onEmbeddedFormSave: fieldContext.onEmbeddedFormSave,
            inlineCreateAutoSave: (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.inlineCreateAutoSave,
        };
        setformViewParams(formviewparams);
        setvisibleCreateRelationEntity(true);
    };
    var handlePopupClose = function () {
        var _a;
        var _b, _c, _d;
        setvisibleCreateRelationEntity(false);
        setRefreshList(function (prev) { return !prev; });
        var customFilter = fieldContext.fieldMetadata.relationCoModelFieldName ? fieldContext.fieldMetadata.relationCoModelFieldName : "".concat(fieldContext.modelName);
        var lisviewparams = {
            moduleName: fieldContext.fieldMetadata.relationModelModuleName,
            modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
            inlineCreate: readOnlyPermission === false ? true : false,
            customLayout: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineListLayout,
            embeded: true,
            id: fieldContext.data ? (_c = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.data) === null || _c === void 0 ? void 0 : _c.id : 'new',
            customFilter: (_a = {},
                _a[customFilter] = {
                    id: {
                        $eq: fieldContext.data ? (_d = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.data) === null || _d === void 0 ? void 0 : _d.id : -1
                    }
                },
                _a)
        };
        setListViewParams(lisviewparams);
    };
    //Intial Params 
    useEffect(function () {
        var _a, _b;
        var _c, _d, _e, _f, _g, _h;
        var customFilter = fieldContext.fieldMetadata.relationCoModelFieldName ? fieldContext.fieldMetadata.relationCoModelFieldName : "".concat(fieldContext.modelName);
        var listviewparams = {
            moduleName: fieldContext.fieldMetadata.relationModelModuleName,
            modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
            inlineCreate: readOnlyPermission === false ? true : false,
            customLayout: (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.inlineListLayout,
            embeded: true,
            id: fieldContext.data ? (_d = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.data) === null || _d === void 0 ? void 0 : _d.id : 'new',
            customFilter: (_a = {},
                _a[customFilter] = {
                    id: {
                        $eq: ((_e = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.data) === null || _e === void 0 ? void 0 : _e.id) !== undefined ? (_f = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.data) === null || _f === void 0 ? void 0 : _f.id : -1
                    }
                },
                _a)
        };
        setListViewParams(listviewparams);
        var formviewparams = {
            moduleName: fieldContext.fieldMetadata.relationModelModuleName,
            id: "new",
            embeded: true,
            inlineCreateAutoSave: (_g = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.inlineCreateAutoSave,
            customLayout: (_h = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.inlineCreateLayout,
            modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
            parentFieldName: fieldContext.fieldMetadata.relationCoModelFieldName,
            parentData: userKeyField ? (_b = {}, _b[userKeyField] = { solidManyToOneLabel: fieldContext.data[userKeyField], solidManyToOneValue: fieldContext.data['id'] }, _b) : {},
            onEmbeddedFormSave: fieldContext.onEmbeddedFormSave,
            isCustomCreate: false
        };
        setformViewParams(formviewparams);
    }, [readOnlyPermission]);
    var fieldDisabled = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.disabled;
    var fieldReadonly = (_h = fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.readonly;
    var formDisabled = (_l = (_k = (_j = solidFormViewMetaData.data.solidView) === null || _j === void 0 ? void 0 : _j.layout) === null || _k === void 0 ? void 0 : _k.attrs) === null || _l === void 0 ? void 0 : _l.disabled;
    var formReadonly = (_p = (_o = (_m = solidFormViewMetaData.data.solidView) === null || _m === void 0 ? void 0 : _m.layout) === null || _o === void 0 ? void 0 : _o.attrs) === null || _p === void 0 ? void 0 : _p.readonly;
    return (_jsxs("div", { children: [showFieldLabel != false &&
                _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), listViewParams &&
                _jsx(SolidListView, __assign({}, listViewParams, { handlePopUpOpen: handlePopupOpen }), refreshList.toString()), readOnlyPermission !== true && formViewParams &&
                _jsx(RenderSolidFormEmbededView, { fieldLayoutInfo: fieldLayoutInfo, visibleCreateRelationEntity: visibleCreateRelationEntity, setvisibleCreateRelationEntity: setvisibleCreateRelationEntity, formViewParams: formViewParams, handlePopupClose: handlePopupClose })] }));
};
export var RenderSolidFormEmbededView = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j;
    var fieldLayoutInfo = _a.fieldLayoutInfo, customCreateHandler = _a.customCreateHandler, visibleCreateRelationEntity = _a.visibleCreateRelationEntity, setvisibleCreateRelationEntity = _a.setvisibleCreateRelationEntity, formViewParams = _a.formViewParams, handlePopupClose = _a.handlePopupClose;
    var router = useRouter();
    var params = {
        moduleName: formViewParams.moduleName,
        id: formViewParams === null || formViewParams === void 0 ? void 0 : formViewParams.id,
        modelName: camelCase(formViewParams.modelName),
        embeded: formViewParams.embeded,
        customLayout: formViewParams.customLayout,
        customCreateHandler: (function (values) {
            setvisibleCreateRelationEntity(false);
            customCreateHandler(values);
        }),
        inlineCreateAutoSave: formViewParams.inlineCreateAutoSave,
        handlePopupClose: handlePopupClose,
        parentFieldName: formViewParams.parentFieldName,
        parentData: formViewParams.parentData,
        onEmbeddedFormSave: formViewParams.onEmbeddedFormSave,
        isCustomCreate: formViewParams.isCustomCreate
    };
    return (_jsx("div", { className: "many-to-many-add", children: _jsx(Dialog, { header: "", showHeader: false, visible: visibleCreateRelationEntity, className: "solid-dialog", style: {
                width: (_e = (_d = (_c = (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineCreateLayout) === null || _c === void 0 ? void 0 : _c.attrs) === null || _d === void 0 ? void 0 : _d.width) !== null && _e !== void 0 ? _e : "60vw",
                height: (_j = (_h = (_g = (_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.inlineCreateLayout) === null || _g === void 0 ? void 0 : _g.attrs) === null || _h === void 0 ? void 0 : _h.height) !== null && _j !== void 0 ? _j : "auto"
            }, onHide: function () {
                if (!visibleCreateRelationEntity)
                    return;
                setvisibleCreateRelationEntity(false);
            }, breakpoints: { '1199px': '35rem', "767px": '85vw', "550px": '90vw' }, children: params &&
                _jsx(SolidFormView, __assign({}, params)) }) }));
};
export var PseudoRelationOneToManyFormWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j;
    var formData = _a.formData, field = _a.field, fieldsMetadata = _a.fieldsMetadata, viewMetadata = _a.viewMetadata, formViewData = _a.formViewData;
    var fieldLayoutInfo = field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = fieldLayoutInfo.attrs.label;
    var fieldDescription = fieldLayoutInfo.attrs.description;
    var _k = useState(false), visibleCreateRelationEntity = _k[0], setvisibleCreateRelationEntity = _k[1];
    var _l = useState(), listViewParams = _l[0], setListViewParams = _l[1];
    var _m = useState(null), formViewParams = _m[0], setformViewParams = _m[1];
    var _o = useState(false), refreshList = _o[0], setRefreshList = _o[1]; // Added state for rerender
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    var readOnlyPermission = fieldLayoutInfo.readOnly ? fieldLayoutInfo.readOnly : false;
    var pathname = usePathname();
    var parentModelName = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.parentModelName;
    var childModelName = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.childModelName;
    var parentFieldName = (_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.parentFieldName;
    var childFieldName = (_g = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.childFieldName;
    var parentModuleName = (_h = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.parentModuleName;
    var childModuleName = (_j = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _j === void 0 ? void 0 : _j.childModuleName;
    var handlePopupOpen = function (id) {
        var _a;
        var _b, _c;
        var formviewparams = {
            moduleName: childModuleName,
            modelName: camelCase(childModelName),
            id: id,
            embeded: true,
            isCustomCreate: false,
            inlineCreateAutoSave: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineCreateAutoSave,
            customLayout: (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.inlineCreateLayout,
            parentFieldName: parentFieldName,
            parentData: childFieldName && (formViewData === null || formViewData === void 0 ? void 0 : formViewData.data) && (formViewData === null || formViewData === void 0 ? void 0 : formViewData.data[parentFieldName]) ? (_a = {}, _a[childFieldName] = formViewData.data[parentFieldName], _a) : {},
            onEmbeddedFormSave: fieldLayoutInfo.onEmbeddedFormSave ? fieldLayoutInfo : false
        };
        setformViewParams(formviewparams);
        setvisibleCreateRelationEntity(true);
    };
    var handlePopupClose = function () {
        var _a;
        var _b, _c;
        setvisibleCreateRelationEntity(false);
        setRefreshList(function (prev) { return !prev; });
        var customFilter = childModelName;
        var lisviewparams = {
            moduleName: childModuleName,
            modelName: camelCase(childModelName),
            inlineCreate: readOnlyPermission === false ? true : false,
            customLayout: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineListLayout,
            embeded: true,
            id: formViewData && (formViewData === null || formViewData === void 0 ? void 0 : formViewData.data) ? (_c = formViewData === null || formViewData === void 0 ? void 0 : formViewData.data) === null || _c === void 0 ? void 0 : _c.id : 'new',
            customFilter: childFieldName && (formViewData === null || formViewData === void 0 ? void 0 : formViewData.data)
                ? (_a = {},
                    _a[childFieldName] = {
                        $eq: formViewData.data[parentFieldName]
                    },
                    _a) : {
                id: {
                    $eq: -1
                }
            }
        };
        setListViewParams(lisviewparams);
    };
    //Intial Params 
    useEffect(function () {
        var _a, _b;
        var _c, _d, _e, _f;
        var customFilter = childModelName;
        var listviewparams = {
            moduleName: childModuleName,
            modelName: camelCase(childModelName),
            inlineCreate: readOnlyPermission === false ? true : false,
            customLayout: (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.inlineListLayout,
            embeded: true,
            id: formViewData && (formViewData === null || formViewData === void 0 ? void 0 : formViewData.data) ? (_d = formViewData === null || formViewData === void 0 ? void 0 : formViewData.data) === null || _d === void 0 ? void 0 : _d.id : 'new',
            customFilter: childFieldName && (formViewData === null || formViewData === void 0 ? void 0 : formViewData.data)
                ? (_a = {},
                    _a[childFieldName] = {
                        $eq: formViewData.data[parentFieldName]
                    },
                    _a) : {
                id: {
                    $eq: -1
                }
            }
        };
        setListViewParams(listviewparams);
        var formviewparams = {
            moduleName: childModuleName,
            modelName: camelCase(childModelName),
            id: "new",
            embeded: true,
            isCustomCreate: false,
            inlineCreateAutoSave: (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.inlineCreateAutoSave,
            customLayout: (_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.inlineCreateLayout,
            parentFieldName: parentFieldName,
            parentData: childFieldName && formViewData && (formViewData === null || formViewData === void 0 ? void 0 : formViewData.data) && (formViewData === null || formViewData === void 0 ? void 0 : formViewData.data[parentFieldName]) ? (_b = {}, _b[childFieldName] = formViewData.data[parentFieldName], _b) : {},
            onEmbeddedFormSave: formViewData && formViewData.onEmbeddedFormSave
        };
        setformViewParams(formviewparams);
    }, [readOnlyPermission, formViewData]);
    return (_jsxs("div", { children: [showFieldLabel != false &&
                _jsx("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: fieldLabel }), listViewParams &&
                _jsx(SolidListView, __assign({}, listViewParams, { handlePopUpOpen: handlePopupOpen }), refreshList.toString()), readOnlyPermission !== true && formViewParams &&
                _jsx(RenderSolidFormEmbededView, { fieldLayoutInfo: fieldLayoutInfo, visibleCreateRelationEntity: visibleCreateRelationEntity, setvisibleCreateRelationEntity: setvisibleCreateRelationEntity, formViewParams: formViewParams, handlePopupClose: handlePopupClose })] }));
};
