// @solidxai/core-ui/dist/components/core/form/fields/relations/SolidRelationManyToOneField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { createSolidEntityApi } from "../../../../../redux/api/solidEntityApi";
import { AutoComplete } from "primereact/autocomplete";
import { Message } from "primereact/message";
import qs from "qs";
import { useEffect, useRef, useState } from "react";
import * as Yup from 'yup';
import { camelCase } from "lodash";
import { Button } from "primereact/button";
import { Dialog } from "primereact/dialog";
import SolidFormView from '../../../../../components/core/form/SolidFormView';
import { getExtensionComponent } from "../../../../../helpers/registry";
import showToast from "../../../../../helpers/showToast";
// import Handlebars from "handlebars/dist/handlebars";
import * as Handlebars from "handlebars";
import { Toast } from "primereact/toast";
import { SolidFieldTooltip } from "../../../../../components/common/SolidFieldTooltip";
import { ERROR_MESSAGES } from "../../../../../constants/error-messages";
import { getVirtualScrollerOptions } from "../../../../../helpers/autoCompleteVirtualScroll";
var SolidRelationManyToOneField = /** @class */ (function () {
    function SolidRelationManyToOneField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidRelationManyToOneField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b, _c, _d, _e, _f, _g;
            var manyToOneFieldData = (_a = this.fieldContext) === null || _a === void 0 ? void 0 : _a.data[(_d = (_c = (_b = this.fieldContext) === null || _b === void 0 ? void 0 : _b.field) === null || _c === void 0 ? void 0 : _c.attrs) === null || _d === void 0 ? void 0 : _d.name];
            var fieldMetadata = (_e = this.fieldContext) === null || _e === void 0 ? void 0 : _e.fieldMetadata;
            var userKeyField = (_g = (_f = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _f === void 0 ? void 0 : _f.userKeyField) === null || _g === void 0 ? void 0 : _g.name;
            var manyToOneColVal = manyToOneFieldData ? manyToOneFieldData[userKeyField] : '';
            if (manyToOneColVal) {
                return __assign({ solidManyToOneLabel: manyToOneColVal || '', solidManyToOneValue: (manyToOneFieldData === null || manyToOneFieldData === void 0 ? void 0 : manyToOneFieldData.id) || '' }, manyToOneFieldData);
            }
            if (this.fieldContext.parentData) {
                var parentDataForKey = this.fieldContext.parentData[userKeyField];
                if (parentDataForKey && typeof parentDataForKey === 'object') {
                    var _h = Object.entries(this.fieldContext.parentData)[0] || [], key = _h[0], value = _h[1];
                    if (key && value !== undefined) {
                        return { solidManyToOneLabel: value.solidManyToOneLabel, solidManyToOneValue: value.solidManyToOneValue };
                    }
                }
            }
            return {};
        }
    });
    Object.defineProperty(SolidRelationManyToOneField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            if (value === null || value === void 0 ? void 0 : value.solidManyToOneValue) {
                formData.append("".concat(fieldLayoutInfo.attrs.name, "Id"), value.solidManyToOneValue);
            }
        }
    });
    Object.defineProperty(SolidRelationManyToOneField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            var schema = Yup.mixed();
            // Custom validation for relation field
            if (fieldMetadata.required) {
                schema = schema.test(ERROR_MESSAGES.REQUIRED_REALTION, ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel), function (value) {
                    // Handle empty values
                    if (!value)
                        return false;
                    // If it's an object with solidManyToOneValue, check if it's valid
                    if (typeof value === 'object' && value !== null && value.solidManyToOneValue) {
                        return true;
                    }
                    // If it's a string (user typed but didn't select), it's invalid for required field
                    if (typeof value === 'string') {
                        return false;
                    }
                    return false;
                });
            }
            // Add validation to ensure valid selection
            schema = schema.test(ERROR_MESSAGES.VALIDATE_SELECTION, ERROR_MESSAGES.SELECT_VALID_FROM_DROPDOWN(fieldLabel), function (value) {
                // If not required and empty, it's valid
                if (!fieldMetadata.required && (!value || value === '')) {
                    return true;
                }
                // If it's an object with solidManyToOneValue, it's a valid selection
                if (typeof value === 'object' && value !== null && value.solidManyToOneValue) {
                    return true;
                }
                // If it's a string (user typed but didn't select), it's invalid
                if (typeof value === 'string' && value.trim() !== '') {
                    return false;
                }
                // Empty value for non-required field
                return !fieldMetadata.required;
            });
            return schema;
        }
    });
    Object.defineProperty(SolidRelationManyToOneField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a, _b;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var isVisible = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.visible) || true;
            if (this.fieldContext.parentData && this.fieldContext.parentFieldName === fieldLayoutInfo.attrs.name) {
                isVisible = false;
            }
            // const isVisible = this.fieldContext.parentData ? fieldLayoutInfo.attrs?.visible === true : fieldLayoutInfo.attrs?.visible !== false;
            if (!isVisible) {
                return null;
            }
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultRelationManyToOneFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultRelationManyToOneFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" && this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" && (_jsx(_Fragment, { children: editWidget && this.renderExtensionRenderMode(editWidget, formik) }))] }) }));
        }
    });
    Object.defineProperty(SolidRelationManyToOneField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidRelationManyToOneField;
}());
export { SolidRelationManyToOneField };
export var DefaultRelationManyToOneFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var toast = useRef(null);
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var _p = useState(false), visibleCreateRelationEntity = _p[0], setvisibleCreateRelationEntity = _p[1];
    var _q = useState(), formViewParams = _q[0], setformViewParams = _q[1];
    // auto complete specific code. 
    var entityApi = createSolidEntityApi(fieldMetadata.relationCoModelSingularName);
    var useLazyGetSolidEntitiesQuery = entityApi.useLazyGetSolidEntitiesQuery;
    var triggerGetSolidEntities = useLazyGetSolidEntitiesQuery()[0];
    var fieldDisabled = (_e = fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.disabled;
    var fieldReadonly = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.readonly;
    var formDisabled = (_j = (_h = (_g = solidFormViewMetaData.data.solidView) === null || _g === void 0 ? void 0 : _g.layout) === null || _h === void 0 ? void 0 : _h.attrs) === null || _j === void 0 ? void 0 : _j.disabled;
    var formReadonly = (_m = (_l = (_k = solidFormViewMetaData.data.solidView) === null || _k === void 0 ? void 0 : _k.layout) === null || _l === void 0 ? void 0 : _l.attrs) === null || _m === void 0 ? void 0 : _m.readonly;
    var whereClause = fieldLayoutInfo.attrs.whereClause;
    var _r = useState([]), autoCompleteItems = _r[0], setAutoCompleteItems = _r[1];
    var _s = useState(0), offset = _s[0], setOffset = _s[1];
    var _t = useState(true), hasMore = _t[0], setHasMore = _t[1];
    var _u = useState(""), currentQuery = _u[0], setCurrentQuery = _u[1];
    var LIMIT = 50;
    var _v = useState(false), loading = _v[0], setLoading = _v[1];
    useEffect(function () {
        var _a, _b;
        var formviewparams = {
            moduleName: fieldContext.fieldMetadata.relationModelModuleName,
            modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName),
            id: "new",
            embeded: true,
            inlineCreateAutoSave: (_a = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.inlineCreateAutoSave,
            layout: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineCreateLayout
        };
        setformViewParams(formviewparams);
    }, []);
    var autoCompleteSearch = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, fixedFilterToBeApplied, fixedFilterParsed, convertedFixedFilter, fixedFilterTemplate, renderedFilter, parsedFilter, isValid, autocompleteQs, autocompleteResponse, autocompleteData, autoCompleteItems_1;
        var _a, _b;
        var _c, _d, _e, _f, _g, _h, _j, _k, _l;
        return __generator(this, function (_m) {
            switch (_m.label) {
                case 0:
                    setOffset(0);
                    setCurrentQuery(event.query || "");
                    setHasMore(true);
                    queryData = {
                        offset: 0,
                        limit: LIMIT,
                        filters: {
                            $and: [
                                (_a = {},
                                    _a[(_d = (_c = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _c === void 0 ? void 0 : _c.userKeyField) === null || _d === void 0 ? void 0 : _d.name] = (_b = {},
                                        _b[((_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.autocompleteMatchMode) || '$containsi'] = event.query,
                                        _b),
                                    _a)
                            ]
                        }
                    };
                    fixedFilterToBeApplied = false;
                    fixedFilterParsed = false;
                    if (((_h = (_g = (_f = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _f === void 0 ? void 0 : _f.solidView) === null || _g === void 0 ? void 0 : _g.model) === null || _h === void 0 ? void 0 : _h.singularName) === "listOfValues" && fieldContext.fieldMetadata.relationCoModelSingularName !== "moduleMetadata") {
                        fixedFilterToBeApplied = true;
                    }
                    if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter) || ((_j = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _j === void 0 ? void 0 : _j.whereClause)) {
                        convertedFixedFilter = ((_k = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _k === void 0 ? void 0 : _k.whereClause) ? (_l = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _l === void 0 ? void 0 : _l.whereClause : fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter;
                        fixedFilterToBeApplied = true;
                        fixedFilterTemplate = Handlebars.compile(convertedFixedFilter);
                        renderedFilter = fixedFilterTemplate(formik.values);
                        parsedFilter = void 0;
                        try {
                            parsedFilter = JSON.parse(renderedFilter);
                            isValid = function (obj) {
                                if (!obj || typeof obj !== 'object')
                                    return false;
                                // Recursively check all nested values
                                var hasValidValue = function (val) {
                                    if (val === null || val === undefined || val === '')
                                        return false;
                                    if (typeof val === 'object') {
                                        return Object.values(val).some(hasValidValue);
                                    }
                                    return true;
                                };
                                return hasValidValue(obj);
                            };
                            if (isValid(parsedFilter)) {
                                queryData.filters.$and.push(parsedFilter);
                                fixedFilterParsed = true;
                            }
                            else {
                                console.warn(ERROR_MESSAGES.SKIPPING_EMPTY_FIXED_FILTER, parsedFilter);
                            }
                        }
                        catch (e) {
                            console.error(ERROR_MESSAGES.INVALID_JSON_WHERECLAUSE, renderedFilter);
                            parsedFilter = {}; // fallback or throw error as needed
                        }
                    }
                    autocompleteQs = qs.stringify(queryData, {
                        encodeValuesOnly: true,
                    });
                    if (!(fixedFilterToBeApplied && !fixedFilterParsed)) return [3 /*break*/, 1];
                    showToast(toast, "error", ERROR_MESSAGES.SELECT_RELEVANT_FIELD, ERROR_MESSAGES.FIELD_NOT_SELECT);
                    return [3 /*break*/, 4];
                case 1:
                    _m.trys.push([1, , 3, 4]);
                    setLoading(true);
                    return [4 /*yield*/, triggerGetSolidEntities(autocompleteQs)];
                case 2:
                    autocompleteResponse = _m.sent();
                    autocompleteData = autocompleteResponse.data;
                    if (autocompleteData) {
                        autoCompleteItems_1 = autocompleteData.records.map(function (item) {
                            var _a, _b;
                            return __assign({ solidManyToOneLabel: item[(_b = (_a = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _a === void 0 ? void 0 : _a.userKeyField) === null || _b === void 0 ? void 0 : _b.name], solidManyToOneValue: item['id'] }, item);
                        });
                        setAutoCompleteItems(autoCompleteItems_1);
                        setOffset(autocompleteData.records.length);
                        setHasMore(autocompleteData.records.length === LIMIT);
                    }
                    return [3 /*break*/, 4];
                case 3:
                    setLoading(false);
                    return [7 /*endfinally*/];
                case 4: return [2 /*return*/];
            }
        });
    }); };
    var onLazyLoad = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, fixedFilterToBeApplied, fixedFilterParsed, convertedFixedFilter, fixedFilterTemplate, renderedFilter, parsedFilter, isValid, autocompleteQs, response, records_1;
        var _a, _b;
        var _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
        return __generator(this, function (_o) {
            switch (_o.label) {
                case 0:
                    if (!hasMore || loading)
                        return [2 /*return*/];
                    if (event.last < autoCompleteItems.length - 1)
                        return [2 /*return*/]; // Only load if user reaches bottom
                    queryData = {
                        offset: offset,
                        limit: LIMIT,
                        filters: {
                            $and: [
                                (_a = {},
                                    _a[(_d = (_c = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _c === void 0 ? void 0 : _c.userKeyField) === null || _d === void 0 ? void 0 : _d.name] = (_b = {},
                                        _b[((_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.autocompleteMatchMode) || '$containsi'] = currentQuery,
                                        _b),
                                    _a)
                            ]
                        }
                    };
                    fixedFilterToBeApplied = false;
                    fixedFilterParsed = false;
                    if (((_h = (_g = (_f = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _f === void 0 ? void 0 : _f.solidView) === null || _g === void 0 ? void 0 : _g.model) === null || _h === void 0 ? void 0 : _h.singularName) === "listOfValues" && fieldContext.fieldMetadata.relationCoModelSingularName !== "moduleMetadata") {
                        fixedFilterToBeApplied = true;
                    }
                    if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter) || ((_j = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _j === void 0 ? void 0 : _j.whereClause)) {
                        convertedFixedFilter = ((_k = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _k === void 0 ? void 0 : _k.whereClause) ? (_l = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _l === void 0 ? void 0 : _l.whereClause : fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter;
                        fixedFilterToBeApplied = true;
                        fixedFilterTemplate = Handlebars.compile(convertedFixedFilter);
                        renderedFilter = fixedFilterTemplate(formik.values);
                        parsedFilter = void 0;
                        try {
                            parsedFilter = JSON.parse(renderedFilter);
                            isValid = function (obj) {
                                if (!obj || typeof obj !== 'object')
                                    return false;
                                // Recursively check all nested values
                                var hasValidValue = function (val) {
                                    if (val === null || val === undefined || val === '')
                                        return false;
                                    if (typeof val === 'object') {
                                        return Object.values(val).some(hasValidValue);
                                    }
                                    return true;
                                };
                                return hasValidValue(obj);
                            };
                            if (isValid(parsedFilter)) {
                                queryData.filters.$and.push(parsedFilter);
                                fixedFilterParsed = true;
                            }
                            else {
                                console.warn(ERROR_MESSAGES.SKIPPING_EMPTY_FIXED_FILTER, parsedFilter);
                            }
                        }
                        catch (e) {
                            console.error(ERROR_MESSAGES.INVALID_JSON_WHERECLAUSE, renderedFilter);
                            parsedFilter = {}; // fallback or throw error as needed
                        }
                    }
                    autocompleteQs = qs.stringify(queryData, { encodeValuesOnly: true });
                    if (!(fixedFilterToBeApplied && !fixedFilterParsed)) return [3 /*break*/, 1];
                    showToast(toast, "error", ERROR_MESSAGES.SELECT_RELEVANT_FIELD, ERROR_MESSAGES.FIELD_NOT_SELECT);
                    return [3 /*break*/, 4];
                case 1:
                    _o.trys.push([1, , 3, 4]);
                    setLoading(true);
                    return [4 /*yield*/, triggerGetSolidEntities(autocompleteQs)];
                case 2:
                    response = _o.sent();
                    records_1 = ((_m = response.data) === null || _m === void 0 ? void 0 : _m.records) || [];
                    if (records_1.length < LIMIT)
                        setHasMore(false);
                    setAutoCompleteItems(function (prev) { return __spreadArray(__spreadArray([], prev, true), records_1.map(function (item) {
                        var _a, _b;
                        return (__assign({ solidManyToOneLabel: item[(_b = (_a = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _a === void 0 ? void 0 : _a.userKeyField) === null || _b === void 0 ? void 0 : _b.name], solidManyToOneValue: item.id }, item));
                    }), true); });
                    setOffset(function (prev) { return prev + records_1.length; });
                    return [3 /*break*/, 4];
                case 3:
                    setLoading(false);
                    return [7 /*endfinally*/];
                case 4: return [2 /*return*/];
            }
        });
    }); };
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var customCreateHandler = function (values) {
        var _a, _b;
        var currentRelationData = formik.values[fieldLayoutInfo.attrs.name] || [];
        var jsonValues = Object.fromEntries(values.entries());
        var updatedRelationData = __spreadArray(__spreadArray([], currentRelationData, true), [
            {
                solidManyToOneLabel: jsonValues[(_b = (_a = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _a === void 0 ? void 0 : _a.userKeyField) === null || _b === void 0 ? void 0 : _b.name],
                solidManyToOneValue: "new",
                original: jsonValues,
            },
        ], false);
        formik.setFieldValue(fieldLayoutInfo.attrs.name, updatedRelationData);
    };
    return (_jsxs("div", { className: "relative", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsxs("div", { className: "flex align-items-center gap-3", children: [_jsx(AutoComplete, __assign({ readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled || readOnlyPermission }, formik.getFieldProps(fieldLayoutInfo.attrs.name), { id: fieldLayoutInfo.attrs.name, field: "solidManyToOneLabel", value: formik.values[fieldLayoutInfo.attrs.name] || '', dropdown: !readOnlyPermission, suggestions: autoCompleteItems, completeMethod: autoCompleteSearch, onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, onFocus: function (e) { return e.target.select(); }, className: "w-full solid-standard-autocomplete", 
                                // virtualScrollerOptions={{
                                //     itemSize: 38,
                                //     lazy: true,
                                //     onLazyLoad
                                // }}
                                virtualScrollerOptions: getVirtualScrollerOptions({
                                    itemsLength: autoCompleteItems.length,
                                    lazy: true,
                                    onLazyLoad: onLazyLoad,
                                }) })), fieldLayoutInfo.attrs.inlineCreate === "true" && readOnlyPermission === false && formViewParams &&
                                _jsx(RenderSolidFormEmbededView, { formik: formik, fieldContext: fieldContext, customCreateHandler: customCreateHandler, visibleCreateRelationEntity: visibleCreateRelationEntity, setvisibleCreateRelationEntity: setvisibleCreateRelationEntity, formViewParams: formViewParams })] })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_o = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _o === void 0 ? void 0 : _o.toString() }) }))] }));
};
export var RenderSolidFormEmbededView = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r;
    var formik = _a.formik, fieldContext = _a.fieldContext, customCreateHandler = _a.customCreateHandler, visibleCreateRelationEntity = _a.visibleCreateRelationEntity, setvisibleCreateRelationEntity = _a.setvisibleCreateRelationEntity, formViewParams = _a.formViewParams;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-6 flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var parentModelName = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.parentModelName;
    var childModelName = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.childModelName;
    var parentFieldName = (_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.parentFieldName;
    var childFieldName = (_g = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.childFieldName;
    var parentModuleName = (_h = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.parentModuleName;
    var params = {
        moduleName: formViewParams.moduleName,
        modelName: camelCase(formViewParams.modelName),
        id: formViewParams.id,
        embeded: formViewParams.embeded,
        inlineCreateAutoSave: formViewParams.inlineCreateAutoSave,
        layout: formViewParams.layout,
        customCreateHandler: (function (values) {
            setvisibleCreateRelationEntity(false);
            customCreateHandler(values);
        }),
        handlePopupClose: (function () {
            setvisibleCreateRelationEntity(false);
        }),
    };
    return (_jsxs("div", { children: [_jsx("div", { children: _jsx(Button, { icon: "pi pi-plus", rounded: true, outlined: true, "aria-label": "Filter", type: "button", size: "small", onClick: function () { return setvisibleCreateRelationEntity(true); }, className: "custom-add-button" }) }), _jsx(Dialog, { header: "", showHeader: false, visible: visibleCreateRelationEntity, style: {
                    width: (_m = (_l = (_k = (_j = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _j === void 0 ? void 0 : _j.inlineCreateLayout) === null || _k === void 0 ? void 0 : _k.attrs) === null || _l === void 0 ? void 0 : _l.width) !== null && _m !== void 0 ? _m : "60vw",
                    height: (_r = (_q = (_p = (_o = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _o === void 0 ? void 0 : _o.inlineCreateLayout) === null || _p === void 0 ? void 0 : _p.attrs) === null || _q === void 0 ? void 0 : _q.height) !== null && _r !== void 0 ? _r : "auto"
                }, onHide: function () {
                    if (!visibleCreateRelationEntity)
                        return;
                    setvisibleCreateRelationEntity(false);
                }, className: "solid-dialog", breakpoints: { '1199px': '35rem', "767px": '85vw', "550px": '90vw' }, children: _jsx(SolidFormView, __assign({}, params)) })] }));
};
export var DefaultRelationManyToOneFormViewWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    var value = formik.values[fieldLayoutInfo.attrs.name];
    var userKeyField = ((_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.coModelFieldToDisplay) ? (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.coModelFieldToDisplay : (_g = (_f = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _f === void 0 ? void 0 : _f.userKeyField) === null || _g === void 0 ? void 0 : _g.name;
    var displayValue = value === null || value === void 0 ? void 0 : value[userKeyField];
    return (_jsxs("div", { className: "mt-2 flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: displayValue })] }));
};
export var PseudoRelationManyToOneFormWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var toast = useRef(null);
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var readOnlyPermission = fieldContext.readOnly ? fieldContext.readOnly : false;
    var parentModelName = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.parentModelName;
    var childModelName = (_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.childModelName;
    var parentFieldName = (_g = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.parentFieldName;
    var childFieldName = (_h = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.childFieldName;
    var parentModuleName = (_j = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _j === void 0 ? void 0 : _j.parentModuleName;
    var parentFieldLabels = (_k = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _k === void 0 ? void 0 : _k.parentFieldLabels;
    var parentSearchFields = (_l = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _l === void 0 ? void 0 : _l.parentSearchFields;
    var _w = useState(), formViewParams = _w[0], setformViewParams = _w[1];
    var viewMode = fieldContext.viewMode;
    useEffect(function () {
        var _a, _b;
        var formviewparams = {
            moduleName: parentModuleName,
            modelName: camelCase(parentModelName),
            id: "new",
            embeded: true,
            inlineCreateAutoSave: (_a = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.inlineCreateAutoSave,
            layout: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineCreateLayout
        };
        setformViewParams(formviewparams);
    }, []);
    var _x = useState(false), visibleCreateRelationEntity = _x[0], setvisibleCreateRelationEntity = _x[1];
    // auto complete specific code. 
    var entityApi = createSolidEntityApi(parentModelName || "userViewMetadata");
    var useLazyGetSolidEntitiesQuery = entityApi.useLazyGetSolidEntitiesQuery;
    var triggerGetSolidEntities = useLazyGetSolidEntitiesQuery()[0];
    var fieldDisabled = (_m = fieldLayoutInfo.attrs) === null || _m === void 0 ? void 0 : _m.disabled;
    var fieldReadonly = (_o = fieldLayoutInfo.attrs) === null || _o === void 0 ? void 0 : _o.readonly;
    var formDisabled = (_r = (_q = (_p = solidFormViewMetaData.data.solidView) === null || _p === void 0 ? void 0 : _p.layout) === null || _q === void 0 ? void 0 : _q.attrs) === null || _r === void 0 ? void 0 : _r.disabled;
    var formReadonly = (_u = (_t = (_s = solidFormViewMetaData.data.solidView) === null || _s === void 0 ? void 0 : _s.layout) === null || _t === void 0 ? void 0 : _t.attrs) === null || _u === void 0 ? void 0 : _u.readonly;
    var whereClause = fieldLayoutInfo.attrs.whereClause;
    var _y = useState([]), autoCompleteItems = _y[0], setAutoCompleteItems = _y[1];
    var _z = useState(0), offset = _z[0], setOffset = _z[1];
    var _0 = useState(true), hasMore = _0[0], setHasMore = _0[1];
    var _1 = useState(""), currentQuery = _1[0], setCurrentQuery = _1[1];
    var LIMIT = 50;
    var _2 = useState(false), loading = _2[0], setLoading = _2[1];
    var _3 = useState(null), viewDisplayValue = _3[0], setViewDisplayValue = _3[1];
    useEffect(function () {
        var fn = function () { return __awaiter(void 0, void 0, void 0, function () {
            var queryData, autocompleteQs, autocompleteResponse, autocompleteData, formattedRecords;
            var _a, _b;
            var _c, _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        if (!fieldContext.data[fieldLayoutInfo.attrs.name]) return [3 /*break*/, 2];
                        queryData = {
                            offset: 0,
                            limit: LIMIT,
                            filters: {
                                $and: [
                                    (_a = {},
                                        _a[parentFieldName] = (_b = {},
                                            _b[((_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.autocompleteMatchMode) || '$eqi'] = (_d = fieldContext.data[fieldLayoutInfo.attrs.name]) !== null && _d !== void 0 ? _d : formik.values[fieldLayoutInfo.attrs.name],
                                            _b),
                                        _a)
                                ]
                            }
                        };
                        autocompleteQs = qs.stringify(queryData, {
                            encodeValuesOnly: true,
                        });
                        return [4 /*yield*/, triggerGetSolidEntities(autocompleteQs)];
                    case 1:
                        autocompleteResponse = _e.sent();
                        autocompleteData = autocompleteResponse.data;
                        if (autocompleteData) {
                            formattedRecords = autocompleteData.records.map(function (item) {
                                var label = parentFieldLabels && parentFieldLabels.length > 0 ? parentFieldLabels.map(function (label) { return item[label]; }).join(" - ") : item[parentFieldName];
                                return __assign({ solidManyToOneLabel: label, solidManyToOneValue: item[parentFieldName] }, item);
                            });
                            if (formattedRecords.length > 0) {
                                formik.setFieldValue(fieldLayoutInfo.attrs.name, formattedRecords[0]);
                            }
                        }
                        _e.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        }); };
        fn();
    }, []);
    // Add this NEW useEffect - only for view mode refresh
    useEffect(function () {
        var fn = function () { return __awaiter(void 0, void 0, void 0, function () {
            var rawValue, searchValue, queryData, qsStr, res, record, label;
            var _a, _b;
            var _c, _d, _e, _f;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        if (viewMode !== "view") {
                            setViewDisplayValue(null);
                            return [2 /*return*/];
                        }
                        rawValue = (_c = fieldContext.data[fieldLayoutInfo.attrs.name]) !== null && _c !== void 0 ? _c : formik.values[fieldLayoutInfo.attrs.name];
                        if (!rawValue)
                            return [2 /*return*/];
                        searchValue = typeof rawValue === "object"
                            ? rawValue[parentFieldName]
                            : rawValue;
                        queryData = {
                            offset: 0,
                            limit: 1,
                            filters: {
                                $and: [
                                    (_a = {},
                                        _a[parentFieldName] = (_b = {},
                                            _b[((_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.autocompleteMatchMode) || "$eqi"] = searchValue,
                                            _b),
                                        _a),
                                ],
                            },
                        };
                        qsStr = qs.stringify(queryData, { encodeValuesOnly: true });
                        return [4 /*yield*/, triggerGetSolidEntities(qsStr)];
                    case 1:
                        res = _g.sent();
                        record = (_f = (_e = res.data) === null || _e === void 0 ? void 0 : _e.records) === null || _f === void 0 ? void 0 : _f[0];
                        if (!record)
                            return [2 /*return*/];
                        label = (parentFieldLabels === null || parentFieldLabels === void 0 ? void 0 : parentFieldLabels.length) > 0
                            ? parentFieldLabels.map(function (l) { return record[l]; }).join(" - ")
                            : record[parentFieldName];
                        console.log("Final View Label", label, record[parentFieldName]);
                        setViewDisplayValue(__assign({ solidManyToOneLabel: label, solidManyToOneValue: record[parentFieldName] }, record));
                        return [2 /*return*/];
                }
            });
        }); };
        fn();
    }, [viewMode, formik.values[fieldLayoutInfo.attrs.name]]);
    var resolvedValue = viewMode === "view" ? viewDisplayValue : formik.values[fieldLayoutInfo.attrs.name];
    var autoCompleteSearch = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, fixedFilterToBeApplied, fixedFilterParsed, convertedFixedFilter, fixedFilterTemplate, renderedFilter, parsedFilter, isValid, autocompleteQs, autocompleteResponse, autocompleteData, autoCompleteItems_2;
        var _a, _b;
        var _c, _d, _e, _f, _g, _h, _j;
        return __generator(this, function (_k) {
            switch (_k.label) {
                case 0:
                    setOffset(0);
                    setCurrentQuery(event.query || "");
                    setHasMore(true);
                    queryData = {
                        offset: 0,
                        limit: LIMIT,
                        filters: {
                            $and: __spreadArray([], ((parentSearchFields === null || parentSearchFields === void 0 ? void 0 : parentSearchFields.length) > 0
                                ? [
                                    {
                                        $or: parentSearchFields.map(function (field) {
                                            var _a, _b;
                                            var _c;
                                            return (_a = {},
                                                _a[field] = (_b = {},
                                                    _b[((_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.autocompleteMatchMode) || '$containsi'] = event.query,
                                                    _b),
                                                _a);
                                        })
                                    }
                                ]
                                : [
                                    (_a = {},
                                        _a[parentFieldName] = (_b = {},
                                            _b[((_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.autocompleteMatchMode) || '$containsi'] = event.query,
                                            _b),
                                        _a)
                                ]), true)
                        }
                    };
                    fixedFilterToBeApplied = false;
                    fixedFilterParsed = false;
                    if (((_f = (_e = (_d = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _d === void 0 ? void 0 : _d.solidView) === null || _e === void 0 ? void 0 : _e.model) === null || _f === void 0 ? void 0 : _f.singularName) === "listOfValues" && fieldContext.fieldMetadata.relationCoModelSingularName !== "moduleMetadata") {
                        fixedFilterToBeApplied = true;
                    }
                    if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter) || ((_g = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.whereClause)) {
                        convertedFixedFilter = ((_h = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.whereClause) ? (_j = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _j === void 0 ? void 0 : _j.whereClause : fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter;
                        fixedFilterToBeApplied = true;
                        fixedFilterTemplate = Handlebars.compile(convertedFixedFilter);
                        renderedFilter = fixedFilterTemplate(formik.values);
                        parsedFilter = void 0;
                        try {
                            parsedFilter = JSON.parse(renderedFilter);
                            isValid = function (obj) {
                                if (!obj || typeof obj !== 'object')
                                    return false;
                                // Recursively check all nested values
                                var hasValidValue = function (val) {
                                    if (val === null || val === undefined || val === '')
                                        return false;
                                    if (typeof val === 'object') {
                                        return Object.values(val).some(hasValidValue);
                                    }
                                    return true;
                                };
                                return hasValidValue(obj);
                            };
                            if (isValid(parsedFilter)) {
                                queryData.filters.$and.push(parsedFilter);
                                fixedFilterParsed = true;
                            }
                            else {
                                console.warn(ERROR_MESSAGES.SKIPPING_EMPTY_FIXED_FILTER, parsedFilter);
                            }
                        }
                        catch (e) {
                            console.error(ERROR_MESSAGES.INVALID_JSON_WHERECLAUSE, renderedFilter);
                            parsedFilter = {}; // fallback or throw error as needed
                        }
                    }
                    autocompleteQs = qs.stringify(queryData, {
                        encodeValuesOnly: true,
                    });
                    if (!(fixedFilterToBeApplied && !fixedFilterParsed)) return [3 /*break*/, 1];
                    showToast(toast, "error", ERROR_MESSAGES.SELECT_RELEVANT_FIELD, ERROR_MESSAGES.FIELD_NOT_SELECT);
                    return [3 /*break*/, 4];
                case 1:
                    _k.trys.push([1, , 3, 4]);
                    setLoading(true);
                    return [4 /*yield*/, triggerGetSolidEntities(autocompleteQs)];
                case 2:
                    autocompleteResponse = _k.sent();
                    autocompleteData = autocompleteResponse.data;
                    if (autocompleteData) {
                        autoCompleteItems_2 = autocompleteData.records.map(function (item) {
                            var label = parentFieldLabels && parentFieldLabels.length > 0 ? parentFieldLabels.map(function (label) { return item[label]; }).join(" - ") : item[parentFieldName];
                            return __assign({ solidManyToOneLabel: label, solidManyToOneValue: item[parentFieldName] }, item);
                        });
                        // const autoCompleteItems = autocompleteData.records.map((item: any) => {label: item[parentFieldName],value});
                        setAutoCompleteItems(autoCompleteItems_2);
                        setOffset(autocompleteData.records.length);
                        setHasMore(autocompleteData.records.length === LIMIT);
                    }
                    return [3 /*break*/, 4];
                case 3:
                    setLoading(false);
                    return [7 /*endfinally*/];
                case 4: return [2 /*return*/];
            }
        });
    }); };
    var onLazyLoad = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, fixedFilterToBeApplied, fixedFilterParsed, convertedFixedFilter, fixedFilterTemplate, renderedFilter, parsedFilter, isValid, autocompleteQs, response, records_2;
        var _a, _b;
        var _c, _d, _e, _f, _g, _h, _j, _k;
        return __generator(this, function (_l) {
            switch (_l.label) {
                case 0:
                    if (!hasMore || loading)
                        return [2 /*return*/];
                    if (event.last < autoCompleteItems.length - 1)
                        return [2 /*return*/]; // Only load if user reaches bottom
                    queryData = {
                        offset: offset,
                        limit: LIMIT,
                        filters: {
                            $and: __spreadArray([], ((parentSearchFields === null || parentSearchFields === void 0 ? void 0 : parentSearchFields.length) > 0
                                ? [
                                    {
                                        $or: parentSearchFields.map(function (field) {
                                            var _a, _b;
                                            var _c;
                                            return (_a = {},
                                                _a[field] = (_b = {},
                                                    _b[((_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.autocompleteMatchMode) || '$containsi'] = currentQuery,
                                                    _b),
                                                _a);
                                        })
                                    }
                                ]
                                : [
                                    (_a = {},
                                        _a[parentFieldName] = (_b = {},
                                            _b[((_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.autocompleteMatchMode) || '$containsi'] = currentQuery,
                                            _b),
                                        _a)
                                ]), true)
                        }
                    };
                    fixedFilterToBeApplied = false;
                    fixedFilterParsed = false;
                    if (((_f = (_e = (_d = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _d === void 0 ? void 0 : _d.solidView) === null || _e === void 0 ? void 0 : _e.model) === null || _f === void 0 ? void 0 : _f.singularName) === "listOfValues" && fieldContext.fieldMetadata.relationCoModelSingularName !== "moduleMetadata") {
                        fixedFilterToBeApplied = true;
                    }
                    if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter) || ((_g = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.whereClause)) {
                        convertedFixedFilter = ((_h = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.whereClause) ? (_j = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _j === void 0 ? void 0 : _j.whereClause : fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationFieldFixedFilter;
                        fixedFilterToBeApplied = true;
                        fixedFilterTemplate = Handlebars.compile(convertedFixedFilter);
                        renderedFilter = fixedFilterTemplate(formik.values);
                        parsedFilter = void 0;
                        try {
                            parsedFilter = JSON.parse(renderedFilter);
                            isValid = function (obj) {
                                if (!obj || typeof obj !== 'object')
                                    return false;
                                // Recursively check all nested values
                                var hasValidValue = function (val) {
                                    if (val === null || val === undefined || val === '')
                                        return false;
                                    if (typeof val === 'object') {
                                        return Object.values(val).some(hasValidValue);
                                    }
                                    return true;
                                };
                                return hasValidValue(obj);
                            };
                            if (isValid(parsedFilter)) {
                                queryData.filters.$and.push(parsedFilter);
                                fixedFilterParsed = true;
                            }
                            else {
                                console.warn(ERROR_MESSAGES.SKIPPING_EMPTY_FIXED_FILTER, parsedFilter);
                            }
                        }
                        catch (e) {
                            console.error(ERROR_MESSAGES.INVALID_JSON_WHERECLAUSE, renderedFilter);
                            parsedFilter = {}; // fallback or throw error as needed
                        }
                    }
                    autocompleteQs = qs.stringify(queryData, { encodeValuesOnly: true });
                    if (!(fixedFilterToBeApplied && !fixedFilterParsed)) return [3 /*break*/, 1];
                    showToast(toast, "error", ERROR_MESSAGES.SELECT_RELEVANT_FIELD, ERROR_MESSAGES.FIELD_NOT_SELECT);
                    return [3 /*break*/, 4];
                case 1:
                    _l.trys.push([1, , 3, 4]);
                    setLoading(true);
                    return [4 /*yield*/, triggerGetSolidEntities(autocompleteQs)];
                case 2:
                    response = _l.sent();
                    records_2 = ((_k = response.data) === null || _k === void 0 ? void 0 : _k.records) || [];
                    if (records_2.length < LIMIT)
                        setHasMore(false);
                    setAutoCompleteItems(function (prev) { return __spreadArray(__spreadArray([], prev, true), records_2.map(function (item) {
                        var label = parentFieldLabels && parentFieldLabels.length > 0 ? parentFieldLabels.map(function (label) { return item[label]; }).join(" - ") : item[parentFieldName];
                        return __assign({ solidManyToOneLabel: label, solidManyToOneValue: item[parentFieldName] }, item);
                    }), true); });
                    setOffset(function (prev) { return prev + records_2.length; });
                    return [3 /*break*/, 4];
                case 3:
                    setLoading(false);
                    return [7 /*endfinally*/];
                case 4: return [2 /*return*/];
            }
        });
    }); };
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var customCreateHandler = function (values) {
        var currentRelationData = formik.values[fieldLayoutInfo.attrs.name] || [];
        var jsonValues = Object.fromEntries(values.entries());
        console.log("jsonValues", jsonValues);
        var label = parentFieldLabels && parentFieldLabels.length > 0 ? parentFieldLabels.map(function (label) { return jsonValues[label]; }).join(" - ") : jsonValues[parentFieldName];
        var updatedRelationData = __spreadArray(__spreadArray([], currentRelationData, true), [
            // jsonValues[parentFieldName]
            {
                solidManyToOneLabel: label,
                solidManyToOneValue: "new",
                original: jsonValues,
            },
        ], false);
        formik.setFieldValue(fieldLayoutInfo.attrs.name, updatedRelationData);
    };
    return (_jsxs("div", { className: "relative", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " * " }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsxs("div", { className: "flex align-items-center gap-3", children: [_jsx(AutoComplete, __assign({ readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled || readOnlyPermission || viewMode === "view", field: "solidManyToOneLabel" }, formik.getFieldProps(fieldLayoutInfo.attrs.name), { id: fieldLayoutInfo.attrs.name, value: resolvedValue || null, dropdown: !readOnlyPermission, suggestions: autoCompleteItems, completeMethod: autoCompleteSearch, onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, onFocus: function (e) { return e.target.select(); }, className: "w-full solid-standard-autocomplete", 
                                // virtualScrollerOptions={{
                                //     itemSize: 38,
                                //     lazy: true,
                                //     onLazyLoad
                                // }}
                                virtualScrollerOptions: getVirtualScrollerOptions({
                                    itemsLength: autoCompleteItems.length,
                                    lazy: true,
                                    onLazyLoad: onLazyLoad,
                                }) })), fieldLayoutInfo.attrs.inlineCreate === "true" && readOnlyPermission === false && formViewParams && viewMode !== "view" &&
                                _jsx(RenderSolidFormEmbededView, { formik: formik, fieldContext: fieldContext, customCreateHandler: customCreateHandler, visibleCreateRelationEntity: visibleCreateRelationEntity, setvisibleCreateRelationEntity: setvisibleCreateRelationEntity, formViewParams: formViewParams, children: " " })] })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_v = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _v === void 0 ? void 0 : _v.toString() }) }))] }));
};
