// @solidxai/core-ui/dist/components/core/common/AvatarWidget.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
export var AvatarWidget = function (_a) {
    // use "widget": "SolidUserNameAvatarWidget" in the list view field 
    var value = _a.value;
    var getInitials = function (value) {
        if (value) {
            var names = value === null || value === void 0 ? void 0 : value.trim().split(' ');
            var initials_1 = names.length === 1
                ? names[0][0]
                : names[0][0] + names[names.length - 1][0];
            return initials_1.toUpperCase();
        }
        else {
            return "";
        }
    };
    var getColorFromInitials = function (initials) {
        var hash = 0;
        for (var i = 0; i < initials.length; i++) {
            hash = initials.charCodeAt(i) + ((hash << 5) - hash);
        }
        var hue = Math.abs(hash) % 360;
        return "hsl(".concat(hue, ", 60%, 60%)"); // nice pastel color
    };
    var initials = getInitials(value);
    var bgColor = getColorFromInitials(initials);
    return (_jsx("div", { className: "solid-table-row", children: value &&
            _jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '0.5rem' }, children: [_jsx("div", { style: {
                            width: 20,
                            height: 20,
                            borderRadius: '50%',
                            backgroundColor: bgColor,
                            color: 'white',
                            display: 'flex',
                            alignItems: 'center',
                            justifyContent: 'center',
                            fontWeight: 600,
                            fontSize: 12
                        }, children: initials }), _jsx("span", { children: value })] }) }));
};
