// @solidxai/core-ui/dist/components/core/common/FilterComponent.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { InputText } from 'primereact/inputtext';
import { useRef, useState } from 'react';
import { SolidFilterFields } from '../filter/SolidFilterFields';
import { Button } from 'primereact/button';
import { Fieldset } from 'primereact/fieldset';
import { OverlayPanel } from 'primereact/overlaypanel';
import { AutoComplete } from 'primereact/autocomplete';
export var FilterRuleType;
(function (FilterRuleType) {
    FilterRuleType["RULE"] = "rule";
    FilterRuleType["RULE_GROUP"] = "rule_group";
})(FilterRuleType || (FilterRuleType = {}));
export var FilterOperator;
(function (FilterOperator) {
    FilterOperator["AND"] = "and";
    FilterOperator["OR"] = "or";
})(FilterOperator || (FilterOperator = {}));
export var FilterMatchMode;
(function (FilterMatchMode) {
    FilterMatchMode["STARTS_WITH"] = "startsWith";
    FilterMatchMode["CONTAINS"] = "contains";
    FilterMatchMode["EQUALS"] = "equals";
    FilterMatchMode["GREATER_THAN"] = "gt";
    FilterMatchMode["LESS_THAN"] = "lt";
})(FilterMatchMode || (FilterMatchMode = {}));
// const fields: Field[] = [
//   { name: 'name', type: 'string' },
//   { name: 'dob', type: 'date' },
//   { name: 'lastName', type: 'string' },
//   { name: 'amount', type: 'number' }
// ];
var operatorOptions = {
    // string: [FilterMatchMode.CONTAINS, FilterMatchMode.EQUALS],
    // date: [FilterMatchMode.EQUALS, FilterMatchMode.GREATER_THAN, FilterMatchMode.LESS_THAN],
    // number: [FilterMatchMode.GREATER_THAN, FilterMatchMode.LESS_THAN],
    id: [FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.LESS_THAN, FilterMatchMode.LESS_THAN_OR_EQUAL_TO, FilterMatchMode.GREATER_THAN, FilterMatchMode.GREATER_THAN_OR_EQUAL_TO, FilterMatchMode.IN, FilterMatchMode.NOT_IN, FilterMatchMode.BETWEEN],
    int: [FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.LESS_THAN, FilterMatchMode.LESS_THAN_OR_EQUAL_TO, FilterMatchMode.GREATER_THAN, FilterMatchMode.GREATER_THAN_OR_EQUAL_TO, FilterMatchMode.IN, FilterMatchMode.NOT_IN, FilterMatchMode.BETWEEN],
    bigint: [FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.LESS_THAN, FilterMatchMode.LESS_THAN_OR_EQUAL_TO, FilterMatchMode.GREATER_THAN, FilterMatchMode.GREATER_THAN_OR_EQUAL_TO, FilterMatchMode.IN, FilterMatchMode.NOT_IN, FilterMatchMode.BETWEEN],
    float: [FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.LESS_THAN, FilterMatchMode.LESS_THAN_OR_EQUAL_TO, FilterMatchMode.GREATER_THAN, FilterMatchMode.GREATER_THAN_OR_EQUAL_TO, FilterMatchMode.IN, FilterMatchMode.NOT_IN, FilterMatchMode.BETWEEN],
    decimal: [FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.LESS_THAN, FilterMatchMode.LESS_THAN_OR_EQUAL_TO, FilterMatchMode.GREATER_THAN, FilterMatchMode.GREATER_THAN_OR_EQUAL_TO, FilterMatchMode.IN, FilterMatchMode.NOT_IN, FilterMatchMode.BETWEEN],
    shortText: [FilterMatchMode.STARTS_WITH, FilterMatchMode.CONTAINS, FilterMatchMode.NOT_CONTAINS, FilterMatchMode.ENDS_WITH, FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.IN, FilterMatchMode.NOT_IN],
    longText: [FilterMatchMode.STARTS_WITH, FilterMatchMode.CONTAINS, FilterMatchMode.NOT_CONTAINS, FilterMatchMode.ENDS_WITH, FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.IN, FilterMatchMode.NOT_IN],
    richText: [FilterMatchMode.STARTS_WITH, FilterMatchMode.CONTAINS, FilterMatchMode.NOT_CONTAINS, FilterMatchMode.ENDS_WITH, FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.IN, FilterMatchMode.NOT_IN],
    boolean: [FilterMatchMode.EQUALS],
    date: [FilterMatchMode.EQUALS, FilterMatchMode.NOT_EQUALS, FilterMatchMode.LESS_THAN, FilterMatchMode.LESS_THAN_OR_EQUAL_TO, FilterMatchMode.GREATER_THAN, FilterMatchMode.GREATER_THAN_OR_EQUAL_TO, FilterMatchMode.IN, FilterMatchMode.NOT_IN, FilterMatchMode.BETWEEN,],
    datetime: [FilterMatchMode.EQUALS],
    time: [FilterMatchMode.EQUALS],
    relation: [FilterMatchMode.IN, FilterMatchMode.NOT_IN],
    mediaSingle: [],
    mediaMultiple: [],
    selectionStatic: [FilterMatchMode.IN, FilterMatchMode.NOT_IN],
    selectionDynamic: [FilterMatchMode.IN, FilterMatchMode.NOT_IN],
    computed: [],
    externalId: [FilterMatchMode.IN, FilterMatchMode.NOT_IN],
    uuid: [FilterMatchMode.IN, FilterMatchMode.NOT_IN],
};
var getRandomInt = function (min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
};
// Component to render a single filter rule
var FilterRuleComponent = function (_a) {
    // const applicableOperators = rule.fieldName ? operatorOptions[fields.find(f => f.name === rule.fieldName)?.value.type].map(e => { return { name: e } }) : [];
    // const applicableInputField = rule.fieldName ? fields.find(f => f.name === rule.fieldName)?.type : "";
    var viewData = _a.viewData, fields = _a.fields, rule = _a.rule, onChange = _a.onChange, onAddRule = _a.onAddRule, onAddGroup = _a.onAddGroup, onDelete = _a.onDelete, level = _a.level;
    var autoCompleteRef = useRef(null);
    var _b = useState({ name: rule.fieldName }), fieldName = _b[0], setFieldName = _b[1];
    var _c = useState({ name: rule.matchMode }), matchMode = _c[0], setMatchMode = _c[1];
    var _d = useState([]), filteredFields = _d[0], setFilteredFields = _d[1];
    var searchFields = function (event) {
        var query = event.query.toLowerCase();
        var filtered = !query
            ? fields
            : fields.filter(function (item) {
                return item.name.toLowerCase().startsWith(query);
            });
        setFilteredFields(filtered);
        setTimeout(function () {
            var _a;
            (_a = autoCompleteRef.current) === null || _a === void 0 ? void 0 : _a.show();
        }, 0);
    };
    return (
    // <div style={{ marginLeft: (level - 1) * 10 + 'px' }} className="filter-rule">
    _jsxs("div", { className: 'mt-2', children: [_jsxs("div", { className: 'flex flex-column md:flex-row  align-items-start md:align-items-center gap-2 md:gap-3', children: [_jsxs("div", { className: 'formgrid grid w-full', children: [_jsx("div", { className: 'col-12 md:col-4 pb-2 md:pb-0', children: _jsx(AutoComplete, { ref: autoCompleteRef, value: fieldName.name, suggestions: filteredFields, completeMethod: searchFields, field: "name", dropdown: true, forceSelection // only values from list
                                    : true, placeholder: "Select Field", className: "w-full p-inputtext-sm solid-filter-auto-complete-field", onChange: function (e) {
                                        setFieldName({ name: e.value }); // e.value will be an object or null
                                        if (e.value) {
                                            onChange(rule.id, 'fieldName', e.value.value); // send value to parent
                                        }
                                        else {
                                            onChange(rule.id, 'fieldName', '');
                                        }
                                    } }) }), _jsx("div", { className: 'col-12 md:col-8', children: _jsx("div", { className: 'formgrid grid w-full', children: rule.fieldName ?
                                        _jsx("div", { className: 'col-12', children: _jsx(SolidFilterFields, { viewData: viewData, fieldMetadata: viewData.data.solidFieldsMetadata[rule.fieldName], onChange: onChange, index: rule.id, rule: rule }) })
                                        : _jsxs(_Fragment, { children: [_jsx("div", { className: 'col-12 md:col-6 pb-2 md:pb-0', children: _jsx(InputText, { disabled: true, value: rule.value || '', placeholder: "operator", className: 'w-full p-inputtext-sm ' }) }), _jsx("div", { className: 'col-12 md:col-6', children: _jsx(InputText, { disabled: true, value: rule.value || '', placeholder: "value", className: 'w-full p-inputtext-sm' }) })] }) }) })] }), _jsxs("div", { className: 'formgrid grid', children: [_jsx("div", { className: 'col-4 px-0 flex align-items-center', children: _jsx(Button, { text: true, severity: 'secondary', icon: "pi pi-plus", size: 'small', onClick: function () { return onAddRule(rule.parentRule); }, className: 'solid-filter-action-btn' }) }), _jsx("div", { className: 'col-4 px-0 flex align-items-center', children: _jsx(Button, { text: true, severity: 'secondary', icon: "pi pi-folder-plus", size: 'small', onClick: function () { return onAddGroup(rule.id); }, className: 'solid-filter-action-btn' }) }), _jsx("div", { className: 'col-4 px-0 flex align-items-center', children: _jsx(Button, { text: true, severity: 'secondary', icon: "pi pi-trash", size: 'small', onClick: function () { return onDelete(rule.id); }, className: 'solid-filter-action-btn' }) })] })] }), rule.children && rule.children.map(function (nestedRule) { return (_jsx("div", { className: 'py-3 nested-custom-filter', children: nestedRule.type === FilterRuleType.RULE
                    ? _jsx(FilterRuleComponent, { viewData: viewData, fields: fields, rule: nestedRule, onChange: onChange, onAddRule: onAddRule, onAddGroup: onAddGroup, onDelete: onDelete, level: level + 1 }, nestedRule.id)
                    : _jsx(FilterGroupComponent, { viewData: viewData, fields: fields, group: nestedRule, onChange: onChange, onAddRule: onAddRule, onAddGroup: onAddGroup, onDelete: onDelete, level: level + 1 }, nestedRule.id) }, nestedRule.id)); })] }));
};
// Component to render a group of filter rules
var FilterGroupComponent = function (_a) {
    var viewData = _a.viewData, fields = _a.fields, group = _a.group, onChange = _a.onChange, onAddRule = _a.onAddRule, onAddGroup = _a.onAddGroup, onDelete = _a.onDelete, level = _a.level;
    var op = useRef(null);
    var legendTemplate = (_jsxs(_Fragment, { children: [_jsx(Button, { size: 'small', label: group.matchOperator, className: 'small-button', icon: "pi pi-angle-down", iconPos: 'right', onClick: function (e) { return op.current.toggle(e); }, style: {
                    textTransform: "uppercase"
                } }), _jsx(OverlayPanel, { ref: op, className: 'm-0', children: _jsxs("div", { className: 'flex flex-column', children: [_jsx(Button, { size: "small", label: "AND", text: true, className: 'small-button', onClick: function (e) { onChange(group.id, 'matchOperator', FilterOperator.AND); op.current.hide(e); } }), _jsx(Button, { size: "small", label: "OR", text: true, className: 'small-button', onClick: function (e) { onChange(group.id, 'matchOperator', FilterOperator.OR); op.current.hide(e); } })] }) })] })
    // <select className='filter-select'
    //   value={group.matchOperator}
    //   onChange={e => onChange(group.id, 'matchOperator', e.target.value)}
    // >
    //   <option value={FilterOperator.AND}>AND</option>
    //   <option value={FilterOperator.OR}>OR</option>
    // </select>
    );
    return (_jsxs(Fieldset, { legend: legendTemplate, className: 'primary-filter-fieldset', children: [group.children && group.children.map(function (rule) { return (rule.type === FilterRuleType.RULE
                ? _jsx(FilterRuleComponent, { viewData: viewData, fields: fields, rule: rule, onChange: onChange, onAddRule: onAddRule, onAddGroup: onAddGroup, onDelete: onDelete, level: level + 1 }, rule.id)
                : _jsx(FilterGroupComponent, { viewData: viewData, fields: fields, group: rule, onChange: onChange, onAddRule: onAddRule, onAddGroup: onAddGroup, onDelete: onDelete, level: level + 1 }, rule.id)); }), level === 0 &&
                _jsx(Button, { text: true, label: 'Add Condition', icon: "pi pi-plus", size: 'small', onClick: function () { return onAddRule(group.id); }, className: 'px-0 mt-2' })] }));
};
// Main Filter component
var FilterComponent = function (_a) {
    // const initialState: FilterRule[] = [
    //   {
    //     id: 1,
    //     type: FilterRuleType.RULE_GROUP,
    //     matchOperator: FilterOperator.OR,
    //     parentRule: null,
    //     children: [
    //       {
    //         id: Date.now() + getRandomInt(1, 500),
    //         type: FilterRuleType.RULE,
    //         fieldName: null,
    //         matchMode: null,
    //         value: null,
    //         parentRule: 1,
    //         children: []
    //       }
    //     ]
    //   }
    // ];
    var viewData = _a.viewData, fields = _a.fields, filterRules = _a.filterRules, setFilterRules = _a.setFilterRules, transformFilterRules = _a.transformFilterRules, closeDialog = _a.closeDialog;
    // const [filterRules, setFilterRules] = useState<FilterRule[]>(initialState);
    var _b = useState(''), printedState = _b[0], setPrintedState = _b[1];
    var addChild = function (rules, parentId, newChild) {
        return rules.map(function (rule) {
            if (rule.id === parentId) {
                return __assign(__assign({}, rule), { children: __spreadArray(__spreadArray([], (rule.children || []), true), [newChild], false) });
            }
            else if (rule.children) {
                return __assign(__assign({}, rule), { children: addChild(rule.children, parentId, newChild) });
            }
            return rule;
        });
    };
    var handleChange = function (id, key, value) {
        var updateRuleRecursively = function (rules) {
            return rules.map(function (rule) {
                var _a;
                if (rule.id === id) {
                    return __assign(__assign({}, rule), (_a = {}, _a[key] = value, _a));
                }
                else if (rule.children) {
                    return __assign(__assign({}, rule), { children: updateRuleRecursively(rule.children) });
                }
                return rule;
            });
        };
        setFilterRules(function (prev) { return updateRuleRecursively(prev); });
    };
    var handleAddRule = function (parentGroupId) {
        var newRule = {
            id: Date.now(),
            type: FilterRuleType.RULE,
            fieldName: null,
            matchMode: null,
            value: null,
            parentRule: parentGroupId,
            children: []
        };
        setFilterRules(function (prev) { return addChild(prev, parentGroupId, newRule); });
    };
    var handleAddGroup = function (parentRuleId) {
        var newGroupId = Date.now();
        var newGroup = {
            id: newGroupId,
            type: FilterRuleType.RULE_GROUP,
            matchOperator: FilterOperator.AND,
            parentRule: parentRuleId,
            children: [
                {
                    id: Date.now() + getRandomInt(1, 500),
                    type: FilterRuleType.RULE,
                    fieldName: null,
                    matchMode: null,
                    value: null,
                    parentRule: newGroupId,
                    children: []
                },
                {
                    id: Date.now() + getRandomInt(1, 500),
                    type: FilterRuleType.RULE,
                    fieldName: null,
                    matchMode: null,
                    value: null,
                    parentRule: newGroupId,
                    children: []
                }
            ]
        };
        setFilterRules(function (prev) { return addChild(prev, parentRuleId, newGroup); });
    };
    // const handleDeleteRule = (id) => {
    //   const deleteRecursively = (rules, id) => {
    //     return rules.filter(rule => rule.id !== id).map(rule => {
    //       if (rule.children) {
    //         return {
    //           ...rule,
    //           children: deleteRecursively(rule.children, id)
    //         };
    //       }
    //       return rule;
    //     });
    //   };
    //   setFilterRules(prev => deleteRecursively(prev, id));
    // };
    var removeEmptyGroups = function (rule) {
        if (rule.type === "rule_group" && rule.children.length === 0) {
            return null;
        }
        rule.children = rule.children
            .map(removeEmptyGroups) // Recursively process children
            .filter(function (child) { return child !== null; }); // Remove null values
        return rule;
    };
    var handleDeleteRule = function (id) {
        if (filterRules[0].children.length === 1 && filterRules[0].children[0].id === id) {
            return;
        }
        var deleteRecursively = function (rules, id, isRoot) {
            if (isRoot === void 0) { isRoot = false; }
            var mappedData = rules
                .map(function (rule) {
                if (rule.id === id) {
                    return null; // Remove the target rule
                }
                if (rule.children) {
                    var updatedChildren = deleteRecursively(rule.children, id);
                    // If this rule had children and now has none, keep it but set children to []
                    if (rule.children.length > 0 && updatedChildren.length === 0) {
                        return __assign(__assign({}, rule), { children: [] });
                    }
                    return __assign(__assign({}, rule), { children: updatedChildren });
                }
                return rule;
            })
                .filter(function (rule) { return rule !== null; }); // Remove null values
            return mappedData;
        };
        setFilterRules(function (prevRules) { return [removeEmptyGroups(deleteRecursively(prevRules, id, true)[0])]; });
    };
    // const handlePrintState = () => {
    //   // setPrintedState(JSON.stringify(filterRules, null, 2)); // Pretty format the state
    //   transformFilterRules(filterRules)
    // };
    return (_jsxs("div", { className: '', children: [filterRules.map(function (rule) { return (_jsx(FilterGroupComponent, { viewData: viewData, fields: fields, group: rule, onChange: handleChange, onAddRule: handleAddRule, onAddGroup: handleAddGroup, onDelete: handleDeleteRule, level: 0 }, rule.id)); }), _jsxs("div", { className: 'flex gap-3 mt-3', children: [_jsx(Button, { label: "Apply", size: "small", onClick: function () { return transformFilterRules(filterRules); }, type: "submit" }), _jsx(Button, { type: 'button', label: 'Cancel', outlined: true, size: 'small', onClick: closeDialog })] })] }));
};
export default FilterComponent;
