// @solidxai/core-ui/dist/components/core/list/columns/relations/SolidRelationManyToOneColumn.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { FilterMatchMode } from 'primereact/api';
import { Column } from "primereact/column";
import { InputTypes, SolidVarInputsFilterElement } from '../../../../../components/core/list/SolidVarInputsFilterElement';
import { getExtensionComponent } from '../../../../../helpers/registry';
import { Button } from 'primereact/button';
import { useRouter } from "../../../../../hooks/useRouter";
import { kebabCase } from 'change-case';
var SolidRelationManyToOneColumn = function (_a) {
    var _b;
    var solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column;
    var filterable = column.attrs.filterable;
    var showFilterOperator = false;
    var filterMatchModeOptions = [
        { label: 'In', value: FilterMatchMode.IN },
        { label: 'Not In', value: FilterMatchMode.NOT_IN },
    ];
    var columnDataType = undefined;
    var filterTemplate = function (options) {
        return (_jsx(SolidVarInputsFilterElement, { values: options.value, onChange: function (e) { return options.filterCallback(e, options.index); }, inputType: InputTypes.RelationManyToOne, solidListViewMetaData: solidListViewMetaData, fieldMetadata: fieldMetadata, column: column }));
    };
    var header = (_b = column.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    return (_jsx(Column, { field: fieldMetadata.name, header: header, 
        // className="text-sm"
        sortable: column.attrs.sortable, 
        // filter={filterable}
        dataType: columnDataType, showFilterOperator: showFilterOperator, filterMatchModeOptions: filterMatchModeOptions, filterElement: filterTemplate, body: function (rowData) {
            var viewWidget = column.attrs.viewWidget;
            if (!viewWidget) {
                viewWidget = 'DefaultRelationManyToOneListWidget';
            }
            var DynamicWidget = getExtensionComponent(viewWidget);
            var widgetProps = {
                rowData: rowData,
                solidListViewMetaData: solidListViewMetaData,
                fieldMetadata: fieldMetadata,
                column: column
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }, filterPlaceholder: "Search by ".concat(fieldMetadata.displayName), style: { minWidth: "12rem" }, headerClassName: "table-header-fs" }, fieldMetadata.name));
};
export default SolidRelationManyToOneColumn;
export var DefaultRelationManyToOneListWidget = function (_a) {
    var _b, _c, _d, _e;
    var rowData = _a.rowData, solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column;
    var router = useRouter();
    var manyToOneFieldData = rowData[column.attrs.name];
    // This is the userkey that will be present within the rowData.
    if (manyToOneFieldData) {
        // Since this is a many-to-one field, we fetch the user key field of the associated model.
        var userKeyField = ((_b = column === null || column === void 0 ? void 0 : column.attrs) === null || _b === void 0 ? void 0 : _b.coModelFieldToDisplay) ? (_c = column === null || column === void 0 ? void 0 : column.attrs) === null || _c === void 0 ? void 0 : _c.coModelFieldToDisplay : (_e = (_d = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _d === void 0 ? void 0 : _d.userKeyField) === null || _e === void 0 ? void 0 : _e.name;
        var manyToOneColVal = manyToOneFieldData[userKeyField];
        return (_jsx(Button, { label: manyToOneColVal, link: true, onClick: function () {
                var _a;
                // Get current path from browser
                var pathSegments = window.location.pathname.split('/').filter(Boolean);
                // The modelName is the second-last segment before "list"
                var listIndex = pathSegments.lastIndexOf('list');
                if (listIndex > 0) {
                    pathSegments[listIndex - 1] = kebabCase((_a = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _a === void 0 ? void 0 : _a.singularName);
                    pathSegments[listIndex] = "form/".concat(manyToOneFieldData === null || manyToOneFieldData === void 0 ? void 0 : manyToOneFieldData.id);
                }
                var newPath = "/".concat(pathSegments.join('/'), "?viewMode=view");
                router.push(newPath);
            } }));
    }
    else {
        return _jsx("span", {});
    }
};
