// @solidxai/core-ui/dist/routes/pages/auth/SsoPage.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useState } from "react";
import { Card } from "primereact/card";
import { ProgressSpinner } from "primereact/progressspinner";
import { useRouter } from "../../../hooks/useRouter";
import { useSearchParams } from "../../../hooks/useSearchParams";
import { signIn } from "../../../adapters/auth/signIn";
export function SsoPage() {
    var _this = this;
    var router = useRouter();
    var searchParams = useSearchParams();
    var _a = useState(null), error = _a[0], setError = _a[1];
    var _b = useState(false), success = _b[0], setSuccess = _b[1];
    var _c = useState(true), isLoading = _c[0], setIsLoading = _c[1];
    useEffect(function () {
        var fn = function () { return __awaiter(_this, void 0, void 0, function () {
            var email, password, accessToken, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        email = "";
                        password = "";
                        accessToken = searchParams.get("accessToken");
                        return [4 /*yield*/, signIn("credentials", {
                                redirect: false,
                                email: email,
                                password: password,
                                accessToken: accessToken,
                            })];
                    case 1:
                        response = _a.sent();
                        if (response === null || response === void 0 ? void 0 : response.error) {
                            setIsLoading(false);
                            setError(response.error);
                        }
                        else {
                            setIsLoading(false);
                            setSuccess(true);
                            router.push("/admin");
                        }
                        return [2 /*return*/];
                }
            });
        }); };
        fn();
    }, [router, searchParams]);
    return (_jsx("div", { className: "card flex justify-content-center align-items-center", style: { height: "80vh" }, children: _jsxs(Card, { className: "custom-card md:w-25rem", style: {
                boxShadow: "rgba(0, 0, 0, 0.16) 0px 10px 36px 0px, rgba(0, 0, 0, 0.06) 0px 0px 0px 1px",
            }, children: [_jsx("div", { style: {
                        position: "relative",
                        display: "flex",
                        justifyContent: "center",
                        zIndex: 1,
                    }, children: isLoading ? (_jsxs(_Fragment, { children: [_jsx(ProgressSpinner, { style: { width: "50px", height: "50px" } }), _jsx("style", { children: "\n                .p-progress-spinner circle {\n                  stroke: blue !important;\n                }\n              " })] })) : success ? (_jsx("i", { className: "pi pi-check", style: { color: "green", fontSize: "30px", fontWeight: "700" } })) : (_jsx("i", { className: "pi pi-times", style: { color: "red", fontSize: "30px", fontWeight: "700" } })) }), _jsxs("p", { className: "m-20", style: { textAlign: "center" }, children: [isLoading && "Please wait while we authenticate your profile.", error && "".concat(error, " Your not authenticated")] })] }) }));
}
