// @solidxai/core-ui/dist/components/core/form/fields/relations/widgets/helpers/InlineRelationEntityDialog.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { Dialog } from "primereact/dialog";
import { camelCase } from "lodash";
import SolidFormView from "../../../../../../../components/core/form/SolidFormView";
export var InlineRelationEntityDialog = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    var visible = _a.visible, setVisible = _a.setVisible, fieldContext = _a.fieldContext, onCreate = _a.onCreate;
    var fieldLayoutInfo = fieldContext.field;
    var params = {
        moduleName: fieldContext.fieldMetadata.relationModelModuleName,
        id: "new",
        embeded: true,
        customLayout: (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.inlineCreateLayout,
        customCreateHandler: function (values) {
            setVisible(false);
            onCreate(values);
        },
        inlineCreateAutoSave: (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.inlineCreateAutoSave,
        handlePopupClose: function () { return setVisible(false); },
        modelName: camelCase(fieldContext.fieldMetadata.relationCoModelSingularName)
    };
    return (_jsx(Dialog, { visible: visible, showHeader: false, style: {
            width: (_g = (_f = (_e = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.inlineCreateLayout) === null || _e === void 0 ? void 0 : _e.attrs) === null || _f === void 0 ? void 0 : _f.width) !== null && _g !== void 0 ? _g : "60vw",
            height: (_l = (_k = (_j = (_h = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.inlineCreateLayout) === null || _j === void 0 ? void 0 : _j.attrs) === null || _k === void 0 ? void 0 : _k.height) !== null && _l !== void 0 ? _l : "auto"
        }, 
        // breakpoints={{ '1199px': '35rem', "767px": '85vw', "550px": '90vw' }}
        onHide: function () { return setVisible(false); }, className: "solid-dialog", children: _jsx(SolidFormView, __assign({}, params)) }));
};
