// @solidxai/core-ui/dist/components/core/form/fields/relations/widgets/helpers/useRelationEntityHandler.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { useState } from "react";
import { createSolidEntityApi } from "../../../../../../../redux/api/solidEntityApi";
export var useRelationEntityHandler = function (_a) {
    var fieldContext = _a.fieldContext, formik = _a.formik, _b = _a.autoCompleteLimit, autoCompleteLimit = _b === void 0 ? 1000 : _b;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var entityApi = createSolidEntityApi(fieldMetadata.relationCoModelSingularName);
    var useLazyGetSolidEntitiesQuery = entityApi.useLazyGetSolidEntitiesQuery;
    var triggerGetSolidEntities = useLazyGetSolidEntitiesQuery()[0];
    var _c = useState([]), autoCompleteItems = _c[0], setAutoCompleteItems = _c[1];
    var fetchRelationEntities = function (autocompleteQs, limit) {
        if (autocompleteQs === void 0) { autocompleteQs = ""; }
        if (limit === void 0) { limit = autoCompleteLimit; }
        return __awaiter(void 0, void 0, void 0, function () {
            var response, data, mappedItems;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, triggerGetSolidEntities(autocompleteQs)];
                    case 1:
                        response = _a.sent();
                        data = response.data;
                        if (data) {
                            mappedItems = data.records.map(function (item) {
                                var _a, _b;
                                return ({
                                    label: item[(_b = (_a = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _a === void 0 ? void 0 : _a.userKeyField) === null || _b === void 0 ? void 0 : _b.name],
                                    value: item['id'],
                                    original: item
                                });
                            });
                            setAutoCompleteItems(mappedItems);
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    var addNewRelation = function (values) {
        var _a, _b;
        var currentData = formik.values[fieldLayoutInfo.attrs.name] || [];
        var jsonValues = Object.fromEntries(values.entries());
        var newItem = {
            label: jsonValues[(_b = (_a = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.relationModel) === null || _a === void 0 ? void 0 : _a.userKeyField) === null || _b === void 0 ? void 0 : _b.name],
            value: "new",
            original: jsonValues,
        };
        formik.setFieldValue(fieldLayoutInfo.attrs.name, __spreadArray(__spreadArray([], currentData, true), [newItem], false));
        // Optionally add to autocomplete list
        setAutoCompleteItems(function (prev) {
            var exists = prev.some(function (item) { return item.label === newItem.label; });
            return exists ? prev : __spreadArray(__spreadArray([], prev, true), [newItem], false);
        });
    };
    return {
        autoCompleteItems: autoCompleteItems,
        fetchRelationEntities: fetchRelationEntities,
        addNewRelation: addNewRelation
    };
};
