// @solidxai/core-ui/dist/components/core/users/CreateUserRole.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { AutoCompleteField } from "../../../components/common/AutoCompleteField";
import { CancelButton } from "../../../components/common/CancelButton";
import { ERROR_MESSAGES } from "../../../constants/error-messages";
import { useLazyGetrolesQuery } from "../../../redux/api/roleApi";
import { useCreateuserroleMutation } from "../../../redux/api/userApi";
import { useFormik } from "formik";
import { usePathname } from "../../../hooks/usePathname";
import { useRouter } from "../../../hooks/useRouter";
import { Button } from "primereact/button";
import { InputText } from "primereact/inputtext";
import { Message } from "primereact/message";
import { classNames } from "primereact/utils";
import qs from "qs";
import { useEffect, useRef } from "react";
import * as Yup from "yup";
var CreateUserRole = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h;
    var data = _a.data;
    var toast = useRef(null);
    var _j = useLazyGetrolesQuery(), triggerGetRoles = _j[0], _k = _j[1], roleLazy = _k.data, isRoleLazyFetching = _k.isFetching, roleLazyError = _k.error;
    var router = useRouter();
    var _l = useCreateuserroleMutation(), createMenuItem = _l[0], _m = _l[1], isLoading = _m.isLoading, isSuccess = _m.isSuccess, isError = _m.isError, error = _m.error, newInvoice = _m.data;
    var initialValues = {
        username: data ? data.username : "",
        roleName: data ? data.roleName : ""
    };
    var validationSchema = Yup.object({
        username: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED('User Name')),
        roleName: Yup.lazy(function (value) {
            if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
                return Yup.object().required(ERROR_MESSAGES.FIELD_REUQIRED('Role Name'));
            }
            else if (typeof value === 'string') {
                return Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED('Role Name'));
            }
            else {
                return Yup.mixed().required(ERROR_MESSAGES.FIELD_REUQIRED('Role Name'));
            }
        })
    });
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var pathname = usePathname();
    var formik = useFormik({
        initialValues: initialValues,
        validationSchema: validationSchema,
        enableReinitialize: true,
        onSubmit: function (values) { return __awaiter(void 0, void 0, void 0, function () {
            var obj;
            return __generator(this, function (_a) {
                try {
                    obj = {
                        username: values.username,
                        roleName: values.roleName.label
                    };
                    createMenuItem(obj);
                }
                catch (err) {
                    console.error(ERROR_MESSAGES.FAILED_CREATE_INVOICE, err);
                }
                return [2 /*return*/];
            });
        }); },
    });
    var showError = function () { return __awaiter(void 0, void 0, void 0, function () {
        var errors, errorMessages;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, formik.validateForm()];
                case 1:
                    errors = _b.sent();
                    errorMessages = Object.values(errors);
                    if (errorMessages.length > 0) {
                        (_a = toast === null || toast === void 0 ? void 0 : toast.current) === null || _a === void 0 ? void 0 : _a.show({
                            severity: "success",
                            summary: ERROR_MESSAGES.SEND_REPORT,
                            // sticky: true,
                            life: 3000,
                            //@ts-ignore
                            content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: errorMessages.map(function (m, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: String(m) }) }, index)); }) })); },
                        });
                    }
                    return [2 /*return*/];
            }
        });
    }); };
    var searchRole = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var query, queryData, queryString, result, filteredMenu;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    query = event.query;
                    queryData = {};
                    if (query) {
                        queryData = {
                            roleName: query
                        };
                    }
                    queryString = qs.stringify(queryData, {
                        encodeValuesOnly: true
                    });
                    return [4 /*yield*/, triggerGetRoles(queryString).unwrap()];
                case 1:
                    result = _a.sent();
                    // Map the API response to AutoComplete format
                    if (result && result.data) {
                        filteredMenu = result.data.map(function (m) { return ({ label: m.name, value: m.id }); });
                        // Update the suggestions in state
                        return [2 /*return*/, filteredMenu];
                    }
                    else {
                        // Handle the case where no data is returned
                        return [2 /*return*/, []];
                    }
                    return [2 /*return*/];
            }
        });
    }); };
    useEffect(function () {
        if (isSuccess == true) {
            router.push("/admin/iam/users/all");
        }
    }, [isSuccess]);
    return (_jsx("div", { className: "grid", children: _jsx("div", { className: "col-12 xl:col-10 mx-auto", children: _jsx("div", { children: _jsxs("form", { onSubmit: formik.handleSubmit, children: [pathname.includes('create') ? _jsxs("div", { className: "flex gap-3 justify-content-between mb-5", children: [_jsx("div", { className: "form-wrapper-title", children: " Create User Role" }), _jsxs("div", { className: "gap-3 flex", children: [_jsx("div", { children: _jsx(Button, { label: "Save", size: "small", onClick: function () { return showError(); }, type: "submit", className: 'small-button' }) }), _jsx(CancelButton, {})] })] }) :
                            _jsxs("div", { className: "flex gap-3 justify-content-between mb-5", children: [_jsx("h1", { className: "m-0", children: " Edit User" }), _jsxs("div", { className: "gap-3 flex", children: [_jsx("div", { children: _jsx(Button, { label: "Save", onClick: function () { return showError(); }, type: "submit", size: "small", className: 'small-button' }) }), _jsx("div", {}), _jsx(CancelButton, {})] })] }), _jsx("div", { className: "form-wrapper", children: _jsxs("div", { className: "grid formgrid", children: [_jsx("div", { className: "md:col-6 sm:col-12", children: _jsxs("div", { className: "field", children: [_jsx("label", { htmlFor: "username", className: "form-label form-field-label", children: "User Name" }), _jsx(InputText, { type: "text", id: "username", name: "username", onChange: formik.handleChange, value: formik.values.username, className: classNames("p-inputtext-sm w-full small-input", {
                                                        "p-invalid": isFormFieldValid(formik, "username"),
                                                    }) }), isFormFieldValid(formik, "username") && (_jsx(Message, { severity: "error", text: (_b = formik.errors.username) === null || _b === void 0 ? void 0 : _b.toString() }))] }) }), _jsx("div", { className: "md:col-6 sm:col-12", children: _jsxs("div", { className: "field", children: [_jsx("label", { htmlFor: "type", className: "form-label form-field-label", children: "Role Name" }), _jsxs("div", { children: [_jsx(AutoCompleteField, { multiple: false, isFormFieldValid: isFormFieldValid, formik: formik, fieldName: "roleName", searchData: searchRole, existingData: { label: (_c = data === null || data === void 0 ? void 0 : data.parentCategory) === null || _c === void 0 ? void 0 : _c.title, value: (_d = data === null || data === void 0 ? void 0 : data.parentCategory) === null || _d === void 0 ? void 0 : _d.id }, existingDataTitle: (_e = data === null || data === void 0 ? void 0 : data.parentCategory) === null || _e === void 0 ? void 0 : _e.title, existingDataId: (_f = data === null || data === void 0 ? void 0 : data.parentCategory) === null || _f === void 0 ? void 0 : _f.id, className: "solid-standard-autocomplete" }), isFormFieldValid(formik, "roleName") && (_jsx(Message, { severity: "error", text: (_h = (_g = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _g === void 0 ? void 0 : _g.roleName) === null || _h === void 0 ? void 0 : _h.toString() }))] })] }) })] }) })] }) }) }) }));
};
export default CreateUserRole;
