// @solidxai/core-ui/dist/components/core/users/CreateUser.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { BackButton } from "../../../components/common/BackButton";
import { CancelButton } from "../../../components/common/CancelButton";
import { SolidFormHeader } from "../../../components/common/SolidFormHeader";
import { ERROR_MESSAGES } from "../../../constants/error-messages";
import { useRegisterPrivateMutation, useUpdateUserMutation } from "../../../redux/api/authApi";
import { useGetrolesQuery } from "../../../redux/api/roleApi";
import { useDeleteUserMutation } from "../../../redux/api/userApi";
import { useFormik } from "formik";
import { usePathname } from "../../../hooks/usePathname";
import { useRouter } from "../../../hooks/useRouter";
import { Button } from "primereact/button";
import { Checkbox } from "primereact/checkbox";
import { InputText } from "primereact/inputtext";
import { Message } from "primereact/message";
import { Panel } from "primereact/panel";
import { Password } from "primereact/password";
import { Toast } from "primereact/toast";
import { classNames } from "primereact/utils";
import { useEffect, useRef, useState } from "react";
import * as Yup from "yup";
var CreateUser = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
    var data = _a.data, params = _a.params;
    var toast = useRef(null);
    var pathname = usePathname();
    var router = useRouter();
    var _r = useState([]), selectedRoles = _r[0], setSelectedRoles = _r[1];
    var _s = useRegisterPrivateMutation(), registerPrivate = _s[0], _t = _s[1], isLoading = _t.isLoading, userCreateError = _t.error, isSuccess = _t.isSuccess, user = _t.data;
    var _u = useUpdateUserMutation(), updateUser = _u[0], _v = _u[1], isuserUpdating = _v.isLoading, isUpdateuserSuceess = _v.isSuccess, isuserUpdateError = _v.isError, userUpdateError = _v.error, updateduser = _v.data;
    var _w = useDeleteUserMutation(), deleteUser = _w[0], _x = _w[1], isUserDeleted = _x.isLoading, isDeleteUserSuceess = _x.isSuccess, isUserDeleteError = _x.isError, UserDeleteError = _x.error, DeletedUser = _x.data;
    var rolesData = useGetrolesQuery("").data;
    // const [
    //   createrolebulk,
    //   { isLoading, isSuccess, isError, error, data: roles },
    // ] = useCreateuserrolebulkMutation();
    useEffect(function () {
        if (data && data.roles) {
            var userRoles = data.roles.map(function (role) { return role.name; });
            setSelectedRoles(userRoles);
        }
    }, [data]);
    var handleCheckboxChange = function (roleName) {
        var updatedRoles;
        if (selectedRoles.includes(roleName)) {
            updatedRoles = selectedRoles.filter(function (name) { return name !== roleName; });
        }
        else {
            updatedRoles = __spreadArray(__spreadArray([], selectedRoles, true), [roleName], false);
        }
        setSelectedRoles(updatedRoles);
        formik.setFieldTouched('roles', true); // fake "touched"
        formik.setFieldValue('roles', updatedRoles);
    };
    var initialValues = {
        fullName: data ? data.fullName : "",
        username: data ? data.username : "",
        email: data ? data.email : "",
        mobile: data ? data.mobile : "",
        password: "",
        confirmPassword: "",
    };
    var validationSchema = Yup.object({
        fullName: Yup.string().required(),
        username: Yup.string()
            .required('Username is required'),
        // .min(3, 'Username must be at least 3 characters long') // Minimum length
        // .max(20, 'Username cannot be longer than 20 characters') // Maximum length
        // .matches(
        //   /^[a-zA-Z0-9_.-]*$/,
        //   'Username can only contain letters, numbers, underscores, periods, and hyphens'
        // ) // Allowed characters
        // .matches(
        //   /^[a-zA-Z]/,
        //   'Username must start with a letter'
        // ),
        email: Yup
            .string()
            .email(ERROR_MESSAGES.FIELD_INVALID('email address'))
            .required(ERROR_MESSAGES.FIELD_REUQIRED('Email')),
        password: Yup.string().min(6, ERROR_MESSAGES.PASSWORD_CHARACTER(6)).nullable(),
        confirmPassword: Yup.string()
            .when('password', {
            is: function (val) { return !!val; },
            then: function (schema) {
                return schema
                    .oneOf([Yup.ref('password')], ERROR_MESSAGES.FIELD_MUST_MATCH('Password'))
                    .nullable();
            },
            otherwise: function (schema) { return schema.notRequired().nullable(); },
        }),
        mobile: Yup.number().required(ERROR_MESSAGES.FIELD_REUQIRED('Mobile')),
    });
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var showError = function (submit) { return __awaiter(void 0, void 0, void 0, function () {
        var errors, errorMessages;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, formik.validateForm()];
                case 1:
                    errors = _a.sent();
                    errorMessages = Object.values(errors);
                    if (errorMessages.length > 0 && toast.current) {
                        toast.current.show({
                            severity: "error",
                            summary: ERROR_MESSAGES.SEND_REPORT,
                            // sticky: true,
                            life: 3000,
                            //@ts-ignore
                            content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: errorMessages.map(function (m, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: m }) }, index)); }) })); },
                        });
                    }
                    return [2 /*return*/];
            }
        });
    }); };
    function isFetchBaseQueryErrorWithErrorResponse(error) {
        return error && typeof error === 'object' && 'data' in error && 'message' in error.data;
    }
    useEffect(function () {
        var handleError = function (errorToast) {
            var _a;
            var errorMessage = [ERROR_MESSAGES.ERROR_OCCURED];
            if (isFetchBaseQueryErrorWithErrorResponse(errorToast)) {
                errorMessage = errorToast.data.message;
            }
            else {
                errorMessage = [ERROR_MESSAGES.SOMETHING_WRONG];
            }
            (_a = toast.current) === null || _a === void 0 ? void 0 : _a.show({
                severity: 'error',
                summary: ERROR_MESSAGES.ERROR,
                detail: errorMessage,
                sticky: true,
                //@ts-ignore
                content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: Array.isArray(errorMessage) ? (errorMessage.map(function (message, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: message.trim() }) }, index)); })) : (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: errorMessage === null || errorMessage === void 0 ? void 0 : errorMessage.trim() }) })) })); },
            });
        };
        // Check and handle errors from each API operation
        if (userCreateError) {
            handleError(userCreateError);
        }
        if (userUpdateError) {
            handleError(userUpdateError);
        }
    }, [
        userCreateError, userUpdateError
    ]);
    var formik = useFormik({
        initialValues: initialValues,
        validationSchema: validationSchema,
        enableReinitialize: true,
        onSubmit: function (values) { return __awaiter(void 0, void 0, void 0, function () {
            var userData, userData;
            return __generator(this, function (_a) {
                if (data) {
                    userData = {
                        fullName: values.fullName,
                        username: values.username,
                        email: values.email,
                        mobile: values.mobile,
                        roles: selectedRoles,
                    };
                    if (values.password) {
                        userData.password = values.password;
                    }
                    updateUser({ id: data.id, data: userData });
                }
                else {
                    userData = {
                        fullName: values.fullName,
                        username: values.username,
                        email: values.email,
                        mobile: values.mobile,
                        password: values.password,
                        roles: selectedRoles,
                    };
                    registerPrivate(userData);
                }
                return [2 /*return*/];
            });
        }); },
    });
    useEffect(function () {
        if (isSuccess == true || isDeleteUserSuceess === true || isUpdateuserSuceess === true) {
            router.back();
        }
    }, [isSuccess, isDeleteUserSuceess, isUpdateuserSuceess]);
    return (_jsxs("div", { className: "solid-form-wrapper", children: [_jsx(Toast, { ref: toast }), _jsx("div", { className: "solid-form-section", children: _jsxs("form", { onSubmit: formik.handleSubmit, children: [_jsx("div", { className: "solid-form-header", children: params.id === "new" ? (_jsxs(_Fragment, { children: [_jsxs("div", { className: "flex align-items-center gap-3", children: [_jsx(BackButton, {}), _jsx("div", { className: "form-wrapper-title", children: "Create User" })] }), _jsxs("div", { className: "gap-3 flex", children: [formik.dirty &&
                                                _jsx(Button, { label: "Save", size: "small", type: "submit" }), _jsx(CancelButton, {})] })] })) : (_jsxs(_Fragment, { children: [_jsxs("div", { className: "flex align-items-center gap-3", children: [_jsx(BackButton, {}), _jsx("div", { className: "form-wrapper-title", children: "Update User" })] }), _jsx("div", { children: _jsxs("div", { className: "gap-3 flex", children: [formik.dirty &&
                                                    _jsx(Button, { label: "Save", size: "small", type: "submit" }), data &&
                                                    _jsx(Button, { outlined: true, label: "Delete", size: "small", severity: "danger", type: "button", onClick: function () { return deleteUser(data.id); } }), _jsx(CancelButton, {})] }) })] })) }), _jsx(SolidFormHeader, {}), _jsx("div", { className: "px-4 py-3 md:p-4 solid-form-content", children: _jsx("div", { className: "grid", children: _jsxs("div", { className: "col-12 lg:col-10 xl:col-8 mx-auto", children: [_jsx(Panel, { header: "Basic Info", className: "solid-column-panel", children: _jsxs("div", { className: "grid formgrid mt-3", children: [_jsxs("div", { className: "field col-12 md:col-6 flex flex-column gap-2", children: [_jsx("label", { htmlFor: "fullName", className: "form-field-label", children: "Full Name" }), _jsx(InputText, { type: "text", id: "fullName", name: "fullName", autoComplete: "off", onChange: formik.handleChange, value: formik.values.fullName, className: classNames("", {
                                                                    "p-invalid": formik.touched.fullName && formik.errors.fullName,
                                                                }) }), isFormFieldValid(formik, "fullName") && (_jsx(Message, { severity: "error", text: (_c = (_b = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _b === void 0 ? void 0 : _b.fullName) === null || _c === void 0 ? void 0 : _c.toString() }))] }), _jsxs("div", { className: "field col-12 md:col-6 flex flex-column gap-2 mt-3 md:mt-3", children: [_jsx("label", { htmlFor: "username", className: "form-field-label", children: "Username" }), _jsx(InputText, { type: "text", id: "username", name: "username", autoComplete: "off", disabled: data ? true : false, onChange: formik.handleChange, value: formik.values.username, className: classNames("", {
                                                                    "p-invalid": formik.touched.username && formik.errors.username,
                                                                }) }), isFormFieldValid(formik, "username") && (_jsx(Message, { severity: "error", text: (_e = (_d = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _d === void 0 ? void 0 : _d.username) === null || _e === void 0 ? void 0 : _e.toString() }))] }), _jsxs("div", { className: "field col-12 md:col-6 flex flex-column gap-1 mt-4", children: [_jsx("label", { htmlFor: "email", className: "form-field-label", children: "Email" }), _jsx(InputText, { type: "text", id: "email", name: "email", autoComplete: "off", disabled: data ? true : false, onChange: formik.handleChange, value: formik.values.email, className: classNames("", {
                                                                    "p-invalid": formik.touched.email && formik.errors.email,
                                                                }) }), isFormFieldValid(formik, "email") && (_jsx(Message, { severity: "error", text: (_g = (_f = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _f === void 0 ? void 0 : _f.email) === null || _g === void 0 ? void 0 : _g.toString() }))] }), _jsxs("div", { className: "field col-12 md:col-6 flex flex-column gap-1 mt-4", children: [_jsx("label", { htmlFor: "mobile", className: "form-field-label", children: "Mobile" }), _jsx(InputText, { type: "text", id: "mobile", name: "mobile", autoComplete: "off", onChange: formik.handleChange, value: formik.values.mobile, className: classNames("", {
                                                                    "p-invalid": formik.touched.mobile && formik.errors.mobile,
                                                                }) }), isFormFieldValid(formik, "mobile ") && (_jsx(Message, { severity: "error", text: (_j = (_h = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _h === void 0 ? void 0 : _h.mobile) === null || _j === void 0 ? void 0 : _j.toString() }))] }), params.id === "new" && _jsxs(_Fragment, { children: [_jsxs("div", { className: "field col-12 md:col-6 flex flex-column gap-2 py-4", children: [_jsx("label", { htmlFor: "Password", className: "form-field-label", children: "Password" }), _jsx(Password, { id: "password", name: "password", autoComplete: "off", "aria-autocomplete": "none", value: formik.values.password, onChange: formik.handleChange, onBlur: formik.handleBlur, toggleMask: true, invalid: isFormFieldValid(formik, "password"), inputClassName: "w-full", feedback: false }), isFormFieldValid(formik, "password") && (_jsx(Message, { severity: "error", text: (_l = (_k = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _k === void 0 ? void 0 : _k.password) === null || _l === void 0 ? void 0 : _l.toString() }))] }), _jsxs("div", { className: "field col-12 md:col-6 flex flex-column gap-2 pb-4 md:pb-0 md:my-4", children: [_jsx("label", { htmlFor: "Confirm Password", className: "form-field-label", children: "Confirm Password" }), _jsx(Password, { id: "confirmPassword", name: "confirmPassword", autoComplete: "off", "aria-autocomplete": "none", value: formik.values.confirmPassword, onChange: formik.handleChange, onBlur: formik.handleBlur, toggleMask: true, invalid: isFormFieldValid(formik, "confirmPassword"), inputClassName: "w-full", feedback: false }), isFormFieldValid(formik, "confirmPassword") && (_jsx(Message, { severity: "error", text: (_o = (_m = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _m === void 0 ? void 0 : _m.confirmPassword) === null || _o === void 0 ? void 0 : _o.toString() }))] })] })] }) }), _jsx(Panel, { toggleable: true, header: "Roles", className: "solid-column-panel mt-5", children: _jsx("div", { className: "formgrid grid mt-4", children: ((_p = rolesData === null || rolesData === void 0 ? void 0 : rolesData.data) === null || _p === void 0 ? void 0 : _p.records) && ((_q = rolesData === null || rolesData === void 0 ? void 0 : rolesData.data) === null || _q === void 0 ? void 0 : _q.records.map(function (role, i) { return (_jsxs("div", { className: "field col-6 flex gap-2 ".concat(i >= 2 ? 'mt-3' : ''), children: [_jsx(Checkbox, { inputId: role.name, checked: selectedRoles.includes(role.name), onChange: function () { return handleCheckboxChange(role.name); } }), _jsxs("label", { htmlFor: role.name, children: [" ", role.name] })] }, role.name)); })) }) })] }) }) })] }) })] }));
};
export default CreateUser;
