// @solidxai/core-ui/dist/components/core/users/UserListView.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useDeleteMultipleUsersMutation, useLazyGetusersQuery } from '../../../redux/api/userApi';
import Link from "../../common/Link";
import { FilterMatchMode } from "primereact/api";
import { Button } from "primereact/button";
import { Column } from "primereact/column";
import { DataTable, } from "primereact/datatable";
import { useEffect, useRef, useState } from "react";
import { CreateButton } from "../../../components/common/CreateButton";
import { Dialog } from "primereact/dialog";
import { Toast } from "primereact/toast";
import qs from "qs";
import { ERROR_MESSAGES } from '../../../constants/error-messages';
export var UserListView = function () {
    var toast = useRef(null);
    var _a = useState([]), users = _a[0], setUsers = _a[1];
    var _b = useState({
        fullName: { value: null, matchMode: FilterMatchMode.STARTS_WITH },
        username: { value: null, matchMode: FilterMatchMode.STARTS_WITH },
        email: { value: null, matchMode: FilterMatchMode.STARTS_WITH },
        mobile: { value: null, matchMode: FilterMatchMode.STARTS_WITH },
        // "roles.name": { value: null, matchMode: FilterMatchMode.STARTS_WITH },
    }), filters = _b[0], setFilters = _b[1];
    var _c = useState(0), first = _c[0], setFirst = _c[1];
    var _d = useState(25), rows = _d[0], setRows = _d[1];
    var _e = useState(0), totalRecords = _e[0], setTotalRecords = _e[1];
    var _f = useState(""), sortField = _f[0], setSortField = _f[1];
    var _g = useState(0), sortOrder = _g[0], setSortOrder = _g[1];
    var _h = useState([]), selectedUsers = _h[0], setSelectedUsers = _h[1];
    var _j = useState(true), loading = _j[0], setLoading = _j[1];
    var _k = useState(false), isDialogVisible = _k[0], setDialogVisible = _k[1];
    var _l = useLazyGetusersQuery(), triggerGetUser = _l[0], _m = _l[1], user = _m.data, isLoading = _m.isLoading, error = _m.error, isError = _m.isError;
    var _o = useDeleteMultipleUsersMutation(), deleteManyUser = _o[0], _p = _o[1], isUserDeleted = _p.isLoading, isDeleteUserSuceess = _p.isSuccess, isUserDeleteError = _p.isError, UserDeleteError = _p.error, DeletedUser = _p.data;
    useEffect(function () {
        var _a;
        if (user) {
            setUsers(user === null || user === void 0 ? void 0 : user.data.records);
            setTotalRecords((_a = user === null || user === void 0 ? void 0 : user.data.meta) === null || _a === void 0 ? void 0 : _a.totalRecords);
            setLoading(false);
        }
    }, [user]);
    // const initFilters = () => {
    //     setFilters(defaultFilters);
    //     setGlobalFilterValue('');
    // };
    // const clearFilter = () => {
    //     initFilters();
    // };
    useEffect(function () {
        var queryData = {
            offset: 0,
            limit: 25,
            populate: ['roles']
        };
        var queryString = qs.stringify(queryData, {
            encodeValuesOnly: true
        });
        triggerGetUser(queryString);
        setSelectedUsers([]);
    }, [isDeleteUserSuceess]);
    useEffect(function () {
        var _a;
        if (isError || isUserDeleteError) {
            setLoading(false);
            var errorMessage_1 = ERROR_MESSAGES.ERROR_OCCURED;
            var errorToast = isError ? error : UserDeleteError;
            if (isFetchBaseQueryErrorWithErrorResponse(errorToast)) {
                errorMessage_1 = "".concat(errorToast.data.message);
            }
            else {
                errorMessage_1 = ERROR_MESSAGES.SOMETHING_WRONG;
            }
            (_a = toast.current) === null || _a === void 0 ? void 0 : _a.show({
                severity: 'error',
                summary: ERROR_MESSAGES.ERROR,
                detail: errorMessage_1,
                life: 3000,
                //@ts-ignore
                content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: _jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: errorMessage_1 }) }) })); },
            });
        }
    }, [isError, isUserDeleteError]);
    function isFetchBaseQueryErrorWithErrorResponse(error) {
        return error && typeof error === 'object' && 'data' in error && 'message' in error.data;
    }
    var onPageChange = function (event) {
        setFirst(event.first);
        setRows(event.rows);
        setQueryString(event.first, event.rows, sortField, sortOrder, filters);
    };
    var onSort = function (event) {
        var sortField = event.sortField, sortOrder = event.sortOrder;
        var validSortOrder = sortOrder === 1 || sortOrder === -1 ? sortOrder : 0;
        setSortField(sortField);
        setSortOrder(validSortOrder);
        setFirst(0);
        setQueryString(0, rows, sortField, sortOrder === 1 || sortOrder === -1 ? sortOrder : 0, filters);
    };
    var onSelectionChange = function (event) {
        var value = event.value;
        setSelectedUsers(value);
    };
    // Filter fuctions
    var setQueryString = function (offset, row, field, order, filterQuery) { return __awaiter(void 0, void 0, void 0, function () {
        var formattedFilters, queryData, queryString;
        return __generator(this, function (_a) {
            formattedFilters = {};
            Object.keys(filterQuery).forEach(function (field) {
                var _a, _b;
                var filterValue = filterQuery[field].value;
                var matchMode = filterQuery[field].matchMode;
                if (filterValue !== null && filterValue !== undefined) {
                    var operator = matchMode;
                    switch (matchMode) {
                        case FilterMatchMode.CONTAINS:
                            operator = "$containsi";
                            break;
                        case FilterMatchMode.STARTS_WITH:
                            operator = "$startsWithi";
                            break;
                        case FilterMatchMode.EQUALS:
                            operator = "$eqi";
                            break;
                        case FilterMatchMode.NOT_CONTAINS:
                            operator = "$notContainsi";
                            break;
                        case FilterMatchMode.NOT_EQUALS:
                            operator = "$nei";
                            break;
                        case FilterMatchMode.ENDS_WITH:
                            operator = "$endsWithi";
                            break;
                        default:
                            operator = matchMode;
                    }
                    if (field.includes(".")) {
                        var fieldParts = field.split(".");
                        if (!formattedFilters[fieldParts[0]]) {
                            formattedFilters[fieldParts[0]] = {};
                        }
                        formattedFilters[fieldParts[0]][fieldParts[1]] = (_a = {}, _a[operator] = filterValue, _a);
                    }
                    else {
                        formattedFilters[field] = (_b = {}, _b[operator] = filterValue, _b);
                    }
                }
            });
            queryData = {
                offset: offset !== null && offset !== void 0 ? offset : first,
                limit: row !== null && row !== void 0 ? row : rows,
                populate: ['roles'],
                filters: formattedFilters,
            };
            if (field) {
                queryData.sort = [
                    "".concat(field, ":").concat(order == 0 ? null : order == 1 ? "asc" : "desc"),
                ];
            }
            queryString = qs.stringify(queryData, {
                encodeValuesOnly: true,
            });
            triggerGetUser(queryString);
            return [2 /*return*/];
        });
    }); };
    var onFilter = function (e) {
        setFilters(e.filters);
        setQueryString(0, rows, sortField, sortOrder === 1 || sortOrder === -1 ? sortOrder : 0, e.filters);
    };
    var detailsBodyTemplate = function (product) {
        return (_jsx(Link, { href: "".concat(product.id), rel: "noopener noreferrer", className: "text-sm font-bold p-0", style: { color: "#12415D" }, children: _jsx("i", { className: "pi pi-pencil", style: { fontSize: "1rem" } }) }));
    };
    var deleteBulk = function () {
        var deleteList = [];
        selectedUsers.forEach(function (element) {
            deleteList.push(element.id);
        });
        deleteManyUser(deleteList);
        setDialogVisible(false);
    };
    var onDeleteClose = function () {
        setDialogVisible(false);
        setSelectedUsers([]);
    };
    return (_jsxs("div", { className: "", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "flex gap-3 mb-4", children: [_jsx(CreateButton, {}), selectedUsers.length > 0 && _jsx(Button, { type: "button", label: "Delete", size: "small", onClick: function () { return setDialogVisible(true); }, className: "small-button", severity: "danger" })] }), _jsx("style", { children: "\n        .p-datatable .p-datatable-loading-overlay {\n        background-color: rgba(0, 0, 0, 0.0);\n        }\n        " }), _jsxs(DataTable, { value: users, lazy: true, tableStyle: { minWidth: "60rem", margin: "auto" }, size: "small", paginator: true, rows: rows, rowsPerPageOptions: [10, 25, 50], dataKey: "id", filters: filters, emptyMessage: "No users found", onFilter: onFilter, filterDisplay: "row", totalRecords: totalRecords, first: first, onPage: onPageChange, onSort: function (e) { return onSort(e); }, sortField: sortField, sortOrder: sortOrder === 1 || sortOrder === -1 ? sortOrder : 0, loading: loading || isLoading, loadingIcon: "pi pi-spinner", selection: selectedUsers, onSelectionChange: onSelectionChange, selectionMode: "multiple", removableSort: true, children: [_jsx(Column, { selectionMode: "multiple", headerStyle: { width: "3em" } }), _jsx(Column, { field: "id", header: "ID", className: "text-sm", sortable: true, headerClassName: "table-header-fs" }), _jsx(Column, { field: "fullName", header: "Full Name", className: "text-sm", sortable: true, filter: true, filterPlaceholder: "Search user by full name", style: { minWidth: "12rem" }, headerClassName: "table-header-fs" }), _jsx(Column, { field: "username", header: "Username", className: "text-sm", sortable: true, filter: true, filterPlaceholder: "Search user by username", style: { minWidth: "12rem" }, headerClassName: "table-header-fs" }), _jsx(Column, { field: "email", header: "Email", className: "text-sm", sortable: true, filter: true, filterPlaceholder: "Search user by email", style: { minWidth: "12rem" }, headerClassName: "table-header-fs" }), _jsx(Column, { body: detailsBodyTemplate })] }), _jsx(Dialog, { visible: isDialogVisible, header: "Confirm Delete", modal: true, className: "solid-confirm-dialog", footer: function () { return (_jsxs("div", { className: "flex justify-content-center", children: [_jsx(Button, { label: "Yes", icon: "pi pi-check", className: 'small-button', severity: "danger", autoFocus: true, onClick: deleteBulk }), _jsx(Button, { label: "No", icon: "pi pi-times", className: 'small-button', onClick: onDeleteClose })] })); }, onHide: function () { return setDialogVisible(false); }, children: _jsx("p", { children: "Are you sure you want to delete the selected Users?" }) })] }));
};
