// @solidxai/core-ui/dist/components/core/model/CreateModel.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { CancelButton, SolidCancelButton } from "../../../components/common/CancelButton";
import { handleError } from "../../../helpers/ToastContainer";
import { useGetFieldDefaultMetaDataQuery } from "../../../redux/api/fieldApi";
import { useCreatemodelMutation, useDeletemodelMutation, useLazyGetModelsQuery, useUpdatemodelMutation } from "../../../redux/api/modelApi";
import { usePathname } from "../../../hooks/usePathname";
import { useRouter } from "../../../hooks/useRouter";
import { Button } from "primereact/button";
import { TabPanel, TabView } from "primereact/tabview";
import { Toast } from "primereact/toast";
import qs from "qs";
import { useEffect, useRef, useState } from "react";
import FieldMetaData from "./FieldMetaData";
import ModelMetaData from "./ModelMetaData";
import { BackButton } from "../../../components/common/BackButton";
import { Dialog } from "primereact/dialog";
import { Divider } from "primereact/divider";
import { OverlayPanel } from "primereact/overlaypanel";
import { SolidFormHeader } from "../../../components/common/SolidFormHeader";
import { ERROR_MESSAGES } from "../../../constants/error-messages";
var CreateModel = function (_a) {
    var data = _a.data, params = _a.params;
    var toast = useRef(null);
    var router = useRouter();
    var pathname = usePathname();
    var _b = useState(false), deleteEntity = _b[0], setDeleteEntity = _b[1];
    var _c = useState(data ? true : false), isLoadingData = _c[0], setIsLoadingData = _c[1];
    var formikModelMetadataRef = useRef();
    var formikFieldsMetadataRef = useRef();
    var _d = useState(), modelMetaData = _d[0], setModelMetaData = _d[1];
    var _e = useState([]), fieldMetaData = _e[0], setFieldMetaData = _e[1];
    var _f = useGetFieldDefaultMetaDataQuery(null), fieldDefaultMetaData = _f.data, refetch = _f.refetch;
    var _g = useState([]), allModelsNames = _g[0], setAllModelsNames = _g[1];
    var _h = useLazyGetModelsQuery(), triggerGetModels = _h[0], _j = _h[1], allmodels = _j.data, isAllModelsLoaded = _j.isLoading, allModelsError = _j.error;
    var _k = useCreatemodelMutation(), createModel = _k[0], _l = _k[1], isCreateModelLoading = _l.isLoading, isCreateModelSuccess = _l.isSuccess, isCreateModelError = _l.isError, createModelError = _l.error, newModel = _l.data;
    var _m = useUpdatemodelMutation(), updateModel = _m[0], _o = _m[1], isUpdatedModelUpdating = _o.isLoading, isUpdatedModelSuceess = _o.isSuccess, isUpdateModelError = _o.isError, updateModelError = _o.error, updatedModel = _o.data;
    var _p = useDeletemodelMutation(), deleteModel = _p[0], _q = _p[1], isModelDeleted = _q.isLoading, isDeleteModelSuceess = _q.isSuccess, isDeleteModelError = _q.isError, deleteModelError = _q.error, DeletedModel = _q.data;
    var _r = useState(0), activeIndex = _r[0], setActiveIndex = _r[1];
    var _s = useState({}), tabErrors = _s[0], setTabErrors = _s[1];
    var _t = useState({}), formErrors = _t[0], setFormErrors = _t[1];
    var nextTab = function () {
        setActiveIndex(activeIndex + 1);
    };
    useEffect(function () {
        var _a;
        if (data) {
            var isLegacyTable = data.isLegacyTableWithId || data.isLegacyTable;
            var isLegacyTableWithId = data.isLegacyTable;
            var modelData = __assign(__assign({}, data), { moduleId: (_a = data === null || data === void 0 ? void 0 : data.module) === null || _a === void 0 ? void 0 : _a.id, parentModelId: data === null || data === void 0 ? void 0 : data.parentModel, isLegacyTable: isLegacyTable, isLegacyTableWithId: isLegacyTableWithId });
            setIsLoadingData(false);
            setModelMetaData(modelData);
            var fieldData = data.fields.map(function (f) {
                var fieldCopy = __assign({}, f); // Create a shallow copy of the field object
                fieldCopy.identifier = f.id; // Add the identifier property
                if (fieldCopy.type == "mediaSingle" || fieldCopy.type == "mediaMultiple") {
                    fieldCopy.mediaMaxSizeKb = f.mediaMaxSizeKb / 1024; // Add the identifier property
                }
                return fieldCopy;
            });
            setFieldMetaData(fieldData.sort(function (a, b) { return b.id - a.id; }));
        }
    }, [data]);
    useEffect(function () {
        var queryData = {
            offset: 0,
            limit: 1000,
        };
        var queryString = qs.stringify(queryData, {
            encodeValuesOnly: true
        });
        triggerGetModels(queryString);
    }, [isDeleteModelSuceess]);
    useEffect(function () {
        if (allmodels) {
            if (data) {
                setAllModelsNames(allmodels === null || allmodels === void 0 ? void 0 : allmodels.records.filter(function (i) { return i.id !== data.id; }).map(function (e) { return e.singularName; }));
            }
            else {
                setAllModelsNames(allmodels === null || allmodels === void 0 ? void 0 : allmodels.records.map(function (e) { return e.singularName; }));
            }
        }
    }, [allmodels, data]);
    var handleSubmit = function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (!formikModelMetadataRef.current) return [3 /*break*/, 2];
                    return [4 /*yield*/, formikModelMetadataRef.current.submitForm()];
                case 1:
                    _a.sent(); // Call the handleSubmit function from the formik instance
                    // @ts-ignore
                    // formikModelMetadataRef.current.validateForm().then(() => {
                    //   if (Object.keys(formikModelMetadataRef?.current?.errors).length === 0) {
                    //     if (fieldMetaData.length > 0) {
                    //       handleFormSubmit();
                    //     } else {
                    //       if (activeIndex === 0) {
                    //         nextTab();
                    //         // showError({ error: "Please add Atleast one field" });
                    //         handleError(["Please add Atleast one field"])
                    //       } else {
                    //         handleError(["Please add Atleast one field"])
                    //         // showError({ error: "Please add Atleast one field" });
                    //       }
                    //     }
                    //   }
                    // });
                    formikModelMetadataRef.current.validateForm().then(function (errors) {
                        var _a;
                        var firstErrorTab;
                        if (Object.keys(errors).length > 0) {
                            setFormErrors(errors);
                            var errorMessages = Object.values(errors);
                            errorMessages.forEach(function (error) {
                                handleError([error]); // Call handleError for each error separately
                            });
                            firstErrorTab = 0; // Model Metadata tab has errors
                        }
                        else {
                            setFormErrors({});
                        }
                        if (fieldMetaData.length === 0) {
                            handleError([ERROR_MESSAGES.ADD_ATLEAST_ONE_FIELD]);
                            firstErrorTab = firstErrorTab !== null && firstErrorTab !== void 0 ? firstErrorTab : 1; // If no prior error, set to Field tab
                        }
                        if (firstErrorTab !== undefined) {
                            setTabErrors((_a = {}, _a[firstErrorTab] = true, _a)); // Set error only on the first tab with an issue
                            setActiveIndex(firstErrorTab); // Switch to the tab with an error
                        }
                        else {
                            handleFormSubmit();
                        }
                    });
                    _a.label = 2;
                case 2: return [2 /*return*/];
            }
        });
    }); };
    var handleFormSubmit = function () { return __awaiter(void 0, void 0, void 0, function () {
        var legacyTableConfig, hasPrimaryKey, fieldData, module_1, parentModel, createdAt, updatedAt, id, deletedAt, modelData, updateData, fieldData, module_2, parentModel, modelData, data_1;
        var _a;
        return __generator(this, function (_b) {
            legacyTableConfig = {};
            if ((modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isLegacyTable) && (modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isLegacyTableWithId)) {
                // UI: Both checked → Backend: both true
                legacyTableConfig = {
                    isLegacyTable: true,
                    isLegacyTableWithId: false
                };
            }
            else if ((modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isLegacyTable) && !(modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isLegacyTableWithId)) {
                // UI: Only isLegacyTable checked → Backend: only isLegacyTableWithId true
                legacyTableConfig = {
                    isLegacyTable: false,
                    isLegacyTableWithId: true
                };
            }
            else {
                // UI: Neither checked → Backend: both false
                legacyTableConfig = {
                    isLegacyTable: false,
                    isLegacyTableWithId: false
                };
            }
            if ((modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isLegacyTable) || (modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isLegacyTableWithId)) {
                hasPrimaryKey = fieldMetaData.some(function (field) { return field.isPrimaryKey === true; });
                if (!hasPrimaryKey) {
                    (_a = toast === null || toast === void 0 ? void 0 : toast.current) === null || _a === void 0 ? void 0 : _a.show({
                        severity: "error",
                        summary: "Primary Key Required",
                        detail: "Legacy tables set at least one field marked as Primary Key. Please mark a field as Primary Key before proceeding.",
                        life: 5000,
                    });
                    return [2 /*return*/];
                }
            }
            if (data) {
                fieldData = fieldMetaData.map(function (_a) {
                    var createdAt = _a.createdAt, updatedAt = _a.updatedAt, deletedAt = _a.deletedAt, mediaStorageProvider = _a.mediaStorageProvider, identifier = _a.identifier, rest = __rest(_a, ["createdAt", "updatedAt", "deletedAt", "mediaStorageProvider", "identifier"]);
                    if (rest.mediaMaxSizeKb) {
                        rest.mediaMaxSizeKb = rest.mediaMaxSizeKb ? rest.mediaMaxSizeKb * 1024 : 0;
                    }
                    return rest;
                });
                module_1 = modelMetaData.module, parentModel = modelMetaData.parentModel, createdAt = modelMetaData.createdAt, updatedAt = modelMetaData.updatedAt, id = modelMetaData.id, deletedAt = modelMetaData.deletedAt, modelData = __rest(modelMetaData, ["module", "parentModel", "createdAt", "updatedAt", "id", "deletedAt"]);
                updateData = __assign(__assign(__assign({}, modelData), { displayName: modelData.displayName.trim(), fields: fieldData }), legacyTableConfig);
                updateModel({ id: data.id, data: updateData });
            }
            else {
                if (modelMetaData) {
                    fieldData = fieldMetaData.map(function (_a) {
                        var mediaStorageProvider = _a.mediaStorageProvider, identifier = _a.identifier, rest = __rest(_a, ["mediaStorageProvider", "identifier"]);
                        if (rest.mediaMaxSizeKb) {
                            rest.mediaMaxSizeKb = rest.mediaMaxSizeKb ? rest.mediaMaxSizeKb * 1024 : 0;
                        }
                        return rest;
                    });
                    module_2 = modelMetaData.module, parentModel = modelMetaData.parentModel, modelData = __rest(modelMetaData, ["module", "parentModel"]);
                    data_1 = __assign(__assign(__assign({}, modelData), { displayName: modelData.displayName.trim(), fields: fieldData }), legacyTableConfig);
                    createModel(data_1);
                    if (isCreateModelSuccess) {
                    }
                }
            }
            return [2 /*return*/];
        });
    }); };
    var deleteModelFunction = function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            deleteModel(data === null || data === void 0 ? void 0 : data.id);
            return [2 /*return*/];
        });
    }); };
    useEffect(function () {
        if (isCreateModelSuccess == true || isUpdatedModelSuceess == true || isDeleteModelSuceess == true) {
            // router.push(`/admin/app-builder/model/all`);
            router.back();
        }
    }, [isCreateModelSuccess, isUpdatedModelSuceess, isDeleteModelSuceess]);
    var showError = function (errors) { return __awaiter(void 0, void 0, void 0, function () {
        var errorMessages;
        var _a;
        return __generator(this, function (_b) {
            errorMessages = Object.values(errors);
            if (errorMessages.length > 0) {
                (_a = toast === null || toast === void 0 ? void 0 : toast.current) === null || _a === void 0 ? void 0 : _a.show({
                    severity: "error",
                    summary: ERROR_MESSAGES.SEND_REPORT,
                    // sticky: true,
                    life: 3000,
                    //@ts-ignore
                    content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: errorMessages.map(function (m, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: String(m) }) }, index)); }) })); },
                });
            }
            return [2 /*return*/];
        });
    }); };
    var isFetchBaseQueryErrorWithErrorResponse = function (error) {
        return error && typeof error === 'object' && 'data' in error && 'message' in error.data;
    };
    // const handleError = (errorToast: any) => {
    //   let errorMessage: any = ['An error occurred'];
    //   if (isFetchBaseQueryErrorWithErrorResponse(errorToast)) {
    //     errorMessage = errorToast.data.message;
    //   } else {
    //     errorMessage = ['Something went wrong'];
    //   }
    //   toast?.current?.show({
    //     severity: 'error',
    //     summary: 'Error',
    //     detail: errorMessage,
    //     life: 3000,
    //     //@ts-ignore
    //     content: (props) => (
    //       <div className="flex flex-column align-items-left" style={{ flex: "1" }}>
    //         {Array.isArray(errorMessage) ? (
    //           errorMessage.map((message, index) => (
    //             <div className="flex align-items-center gap-2" key={index}>
    //               <span className="font-bold text-900">{message.trim()}</span>
    //             </div>
    //           ))
    //         ) : (
    //           <div className="flex align-items-center gap-2">
    //             <span className="font-bold text-900">{errorMessage?.trim()}</span>
    //           </div>
    //         )}
    //       </div>
    //     ),
    //   });
    // };
    // Error Handler 
    // Added useEffect to remove active tab class border color if error is there.
    useEffect(function () {
        var tabHeaders = document.querySelectorAll(".p-tabview-nav li");
        if (tabHeaders.length > 0) {
            if (tabErrors[0]) {
                tabHeaders[0].classList.remove("p-highlight");
            }
            if (tabErrors[1]) {
                tabHeaders[1].classList.remove("p-highlight");
            }
        }
    }, [tabErrors]);
    useEffect(function () {
        var errors = [
            { isError: isCreateModelError, error: createModelError },
            { isError: isDeleteModelError, error: deleteModelError },
            { isError: isUpdateModelError, error: updateModelError },
        ];
        if (errors.length > 0) {
            // Handle any error
            errors.forEach(function (_a) {
                var isError = _a.isError, error = _a.error;
                if (isError && error) {
                    handleError(error); // Call the centralized error handler
                }
            });
        }
    }, [isCreateModelError, isDeleteModelError, isUpdateModelError]);
    var op = useRef(null);
    var formActionDropdown = function () {
        return (_jsxs("div", { children: [_jsx(Button, { outlined: true, severity: "secondary", type: "button", icon: 'pi pi-cog', size: "small", className: "surface-card p-0 hidden md:flex", style: {
                        height: 33.06,
                        width: 33.06
                    }, onClick: function (e) {
                        // @ts-ignore 
                        return op.current.toggle(e);
                    } }), _jsx(Button, { outlined: true, type: "button", icon: 'pi pi-cog', size: "small", className: "surface-card p-0 solid-icon-button md:hidden", 
                    // style={{
                    //   height: 33.06,
                    //   width: 33.06
                    // }}
                    onClick: function (e) {
                        // @ts-ignore 
                        return op.current.toggle(e);
                    } }), _jsx(OverlayPanel, { ref: op, className: "solid-custom-overlay", children: _jsx("div", { className: "flex flex-column gap-1 p-1", children: _jsx(Button, { text: true, type: "button", className: "w-8rem text-left gap-2", label: "Delete", size: "small", iconPos: "left", severity: "danger", icon: 'pi pi-trash', onClick: function () { return setDeleteEntity(true); } }) }) })] }));
    };
    var _u = useState(false), isDirty = _u[0], setIsDirty = _u[1];
    return (_jsxs("div", { className: "solid-form-wrapper", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { style: { width: '100%', borderRight: '1px solid var(--primary-light-color' }, children: [_jsx("div", { className: "solid-form-header", children: params.id === "new" ?
                            _jsxs(_Fragment, { children: [_jsxs("div", { className: "flex align-items-center gap-3", children: [_jsx(BackButton, {}), _jsxs("div", { className: "form-wrapper-title solid-long-text-wrapper  ", children: ["Create Model ", (modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.displayName) ? "- ".concat(modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.displayName) : ""] })] }), _jsxs("div", { className: "gap-3 flex", children: [isDirty &&
                                                _jsx(Button, { label: "Save", size: "small", onClick: handleSubmit, type: "submit" }), _jsx(CancelButton, {})] })] })
                            :
                                _jsxs(_Fragment, { children: [_jsxs("div", { className: "flex align-items-center gap-3", children: [_jsx(BackButton, {}), _jsxs("div", { className: "form-wrapper-title solid-long-text-wrapper ", children: ["Edit Model ", (modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.displayName) ? "- ".concat(modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.displayName) : ""] })] }), _jsxs("div", { className: "gap-3 flex", children: [(data === null || data === void 0 ? void 0 : data.isSystem) !== true &&
                                                    _jsx(_Fragment, { children: _jsx("div", { children: isDirty &&
                                                                _jsxs(_Fragment, { children: [_jsx(Button, { label: "Save", size: "small", type: "submit", className: "hidden lg:flex", onClick: handleSubmit }), _jsx(Button, { icon: "pi pi-check", size: "small", type: "submit", className: " lg:hidden solid-icon-button", onClick: handleSubmit })] }) }) }), _jsx(SolidCancelButton, {}), formActionDropdown()] })] }) }), _jsx(SolidFormHeader, {}), _jsx("div", { className: "px-4 py-3 md:p-4 solid-form-content", children: _jsxs(TabView, { activeIndex: activeIndex, onTabChange: function (e) {
                                if (activeIndex == 0) {
                                    formikModelMetadataRef.current.handleSubmit();
                                    setActiveIndex(e.index);
                                }
                                if (activeIndex == 1) {
                                    setActiveIndex(e.index);
                                }
                            }, panelContainerClassName: "px-0", className: "relative", children: [_jsx(TabPanel, { header: "Model", headerClassName: tabErrors[0] ? "tab-error-heading " : "", children: _jsx(ModelMetaData, { modelMetaData: modelMetaData, setModelMetaData: setModelMetaData, allModelsNames: allModelsNames, deleteModelFunction: deleteModelFunction, nextTab: nextTab, formikModelMetadataRef: formikModelMetadataRef, params: params, formErrors: formErrors, setIsDirty: setIsDirty }) }), _jsx(TabPanel, { header: "Fields", headerClassName: tabErrors[1] ? "tab-error-heading " : "", children: _jsx(FieldMetaData, { modelMetaData: modelMetaData, fieldMetaData: fieldMetaData, setFieldMetaData: setFieldMetaData, deleteModelFunction: deleteModelFunction, nextTab: nextTab, formikFieldsMetadataRef: formikFieldsMetadataRef, params: params, setIsDirty: setIsDirty }) })] }) })] }), _jsxs(Dialog, { header: "Delete Field", headerClassName: "py-2", className: "solid-confirm-dialog", contentClassName: "px-0 pb-0", visible: deleteEntity, style: { width: '20vw' }, onHide: function () { if (!deleteEntity)
                    return; setDeleteEntity(false); }, children: [_jsx(Divider, { className: "m-0" }), _jsxs("div", { className: "p-4", children: [_jsx("p", { className: "m-0 solid-primary-title", style: { fontSize: 16 }, children: "Are you sure you want to delete this Field ?" }), _jsx("p", { className: "", style: { color: 'var{--solid-grey-500}' }, children: modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.singularName }), _jsxs("div", { className: "flex align-items-center gap-2 mt-3", children: [_jsx(Button, { label: "Delete", size: "small", onClick: deleteModelFunction }), _jsx(Button, { label: "Cancel", size: "small", onClick: function () { return setDeleteEntity(false); }, outlined: true })] })] })] })] }));
};
export default CreateModel;
