// @solidxai/core-ui/dist/components/core/model/FieldMetaData.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { usePathname } from "../../../hooks/usePathname";
import { Button } from "primereact/button";
import { Column } from "primereact/column";
import { DataTable } from "primereact/datatable";
import { Dialog } from "primereact/dialog";
import { useMountEffect } from "primereact/hooks";
import { Messages } from "primereact/messages";
import { useRef, useState } from "react";
import FieldMetaDataForm from "./FieldMetaDataForm";
import { ERROR_MESSAGES } from "../../../constants/error-messages";
var FieldMetaData = function (_a) {
    var setIsDirty = _a.setIsDirty, modelMetaData = _a.modelMetaData, fieldMetaData = _a.fieldMetaData, setFieldMetaData = _a.setFieldMetaData, deleteModelFunction = _a.deleteModelFunction, nextTab = _a.nextTab, formikFieldsMetadataRef = _a.formikFieldsMetadataRef, params = _a.params;
    var pathname = usePathname();
    var msgs = useRef(null);
    useMountEffect(function () {
        if (msgs.current) {
            msgs.current.clear();
            msgs.current.show({
                id: '1',
                sticky: true,
                severity: 'info',
                detail: "Please select Module and Datasource from the models Tab.",
                closable: false,
            });
        }
    });
    var toast = useRef(null);
    var _b = useState(false), visiblePopup = _b[0], setVisiblePopup = _b[1];
    var _c = useState(false), isRequiredPopUp = _c[0], setIsRequiredPopUp = _c[1];
    var _d = useState(), currentPopup = _d[0], setCurrentPopup = _d[1];
    var _e = useState(null), selectedFieldMetaData = _e[0], setSelectedFieldMetaData = _e[1];
    var _f = useState(false), deleteAlertPopup = _f[0], setDeleteAlertPopup = _f[1];
    var _g = useState(null), rowToDelete = _g[0], setRowToDelete = _g[1];
    var onRowSelect = function (event) {
        setVisiblePopup(true);
    };
    // Template for the pencil icon column
    var editTemplate = function (rowData) {
        return (_jsx(_Fragment, { children: rowData.isSystem !== true && rowData.isMarkedForRemoval !== true &&
                _jsx(Button, { icon: "pi pi-pencil", text: true, onClick: function () {
                        setSelectedFieldMetaData(rowData);
                        setVisiblePopup(true);
                    }, size: "small" }) }));
    };
    var bodyTemplate = function (rowData) {
        return (_jsx(_Fragment, { children: rowData.displayName &&
                _jsx(_Fragment, { children: _jsxs("p", { children: [rowData.displayName, rowData.isMarkedForRemoval === true &&
                                _jsxs(_Fragment, { children: [_jsx("br", {}), "  ", _jsx("span", { style: { fontSize: '11px', color: 'red' }, children: "This field will be removed next time you generate code for this model." })] })] }) }) }));
    };
    // Function to delete a row
    var deleteRow = function (rowData) {
        //show dialog bbefore deletion & which will have to option ok and cancel
        setFieldMetaData(function (prevData) {
            var updatedData = prevData.filter(function (item) { return item.name !== rowData.name; });
            setIsDirty(true); // Ensure dirty state is updated immediately
            return updatedData;
        });
    };
    // Template for the trash (delete) icon column
    var deleteTemplate = function (rowData) {
        return (_jsx(_Fragment, { children: (pathname.includes('create') || (rowData.isSystem !== true && rowData.isMarkedForRemoval !== true)) &&
                _jsx(Button, { icon: "pi pi-trash", text: true, severity: "danger", onClick: function () { setRowToDelete(rowData); setDeleteAlertPopup(true); }, size: "small" }) }));
    };
    var showToaster = function (message, severity) { return __awaiter(void 0, void 0, void 0, function () {
        var errorMessages;
        var _a;
        return __generator(this, function (_b) {
            errorMessages = Object.values(message);
            if (errorMessages.length > 0) {
                (_a = toast === null || toast === void 0 ? void 0 : toast.current) === null || _a === void 0 ? void 0 : _a.show({
                    severity: severity,
                    summary: ERROR_MESSAGES.SEND_REPORT,
                    life: 3000,
                    //@ts-ignore
                    content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: errorMessages.map(function (m, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: String(m) }) }, index)); }) })); },
                });
            }
            return [2 /*return*/];
        });
    }); };
    // const rowClass = (data: any) => {
    //   return {
    //     'bg-red-row': data.isMarkedForRemoval === true
    //   };
    // };
    return (_jsx(_Fragment, { children: !modelMetaData.moduleId || !modelMetaData.dataSource ?
            _jsx("div", { className: "card flex justify-content-center", children: _jsx(Messages, { ref: msgs }) })
            :
                _jsxs(_Fragment, { children: [_jsx("div", { className: "absolute", style: { top: -3, right: 0 }, children: modelMetaData.isSystem !== true &&
                                _jsx(Button, { label: "Add", 
                                    // icon="pi pi-external-link"
                                    onClick: function () {
                                        var _a;
                                        if (!(modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.dataSourceType)) {
                                            (_a = toast.current) === null || _a === void 0 ? void 0 : _a.show({
                                                severity: 'error',
                                                summary: ERROR_MESSAGES.ERROR,
                                                detail: ERROR_MESSAGES.ORM_TYPE_REQUIRED,
                                                sticky: true,
                                            });
                                        }
                                        else {
                                            setSelectedFieldMetaData(null);
                                            setVisiblePopup(true);
                                        }
                                    }, size: "small" }) }), _jsxs(DataTable, { value: fieldMetaData, dataKey: "id", tableStyle: { minWidth: '50rem' }, size: "small", children: [_jsx(Column, { field: "displayName", header: "Display Name", body: bodyTemplate, headerClassName: "table-header-fs" }), _jsx(Column, { field: "name", header: "Name", headerClassName: "table-header-fs" }), _jsx(Column, { field: "type", header: "Type", headerClassName: "table-header-fs" }), modelMetaData.isSystem !== true &&
                                    _jsx(Column, { body: editTemplate, header: "Edit", headerClassName: "table-header-fs", style: { width: '10%' } }), modelMetaData.isSystem !== true &&
                                    _jsx(Column, { body: deleteTemplate, header: "Delete", headerClassName: "table-header-fs", style: { width: '10%' } })] }), _jsx(Dialog, { header: "", visible: visiblePopup, style: { width: "40vw" }, className: "solid-dialog solid-field-dialog", onHide: function () {
                                if (!visiblePopup)
                                    return;
                                setVisiblePopup(false);
                            }, showHeader: false, children: _jsx(FieldMetaDataForm, { setIsDirty: setIsDirty, modelMetaData: modelMetaData, fieldMetaData: selectedFieldMetaData, allFields: fieldMetaData, setFieldMetaData: setFieldMetaData, deleteModelFunction: deleteModelFunction, setVisiblePopup: setVisiblePopup, formikFieldsMetadataRef: formikFieldsMetadataRef, params: params, setIsRequiredPopUp: setIsRequiredPopUp, showToaster: showToaster }) }), _jsxs(Dialog, { visible: isRequiredPopUp, header: (_jsxs("div", { className: "flex align-items-center", children: [_jsx("i", { className: "pi pi-exclamation-triangle text-yellow-500 text-xl mr-2" }), _jsx("span", { children: "Warning" })] })), headerClassName: "text-center warning-header-popup", modal: true, style: { width: '20vw' }, onHide: function () { return setIsRequiredPopUp(false); }, className: "solid-dialog solid-confirm-dialog", children: [_jsx("p", { className: "p-3 mb-0", children: "If there is data against this model this operation might not work and manual intervention will be required" }), _jsx("div", { className: "flex justify-content-start p-3", children: _jsx(Button, { label: "Ok", className: 'small-button', onClick: function () { return setIsRequiredPopUp(false); } }) })] }), _jsxs(Dialog, { visible: deleteAlertPopup, header: (_jsxs("div", { className: "flex align-items-center justify-content-center", children: [_jsx("i", { className: "pi pi-exclamation-triangle text-yellow-500 text-xl mr-2" }), _jsx("span", { children: "Warning" })] })), headerClassName: "text-center warning-header-popup", modal: true, style: { width: '20vw' }, onHide: function () {
                                setDeleteAlertPopup(false);
                                setRowToDelete(null);
                            }, className: "solid-dialog solid-confirm-dialog", children: [_jsx("p", { className: "p-3 mb-0 text-center", children: "Are you sure you want to delete this field?" }), _jsxs("div", { className: "flex justify-content-start p-3 align-items-center justify-content-center gap-3", children: [_jsx(Button, { label: "Ok", className: 'small-button', onClick: function () {
                                                if (rowToDelete) {
                                                    deleteRow(rowToDelete);
                                                }
                                                setDeleteAlertPopup(false);
                                                setRowToDelete(null);
                                            } }), _jsx(Button, { label: "Cancel", className: 'small-button', outlined: true, severity: "danger", onClick: function () {
                                                setDeleteAlertPopup(false);
                                                setRowToDelete(null);
                                            } })] })] })] }) }));
};
export default FieldMetaData;
