// @solidxai/core-ui/dist/components/core/model/ModelMetaData.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { SingleSelectAutoCompleteField } from "../../../components/common/SingleSelectAutoCompleteField";
import { ERROR_MESSAGES } from "../../../constants/error-messages";
import { getSingularAndPlural } from "../../../helpers/helpers";
import { useGetFieldDefaultMetaDataQuery } from "../../../redux/api/fieldApi";
import { useLazyGetModelsQuery } from "../../../redux/api/modelApi";
import { useLazyGetmodulesQuery } from "../../../redux/api/moduleApi";
import { useFormik } from "formik";
import { snakeCase } from "lodash";
import { usePathname } from "../../../hooks/usePathname";
import { useRouter } from "../../../hooks/useRouter";
import { Checkbox } from "primereact/checkbox";
import { InputText } from "primereact/inputtext";
import { InputTextarea } from "primereact/inputtextarea";
import { Message } from "primereact/message";
import { Panel } from "primereact/panel";
import { Toast } from "primereact/toast";
import { classNames } from "primereact/utils";
import qs from "qs";
import React, { useEffect, useRef, useState } from "react";
import * as Yup from "yup";
var ModelMetaData = React.forwardRef(function (_a, ref) {
    // const ModelMetaData = ({ modelMetaData, setModelMetaData, deleteModelFunction, nextTab, formikModelMetadataRef }: any) => {   
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0;
    var modelMetaData = _a.modelMetaData, setModelMetaData = _a.setModelMetaData, allModelsNames = _a.allModelsNames, deleteModelFunction = _a.deleteModelFunction, nextTab = _a.nextTab, formikModelMetadataRef = _a.formikModelMetadataRef, params = _a.params, formErrors = _a.formErrors, setIsDirty = _a.setIsDirty;
    var router = useRouter();
    var toast = useRef(null);
    var pathname = usePathname();
    var _1 = useLazyGetmodulesQuery(), triggerGetModules = _1[0], _2 = _1[1], moduleData = _2.data, isModuleFetching = _2.isFetching, moduleError = _2.error;
    var _3 = useLazyGetModelsQuery(), triggerGetModels = _3[0], _4 = _3[1], modelData = _4.data, isModelFetching = _4.isFetching, modelError = _4.error;
    var _5 = useGetFieldDefaultMetaDataQuery(null), fieldDefaultMetaData = _5.data, isLoading = _5.isLoading, error = _5.error, refetch = _5.refetch;
    var dataSourceTypes = [
        { label: "Mysql", value: "mysql" },
        { label: "Postgres", value: "postgres" },
        { label: "Mssql", value: "mssql" },
        { label: "Oracle", value: "oracle" },
        { label: "Mariadb", value: "mariadb" },
    ];
    var initialValues = {
        singularName: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.singularName : "",
        pluralName: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.pluralName : "",
        displayName: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.displayName : "",
        description: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.description : "",
        dataSource: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.dataSource : "",
        dataSourceType: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.dataSourceType : "",
        tableName: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.tableName : null,
        moduleId: modelMetaData ? (_b = modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.module) === null || _b === void 0 ? void 0 : _b.id : "",
        module: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.module : "",
        isSystem: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isSystem : false,
        enableSoftDelete: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.enableSoftDelete : "",
        enableAuditTracking: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.enableAuditTracking : true,
        internationalisation: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.internationalisation : "",
        draftPublishWorkflow: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.draftPublishWorkflow : "",
        isChild: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isChild : "",
        parentModelId: modelMetaData ? (_c = modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.parentModel) === null || _c === void 0 ? void 0 : _c.id : "",
        parentModel: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.parentModel : "",
        isLegacyTable: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isLegacyTable : false,
        isLegacyTableWithId: modelMetaData ? modelMetaData === null || modelMetaData === void 0 ? void 0 : modelMetaData.isLegacyTableWithId : false,
        //   isLegacyTable: modelMetaData 
        //   ? (modelMetaData.isLegacyTable && modelMetaData.isLegacyTableWithId && params.id !== 'new') 
        //     ? true  
        //     : modelMetaData.isLegacyTableWithId 
        //       ? true  
        //       : false 
        //   : false,
        // isLegacyTableWithId: modelMetaData 
        //   ? (modelMetaData.isLegacyTable && modelMetaData.isLegacyTableWithId &&  params.id !== 'new') 
        //     ? true  
        //     : false  
        //   : false,
    };
    var _6 = useState(false), showTableName = _6[0], setShowTableName = _6[1];
    var _7 = useState(false), showParentModel = _7[0], setShowParentModel = _7[1];
    var _8 = useState(null), selectedModule = _8[0], setSelectedModule = _8[1];
    useEffect(function () {
        if (modelMetaData && modelMetaData.tableName) {
            setShowTableName(true);
        }
        if (modelMetaData && modelMetaData.isChild) {
            setShowParentModel(true);
        }
        else if (modelMetaData && !modelMetaData.isChild) {
            setShowParentModel(false);
        }
    }, [modelMetaData]);
    var validationSchema = Yup.object({
        singularName: Yup.string()
            // .matches(
            //   /^[a-z]+(-[a-z]+)*$/,
            //   "Invalid format. Use lowercase letters and hyphens only"
            // )
            .notOneOf(allModelsNames, ERROR_MESSAGES.FIELD_ALREADY_USE('Name', 'name'))
            .required(ERROR_MESSAGES.FIELD_REUQIRED('Singular Name')),
        pluralName: Yup.string()
            // .matches(
            //   /^[a-z]+(-[a-z]+)*$/,
            //   "Invalid format. Use lowercase letters and hyphens only"
            // )
            .required(ERROR_MESSAGES.FIELD_REUQIRED('Plural Name')),
        // tableName: Yup.string().required().matches(/^[a-z0-9_]+$/, ERROR_MESSAGES.SNAKE_CASE('Tabale')),
        displayName: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED("Display Name")),
        description: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED("Description Name")),
        dataSource: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED("Data Source")),
        dataSourceType: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED("Data Source Type")),
        moduleId: Yup.number().required(ERROR_MESSAGES.FIELD_REUQIRED("Module Id")),
        module: Yup.object().required(ERROR_MESSAGES.FIELD_REUQIRED("Module")),
        isSystem: Yup.boolean(),
        enableSoftDelete: Yup.boolean(),
        enableAuditTracking: Yup.boolean(),
        internationalisation: Yup.boolean(),
        draftPublishWorkflow: Yup.boolean(),
        isChild: Yup.boolean(),
        parentModelId: Yup.number().when("isChild", function (isChild, schema) {
            if (isChild.length > 0 && isChild[0] == true) {
                return schema.required(ERROR_MESSAGES.FIELD_REUQIRED('Parent Model Id'));
            }
            else {
                return schema.notRequired().nullable();
            }
        }),
        parentModel: Yup.object().when("isChild", function (isChild, schema) {
            if (isChild.length > 0 && isChild[0] == true) {
                return schema.required(ERROR_MESSAGES.FIELD_REUQIRED('Parent Model'));
            }
            else {
                return schema.notRequired().nullable();
            }
        }),
        isLegacyTable: Yup.boolean(),
        isLegacyTableWithId: Yup.boolean(),
    });
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var formik = useFormik({
        initialValues: initialValues,
        validationSchema: validationSchema,
        validateOnChange: true,
        enableReinitialize: true,
        innerRef: formikModelMetadataRef,
        onSubmit: function (values) { return __awaiter(void 0, void 0, void 0, function () {
            var tableName, modelData_1;
            return __generator(this, function (_a) {
                tableName = generateTableName(values.module.displayName, values.singularName);
                try {
                    modelData_1 = __assign(__assign(__assign(__assign({}, modelMetaData), { singularName: values.singularName, pluralName: values.pluralName, displayName: values.displayName, description: values.description, dataSource: values.dataSource, dataSourceType: values.dataSourceType, tableName: (values === null || values === void 0 ? void 0 : values.tableName) || tableName, moduleId: values.moduleId, module: values.module, isSystem: values.isSystem ? values.isSystem === true : '', enableSoftDelete: values.enableSoftDelete === true ? true : '', enableAuditTracking: values.enableAuditTracking === true ? true : '', internationalisation: values.internationalisation === true ? true : '', draftPublishWorkflow: values.draftPublishWorkflow === true ? true : '', isChild: values.isChild === true ? true : '' }), (values.isChild == true && {
                        parentModelId: values.parentModelId,
                        parentModel: values.parentModel,
                    })), { isLegacyTable: values.isLegacyTable === true ? true : false, isLegacyTableWithId: values.isLegacyTableWithId === true ? true : false });
                    setModelMetaData(modelData_1);
                    nextTab();
                }
                catch (err) {
                    console.error(ERROR_MESSAGES.CREATE_MODEL, err);
                }
                return [2 /*return*/];
            });
        }); },
    });
    function generateTableName(moduleSlug, modelName) {
        var snakeCaseModelName = snakeCase(modelName);
        var sankeCaseModuleName = snakeCase(moduleSlug);
        return "".concat(sankeCaseModuleName, "_").concat(snakeCaseModelName);
    }
    var showError = function () { return __awaiter(void 0, void 0, void 0, function () {
        var errors, errorMessages;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, formik.validateForm()];
                case 1:
                    errors = _b.sent();
                    errorMessages = Object.values(errors);
                    if (errorMessages.length > 0) {
                        (_a = toast === null || toast === void 0 ? void 0 : toast.current) === null || _a === void 0 ? void 0 : _a.show({
                            severity: "error",
                            summary: ERROR_MESSAGES.SEND_REPORT,
                            // sticky: true,
                            life: 3000,
                            //@ts-ignore
                            content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: errorMessages.map(function (m, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: String(m) }) }, index)); }) })); },
                        });
                    }
                    return [2 /*return*/];
            }
        });
    }); };
    var searchModule = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var query, queryData, queryString, result, updatedSuggestion, error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    query = event.query;
                    queryData = {
                        limit: 10,
                        offset: 0,
                        filters: {
                            name: {
                                $containsi: query,
                            },
                        },
                    };
                    queryString = qs.stringify(queryData, {
                        encodeValuesOnly: true,
                    });
                    return [4 /*yield*/, triggerGetModules(queryString).unwrap()];
                case 1:
                    result = _a.sent();
                    if (result && result.records) {
                        updatedSuggestion = __spreadArray([], result.records, true);
                        return [2 /*return*/, updatedSuggestion];
                    }
                    else {
                        return [2 /*return*/, []];
                    }
                    return [3 /*break*/, 3];
                case 2:
                    error_1 = _a.sent();
                    return [2 /*return*/, []];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    var searchModel = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var query, queryData, queryString, result, updatedSuggestion, error_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    query = event.query;
                    queryData = {
                        limit: 10,
                        offset: 0,
                        filters: {
                            singularName: {
                                $containsi: query,
                            },
                        },
                    };
                    // Add module filter dynamically
                    if (selectedModule === null || selectedModule === void 0 ? void 0 : selectedModule.name) {
                        queryData.filters["module"] = {
                            name: { $containsi: selectedModule.name },
                        };
                    }
                    queryString = qs.stringify(queryData, {
                        encodeValuesOnly: true,
                    });
                    return [4 /*yield*/, triggerGetModels(queryString).unwrap()];
                case 1:
                    result = _a.sent();
                    if (result && result.records) {
                        updatedSuggestion = __spreadArray([], result.records, true);
                        return [2 /*return*/, updatedSuggestion];
                    }
                    else {
                        return [2 /*return*/, []];
                    }
                    return [3 /*break*/, 3];
                case 2:
                    error_2 = _a.sent();
                    return [2 /*return*/, []];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    var serachDataSource = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var query, filterredData, transformedData;
        return __generator(this, function (_a) {
            query = event.query;
            try {
                filterredData = fieldDefaultMetaData.data.dataSource.filter(function (t) { return t.name.toLowerCase().startsWith(query.toLowerCase()); });
                transformedData = filterredData.map(function (e) { return ({ label: e.name, value: e.name, type: e.type }); });
                return [2 /*return*/, transformedData];
            }
            catch (error) {
                console.error(ERROR_MESSAGES.FETCHING_ITEMS, error);
                return [2 /*return*/, []];
            }
            return [2 /*return*/];
        });
    }); };
    var serachDataSourceType = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var query, filterredData, transformedData;
        return __generator(this, function (_a) {
            query = event.query;
            try {
                filterredData = fieldDefaultMetaData.data.dataSource.filter(function (t) { return t.type.toLowerCase().startsWith(query.toLowerCase()); });
                transformedData = filterredData.map(function (e) { return ({ label: e.type, value: e.type }); });
                return [2 /*return*/, transformedData];
            }
            catch (error) {
                console.error(ERROR_MESSAGES.FETCHING_ITEMS, error);
                return [2 /*return*/, []];
            }
            return [2 /*return*/];
        });
    }); };
    // useEffect(() => {
    //   if (modelMetaData) {
    //     setModelMetaData(modelMetaData.parentCategoryId);
    //     // formik.setFieldValue("parentCategoryId", modelMetaData.parentCategoryId);
    //   }
    // }, [modelMetaData])
    // Set the formik reference to the formik instance
    // Set the formik reference to the formik instance
    useEffect(function () {
        if (formikModelMetadataRef) {
            formikModelMetadataRef.current = formik; // Assign the formik instance to the ref
        }
    }, [formik, formikModelMetadataRef]);
    var renderCount = useRef(0);
    useEffect(function () {
        renderCount.current += 1; // Increment count on every render
    });
    useEffect(function () {
        formik.validateForm();
        setModelMetaData(formik.values);
    }, [formik.values]);
    useEffect(function () {
        if (formik.dirty) {
            setIsDirty(true);
        }
    }, [formik.dirty]);
    return (_jsxs(_Fragment, { children: [_jsx(Toast, { ref: toast }), _jsx("form", { onSubmit: formik.handleSubmit, children: _jsx("div", { className: "", children: _jsxs("div", { className: "grid formgrid", children: [_jsxs("div", { className: "field col-12 lg:col-6 lg:pr-3", children: [_jsxs(Panel, { header: "Basic Info", className: "solid-column-panel", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { htmlFor: "type", className: "form-field-label", children: "Module" }), _jsx(SingleSelectAutoCompleteField, { disabled: params.id !== 'new', formik: formik, isFormFieldValid: isFormFieldValid, relationField: true, fieldName: "module", fieldNameId: "moduleId", labelKey: "displayName", valueKey: "id", searchData: searchModule, existingData: formik.values.module, formErrors: formErrors, additionalAction: function (e) { return setSelectedModule(e.value); } }, "module"), (isFormFieldValid(formik, "module") || (formErrors["module"])) && (_jsx(Message, { severity: "error", text: (_e = (_d = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _d === void 0 ? void 0 : _d.moduleId) === null || _e === void 0 ? void 0 : _e.toString() }))] }), _jsxs("div", { className: "flex flex-column gap-2 mt-3 lg:mt-4", children: [_jsx("label", { htmlFor: "dataSource", className: "form-field-label", children: "Data Source" }), _jsx(SingleSelectAutoCompleteField, { disabled: params.id !== 'new', formik: formik, isFormFieldValid: isFormFieldValid, 
                                                        // relationField={false}
                                                        fieldName: "dataSource", fieldNameId: null, labelKey: "label", valueKey: "value", searchData: serachDataSource, existingData: formik.values.dataSource, additionalAction: function (e) { return formik.setFieldValue("dataSourceType", e.target.value.type); }, formErrors: formErrors }, "dataSource"), formik.values.dataSourceType && _jsxs("p", { className: "form-field-label text-sm", children: ["Your datasource will connect to a ", formik.values.dataSourceType, " database"] }), (isFormFieldValid(formik, "dataSource") || (formErrors["dataSource"])) && (_jsx(Message, { severity: "error", text: (_g = (_f = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _f === void 0 ? void 0 : _f.dataSource) === null || _g === void 0 ? void 0 : _g.toString() }))] })] }), _jsxs(Panel, { header: "Configurations", className: "solid-column-panel mt-3 mb-3 lg:mt-5", children: [_jsxs("div", { className: "flex align-items-center gap-2 mt-3", children: [_jsx(Checkbox, { name: "enableSoftDelete", onChange: function (e) {
                                                            formik.setFieldValue("enableSoftDelete", e.checked);
                                                        }, disabled: params.id !== 'new', checked: formik.values.enableSoftDelete }), _jsx("label", { htmlFor: "enableSoftDelete", className: "form-field-label", children: "Enable Soft Delete" })] }), params.id !== 'new' && _jsx("p", { className: "form-field-label mt-1 text-sm", children: "Soft-delete can only be set during initial creation to maintain data integrity" }), (isFormFieldValid(formik, "enableSoftDelete") || (formErrors["enableSoftDelete"])) && (_jsx(Message, { severity: "error", text: (_j = (_h = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _h === void 0 ? void 0 : _h.enableSoftDelete) === null || _j === void 0 ? void 0 : _j.toString(), className: "mt-2" })), _jsxs("div", { className: "flex align-items-center gap-2 mt-3", children: [_jsx(Checkbox, { name: "isChild", onChange: function (e) {
                                                            formik.setFieldValue("isChild", e.checked);
                                                        }, disabled: params.id !== 'new', checked: formik.values.isChild }), _jsx("label", { htmlFor: "isChild", className: "form-field-label", children: "Is Child" })] }), params.id !== 'new' && _jsx("p", { className: "form-field-label mt-1 text-sm", children: "Is Current Model child of another Model" }), (isFormFieldValid(formik, "isChild") || (formErrors["isChild"])) && (_jsx(Message, { severity: "error", text: (_l = (_k = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _k === void 0 ? void 0 : _k.isChild) === null || _l === void 0 ? void 0 : _l.toString(), className: "mt-2" })), showParentModel &&
                                                _jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { htmlFor: "type", className: "form-field-label", children: "Parent Model" }), _jsx(SingleSelectAutoCompleteField, { disabled: params.id !== 'new', formik: formik, isFormFieldValid: isFormFieldValid, relationField: true, fieldName: "parentModel", fieldNameId: "parentModelId", labelKey: "displayName", valueKey: "id", searchData: searchModel, existingData: formik.values.parentModel, formErrors: formErrors }, "parentModel"), (isFormFieldValid(formik, "parentModel") || (formErrors["parentModel"])) && (_jsx(Message, { severity: "error", text: (_o = (_m = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _m === void 0 ? void 0 : _m.parentModelId) === null || _o === void 0 ? void 0 : _o.toString() }))] }), _jsxs("div", { className: "flex align-items-center gap-2 mt-3", children: [_jsx(Checkbox, { name: "enableAuditTracking", onChange: function (e) {
                                                            formik.setFieldValue("enableAuditTracking", e.checked);
                                                        }, checked: formik.values.enableAuditTracking }), _jsx("label", { htmlFor: "enableAuditTracking", className: "form-field-label", children: "Enable Audit Tracking" })] }), (isFormFieldValid(formik, "enableAuditTracking") || (formErrors["enableAuditTracking"])) && (_jsx(Message, { severity: "error", text: (_q = (_p = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _p === void 0 ? void 0 : _p.enableAuditTracking) === null || _q === void 0 ? void 0 : _q.toString(), className: "mt-2" })), _jsxs("div", { className: "flex align-items-center gap-2 mt-3", children: [_jsx(Checkbox, { name: "internationalisation", onChange: function (e) {
                                                            formik.setFieldValue("internationalisation", e.checked);
                                                        }, checked: formik.values.internationalisation }), _jsx("label", { htmlFor: "internationalisation", className: "form-field-label", children: "Is Internationalisation Enabled" })] }), _jsxs("div", { className: "flex align-items-center gap-2 mt-3", children: [_jsx(Checkbox, { name: "draftPublishWorkflow", onChange: function (e) {
                                                            formik.setFieldValue("draftPublishWorkflow", e.checked);
                                                        }, checked: formik.values.draftPublishWorkflow }), _jsx("label", { htmlFor: "draftPublishWorkflow", className: "form-field-label", children: "Draft/Publish Workflow" })] }), _jsxs("div", { className: "flex align-items-center gap-2 mt-3", children: [_jsx(Checkbox, { name: "isLegacyTable", onChange: function (e) {
                                                            formik.setFieldValue("isLegacyTable", e.checked);
                                                            // Reset isLegacyTableWithId when isLegacyTable is unchecked
                                                            if (!e.checked) {
                                                                formik.setFieldValue("isLegacyTableWithId", false);
                                                            }
                                                        }, checked: formik.values.isLegacyTable }), _jsx("label", { htmlFor: "isLegacyTable", className: "form-field-label", children: "Is Legacy Table" })] }), formik.values.isLegacyTable && (_jsxs("div", { className: "flex align-items-center gap-2 mt-2 ml-4", children: [_jsx(Checkbox, { name: "isLegacyTableWithId", onChange: function (e) {
                                                            formik.setFieldValue("isLegacyTableWithId", e.checked);
                                                        }, checked: formik.values.isLegacyTableWithId }), _jsx("label", { htmlFor: "isLegacyTableWithId", className: "form-field-label", children: "Has existing Id" })] })), formik.values.isLegacyTable && (_jsx("p", { className: "form-field-label mt-2 text-sm", children: "Note: Legacy tables require at least one field marked as Primary Key during model creation" }))] })] }), _jsx("div", { className: "field col-12 lg:col-6 lg:pl-3", children: _jsxs(Panel, { header: "Basic Settings", className: "solid-column-panel", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { htmlFor: "displayName", className: "form-field-label", children: "Display Name" }), _jsx(InputText, { type: "text", id: "displayName", name: "displayName", onChange: function (e) {
                                                        formik.handleChange(e);
                                                        var _a = getSingularAndPlural(e.target.value), toCamelCase = _a.toCamelCase, toSnakeCase = _a.toSnakeCase, toPluralCamelCase = _a.toPluralCamelCase;
                                                        if (params.id === 'new') {
                                                            formik.setFieldValue("singularName", toCamelCase);
                                                            formik.setFieldValue("pluralName", toPluralCamelCase);
                                                        }
                                                        if (params.id === 'new') {
                                                            formik.setFieldValue("tableName", toSnakeCase);
                                                        }
                                                    }, value: formik.values.displayName, className: classNames("", {
                                                        "p-invalid": isFormFieldValid(formik, "displayName") || formErrors["displayName"],
                                                    }) }), (isFormFieldValid(formik, "displayName") || (formErrors["displayName"])) && (_jsx(Message, { severity: "error", text: (_s = (_r = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _r === void 0 ? void 0 : _r.displayName) === null || _s === void 0 ? void 0 : _s.toString() }))] }), _jsxs("div", { className: "flex flex-column gap-1 mt-4", children: [_jsx("label", { htmlFor: "singularName", className: "form-field-label", children: "Singular Name" }), _jsx(InputText, { disabled: true, type: "text", id: "singularName", name: "singularName", onChange: formik.handleChange, value: formik.values.singularName, className: classNames("", {
                                                        "p-invalid": isFormFieldValid(formik, "singularName") || formErrors["singularName"],
                                                    }) }), (isFormFieldValid(formik, "singularName") || (formErrors["singularName"])) && (_jsx(Message, { severity: "error", text: (_u = (_t = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _t === void 0 ? void 0 : _t.singularName) === null || _u === void 0 ? void 0 : _u.toString() }))] }), _jsxs("div", { className: "flex flex-column gap-1 mt-4", children: [_jsx("label", { htmlFor: "pluralName", className: "form-field-label", children: "Plural Name" }), _jsx(InputText, { disabled: true, type: "text", id: "pluralName", name: "pluralName", onChange: formik.handleChange, value: formik.values.pluralName, className: classNames("", {
                                                        "p-invalid": isFormFieldValid(formik, "pluralName") || formErrors["pluralName"],
                                                    }) }), (isFormFieldValid(formik, "pluralName") || (formErrors["pluralName"])) && (_jsx(Message, { severity: "error", text: (_w = (_v = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _v === void 0 ? void 0 : _v.pluralName) === null || _w === void 0 ? void 0 : _w.toString() }))] }), _jsxs("div", { className: "flex flex-column gap-1 mt-4", children: [_jsx("label", { htmlFor: "tableName", className: "form-field-label", children: "Table Name" }), _jsx(InputText, { disabled: params.id !== 'new', type: "text", id: "tableName", name: "tableName", onChange: formik.handleChange, value: formik.values.tableName, className: classNames("", {
                                                        "p-invalid": isFormFieldValid(formik, "tableName") || formErrors["tableName"],
                                                    }) }), (isFormFieldValid(formik, "tableName") || (formErrors["tableName"])) && (_jsx(Message, { severity: "error", text: (_y = (_x = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _x === void 0 ? void 0 : _x.tableName) === null || _y === void 0 ? void 0 : _y.toString() }))] }), _jsxs("div", { className: "flex flex-column gap-1 mt-4", children: [_jsx("label", { htmlFor: "description", className: "form-field-label", children: "Description" }), _jsx(InputTextarea, { id: "description", name: "description", onChange: formik.handleChange, value: formik.values.description, className: classNames("", {
                                                        "p-invalid": isFormFieldValid(formik, "description") || formErrors["description"],
                                                    }), rows: 5, cols: 30 }), (isFormFieldValid(formik, "description") || (formErrors["description"])) && (_jsx(Message, { severity: "error", text: (_0 = (_z = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _z === void 0 ? void 0 : _z.description) === null || _0 === void 0 ? void 0 : _0.toString() }))] })] }) })] }) }) })] }));
});
export default ModelMetaData;
