// @solidxai/core-ui/dist/components/core/list/columns/SolidShortTextColumn.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { FilterMatchMode } from 'primereact/api';
import { Column } from "primereact/column";
import { getNumberOfInputs } from '../SolidListViewColumn';
import { InputTypes, SolidVarInputsFilterElement } from "../SolidVarInputsFilterElement";
import SolidTableRowCell from '../SolidTableRowCell';
import { getExtensionComponent } from '../../../../helpers/registry';
import dayjs from 'dayjs';
var SolidShortTextColumn = function (_a) {
    var _b, _c, _d, _e, _f;
    var solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox;
    var filterable = column.attrs.filterable;
    var showFilterOperator = false;
    var columnDataType = 'text';
    var filterMatchModeOptions = [
        { label: 'Starts With', value: FilterMatchMode.STARTS_WITH },
        { label: 'Contains', value: FilterMatchMode.CONTAINS },
        { label: 'Not Contains', value: FilterMatchMode.NOT_CONTAINS },
        { label: 'Ends With', value: FilterMatchMode.ENDS_WITH },
        { label: 'Equals', value: FilterMatchMode.EQUALS },
        { label: 'Not Equals', value: FilterMatchMode.NOT_EQUALS },
        { label: 'In', value: FilterMatchMode.IN },
        { label: 'Not In', value: FilterMatchMode.NOT_IN }
    ];
    var filterTemplate = function (options) {
        var numberOfInputs = getNumberOfInputs(options.filterModel.matchMode);
        return (_jsx(SolidVarInputsFilterElement, { values: options.value, onChange: function (e) { return options.filterCallback(e, options.index); }, numberOfInputs: numberOfInputs, inputType: InputTypes.Text, solidListViewMetaData: solidListViewMetaData, fieldMetadata: fieldMetadata, column: column }));
    };
    var truncateAfter = (_e = (_d = (_c = (_b = solidListViewMetaData === null || solidListViewMetaData === void 0 ? void 0 : solidListViewMetaData.data) === null || _b === void 0 ? void 0 : _b.solidView) === null || _c === void 0 ? void 0 : _c.layout) === null || _d === void 0 ? void 0 : _d.attrs) === null || _e === void 0 ? void 0 : _e.truncateAfter;
    var header = (_f = column.attrs.label) !== null && _f !== void 0 ? _f : fieldMetadata.displayName;
    return (_jsx(Column, { field: fieldMetadata.name, 
        // header={header}
        // className="text-sm"
        sortable: column.attrs.sortable, 
        // filter={filterable}
        dataType: columnDataType, showFilterOperator: showFilterOperator, filterMatchModeOptions: filterMatchModeOptions, filterElement: filterTemplate, filterPlaceholder: "Search by ".concat(fieldMetadata.displayName), 
        // style={{ minWidth: "12rem" }}
        // headerClassName="table-header-fs"
        header: function () {
            return (_jsx("div", { style: { maxWidth: truncateAfter ? "".concat(truncateAfter, "ch") : '30ch', whiteSpace: 'nowrap', textOverflow: 'ellipsis', overflow: 'hidden' }, children: header }));
        }, body: function (rowData) {
            //add this widget to render media image in image format  - SolidShortTextFieldImageRenderModeWidget 
            var viewWidget = column.attrs.viewWidget;
            if (!viewWidget) {
                viewWidget = 'DefaultTextListWidget';
            }
            var DynamicWidget = getExtensionComponent(viewWidget);
            var widgetProps = {
                rowData: rowData,
                solidListViewMetaData: solidListViewMetaData,
                fieldMetadata: fieldMetadata,
                column: column,
                setLightboxUrls: setLightboxUrls,
                setOpenLightbox: setOpenLightbox
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        } }, fieldMetadata.name));
};
export default SolidShortTextColumn;
function parseIsoDate(value) {
    if (!value || typeof value !== "string")
        return null;
    // Fast reject (avoids false positives)
    if (!value.includes("T"))
        return null;
    var d = new Date(value);
    return isNaN(d.getTime()) ? null : d;
}
export var DefaultTextListWidget = function (_a) {
    var _b, _c, _d, _e, _f;
    var rowData = _a.rowData, solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column;
    var truncateAfter = (_e = (_d = (_c = (_b = solidListViewMetaData === null || solidListViewMetaData === void 0 ? void 0 : solidListViewMetaData.data) === null || _b === void 0 ? void 0 : _b.solidView) === null || _c === void 0 ? void 0 : _c.layout) === null || _d === void 0 ? void 0 : _d.attrs) === null || _e === void 0 ? void 0 : _e.truncateAfter;
    var displayValue = rowData[fieldMetadata.name];
    //    if (typeof displayValue === "string" && displayValue.includes("T") && displayValue.endsWith("Z")) {
    //         const d = new Date(displayValue);
    //         if (!isNaN(d.getTime())) {
    //             const year = d.getFullYear();
    //             const month = String(d.getMonth() + 1).padStart(2, "0");
    //             const day = String(d.getDate()).padStart(2, "0");
    //             const hours = String(d.getHours()).padStart(2, "0");
    //             const minutes = String(d.getMinutes()).padStart(2, "0");
    //             const seconds = String(d.getSeconds()).padStart(2, "0");
    //             displayValue = `${year}-${month}-${day} ${hours}:${minutes}:${seconds}.000`;
    //         }
    //     }
    if (typeof displayValue === "string") {
        var parsedDate = parseIsoDate(displayValue);
        if (parsedDate) {
            var format = (_f = column === null || column === void 0 ? void 0 : column.attrs) === null || _f === void 0 ? void 0 : _f.format;
            // 🔵 Explicit format → dayjs
            if (format) {
                displayValue = dayjs(parsedDate).format(format);
            }
            // 🟢 Default → user locale
            else {
                displayValue = parsedDate.toLocaleString();
            }
        }
    }
    if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.selectionStaticValues) && displayValue) {
        var mapping_1 = {};
        fieldMetadata.selectionStaticValues.forEach(function (entry) {
            var _a = entry.split(":"), val = _a[0], label = _a[1];
            mapping_1[val] = label;
        });
        var values = displayValue.split(",").map(function (v) { return v.trim(); });
        displayValue = values.map(function (v) { return mapping_1[v] || v; }).join(", ");
    }
    return (_jsx(SolidTableRowCell, { value: displayValue, truncateAfter: truncateAfter }));
};
