// @solidxai/core-ui/dist/components/core/list/columns/SolidMediaMultipleColumn.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useState } from 'react';
import { Column } from "primereact/column";
import { Button } from 'primereact/button';
import { getExtensionComponent } from '../../../../helpers/registry';
import { classNames } from 'primereact/utils';
import { FileReaderExt } from '../../../../components/common/FileReaderExt';
import { Dialog } from 'primereact/dialog';
// Helpers
var isImageFile = function (url) { return /\.(jpg|jpeg|png|gif|bmp|webp)$/i.test(url); };
var isVideoFile = function (url) { return /\.(mp4|webm|ogg)$/i.test(url); };
var isAudioFile = function (url) { return /\.(mp3|wav|ogg)$/i.test(url); };
var downloadOnlyExt = [
    "txt", "zip", "rar",
    "doc", "docx",
    "xls", "xlsx",
    "ppt", "pptx",
    "pdf", "csv"
];
var isDocumentType = function (url) {
    var _a;
    var ext = (_a = url.split("?")[0].split(".").pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    return ext ? downloadOnlyExt.includes(ext) : false;
};
var downloadFile = function (url, name) {
    if (name === void 0) { name = ""; }
    var link = document.createElement("a");
    link.href = url;
    link.download = name;
    link.target = "_blank";
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
};
// Thumbnail preview component
var MediaPreview = function (_a) {
    var src = _a.src, onClick = _a.onClick;
    var _b = useState(false), isBroken = _b[0], setIsBroken = _b[1];
    var handleClick = function (event) {
        onClick(event);
    };
    if (!isBroken) {
        if (isImageFile(src)) {
            return (_jsx("img", { src: src, alt: "media", className: "shadow-2 border-round", width: 40, height: 40, style: { objectFit: "cover" }, onError: function () { return setIsBroken(true); }, onClick: handleClick }));
        }
        if (isVideoFile(src)) {
            return (_jsx("video", { src: src, width: 40, height: 40, className: "shadow-2 border-round", style: { objectFit: "cover" }, onError: function () { return setIsBroken(true); }, onClick: handleClick, muted: true }));
        }
        if (isAudioFile(src)) {
            return (_jsx("div", { className: "shadow-2 border-round flex align-items-center justify-content-center bg-gray-100", style: { width: 40, height: 40 }, onClick: handleClick, children: _jsx("i", { className: "pi pi-volume-up text-xl text-gray-600" }) }));
        }
    }
    return (_jsx("div", { style: { width: 40, height: 40, display: "flex", alignItems: "center", justifyContent: "center" }, onClick: handleClick, children: _jsx("i", { className: classNames("pi pi-file", "text-3xl text-gray-400") }) }));
};
// Column Component
var SolidMediaMultipleColumn = function (_a) {
    var _b;
    var solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox;
    var filterable = false;
    var showFilterOperator = false;
    var columnDataType = undefined;
    var header = (_b = column.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    return (_jsx(Column, { field: fieldMetadata.name, header: header, body: function (rowData) {
            var viewWidget = column.attrs.viewWidget;
            if (!viewWidget) {
                viewWidget = 'DefaultMediaMultipleListWidget';
            }
            var DynamicWidget = getExtensionComponent(viewWidget);
            var widgetProps = {
                rowData: rowData,
                solidListViewMetaData: solidListViewMetaData,
                fieldMetadata: fieldMetadata,
                column: column,
                setLightboxUrls: setLightboxUrls,
                setOpenLightbox: setOpenLightbox
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }, sortable: column.attrs.sortable, dataType: columnDataType, showFilterOperator: showFilterOperator, filterPlaceholder: "Search by ".concat(fieldMetadata.displayName), style: { minWidth: "12rem" }, headerClassName: "table-header-fs" }, fieldMetadata.name));
};
export default SolidMediaMultipleColumn;
// Default multiple widget
export var DefaultMediaMultipleListWidget = function (_a) {
    var _b, _c, _d;
    var rowData = _a.rowData, fieldMetadata = _a.fieldMetadata, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox;
    var _e = useState(false), isShowAllFiles = _e[0], setShowAllFiles = _e[1];
    if (!((_b = rowData === null || rowData === void 0 ? void 0 : rowData._media) === null || _b === void 0 ? void 0 : _b[fieldMetadata.name]))
        return null;
    var fullrecord = (_c = rowData._media[fieldMetadata.name]) === null || _c === void 0 ? void 0 : _c.map(function (file) { return ({
        name: file.originalFileName,
        type: file.mimeType,
        size: file.fileSize,
        id: file.id,
        fileUrl: file._full_url
    }); });
    var formatFileSize = function (size) {
        return size >= 1024 * 1024 ? "".concat((size / (1024 * 1024)).toFixed(1), " MB") : "".concat((size / 1024).toFixed(1), " KB");
    };
    var handleFileView = function (file) {
        var _a;
        var fileUrl = (file === null || file === void 0 ? void 0 : file.fileUrl) || "";
        var cleanUrl = fileUrl.split("?")[0];
        var ext = (_a = cleanUrl.split(".").pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
        if (ext && downloadOnlyExt.includes(ext)) {
            downloadFile(file === null || file === void 0 ? void 0 : file.fileUrl, "");
        }
        else {
            setLightboxUrls === null || setLightboxUrls === void 0 ? void 0 : setLightboxUrls([{ src: file.fileUrl, downloadUrl: file.fileUrl }]);
            setOpenLightbox === null || setOpenLightbox === void 0 ? void 0 : setOpenLightbox(true);
        }
    };
    return fullrecord.length > 0 ? (_jsxs("div", { className: 'flex gap-2 align-items-end', children: [_jsx(MediaPreview, { src: (_d = fullrecord[0]) === null || _d === void 0 ? void 0 : _d.fileUrl, onClick: function (event) {
                    var _a, _b, _c;
                    event.stopPropagation();
                    var cleanUrl = (_a = fullrecord[0]) === null || _a === void 0 ? void 0 : _a.fileUrl.split("?")[0];
                    var ext = (_b = cleanUrl.split(".").pop()) === null || _b === void 0 ? void 0 : _b.toLowerCase();
                    if (ext && downloadOnlyExt.includes(ext) && (fullrecord === null || fullrecord === void 0 ? void 0 : fullrecord.length) > 1) {
                        setShowAllFiles(true);
                        return;
                    }
                    else if (ext && downloadOnlyExt.includes(ext)) {
                        downloadFile((_c = fullrecord[0]) === null || _c === void 0 ? void 0 : _c.fileUrl, "");
                        return;
                    }
                    // FIRST FILE IS MEDIA ⇒ OPEN LIGHTBOX
                    var urlsWithType = fullrecord.map(function (file) { return ({
                        src: file.fileUrl,
                        downloadUrl: file.fileUrl,
                    }); });
                    setLightboxUrls(urlsWithType);
                    setOpenLightbox(true);
                } }), (fullrecord === null || fullrecord === void 0 ? void 0 : fullrecord.length) > 1 && _jsxs("span", { style: {
                    color: "#0895CD",
                    fontWeight: "bold",
                    cursor: "pointer",
                    marginLeft: "4px"
                }, onClick: function (event) {
                    var _a;
                    event.stopPropagation();
                    if (isDocumentType((_a = fullrecord[0]) === null || _a === void 0 ? void 0 : _a.fileUrl)) {
                        setShowAllFiles(true);
                    }
                    else {
                        var urlsWithType = fullrecord.map(function (file) { return ({
                            src: file.fileUrl,
                            downloadUrl: file.fileUrl
                        }); });
                        setLightboxUrls(urlsWithType);
                        setOpenLightbox(true);
                    }
                }, children: ["+", fullrecord.length - 1] }), _jsx(Dialog, { visible: isShowAllFiles, header: "Items Uploaded", modal: true, onHide: function () { return setShowAllFiles(false); }, 
                // style={{ minWidth: 450 }}
                style: { width: '32rem' }, onClick: function (event) { return event.stopPropagation(); }, breakpoints: { '591px': '94vw' }, children: fullrecord === null || fullrecord === void 0 ? void 0 : fullrecord.map(function (file) {
                    var fileId = "".concat(file.name, "-").concat(file.size);
                    return (_jsx("div", { className: "solid-file-upload-wrapper mb-3", children: _jsxs("div", { className: "flex align-items-center gap-2", children: [_jsx(FileReaderExt, { fileDetails: file }), _jsxs("div", { className: "w-full flex flex-column gap-1", children: [_jsxs("div", { className: "flex align-items-center justify-content-between", children: [_jsx("p", { className: "font-normal w-11 text-primary m-0", style: { cursor: "pointer" }, onClick: function () { return handleFileView(file); }, children: file === null || file === void 0 ? void 0 : file.name }), _jsx(Button, { type: "button", text: true, icon: "pi pi-download", size: "small", onClick: function () {
                                                        downloadFile(file === null || file === void 0 ? void 0 : file.fileUrl, "");
                                                    }, style: { height: 16, width: 16 } })] }), _jsx("div", { className: "text-sm", children: formatFileSize(file.size) })] })] }) }, fileId));
                }) })] })) : (_jsx("div", { style: { height: 40, width: 40 } }));
};
