// @solidxai/core-ui/dist/components/core/list/columns/SolidBooleanColumn.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { Column } from "primereact/column";
import { Dropdown } from 'primereact/dropdown';
import { getExtensionComponent } from "../../../../helpers/registry";
var SolidBooleanColumn = function (_a) {
    var _b;
    var solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column;
    var filterable = column.attrs.filterable;
    var showFilterOperator = false;
    var columnDataType = 'boolean';
    var booleanOptions = [
        { label: 'Yes', value: true },
        { label: 'No', value: false },
    ];
    // Custom filter template for Boolean column
    var filterTemplate = function (options) {
        return (_jsx(Dropdown, { value: options.value, options: booleanOptions, onChange: function (e) { return options.filterCallback(e.value); }, placeholder: "Select", className: "p-column-filter" }));
    };
    // Custom body template for Boolean column
    var bodyTemplate = function (rowData) {
        var _a;
        var fieldKey = (_a = column.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.name;
        return rowData[fieldKey] ? 'Yes' : 'No';
    };
    var header = (_b = column.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    // TODO: filterTemplate, bodyTemplate & filterMatchModeOptions to be provided.
    return (_jsx(Column, { field: fieldMetadata.name, header: header, 
        // className="text-sm"
        sortable: column.attrs.sortable, 
        // filter={filterable}
        dataType: columnDataType, showFilterOperator: showFilterOperator, 
        // filterMatchModeOptions={filterMatchModeOptions}
        filterElement: filterTemplate, body: function (rowData) {
            var viewWidget = column.attrs.viewWidget;
            if (!viewWidget) {
                viewWidget = 'DefaultBooleanListWidget';
            }
            var DynamicWidget = getExtensionComponent(viewWidget);
            var widgetProps = {
                rowData: rowData,
                solidListViewMetaData: solidListViewMetaData,
                fieldMetadata: fieldMetadata,
                column: column
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }, filterPlaceholder: "Search by ".concat(fieldMetadata.displayName), style: { minWidth: "12rem" }, headerClassName: "table-header-fs" }, fieldMetadata.name));
};
export default SolidBooleanColumn;
export var DefaultBooleanListWidget = function (_a) {
    var _b, _c, _d;
    var rowData = _a.rowData, solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column;
    var fieldKey = (_b = column.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.name;
    var trueLabel = (_c = column.attrs.trueLabel) !== null && _c !== void 0 ? _c : 'Yes';
    var falseLabel = (_d = column.attrs.falseLabel) !== null && _d !== void 0 ? _d : 'No';
    console.log("rendering boolean column ".concat(fieldKey, " with value ").concat(rowData[fieldKey], ", trueLabel=").concat(trueLabel, ", falseLabel=").concat(falseLabel));
    return rowData[fieldKey] === true ? trueLabel : falseLabel;
};
