// @solidxai/core-ui/dist/components/core/list/columns/SolidSelectionDynamicColumn.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { FilterMatchMode } from 'primereact/api';
import { Column } from "primereact/column";
import { InputTypes, SolidVarInputsFilterElement } from "../SolidVarInputsFilterElement";
import { getExtensionComponent } from '../../../../helpers/registry';
var SolidSelectionDynamicColumn = function (_a) {
    var _b, _c, _d, _e, _f;
    var solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column;
    var filterable = column.attrs.filterable;
    var showFilterOperator = false;
    var columnDataType = fieldMetadata.selectionValueType === 'int' ? 'numeric' : 'text';
    var filterMatchModeOptions = [
        { label: 'In', value: FilterMatchMode.IN },
        { label: 'Not In', value: FilterMatchMode.NOT_IN },
    ];
    var filterTemplate = function (options) {
        return (_jsx(SolidVarInputsFilterElement, { values: options.value, onChange: function (e) { return options.filterCallback(e, options.index); }, inputType: InputTypes.SelectionDynamic, solidListViewMetaData: solidListViewMetaData, fieldMetadata: fieldMetadata, column: column }));
    };
    var truncateAfter = (_e = (_d = (_c = (_b = solidListViewMetaData === null || solidListViewMetaData === void 0 ? void 0 : solidListViewMetaData.data) === null || _b === void 0 ? void 0 : _b.solidView) === null || _c === void 0 ? void 0 : _c.layout) === null || _d === void 0 ? void 0 : _d.attrs) === null || _e === void 0 ? void 0 : _e.truncateAfter;
    var header = (_f = column.attrs.label) !== null && _f !== void 0 ? _f : fieldMetadata.displayName;
    return (_jsx(Column, { field: fieldMetadata.name, 
        // header={header}
        // className="text-sm"
        sortable: column.attrs.sortable, 
        // filter={filterable}
        dataType: columnDataType, showFilterOperator: showFilterOperator, filterMatchModeOptions: filterMatchModeOptions, filterElement: filterTemplate, filterPlaceholder: "Search by ".concat(fieldMetadata.displayName), 
        // style={{ minWidth: "12rem" }}
        // headerClassName="table-header-fs"
        header: function () {
            return (_jsx("div", { style: { maxWidth: truncateAfter ? "".concat(truncateAfter, "ch") : '30ch', whiteSpace: 'nowrap', textOverflow: 'ellipsis', overflow: 'hidden' }, children: header }));
        }, body: function (rowData) {
            var viewWidget = column.attrs.viewWidget;
            if (!viewWidget) {
                viewWidget = 'DefaultTextListWidget';
            }
            var DynamicWidget = getExtensionComponent(viewWidget);
            var widgetProps = {
                rowData: rowData,
                solidListViewMetaData: solidListViewMetaData,
                fieldMetadata: fieldMetadata,
                column: column
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        } }, fieldMetadata.name));
};
export default SolidSelectionDynamicColumn;
