// @solidxai/core-ui/dist/components/core/list/columns/SolidMediaSingleColumn.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment } from "react/jsx-runtime";
import { useState } from 'react';
import { Column } from "primereact/column";
import { classNames } from 'primereact/utils';
import { getExtensionComponent } from '../../../../helpers/registry';
// Helpers for file type detection
var isImageFile = function (url) { return /\.(jpg|jpeg|png|gif|bmp|webp)$/i.test(url); };
var isVideoFile = function (url) { return /\.(mp4|webm|ogg)$/i.test(url); };
var isAudioFile = function (url) { return /\.(mp3|wav|ogg)$/i.test(url); };
// Extensions that should be downloaded directly
var downloadOnlyExt = [
    "txt", "zip", "rar",
    "doc", "docx",
    "xls", "xlsx",
    "ppt", "pptx",
    "pdf"
];
// Media component with fallback for broken links
var MediaWithFallback = function (_a) {
    var src = _a.src, alt = _a.alt, onClick = _a.onClick;
    var _b = useState(false), isBroken = _b[0], setIsBroken = _b[1];
    var handleClick = function (event) {
        var _a;
        event.stopPropagation();
        var cleanUrl = src.split("?")[0];
        var ext = (_a = cleanUrl.split(".").pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
        if (ext && downloadOnlyExt.includes(ext)) {
            // Trigger download for docs/archives
            var link = document.createElement("a");
            link.href = src;
            link.download = "";
            link.target = "_blank";
            document.body.appendChild(link);
            link.click();
            document.body.removeChild(link);
        }
        else {
            onClick(event);
        }
    };
    if (!isBroken) {
        if (isImageFile(src)) {
            return (_jsx("img", { src: src, alt: alt, className: "shadow-2 border-round", width: 40, height: 40, style: { objectFit: "cover", cursor: "pointer" }, onError: function () { return setIsBroken(true); }, onClick: handleClick }));
        }
        if (isVideoFile(src)) {
            return (_jsx("video", { src: src, width: 40, height: 40, className: "shadow-2 border-round", style: { objectFit: "cover", cursor: "pointer" }, onError: function () { return setIsBroken(true); }, onClick: handleClick, muted: true }));
        }
        if (isAudioFile(src)) {
            return (_jsx("div", { className: "shadow-2 border-round flex align-items-center justify-content-center bg-gray-100", style: { width: 40, height: 40, cursor: "pointer" }, onClick: handleClick, children: _jsx("i", { className: "pi pi-volume-up text-xl text-gray-600" }) }));
        }
    }
    // fallback icon (docs/others)
    return (_jsx("div", { style: { width: 40, height: 40, display: "flex", alignItems: "center", justifyContent: "center", cursor: "pointer" }, onClick: handleClick, children: _jsx("i", { className: classNames("pi pi-file", "text-3xl text-gray-400") }) }));
};
// Main column renderer
var SolidMediaSingleColumn = function (_a) {
    var _b;
    var solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox;
    var header = (_b = column.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    return (_jsx(Column, { field: fieldMetadata.name, header: header, body: function (rowData) {
            var viewWidget = column.attrs.viewWidget;
            if (!viewWidget) {
                viewWidget = 'DefaultMediaSingleListWidget';
            }
            var DynamicWidget = getExtensionComponent(viewWidget);
            var widgetProps = {
                rowData: rowData,
                solidListViewMetaData: solidListViewMetaData,
                fieldMetadata: fieldMetadata,
                column: column,
                setLightboxUrls: setLightboxUrls,
                setOpenLightbox: setOpenLightbox
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }, sortable: column.attrs.sortable, showFilterOperator: false, filterPlaceholder: "Search by ".concat(fieldMetadata.displayName), style: { minWidth: "12rem" }, headerClassName: "table-header-fs" }, fieldMetadata.name));
};
export default SolidMediaSingleColumn;
// Default widget for single media field
export var DefaultMediaSingleListWidget = function (_a) {
    var _b;
    var rowData = _a.rowData, solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox;
    if (!((_b = rowData === null || rowData === void 0 ? void 0 : rowData._media) === null || _b === void 0 ? void 0 : _b[fieldMetadata.name]))
        return null;
    var mediaUrls = rowData._media[fieldMetadata.name].map(function (i) { return i._full_url; });
    var firstUrl = mediaUrls[0];
    if (!firstUrl)
        return _jsx("div", { style: { height: 40, width: 40 } });
    return (_jsx(MediaWithFallback, { src: firstUrl, alt: "media", onClick: function (event) {
            // Only open lightbox for image, video, or audio
            event.stopPropagation();
            if (isImageFile(firstUrl) || isVideoFile(firstUrl) || isAudioFile(firstUrl)) {
                setLightboxUrls([{ src: firstUrl, downloadUrl: firstUrl }]);
                setOpenLightbox(true);
            }
        } }));
};
