// @solidxai/core-ui/dist/components/core/common/SolidAccountSettings/SolidPersonalInfo.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useDeleteMediaMutation } from "../../../../redux/api/mediaApi";
import { useGetUserQuery, useUpdateUserProfileMutation } from "../../../../redux/api/userApi";
import { useFormik } from "formik";
import { AutoComplete } from "primereact/autocomplete";
import { Avatar } from "primereact/avatar";
import { Button } from "primereact/button";
import { Dialog } from "primereact/dialog";
import { InputText } from "primereact/inputtext";
import { Toast } from "primereact/toast";
import { useRef, useState } from "react";
import styles from './SolidAccountSettings.module.css';
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
import showToast from "../../../../helpers/showToast";
import { useSession } from '../../../../hooks/useSession';
export var SolidPersonalInfo = function () {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
    var fileInputRef = useRef(null);
    var toast = useRef(null);
    var _x = useState(null), previewImage = _x[0], setPreviewImage = _x[1];
    var _y = useState(false), deleteDialogVisible = _y[0], setDeleteDialogVisible = _y[1];
    var _z = useState(false), replaceDialogVisible = _z[0], setReplaceDialogVisible = _z[1];
    var _0 = useState(null), pendingFile = _0[0], setPendingFile = _0[1];
    var _1 = useSession(), session = _1.data, status = _1.status;
    var user = session === null || session === void 0 ? void 0 : session.user;
    var userId = user === null || user === void 0 ? void 0 : user.id;
    var _2 = useGetUserQuery(userId), userData = _2.data, refetch = _2.refetch;
    var deleteMedia = useDeleteMediaMutation()[0];
    var updateUser = useUpdateUserProfileMutation()[0];
    var initialValues = {
        fullName: (_b = (_a = userData === null || userData === void 0 ? void 0 : userData.data) === null || _a === void 0 ? void 0 : _a.fullName) !== null && _b !== void 0 ? _b : "",
        profilePicture: (_f = (_e = (_d = (_c = userData === null || userData === void 0 ? void 0 : userData.data) === null || _c === void 0 ? void 0 : _c._media) === null || _d === void 0 ? void 0 : _d.profilePicture[0]) === null || _e === void 0 ? void 0 : _e._full_url) !== null && _f !== void 0 ? _f : null,
    };
    var formik = useFormik({
        initialValues: initialValues,
        enableReinitialize: true,
        onSubmit: function (values) { return __awaiter(void 0, void 0, void 0, function () {
            var formData, response, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        formData = new FormData();
                        if (values.fullName !== initialValues.fullName) {
                            formData.append("fullName", values.fullName);
                        }
                        if (values.profilePicture && values.profilePicture instanceof File) {
                            formData.append("profilePicture", values.profilePicture);
                        }
                        if (!formData.has("fullName") && !formData.has("profilePicture")) {
                            showToast(toast, "info", ERROR_MESSAGES.NO_CHANGE, ERROR_MESSAGES.NO_UPDATE_MADE);
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, updateUser({ data: formData }).unwrap()];
                    case 1:
                        response = _a.sent();
                        if ((response === null || response === void 0 ? void 0 : response.statusCode) === 200) {
                            showToast(toast, "success", ERROR_MESSAGES.PROFILE_SAVED, ERROR_MESSAGES.PROFILE_SAVED_SUCCESSFULLY);
                            refetch();
                            formik.resetForm();
                            setPreviewImage(null);
                        }
                        else {
                            showToast(toast, "error", ERROR_MESSAGES.FAILED, ERROR_MESSAGES.FAILED_UPDATED_PROFILE);
                        }
                        return [3 /*break*/, 3];
                    case 2:
                        error_1 = _a.sent();
                        showToast(toast, "error", ERROR_MESSAGES.FAILED, ERROR_MESSAGES.SOMETHING_WRONG);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        }); },
    });
    var getInitials = function (value) {
        var _a;
        if (!value)
            return "";
        var email = value.includes('@') ? value.split('@')[0] : value;
        return ((_a = email[0]) === null || _a === void 0 ? void 0 : _a.toUpperCase()) || "";
    };
    var getColorFromInitials = function (initials) {
        var hash = 0;
        for (var i = 0; i < initials.length; i++) {
            hash = initials.charCodeAt(i) + ((hash << 5) - hash);
        }
        var hue = Math.abs(hash) % 360;
        return "hsl(".concat(hue, ", 60%, 60%)");
    };
    var value = (_g = userData === null || userData === void 0 ? void 0 : userData.data) === null || _g === void 0 ? void 0 : _g.email;
    var initials = getInitials(value);
    var bgColor = getColorFromInitials(initials);
    var handleFileChange = function (file) {
        var _a, _b, _c;
        var existing = (_c = (_b = (_a = userData === null || userData === void 0 ? void 0 : userData.data) === null || _a === void 0 ? void 0 : _a._media) === null || _b === void 0 ? void 0 : _b.profilePicture) === null || _c === void 0 ? void 0 : _c[0];
        if (existing) {
            setPendingFile(file);
            setReplaceDialogVisible(true);
        }
        else {
            formik.setFieldValue("profilePicture", file);
            setPreviewImage(URL.createObjectURL(file));
        }
    };
    var handleDeleteAvatar = function () {
        var _a, _b, _c;
        var existing = (_c = (_b = (_a = userData === null || userData === void 0 ? void 0 : userData.data) === null || _a === void 0 ? void 0 : _a._media) === null || _b === void 0 ? void 0 : _b.profilePicture) === null || _c === void 0 ? void 0 : _c[0];
        if (existing) {
            setDeleteDialogVisible(true);
        }
        else {
            formik.setFieldValue("profilePicture", null);
            setPreviewImage(null);
        }
    };
    var confirmReplace = function () { return __awaiter(void 0, void 0, void 0, function () {
        var existing, error_2;
        var _a, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    existing = (_c = (_b = (_a = userData === null || userData === void 0 ? void 0 : userData.data) === null || _a === void 0 ? void 0 : _a._media) === null || _b === void 0 ? void 0 : _b.profilePicture) === null || _c === void 0 ? void 0 : _c[0];
                    _d.label = 1;
                case 1:
                    _d.trys.push([1, 4, , 5]);
                    if (!(existing === null || existing === void 0 ? void 0 : existing.id)) return [3 /*break*/, 3];
                    return [4 /*yield*/, deleteMedia(existing.id).unwrap()];
                case 2:
                    _d.sent();
                    _d.label = 3;
                case 3:
                    if (pendingFile) {
                        formik.setFieldValue("profilePicture", pendingFile);
                        setTimeout(function () {
                            formik.submitForm();
                        }, 0);
                    }
                    return [3 /*break*/, 5];
                case 4:
                    error_2 = _d.sent();
                    showToast(toast, "error", ERROR_MESSAGES.FAILED, ERROR_MESSAGES.FAILED_DELETED_IMAGE);
                    return [3 /*break*/, 5];
                case 5:
                    setReplaceDialogVisible(false);
                    setPendingFile(null);
                    return [2 /*return*/];
            }
        });
    }); };
    var confirmDelete = function () { return __awaiter(void 0, void 0, void 0, function () {
        var existing, _a;
        var _b, _c, _d;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    existing = (_d = (_c = (_b = userData === null || userData === void 0 ? void 0 : userData.data) === null || _b === void 0 ? void 0 : _b._media) === null || _c === void 0 ? void 0 : _c.profilePicture) === null || _d === void 0 ? void 0 : _d[0];
                    if (!(existing === null || existing === void 0 ? void 0 : existing.id)) return [3 /*break*/, 4];
                    _e.label = 1;
                case 1:
                    _e.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, deleteMedia(existing.id).unwrap()];
                case 2:
                    _e.sent();
                    showToast(toast, "success", ERROR_MESSAGES.DELETED, ERROR_MESSAGES.PROFILE_PICTURE_REMOVE);
                    refetch();
                    return [3 /*break*/, 4];
                case 3:
                    _a = _e.sent();
                    showToast(toast, "error", ERROR_MESSAGES.ERROR, ERROR_MESSAGES.FAILED_DELETED_IMAGE);
                    return [3 /*break*/, 4];
                case 4:
                    formik.setFieldValue("profilePicture", null);
                    setPreviewImage(null);
                    setDeleteDialogVisible(false);
                    return [2 /*return*/];
            }
        });
    }); };
    return (_jsxs("form", { onSubmit: formik.handleSubmit, className: "h-full flex flex-column justify-content-between", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { children: [_jsxs("div", { children: [_jsx("label", { className: "form-field-label mb-2 font-bold", children: "Profile Picture" }), _jsxs("div", { className: "flex align-items-center gap-3", children: [previewImage ? (_jsx(Avatar, { image: previewImage, shape: "circle", style: {
                                            height: '5rem',
                                            width: '5rem'
                                        } })) : ((_l = (_k = (_j = (_h = userData === null || userData === void 0 ? void 0 : userData.data) === null || _h === void 0 ? void 0 : _h._media) === null || _j === void 0 ? void 0 : _j.profilePicture) === null || _k === void 0 ? void 0 : _k[0]) === null || _l === void 0 ? void 0 : _l._full_url) ? (_jsxs("div", { className: "relative", children: [_jsx(Avatar, { image: (_q = (_p = (_o = (_m = userData === null || userData === void 0 ? void 0 : userData.data) === null || _m === void 0 ? void 0 : _m._media) === null || _o === void 0 ? void 0 : _o.profilePicture) === null || _p === void 0 ? void 0 : _p[0]) === null || _q === void 0 ? void 0 : _q._full_url, shape: "circle", style: {
                                                    height: '5rem',
                                                    width: '5rem'
                                                } }), _jsx("div", { className: styles.SolidRemoveProfile, onClick: handleDeleteAvatar, children: _jsx("span", { className: "pi pi-times" }) })] })) : (_jsx(Avatar, { label: initials, size: "xlarge", shape: "circle", style: {
                                            backgroundColor: bgColor,
                                            color: '#ffffff',
                                            height: '5rem',
                                            width: '5rem'
                                        } })), _jsxs("div", { className: "flex gap-2", children: [_jsx(Button, { type: "button", size: "small", severity: "secondary", label: "Upload Avatar", outlined: true, className: "small-button", onClick: function () { var _a; return (_a = fileInputRef.current) === null || _a === void 0 ? void 0 : _a.click(); } }), _jsx("input", { ref: fileInputRef, type: "file", accept: ".jpg,.jpeg,.png,.svg", hidden: true, onChange: function (e) {
                                                    var _a;
                                                    var file = (_a = e.target.files) === null || _a === void 0 ? void 0 : _a[0];
                                                    if (file)
                                                        handleFileChange(file);
                                                } })] })] })] }), _jsx("div", { className: 'mt-4 mb-2 sm:mb-0', style: { borderBottom: '1px dashed #D8E2EA' } }), _jsxs("div", { className: "mt-4", children: [_jsx("label", { className: "form-field-label mb-2 font-bold", children: "Details" }), _jsxs("div", { className: "grid", children: [_jsxs("div", { className: "col-12 md:col-8 lg:col-5", children: [_jsx("label", { className: "form-field-label mb-2", children: "Name" }), _jsx(InputText, { name: "fullName", value: formik.values.fullName, onChange: formik.handleChange, className: "w-full" })] }), _jsxs("div", { className: "col-12 md:col-8 lg:col-5", children: [_jsx("label", { className: "form-field-label mb-2", children: "Email" }), _jsx(InputText, { disabled: true, placeholder: (_r = userData === null || userData === void 0 ? void 0 : userData.data) === null || _r === void 0 ? void 0 : _r.email, value: (_s = userData === null || userData === void 0 ? void 0 : userData.data) === null || _s === void 0 ? void 0 : _s.email, className: "w-full" })] }), _jsxs("div", { className: "col-12 md:col-8 lg:col-5", children: [_jsx("label", { className: "form-field-label mb-2", children: "Contact Number" }), _jsx(InputText, { disabled: true, placeholder: (_t = userData === null || userData === void 0 ? void 0 : userData.data) === null || _t === void 0 ? void 0 : _t.mobile, value: (_u = userData === null || userData === void 0 ? void 0 : userData.data) === null || _u === void 0 ? void 0 : _u.mobile, className: "w-full" })] }), _jsxs("div", { className: "col-12 md:col-8 lg:col-5", children: [_jsx("label", { className: "form-field-label mb-2", children: "Role" }), _jsx(AutoComplete, { multiple: true, disabled: true, value: ((_w = (_v = userData === null || userData === void 0 ? void 0 : userData.data) === null || _v === void 0 ? void 0 : _v.roles) === null || _w === void 0 ? void 0 : _w.map(function (role) { return role.name; })) || [] })] })] })] })] }), _jsx("div", { children: _jsx(Button, { type: "submit", size: "small", label: "Save" }) }), _jsx(Dialog, { header: _jsx("h5", { className: 'm-0 font-bold', children: "Replace Profile Picture" }), visible: replaceDialogVisible, headerClassName: "px-4 py-3 secondary-border-bottom", contentClassName: "p-0", style: { width: '25vw' }, className: "solid-confirm-dialog ", modal: true, onHide: function () { return setReplaceDialogVisible(false); }, children: _jsxs("div", { className: "p-4", children: ["Do you want to replace the existing profile picture?", _jsxs("div", { className: "flex justify-content-start mt-4 gap-2", children: [_jsx(Button, { size: "small", label: "Confirm", severity: "danger", onClick: confirmReplace }), _jsx(Button, { size: "small", outlined: true, label: "Cancel", onClick: function () { return setReplaceDialogVisible(false); } })] })] }) }), _jsx(Dialog, { header: _jsx("h5", { className: 'm-0 font-bold', children: "Delete Profile Picture" }), visible: deleteDialogVisible, headerClassName: "px-4 py-3 secondary-border-bottom", contentClassName: "p-0", style: { width: '25vw' }, modal: true, className: "solid-confirm-dialog ", onHide: function () { return setDeleteDialogVisible(false); }, children: _jsxs("div", { className: "p-4", children: ["Do you want to delete your profile picture?", _jsxs("div", { className: "flex justify-content-start mt-4 gap-2", children: [_jsx(Button, { size: "small", label: "Confirm", severity: "danger", onClick: confirmDelete }), _jsx(Button, { size: "small", outlined: true, label: "Cancel", onClick: function () { return setDeleteDialogVisible(false); } })] })] }) })] }));
};
