// @solidxai/core-ui/dist/components/core/common/SolidAccountSettings/SolidChangePassword.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { handleLogout } from "../../../../adapters/auth/handleLogout";
import { useChangePasswordMutation } from '../../../../redux/api/authApi';
import { useFormik } from 'formik';
import { useSession } from "../../../../hooks/useSession";
import { Button } from 'primereact/button';
import { Message } from 'primereact/message';
import { Password } from 'primereact/password';
import { Toast } from 'primereact/toast';
import { useMemo, useRef } from 'react';
import * as Yup from 'yup';
import { SolidPasswordHelperText } from '../SolidPasswordHelperText';
import { ERROR_MESSAGES } from '../../../../constants/error-messages';
import { env } from "../../../../adapters/env";
import showToast from "../../../../helpers/showToast";
export var SolidChangePassword = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
    var solidSettingsData = _a.solidSettingsData;
    var toast = useRef(null);
    var changePassword = useChangePasswordMutation()[0];
    var session = useSession();
    var envPasswordRegex = env("NEXT_PUBLIC_PASSWORD_REGEX");
    // Try backend regex first, then env, then fallback
    var effectiveRegex = useMemo(function () {
        var _a;
        try {
            var backendRegex = (_a = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _a === void 0 ? void 0 : _a.authenticationPasswordRegex;
            if (backendRegex) {
                // const unescaped = JSON.parse(`"${backendRegex}"`);
                return new RegExp(backendRegex);
            }
            if (envPasswordRegex) {
                var unescaped = JSON.parse("\"".concat(envPasswordRegex, "\""));
                return new RegExp(unescaped);
            }
        }
        catch (error) {
            console.error(ERROR_MESSAGES.INVALID_PASSWORD_REGX, error);
        }
        return null;
    }, [solidSettingsData, envPasswordRegex]);
    var validationSchema = useMemo(function () {
        var _a;
        var newPasswordValidation = effectiveRegex
            ? Yup.string()
                .matches(effectiveRegex, ((_a = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _a === void 0 ? void 0 : _a.authenticationPasswordRegexErrorMessage) ||
                ERROR_MESSAGES.PASSWORD_DO_NOT_MEET)
                .required(ERROR_MESSAGES.FIELD_REUQIRED('New password'))
            : Yup.string()
                .min(6, ERROR_MESSAGES.PASSWORD_CHARACTER(6))
                .required(ERROR_MESSAGES.FIELD_REUQIRED('New password'));
        return Yup.object({
            email: Yup.string().email(ERROR_MESSAGES.FIELD_INAVLID_FORMAT('email')).required(ERROR_MESSAGES.FIELD_REUQIRED('Email')),
            currentPassword: Yup.string()
                .min(6, ERROR_MESSAGES.PASSWORD_CHARACTER(6))
                .required(ERROR_MESSAGES.FIELD_REUQIRED('Current password')),
            newPassword: newPasswordValidation,
            confirmPassword: Yup.string()
                .oneOf([Yup.ref("newPassword")], ERROR_MESSAGES.MUST_MATCH)
                .required(ERROR_MESSAGES.FIELD_REUQIRED('Confirm password')),
            id: Yup.number().required(ERROR_MESSAGES.FIELD_INAVLID_FORMAT('User ID')),
        });
    }, [effectiveRegex, solidSettingsData]);
    var formik = useFormik({
        initialValues: {
            email: (_c = (_b = session === null || session === void 0 ? void 0 : session.data) === null || _b === void 0 ? void 0 : _b.user) === null || _c === void 0 ? void 0 : _c.email,
            currentPassword: "",
            newPassword: "",
            confirmPassword: "",
            id: (_e = (_d = session === null || session === void 0 ? void 0 : session.data) === null || _d === void 0 ? void 0 : _d.user) === null || _e === void 0 ? void 0 : _e.id,
        },
        validationSchema: validationSchema,
        onSubmit: function (values, _a) {
            var setErrors = _a.setErrors, resetForm = _a.resetForm;
            return __awaiter(void 0, void 0, void 0, function () {
                var payload, response, err_1;
                var _b, _c, _d, _e, _f, _g, _h, _j;
                return __generator(this, function (_k) {
                    switch (_k.label) {
                        case 0:
                            _k.trys.push([0, 2, , 3]);
                            payload = {
                                id: values.id,
                                email: (_c = (_b = session === null || session === void 0 ? void 0 : session.data) === null || _b === void 0 ? void 0 : _b.user) === null || _c === void 0 ? void 0 : _c.email,
                                currentPassword: values.currentPassword,
                                newPassword: values.newPassword,
                            };
                            return [4 /*yield*/, changePassword(payload).unwrap()];
                        case 1:
                            response = _k.sent();
                            if (response === null || response === void 0 ? void 0 : response.error) {
                                showToast(toast, "error", ERROR_MESSAGES.ERROR, response.error);
                                setErrors({
                                    currentPassword: ERROR_MESSAGES.INCORRECT_CURRENT,
                                    newPassword: ERROR_MESSAGES.MUST_MATCH,
                                    confirmPassword: ERROR_MESSAGES.MUST_MATCH,
                                });
                            }
                            else {
                                showToast(toast, "success", ERROR_MESSAGES.PASSWORD_CHANGE, ERROR_MESSAGES.PASSWORD_CHANGE);
                                handleLogout(toast);
                                resetForm();
                            }
                            return [3 /*break*/, 3];
                        case 2:
                            err_1 = _k.sent();
                            showToast(toast, "error", (_d = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _d === void 0 ? void 0 : _d.message, ((_f = (_e = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _e === void 0 ? void 0 : _e.data) === null || _f === void 0 ? void 0 : _f.message) ? (_h = (_g = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _g === void 0 ? void 0 : _g.data) === null || _h === void 0 ? void 0 : _h.message : (_j = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _j === void 0 ? void 0 : _j.message);
                            return [3 /*break*/, 3];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        },
    });
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    return (_jsxs("form", { onSubmit: formik.handleSubmit, className: "h-full flex flex-column justify-content-between", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { children: [_jsx("div", { className: 'grid', children: _jsxs("div", { className: 'col-12 md-col-8 lg:col-5', children: [_jsxs("div", { className: "flex flex-column gap-2", children: [_jsx("label", { htmlFor: "currentPassword", className: "solid-auth-input-label", children: "Current Password" }), _jsx(Password, { id: "currentPassword", name: "currentPassword", value: formik.values.currentPassword, onChange: formik.handleChange, onBlur: formik.handleBlur, toggleMask: true, invalid: !!formik.errors.currentPassword, inputClassName: "w-full", feedback: false }), isFormFieldValid(formik, "currentPassword") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_g = (_f = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _f === void 0 ? void 0 : _f.currentPassword) === null || _g === void 0 ? void 0 : _g.toString() })] }), _jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [_jsx("label", { htmlFor: "password", className: "solid-auth-input-label", children: "New Password" }), _jsx(Password, { id: "newPassword", name: "newPassword", value: formik.values.newPassword, onChange: formik.handleChange, onBlur: formik.handleBlur, toggleMask: true, invalid: !!formik.errors.newPassword, inputClassName: "w-full", feedback: false }), isFormFieldValid(formik, "newPassword") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_j = (_h = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _h === void 0 ? void 0 : _h.newPassword) === null || _j === void 0 ? void 0 : _j.toString() })] }), _jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [_jsx("label", { htmlFor: "password", className: "solid-auth-input-label", children: "Confirm New Password" }), _jsx(Password, { id: "confirmPassword", name: "confirmPassword", value: formik.values.confirmPassword, onChange: formik.handleChange, onBlur: formik.handleBlur, toggleMask: true, invalid: !!formik.errors.confirmPassword, inputClassName: "w-full", feedback: false }), isFormFieldValid(formik, "confirmPassword") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_l = (_k = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _k === void 0 ? void 0 : _k.confirmPassword) === null || _l === void 0 ? void 0 : _l.toString() })] })] }) }), _jsx(SolidPasswordHelperText, { text: (_m = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _m === void 0 ? void 0 : _m.authenticationPasswordComplexityDescription })] }), _jsx("div", { children: _jsx(Button, { type: 'submit', size: 'small', label: "Change Password", disabled: formik.isSubmitting, loading: formik.isSubmitting }) })] }));
};
