// @solidxai/core-ui/dist/redux/api/authApi.js
import { createApi } from "@reduxjs/toolkit/query/react";
import { baseQueryWithAuth } from "./fetchBaseQuery";
export var authApi = createApi({
    reducerPath: "authApi",
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        initateLogin: builder.mutation({
            query: function (body) {
                return {
                    url: "/iam/otp/login/initiate",
                    method: "POST",
                    body: body,
                };
            },
        }),
        confirmOtpLogin: builder.mutation({
            query: function (body) {
                return {
                    url: "/iam/otp/login/confirm",
                    method: "POST",
                    body: body,
                };
            },
        }),
        register: builder.mutation({
            query: function (body) {
                return {
                    url: "/iam/register",
                    method: "POST",
                    body: body,
                };
            },
        }),
        initateRegister: builder.mutation({
            query: function (body) {
                return {
                    url: "/iam/otp/register/initiate",
                    method: "POST",
                    body: body,
                };
            },
        }),
        confirmOtpRegister: builder.mutation({
            query: function (body) {
                return {
                    url: "/iam/otp/register/confirm",
                    method: "POST",
                    body: body,
                };
            },
        }),
        registerPrivate: builder.mutation({
            query: function (body) {
                return {
                    url: "/iam/register-private",
                    method: "POST",
                    body: body,
                };
            },
        }),
        updateUser: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/user/".concat(id, "/update-user-and-roles"),
                    method: 'PATCH',
                    body: data,
                });
            },
        }),
        forgotPassword: builder.mutation({
            query: function (body) {
                return {
                    url: "/password/forgot",
                    method: "POST",
                    body: body,
                };
            },
        }),
        resetPassword: builder.mutation({
            query: function (_a) {
                var token = _a.token, body = _a.body;
                return {
                    url: "/password/reset/".concat(token),
                    method: "PUT",
                    body: body,
                };
            },
        }),
        initiateChangePassword: builder.mutation({
            query: function (body) {
                return {
                    url: '/iam/initiate/forgot-password',
                    method: "POST",
                    body: body,
                };
            }
        }),
        confirmForgotPassword: builder.mutation({
            query: function (body) {
                return {
                    url: '/iam/confirm/forgot-password',
                    method: "POST",
                    body: body
                };
            }
        }),
        changePassword: builder.mutation({
            query: function (body) {
                return {
                    url: "/iam/change-password",
                    method: "POST",
                    body: body,
                };
            },
        }),
    }); },
});
export var useRegisterMutation = authApi.useRegisterMutation, useRegisterPrivateMutation = authApi.useRegisterPrivateMutation, useUpdateUserMutation = authApi.useUpdateUserMutation, useForgotPasswordMutation = authApi.useForgotPasswordMutation, useResetPasswordMutation = authApi.useResetPasswordMutation, useInitiateChangePasswordMutation = authApi.useInitiateChangePasswordMutation, useConfirmForgotPasswordMutation = authApi.useConfirmForgotPasswordMutation, useChangePasswordMutation = authApi.useChangePasswordMutation, useInitateLoginMutation = authApi.useInitateLoginMutation, useConfirmOtpLoginMutation = authApi.useConfirmOtpLoginMutation, useInitateRegisterMutation = authApi.useInitateRegisterMutation, useConfirmOtpRegisterMutation = authApi.useConfirmOtpRegisterMutation;
