// @solidxai/core-ui/dist/redux/api/dashboardApi.js
import { createApi } from "@reduxjs/toolkit/query/react";
import { baseQueryWithAuth } from "./fetchBaseQuery";
// API Definition
export var dashboardApi = createApi({
    reducerPath: "dashboardApi",
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getDashboard: builder.query({
            query: function (qs) { return "/dashboard?".concat(qs); },
            transformResponse: function (response) {
                // console.log(`Dashboard Response:`, response);
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta,
                };
            },
        }),
        getDashboardVariableSelectionDynamicValues: builder.query({
            query: function (qs) { return "/dashboard/selection-dynamic-values?".concat(qs); },
            transformResponse: function (response) {
                // console.log(`Dashboard Variable Selection Dynamic Values Response:`, response);
                if (response.error) {
                    throw new Error(response.error);
                }
                return response.data;
            },
        }),
    }); },
});
// Hooks
export var useGetDashboardQuery = dashboardApi.useGetDashboardQuery, useLazyGetDashboardQuery = dashboardApi.useLazyGetDashboardQuery, useGetDashboardVariableSelectionDynamicValuesQuery = dashboardApi.useGetDashboardVariableSelectionDynamicValuesQuery, useLazyGetDashboardVariableSelectionDynamicValuesQuery = dashboardApi.useLazyGetDashboardVariableSelectionDynamicValuesQuery;
