// @solidxai/core-ui/dist/redux/api/exportTemplateApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var exportTemplateApi = createApi({
    reducerPath: 'exportTemplateApi',
    baseQuery: baseQueryWithAuth,
    tagTypes: ['ExportTemplates'],
    endpoints: function (builder) { return ({
        createExportTemplate: builder.mutation({
            query: function (templateData) { return ({
                url: '/export-template',
                method: 'POST',
                body: templateData
            }); },
            invalidatesTags: ['ExportTemplates'],
        }),
        getExportTemplates: builder.query({
            query: function () {
                return "/export-template";
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            },
            providesTags: ['ExportTemplates'],
        }),
        createExportSync: builder.mutation({
            query: function (_a) {
                var id = _a.id;
                return ({
                    url: "/export-template/".concat(id, "/startExport/sync"),
                });
            },
            transformResponse: function (response) {
                return {
                    file: response
                };
            }
        }),
        createExportAsync: builder.mutation({
            query: function (_a) {
                var id = _a.id, templateData = _a.templateData;
                return ({
                    url: "/export-template/".concat(id, "/startExport/async"),
                    method: 'POST',
                    body: templateData
                });
            },
        }),
        deleteExportTemplate: builder.mutation({
            query: function (id) { return ({
                url: "/export-template/".concat(id),
                method: 'DELETE',
            }); },
            invalidatesTags: ['ExportTemplates'],
        }),
    }); }
});
export var useCreateExportTemplateMutation = exportTemplateApi.useCreateExportTemplateMutation, useGetExportTemplatesQuery = exportTemplateApi.useGetExportTemplatesQuery, useCreateExportSyncMutation = exportTemplateApi.useCreateExportSyncMutation, useCreateExportAsyncMutation = exportTemplateApi.useCreateExportAsyncMutation, useDeleteExportTemplateMutation = exportTemplateApi.useDeleteExportTemplateMutation;
