// @solidxai/core-ui/dist/redux/api/fetchBaseQuery.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { fetchBaseQuery } from "@reduxjs/toolkit/query/react";
import { getSession } from "../../adapters/auth/index";
import { env } from "../../adapters/env";
import { logger } from "../../helpers/logger";
import { eventBus, AppEvents } from "../../helpers/eventBus";
// Base URL for the API endpoints
var baseUrl = "".concat(env("NEXT_PUBLIC_BACKEND_API_URL"), "/api");
logger.debug("fetchBaseQuery resolved baseUrl to ".concat(baseUrl));
var rawBaseQuery = fetchBaseQuery({
    baseUrl: baseUrl,
    prepareHeaders: function (headers) { return __awaiter(void 0, void 0, void 0, function () {
        var session;
        var _a, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0: return [4 /*yield*/, getSession()];
                case 1:
                    session = _d.sent();
                    if ((_a = session === null || session === void 0 ? void 0 : session.user) === null || _a === void 0 ? void 0 : _a.accessToken) {
                        // Add access token to headers
                        headers.set('authorization', "Bearer ".concat((_b = session === null || session === void 0 ? void 0 : session.user) === null || _b === void 0 ? void 0 : _b.accessToken));
                    }
                    if ((_c = session === null || session === void 0 ? void 0 : session.user) === null || _c === void 0 ? void 0 : _c.accessToken) {
                        headers.set("authorization", "Bearer ".concat(session.user.accessToken));
                        logger.debug("[prepareHeaders] set auth header");
                    }
                    else {
                        logger.debug("[prepareHeaders] no access token");
                    }
                    logger.debug("[prepareHeaders] end");
                    return [2 /*return*/, headers];
            }
        });
    }); },
});
export var baseQueryWithAuth = function (args, api, extraOptions) { return __awaiter(void 0, void 0, void 0, function () {
    var result, status_1, isNetwork;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0: return [4 /*yield*/, rawBaseQuery(args, api, extraOptions)];
            case 1:
                result = _a.sent();
                if (result.error) {
                    status_1 = result.error.status;
                    isNetwork = status_1 === "FETCH_ERROR" ||
                        status_1 === "PARSING_ERROR" ||
                        (typeof status_1 === "number" && status_1 >= 500);
                    if (isNetwork) {
                        eventBus.emit(AppEvents.GlobalError, {
                            status: status_1,
                            message: "Unable to reach the server. Please try again later.",
                            error: result.error,
                        });
                    }
                }
                return [2 /*return*/, result];
        }
    });
}); };
