// @solidxai/core-ui/dist/redux/api/fieldApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var fieldsApi = createApi({
    reducerPath: 'fieldApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getfields: builder.query({
            query: function (qs) {
                return "/field-metadata?populate[0]=model&".concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getfieldById: builder.query({
            query: function (id) { return "/field-metadata/".concat(id); },
        }),
        getFieldDefaultMetaData: builder.query({
            query: function () { return "/field-metadata/default"; },
        }),
        getSelectionDynamicValues: builder.query({
            query: function (qs) { return "/field-metadata/selection-dynamic-values?".concat(qs); },
        }),
        getSelectionDynamicValue: builder.query({
            query: function (qs) { return "/field-metadata/selection-dynamic-value?".concat(qs); },
        }),
        createfield: builder.mutation({
            query: function (field) { return ({
                url: '/field-metadata',
                method: 'POST',
                body: field
            }); },
        }),
        // updatefield: builder.mutation({
        //     query: ({ id, data }) => ({
        //         url: `/field-metadata/${id}`,
        //         method: 'PUT',
        //         body: data,
        //     }), 
        // }),
        // deleteMultiplefields: builder.mutation({
        //     query: (data) => ({
        //         url: `/field/bulk/`,
        //         method: 'DELETE',
        //         body:data
        //     }),
        // }),
        deletefield: builder.mutation({
            query: function (id) { return ({
                url: "/field/".concat(id),
                method: 'DELETE',
            }); },
        }),
        resolveS3Url: builder.mutation({
            query: function (data) { return ({
                url: '/field-metadata/resolve-s3-url',
                method: 'POST',
                body: data
            }); },
        }),
    }); }
});
export var useGetfieldsQuery = fieldsApi.useGetfieldsQuery, useLazyGetfieldsQuery = fieldsApi.useLazyGetfieldsQuery, useLazyGetfieldByIdQuery = fieldsApi.useLazyGetfieldByIdQuery, useGetFieldDefaultMetaDataQuery = fieldsApi.useGetFieldDefaultMetaDataQuery, useGetSelectionDynamicValuesQuery = fieldsApi.useGetSelectionDynamicValuesQuery, useLazyGetSelectionDynamicValuesQuery = fieldsApi.useLazyGetSelectionDynamicValuesQuery, useGetSelectionDynamicValueQuery = fieldsApi.useGetSelectionDynamicValueQuery, useLazyGetSelectionDynamicValueQuery = fieldsApi.useLazyGetSelectionDynamicValueQuery, useLazyGetFieldDefaultMetaDataQuery = fieldsApi.useLazyGetFieldDefaultMetaDataQuery, useGetfieldByIdQuery = fieldsApi.useGetfieldByIdQuery, useCreatefieldMutation = fieldsApi.useCreatefieldMutation, useDeletefieldMutation = fieldsApi.useDeletefieldMutation, useResolveS3UrlMutation = fieldsApi.useResolveS3UrlMutation;
