// @solidxai/core-ui/dist/redux/api/importTransactionApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var importTransactionApi = createApi({
    reducerPath: 'importTransactionApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        // Get instructions
        getImportInstructions: builder.query({
            query: function (_a) {
                var id = _a.id;
                return "/import-transaction/import-instructions/".concat(id);
            },
        }),
        // Get import templates
        getImportTemplate: builder.query({
            query: function (_a) {
                var id = _a.id, format = _a.format;
                return "/import-transaction/import-template/".concat(id, "/").concat(format);
            }
        }),
        // create transaction / upload file here
        createImportTransaction: builder.mutation({
            query: function (templateData) { return ({
                url: '/import-transaction',
                method: 'POST',
                body: templateData
            }); },
        }),
        // After getting response getting id patch that here and pass     "status": "mapping_created",
        patchUpdateImportTransaction: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/import-transaction/".concat(id),
                    method: 'PATCH',
                    body: data,
                });
            },
        }),
        // Get Mattping info
        // Get import templates
        getImportMappingInfo: builder.query({
            query: function (_a) {
                var id = _a.id;
                return "/import-transaction/".concat(id, "/import-mapping-info");
            }
        }),
        createImportSync: builder.mutation({
            query: function (_a) {
                var id = _a.id;
                return ({
                    url: "/import-transaction/".concat(id, "/start-import/sync"),
                    method: 'POST',
                });
            },
        }),
        createImportAsync: builder.mutation({
            query: function (_a) {
                var id = _a.id;
                return ({
                    url: "/import-transaction/".concat(id, "/start-import/async"),
                    method: 'POST',
                });
            },
        }),
    }); }
});
export var useGetImportInstructionsQuery = importTransactionApi.useGetImportInstructionsQuery, useLazyGetImportInstructionsQuery = importTransactionApi.useLazyGetImportInstructionsQuery, useGetImportTemplateQuery = importTransactionApi.useGetImportTemplateQuery, useLazyGetImportTemplateQuery = importTransactionApi.useLazyGetImportTemplateQuery, useCreateImportTransactionMutation = importTransactionApi.useCreateImportTransactionMutation, usePatchUpdateImportTransactionMutation = importTransactionApi.usePatchUpdateImportTransactionMutation, useGetImportMappingInfoQuery = importTransactionApi.useGetImportMappingInfoQuery, useLazyGetImportMappingInfoQuery = importTransactionApi.useLazyGetImportMappingInfoQuery, useCreateImportAsyncMutation = importTransactionApi.useCreateImportAsyncMutation, useCreateImportSyncMutation = importTransactionApi.useCreateImportSyncMutation;
