// @solidxai/core-ui/dist/redux/api/mediaStorageProviderApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var mediaStorageProviderApi = createApi({
    reducerPath: 'mediaStorageProviderApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getMediaStorageProviders: builder.query({
            query: function (qs) {
                return "/media-storage-provider-metadata?".concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getMediaStorageProivderById: builder.query({
            query: function (id) { return "/media-storage-provider-metadata/".concat(id); },
        }),
        createMediaStorageProivder: builder.mutation({
            query: function (MediaStorageProivder) { return ({
                url: '/media-storage-provider-metadata',
                method: 'POST',
                body: MediaStorageProivder
            }); },
        }),
        updateMediaStorageProivder: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/media-storage-provider-metadata/".concat(id),
                    method: 'PUT',
                    body: data,
                });
            },
        }),
        deleteMultipleMediaStorageProviders: builder.mutation({
            query: function (data) { return ({
                url: "/media-storage-provider-metadata/bulk/",
                method: 'DELETE',
                body: data
            }); },
        }),
        deleteMediaStorageProivder: builder.mutation({
            query: function (id) { return ({
                url: "/media-storage-provider-metadata/".concat(id),
                method: 'DELETE',
            }); },
        }),
    }); }
});
export var useGetMediaStorageProvidersQuery = mediaStorageProviderApi.useGetMediaStorageProvidersQuery, useLazyGetMediaStorageProivderByIdQuery = mediaStorageProviderApi.useLazyGetMediaStorageProivderByIdQuery, useLazyGetMediaStorageProvidersQuery = mediaStorageProviderApi.useLazyGetMediaStorageProvidersQuery, useGetMediaStorageProivderByIdQuery = mediaStorageProviderApi.useGetMediaStorageProivderByIdQuery, useCreateMediaStorageProivderMutation = mediaStorageProviderApi.useCreateMediaStorageProivderMutation, useUpdateMediaStorageProivderMutation = mediaStorageProviderApi.useUpdateMediaStorageProivderMutation, useDeleteMediaStorageProivderMutation = mediaStorageProviderApi.useDeleteMediaStorageProivderMutation, useDeleteMultipleMediaStorageProvidersMutation = mediaStorageProviderApi.useDeleteMultipleMediaStorageProvidersMutation;
