// @solidxai/core-ui/dist/redux/api/mediaApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var mediaApi = createApi({
    reducerPath: 'mediaApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getMedias: builder.query({
            query: function (qs) {
                return "/media?".concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getMediaById: builder.query({
            query: function (id) { return "/media/".concat(id); },
        }),
        createMedia: builder.mutation({
            query: function (Media) { return ({
                url: '/media',
                method: 'POST',
                body: Media
            }); },
        }),
        updateMedia: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/media/".concat(id),
                    method: 'PUT',
                    body: data,
                });
            },
        }),
        deleteMultipleMedias: builder.mutation({
            query: function (data) { return ({
                url: "/media/bulk/",
                method: 'DELETE',
                body: data
            }); },
        }),
        deleteMedia: builder.mutation({
            query: function (id) { return ({
                url: "/media/".concat(id),
                method: 'DELETE',
            }); },
        }),
    }); }
});
export var useGetMediasQuery = mediaApi.useGetMediasQuery, useLazyGetMediaByIdQuery = mediaApi.useLazyGetMediaByIdQuery, useLazyGetMediasQuery = mediaApi.useLazyGetMediasQuery, useGetMediaByIdQuery = mediaApi.useGetMediaByIdQuery, useCreateMediaMutation = mediaApi.useCreateMediaMutation, useUpdateMediaMutation = mediaApi.useUpdateMediaMutation, useDeleteMediaMutation = mediaApi.useDeleteMediaMutation, useDeleteMultipleMediasMutation = mediaApi.useDeleteMultipleMediasMutation;
