// @solidxai/core-ui/dist/redux/api/modelApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var modelsApi = createApi({
    reducerPath: 'modelApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getModels: builder.query({
            query: function (qs) {
                return "/model-metadata?&populate[1]=userKeyField&".concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getmodelById: builder.query({
            query: function (id) { return "/model-metadata/".concat(id, "?populate[0]=fields&populate[1]=module&&populate[2]=fields.mediaStorageProvider&populate[3]=parentModel"); },
        }),
        createmodel: builder.mutation({
            query: function (model) { return ({
                url: '/model-metadata',
                method: 'POST',
                body: model
            }); },
        }),
        generateCodeForModel: builder.mutation({
            query: function (model) { return ({
                url: "/model-metadata/".concat(model.id, "/generate-code"),
                method: 'POST',
                body: model
            }); },
        }),
        updatemodel: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/model-metadata/".concat(id),
                    method: 'Put',
                    body: data,
                });
            },
        }),
        // deleteMultiplemodels: builder.mutation({
        //     query: (data) => ({
        //         url: `/model/bulk/`,
        //         method: 'DELETE',
        //         body:data
        //     }),
        // }),
        deleteMultipleModels: builder.mutation({
            query: function (data) { return ({
                url: "/model-metadata/bulk/",
                method: 'DELETE',
                body: data
            }); },
        }),
        deletemodel: builder.mutation({
            query: function (id) { return ({
                url: "/model-metadata/".concat(id),
                method: 'DELETE',
            }); },
        }),
        updateUserKey: builder.mutation({
            query: function (data) { return ({
                url: "/model-metadata/update-user-key",
                method: 'POST',
                body: data,
            }); },
        }),
        navigation: builder.query({
            query: function (qs) { return "/model-metadata/navigation?".concat(qs); },
        }),
    }); }
});
export var useGetModelsQuery = modelsApi.useGetModelsQuery, useLazyGetModelsQuery = modelsApi.useLazyGetModelsQuery, useLazyGetmodelByIdQuery = modelsApi.useLazyGetmodelByIdQuery, useGetmodelByIdQuery = modelsApi.useGetmodelByIdQuery, useCreatemodelMutation = modelsApi.useCreatemodelMutation, useGenerateCodeForModelMutation = modelsApi.useGenerateCodeForModelMutation, useUpdatemodelMutation = modelsApi.useUpdatemodelMutation, useDeleteMultipleModelsMutation = modelsApi.useDeleteMultipleModelsMutation, useDeletemodelMutation = modelsApi.useDeletemodelMutation, useUpdateUserKeyMutation = modelsApi.useUpdateUserKeyMutation, useLazyNavigationQuery = modelsApi.useLazyNavigationQuery;
