// @solidxai/core-ui/dist/redux/api/moduleApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var modulesApi = createApi({
    reducerPath: 'moduleApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getmodules: builder.query({
            query: function (qs) {
                return "/module-metadata?".concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getmoduleById: builder.query({
            query: function (id) { return "/module-metadata/".concat(id); },
        }),
        createmodule: builder.mutation({
            query: function (module) { return ({
                url: '/module-metadata',
                method: 'POST',
                body: module
            }); },
        }),
        generateCodeFormodule: builder.mutation({
            query: function (module) { return ({
                url: "/module-metadata/".concat(module.id, "/generate-code"),
                method: 'POST',
                body: module
            }); },
        }),
        refreshPermissions: builder.mutation({
            query: function (module) { return ({
                url: "/module-metadata/refresh-permission",
                method: 'POST',
                body: module
            }); },
        }),
        updatemodule: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/module-metadata/".concat(id),
                    method: 'PUT',
                    body: data,
                });
            },
        }),
        deleteMultiplemodules: builder.mutation({
            query: function (data) { return ({
                url: "/module-metadata/bulk/",
                method: 'DELETE',
                body: data
            }); },
        }),
        deletemodule: builder.mutation({
            query: function (id) { return ({
                url: "/module-metadata/".concat(id),
                method: 'DELETE',
            }); },
        }),
        getDefaultDataSource: builder.query({
            query: function () { return "/module-metadata/fetch-default-datasource"; },
        }),
    }); }
});
export var useGetmodulesQuery = modulesApi.useGetmodulesQuery, useLazyGetmodulesQuery = modulesApi.useLazyGetmodulesQuery, useLazyGetmoduleByIdQuery = modulesApi.useLazyGetmoduleByIdQuery, useGetmoduleByIdQuery = modulesApi.useGetmoduleByIdQuery, useCreatemoduleMutation = modulesApi.useCreatemoduleMutation, useGenerateCodeFormoduleMutation = modulesApi.useGenerateCodeFormoduleMutation, useRefreshPermissionsMutation = modulesApi.useRefreshPermissionsMutation, useUpdatemoduleMutation = modulesApi.useUpdatemoduleMutation, useDeletemoduleMutation = modulesApi.useDeletemoduleMutation, useDeleteMultiplemodulesMutation = modulesApi.useDeleteMultiplemodulesMutation, useGetDefaultDataSourceQuery = modulesApi.useGetDefaultDataSourceQuery, useLazyGetDefaultDataSourceQuery = modulesApi.useLazyGetDefaultDataSourceQuery;
