// @solidxai/core-ui/dist/redux/api/permissionApi.js
import { createApi } from "@reduxjs/toolkit/query/react";
import { baseQueryWithAuth } from "./fetchBaseQuery";
export var permissionApi = createApi({
    reducerPath: "permissionApi",
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getpermissions: builder.query({
            query: function (qs) {
                return "/permissions?".concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getpermissionsById: builder.query({
            query: function (id) { return "/permissions/".concat(id, "?offset=0&limit=5"); },
        }),
    }); },
});
export var useGetpermissionsByIdQuery = permissionApi.useGetpermissionsByIdQuery, useGetpermissionsQuery = permissionApi.useGetpermissionsQuery, useLazyGetpermissionsQuery = permissionApi.useLazyGetpermissionsQuery;
