// @solidxai/core-ui/dist/redux/api/pincodeApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var pincodesApi = createApi({
    reducerPath: 'pincodesApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getPincodes: builder.query({
            query: function (qs) {
                return "/pincode?populate[]=city&".concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getPincodeById: builder.query({
            query: function (id) { return "/pincode/".concat(id, "?populate[]=city"); },
        }),
        createPincode: builder.mutation({
            query: function (pincode) { return ({
                url: '/pincode',
                method: 'POST',
                body: pincode
            }); },
        }),
        // updatePincode: builder.mutation({
        //     query: ({ id, data }) => ({
        //         url: `/pincode/${id}`,
        //         method: 'Patch',
        //         body: data,
        //     }),
        // }),
        deleteMultiplePincodes: builder.mutation({
            query: function (data) { return ({
                url: "/pincode/bulk/",
                method: 'DELETE',
                body: data
            }); },
        }),
        deletePincode: builder.mutation({
            query: function (id) { return ({
                url: "/pincode/".concat(id),
                method: 'DELETE',
            }); },
        }),
    }); }
});
export var useLazyGetPincodesQuery = pincodesApi.useLazyGetPincodesQuery, useCreatePincodeMutation = pincodesApi.useCreatePincodeMutation, useDeletePincodeMutation = pincodesApi.useDeletePincodeMutation, useDeleteMultiplePincodesMutation = pincodesApi.useDeleteMultiplePincodesMutation, useGetPincodesQuery = pincodesApi.useGetPincodesQuery, useGetPincodeByIdQuery = pincodesApi.useGetPincodeByIdQuery, useLazyGetPincodeByIdQuery = pincodesApi.useLazyGetPincodeByIdQuery, usePrefetch = pincodesApi.usePrefetch;
