// @solidxai/core-ui/dist/redux/api/roleApi.js
import { createApi } from "@reduxjs/toolkit/query/react";
import { baseQueryWithAuth } from "./fetchBaseQuery";
var resourceName = 'role-metadata';
export var roleApi = createApi({
    reducerPath: "roleApi",
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getroles: builder.query({
            query: function (qs) {
                return "/".concat(resourceName, "?offset=0&limit=1000&").concat(qs);
            },
        }),
        getroleById: builder.query({
            query: function (id) { return "/".concat(resourceName, "/").concat(id, "?offset=0&limit=5"); },
        }),
        createrole: builder.mutation({
            query: function (role) { return ({
                url: '/${resourceName}',
                method: 'POST',
                body: role
            }); },
        }),
        updaterole: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/".concat(resourceName, "/").concat(id),
                    method: 'PATCH',
                    body: data,
                });
            },
        }),
        deleteMultipleRoles: builder.mutation({
            query: function (data) { return ({
                url: "/".concat(resourceName, "/bulk/"),
                method: 'DELETE',
                body: data
            }); },
        }),
        deleterole: builder.mutation({
            query: function (id) { return ({
                url: "/".concat(resourceName, "/").concat(id),
                method: 'DELETE',
            }); },
        }),
    }); },
});
export var useGetrolesQuery = roleApi.useGetrolesQuery, useGetroleByIdQuery = roleApi.useGetroleByIdQuery, useCreateroleMutation = roleApi.useCreateroleMutation, useLazyGetrolesQuery = roleApi.useLazyGetrolesQuery, useUpdateroleMutation = roleApi.useUpdateroleMutation, useDeleteroleMutation = roleApi.useDeleteroleMutation;
