// @solidxai/core-ui/dist/redux/api/solidChatterMessageApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var solidChatterMessageApi = createApi({
    reducerPath: 'solidChatterMessageApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getchatterMessage: builder.query({
            query: function (_a) {
                var entityId = _a.entityId, entityName = _a.entityName, qs = _a.qs;
                return "/chatter-message/getChatterMessages/".concat(entityId, "/").concat(entityName, "?populateMedia[0]=messageAttachments&").concat(qs);
            },
        }),
        createChatterMessage: builder.mutation({
            query: function (data) {
                return {
                    url: '/chatter-message',
                    method: 'POST',
                    body: data
                };
            }
        })
    }); }
});
export var useGetchatterMessageQuery = solidChatterMessageApi.useGetchatterMessageQuery, useLazyGetchatterMessageQuery = solidChatterMessageApi.useLazyGetchatterMessageQuery, useCreateChatterMessageMutation = solidChatterMessageApi.useCreateChatterMessageMutation;
