// @solidxai/core-ui/dist/redux/api/solidActionApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
var resourceName = 'action-metadata';
export var solidActionsApi = createApi({
    reducerPath: 'actionMetadataApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getSolidActions: builder.query({
            query: function (qs) {
                return "/".concat(resourceName, "?populate[0]=model&populate[1]=module&").concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getSolidActionById: builder.query({
            query: function (id) { return "/".concat(resourceName, "/").concat(id, "?populate[0]=model&populate[1]=module&populate[2]=view"); },
        }),
        createSolidAction: builder.mutation({
            query: function (solidAction) { return ({
                url: '/${resourceName}',
                method: 'POST',
                body: solidAction
            }); },
        }),
        updateSolidAction: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/".concat(resourceName, "/").concat(id),
                    method: 'PUT',
                    body: data,
                });
            },
        }),
        deleteMultipleSolidActions: builder.mutation({
            query: function (data) { return ({
                url: "/".concat(resourceName, "/bulk/"),
                method: 'DELETE',
                body: data
            }); },
        }),
        deleteSolidAction: builder.mutation({
            query: function (id) { return ({
                url: "/".concat(resourceName, "/").concat(id),
                method: 'DELETE',
            }); },
        }),
    }); },
});
export var useGetSolidActionsQuery = solidActionsApi.useGetSolidActionsQuery, useLazyGetSolidActionsQuery = solidActionsApi.useLazyGetSolidActionsQuery, useLazyGetSolidActionByIdQuery = solidActionsApi.useLazyGetSolidActionByIdQuery, useGetSolidActionByIdQuery = solidActionsApi.useGetSolidActionByIdQuery, useCreateSolidActionMutation = solidActionsApi.useCreateSolidActionMutation, useUpdateSolidActionMutation = solidActionsApi.useUpdateSolidActionMutation, useDeleteSolidActionMutation = solidActionsApi.useDeleteSolidActionMutation, useDeleteMultipleSolidActionsMutation = solidActionsApi.useDeleteMultipleSolidActionsMutation;
