// @solidxai/core-ui/dist/redux/api/solidServiceApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var solidServiceApi = createApi({
    reducerPath: 'solidServiceApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        seeder: builder.mutation({
            query: function (name) { return ({
                url: '/seed',
                method: 'POST',
                body: {
                    "seeder": name
                }
            }); },
        }),
        // Add a mutation for calling /code-generation/post-process, which takes an empty body
        codeGenerationPostProcess: builder.mutation({
            query: function (payload) { return ({
                url: '/code-generation/post-process',
                method: 'POST',
                body: payload
            }); },
        }),
    }); },
});
export var useSeederMutation = solidServiceApi.useSeederMutation, useCodeGenerationPostProcessMutation = solidServiceApi.useCodeGenerationPostProcessMutation;
