// @solidxai/core-ui/dist/redux/api/solidMenuApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
var resourceName = 'menu-item-metadata';
export var solidMenusApi = createApi({
    reducerPath: 'solidMenuApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getSolidMenus: builder.query({
            query: function (qs) {
                return "/".concat(resourceName, "?populate[0]=module&populate[1]=action&populate[2]=parentMenuItem&populate[3]=roles&").concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getSolidMenuById: builder.query({
            query: function (id) { return "/".concat(resourceName, "/").concat(id, "?populate[0]=module&populate[1]=action&populate[2]=parentMenuItem&populate[3]=roles"); },
        }),
        getSolidMenuBasedOnRole: builder.query({
            query: function () { return "/".concat(resourceName, "/me"); },
        }),
        createSolidMenu: builder.mutation({
            query: function (solidMenu) { return ({
                url: '/${resourceName}',
                method: 'POST',
                body: solidMenu
            }); },
        }),
        updateSolidMenu: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/".concat(resourceName, "/").concat(id),
                    method: 'PUT',
                    body: data,
                });
            },
        }),
        deleteMultipleSolidMenus: builder.mutation({
            query: function (data) { return ({
                url: "/".concat(resourceName, "/bulk/"),
                method: 'DELETE',
                body: data
            }); },
        }),
        deleteSolidMenu: builder.mutation({
            query: function (id) { return ({
                url: "/".concat(resourceName, "/").concat(id),
                method: 'DELETE',
            }); },
        }),
    }); }
});
export var useGetSolidMenusQuery = solidMenusApi.useGetSolidMenusQuery, useLazyGetSolidMenusQuery = solidMenusApi.useLazyGetSolidMenusQuery, useLazyGetSolidMenuByIdQuery = solidMenusApi.useLazyGetSolidMenuByIdQuery, useGetSolidMenuBasedOnRoleQuery = solidMenusApi.useGetSolidMenuBasedOnRoleQuery, useLazyGetSolidMenuBasedOnRoleQuery = solidMenusApi.useLazyGetSolidMenuBasedOnRoleQuery, useGetSolidMenuByIdQuery = solidMenusApi.useGetSolidMenuByIdQuery, useCreateSolidMenuMutation = solidMenusApi.useCreateSolidMenuMutation, useUpdateSolidMenuMutation = solidMenusApi.useUpdateSolidMenuMutation, useDeleteSolidMenuMutation = solidMenusApi.useDeleteSolidMenuMutation, useDeleteMultipleSolidMenusMutation = solidMenusApi.useDeleteMultipleSolidMenusMutation;
