// @solidxai/core-ui/dist/redux/api/solidEntityApi.js
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
import { kebabCase } from 'change-case';
export var createSolidEntityApi = function (entityName) {
    var kebabEntityName = kebabCase(entityName);
    return createApi({
        reducerPath: "genericSolid".concat(entityName, "Api"),
        baseQuery: baseQueryWithAuth,
        tagTypes: [entityName],
        endpoints: function (builder) { return ({
            getSolidEntities: builder.query({
                query: function (qs) {
                    return "/".concat(kebabEntityName, "?&").concat(qs);
                },
                transformResponse: function (response) {
                    var _a, _b;
                    if (response.error) {
                        throw new Error(response.error);
                    }
                    return {
                        records: response.data.records,
                        meta: response.data.meta,
                        groupMeta: (_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.groupMeta,
                        groupRecords: ((_b = response === null || response === void 0 ? void 0 : response.data) === null || _b === void 0 ? void 0 : _b.groupRecords) ? response.data.groupRecords : [],
                    };
                },
                providesTags: function (result) {
                    return (result === null || result === void 0 ? void 0 : result.records)
                        ? __spreadArray(__spreadArray([], result.records.map(function (record) { return ({
                            type: entityName,
                            id: record.id,
                        }); }), true), [
                            { type: entityName, id: 'LIST' },
                        ], false) : [{ type: entityName, id: 'LIST' }];
                }
            }),
            getSolidKanbanEntities: builder.query({
                query: function (qs) {
                    return "/".concat(kebabEntityName, "/group-kanban?&").concat(qs);
                },
                transformResponse: function (response) {
                    if (response.error) {
                        throw new Error(response.error);
                    }
                    return {
                        records: response.data.records.groupedData,
                        meta: response.data.meta
                    };
                },
                providesTags: [{ type: entityName, id: 'KANBAN' }],
            }),
            getSolidEntityById: builder.query({
                query: function (_a) {
                    var id = _a.id, qs = _a.qs;
                    return "/".concat(kebabEntityName, "/").concat(id, "?").concat(qs);
                },
                // providesTags: () => [{ type: entityName }],
                providesTags: function (_result, _error, _a) {
                    var id = _a.id;
                    return [{ type: entityName, id: id }];
                },
            }),
            createSolidEntity: builder.mutation({
                query: function (entity) { return ({
                    url: "/".concat(kebabEntityName),
                    method: 'POST',
                    body: entity
                }); },
                invalidatesTags: [{ type: entityName, id: 'LIST' }],
            }),
            upsertSolidEntity: builder.mutation({
                query: function (entity) { return ({
                    url: "/".concat(kebabEntityName, "/upsert"),
                    method: 'POST',
                    body: entity
                }); },
                invalidatesTags: [{ type: entityName, id: 'LIST' }],
            }),
            updateSolidEntity: builder.mutation({
                query: function (_a) {
                    var id = _a.id, data = _a.data;
                    return ({
                        url: "/".concat(kebabEntityName, "/").concat(id),
                        method: 'PUT',
                        body: data,
                    });
                },
                invalidatesTags: function (_result, _error, _a) {
                    var id = _a.id;
                    return [
                        { type: entityName, id: id },
                        { type: entityName, id: 'LIST' },
                    ];
                },
            }),
            deleteMultipleSolidEntities: builder.mutation({
                query: function (data) { return ({
                    url: "/".concat(kebabEntityName, "/bulk/"),
                    method: 'DELETE',
                    body: data
                }); },
                invalidatesTags: [{ type: entityName, id: 'LIST' }]
            }),
            deleteSolidEntity: builder.mutation({
                query: function (id) { return ({
                    url: "/".concat(kebabEntityName, "/").concat(id),
                    method: 'DELETE',
                }); },
                invalidatesTags: function (_result, _error, id) { return [
                    { type: entityName, id: id },
                    { type: entityName, id: 'LIST' },
                ]; },
            }),
            recoverSolidEntityById: builder.query({
                query: function (id) { return "/".concat(kebabEntityName, "/recover/").concat(id); },
                providesTags: function (_result, _error, id) { return [{ type: entityName, id: id }]; },
            }),
            recoverSolidEntity: builder.mutation({
                query: function (data) { return ({
                    url: "/".concat(kebabEntityName, "/bulk-recover/"),
                    method: 'POST',
                    body: data
                }); },
                invalidatesTags: [{ type: entityName, id: 'LIST' }],
            }),
            patchUpdateSolidEntity: builder.mutation({
                query: function (_a) {
                    var id = _a.id, data = _a.data;
                    return ({
                        url: "/".concat(kebabEntityName, "/").concat(id),
                        method: 'PATCH',
                        body: data,
                    });
                },
                invalidatesTags: function (_result, _error, _a) {
                    var id = _a.id;
                    return [
                        { type: entityName, id: id },
                        { type: entityName, id: 'LIST' },
                    ];
                },
            }),
            publishSolidEntity: builder.mutation({
                query: function (id) { return ({
                    url: "/".concat(kebabEntityName, "/").concat(id, "/publish"),
                    method: 'POST'
                }); },
                invalidatesTags: function (_result, _error, id) { return [
                    { type: entityName, id: id },
                    { type: entityName, id: 'LIST' }
                ]; }
            }),
            unpublishSolidEntity: builder.mutation({
                query: function (id) { return ({
                    url: "/".concat(kebabEntityName, "/").concat(id, "/unpublish"),
                    method: 'POST'
                }); },
                invalidatesTags: function (_result, _error, id) { return [
                    { type: entityName, id: id },
                    { type: entityName, id: 'LIST' }
                ]; }
            }),
        }); },
    });
};
// export const {
//     useCreateSolidEntityMutation,
//     useDeleteMultipleSolidEntitiesMutation,
//     useDeleteSolidEntityMutation,
//     useGetSolidEntitiesQuery,
//     useGetSolidEntityByIdQuery,
//     useLazyGetSolidEntitiesQuery,
//     useLazyGetSolidEntityByIdQuery,
//     usePrefetch,
//     useUpdateSolidEntityMutation
// } = solidEntityApi;
