// @solidxai/core-ui/dist/redux/api/solidViewApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
var resourceName = 'view-metadata';
export var solidViewsApi = createApi({
    reducerPath: 'viewMetadataApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getSolidViews: builder.query({
            query: function (qs) {
                return "/".concat(resourceName, "?populate[0]=model&populate[1]=module&").concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getSolidViewLayout: builder.query({
            query: function (qs) {
                return "/".concat(resourceName, "/custom/layout?").concat(qs);
            }
        }),
        getSolidViewById: builder.query({
            query: function (id) { return "/".concat(resourceName, "/").concat(id, "?populate[0]=model&populate[1]=module"); },
        }),
        createSolidView: builder.mutation({
            query: function (solidView) { return ({
                url: '/${resourceName}',
                method: 'POST',
                body: solidView
            }); },
        }),
        updateSolidView: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/".concat(resourceName, "/").concat(id),
                    method: 'PUT',
                    body: data,
                });
            },
        }),
        deleteMultipleSolidViews: builder.mutation({
            query: function (data) { return ({
                url: "/".concat(resourceName, "/bulk/"),
                method: 'DELETE',
                body: data
            }); },
        }),
        deleteSolidView: builder.mutation({
            query: function (id) { return ({
                url: "/".concat(resourceName, "/").concat(id),
                method: 'DELETE',
            }); },
        }),
    }); },
});
export var useGetSolidViewsQuery = solidViewsApi.useGetSolidViewsQuery, useLazyGetSolidViewsQuery = solidViewsApi.useLazyGetSolidViewsQuery, useLazyGetSolidViewByIdQuery = solidViewsApi.useLazyGetSolidViewByIdQuery, useGetSolidViewByIdQuery = solidViewsApi.useGetSolidViewByIdQuery, useGetSolidViewLayoutQuery = solidViewsApi.useGetSolidViewLayoutQuery, useLazyGetSolidViewLayoutQuery = solidViewsApi.useLazyGetSolidViewLayoutQuery, useCreateSolidViewMutation = solidViewsApi.useCreateSolidViewMutation, useUpdateSolidViewMutation = solidViewsApi.useUpdateSolidViewMutation, useDeleteSolidViewMutation = solidViewsApi.useDeleteSolidViewMutation, useDeleteMultipleSolidViewsMutation = solidViewsApi.useDeleteMultipleSolidViewsMutation;
