// @solidxai/core-ui/dist/redux/api/solidSettingsApi.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var solidSettingsApi = createApi({
    reducerPath: 'solidSettingsApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        createSolidSettings: builder.mutation({
            query: function (data) { return ({
                url: '/setting',
                method: 'POST',
                body: data
            }); },
        }),
        getSolidSettings: builder.query({
            query: function () {
                return "/setting";
            },
        }),
        getAuthSettings: builder.query({
            query: function () {
                console.log("builder.query running for endpoint /setting/wrapped");
                return "/setting/wrapped";
            },
            onQueryStarted: function (_arg, _a) {
                var queryFulfilled = _a.queryFulfilled;
                return __awaiter(this, void 0, void 0, function () {
                    var err_1;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0:
                                console.log("[solidSettingsApi] getAuthSettings onQueryStarted");
                                _b.label = 1;
                            case 1:
                                _b.trys.push([1, 3, , 4]);
                                return [4 /*yield*/, queryFulfilled];
                            case 2:
                                _b.sent();
                                console.log("[solidSettingsApi] getAuthSettings queryFulfilled");
                                return [3 /*break*/, 4];
                            case 3:
                                err_1 = _b.sent();
                                console.error("[solidSettingsApi] getAuthSettings queryFulfilled error", err_1);
                                return [3 /*break*/, 4];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            },
        }),
        getSolidSettingsById: builder.query({
            query: function (id) { return "/setting/".concat(id); },
        }),
        updateSolidSettings: builder.mutation({
            query: function (_a) {
                var id = _a.id, data = _a.data;
                return ({
                    url: "/setting/".concat(id),
                    method: 'PUT',
                    body: data,
                });
            },
        }),
        bulkUpdateSolidSettings: builder.mutation({
            query: function (_a) {
                var data = _a.data;
                return ({
                    url: "/setting/bulk",
                    method: 'PUT',
                    body: data,
                });
            },
        }),
        bulkUpdateSolidUserSettings: builder.mutation({
            query: function (_a) {
                var data = _a.data;
                return ({
                    url: "/setting/bulk/user",
                    method: 'POST',
                    body: data,
                });
            },
        }),
        getMcpUrl: builder.query({
            query: function (qs) { return "/setting/get-mcp-url?".concat(qs); },
        }),
    }); }
});
if (typeof window !== "undefined") {
    window.__solidSettingsApi = solidSettingsApi;
}
export var useCreateSolidSettingsMutation = solidSettingsApi.useCreateSolidSettingsMutation, useGetSolidSettingsByIdQuery = solidSettingsApi.useGetSolidSettingsByIdQuery, useGetSolidSettingsQuery = solidSettingsApi.useGetSolidSettingsQuery, useGetAuthSettingsQuery = solidSettingsApi.useGetAuthSettingsQuery, useLazyGetAuthSettingsQuery = solidSettingsApi.useLazyGetAuthSettingsQuery, useLazyGetSolidSettingsQuery = solidSettingsApi.useLazyGetSolidSettingsQuery, useLazyGetSolidSettingsByIdQuery = solidSettingsApi.useLazyGetSolidSettingsByIdQuery, useUpdateSolidSettingsMutation = solidSettingsApi.useUpdateSolidSettingsMutation, useBulkUpdateSolidSettingsMutation = solidSettingsApi.useBulkUpdateSolidSettingsMutation, useBulkUpdateSolidUserSettingsMutation = solidSettingsApi.useBulkUpdateSolidUserSettingsMutation, useGetMcpUrlQuery = solidSettingsApi.useGetMcpUrlQuery, useLazyGetMcpUrlQuery = solidSettingsApi.useLazyGetMcpUrlQuery;
