// @solidxai/core-ui/dist/redux/api/stateApi.js
import { createApi } from '@reduxjs/toolkit/query/react';
import { baseQueryWithAuth } from './fetchBaseQuery';
export var statesApi = createApi({
    reducerPath: 'statesApi',
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getStates: builder.query({
            query: function (qs) {
                return "/state?populate[]=country&".concat(qs);
            },
            transformResponse: function (response) {
                if (response.error) {
                    throw new Error(response.error);
                }
                return {
                    records: response.data.records,
                    meta: response.data.meta
                };
            }
        }),
        getStateById: builder.query({
            query: function (id) { return "/state/".concat(id, "?populate[]=country"); },
        }),
        createState: builder.mutation({
            query: function (state) { return ({
                url: '/state',
                method: 'POST',
                body: state
            }); },
        }),
        // updateState: builder.mutation({
        //     query: ({ id, data }) => ({
        //         url: `/state/${id}`,
        //         method: 'Patch',
        //         body: data,
        //     }),
        // }),
        deleteMultipleStates: builder.mutation({
            query: function (data) { return ({
                url: "/state/bulk/",
                method: 'DELETE',
                body: data
            }); },
        }),
        deleteState: builder.mutation({
            query: function (id) { return ({
                url: "/state/".concat(id),
                method: 'DELETE',
            }); },
        }),
    }); }
});
export var useLazyGetStatesQuery = statesApi.useLazyGetStatesQuery, useCreateStateMutation = statesApi.useCreateStateMutation, useDeleteStateMutation = statesApi.useDeleteStateMutation, useDeleteMultipleStatesMutation = statesApi.useDeleteMultipleStatesMutation, useGetStatesQuery = statesApi.useGetStatesQuery, useGetStateByIdQuery = statesApi.useGetStateByIdQuery, useLazyGetStateByIdQuery = statesApi.useLazyGetStateByIdQuery, usePrefetch = statesApi.usePrefetch;
