// @solidxai/core-ui/dist/redux/api/userApi.js
import { createApi } from "@reduxjs/toolkit/query/react";
import { baseQueryWithAuth } from "./fetchBaseQuery";
var resourceName = 'user';
export var userApi = createApi({
    reducerPath: "userApi",
    baseQuery: baseQueryWithAuth,
    endpoints: function (builder) { return ({
        getusers: builder.query({
            query: function (qs) {
                return "/".concat(resourceName, "?").concat(qs);
            },
            // query: () => '/${resourceName}?offset=0&limit=5',
        }),
        getusersById: builder.query({
            query: function (id) { return "/".concat(resourceName, "/").concat(id, "?populate[0]=roles"); },
        }),
        checkIfPermissionExists: builder.query({
            query: function (query) { return "/".concat(resourceName, "/permissions-exists?").concat(query); },
        }),
        registerUser: builder.mutation({
            query: function (user) { return ({
                url: "/".concat(resourceName, "/roles"),
                method: 'POST',
                body: user
            }); },
        }),
        createuserrole: builder.mutation({
            query: function (role) { return ({
                url: "/".concat(resourceName, "/roles"),
                method: 'POST',
                body: role
            }); },
        }),
        createuserrolebulk: builder.mutation({
            query: function (role) { return ({
                url: "/".concat(resourceName, "/roles/bulk"),
                method: 'POST',
                body: role
            }); },
        }),
        updateProfile: builder.mutation({
            query: function (body) {
                return {
                    url: "/me/update",
                    method: "PUT",
                    body: body,
                };
            },
        }),
        updateSession: builder.query({
            query: function () {
                return {
                    url: "/auth/session?update",
                };
            },
        }),
        updatePassword: builder.mutation({
            query: function (body) {
                return {
                    url: "/me/update_password",
                    method: "PUT",
                    body: body,
                };
            },
        }),
        uploadAvatar: builder.mutation({
            query: function (body) {
                return {
                    url: "/me/upload_avatar",
                    method: "PUT",
                    body: body,
                };
            },
        }),
        updateUser: builder.mutation({
            query: function (_a) {
                var id = _a.id, body = _a.body;
                return {
                    url: "/admin/".concat(resourceName, "/").concat(id),
                    method: "PUT",
                    body: body,
                };
            },
        }),
        deleteMultipleUsers: builder.mutation({
            query: function (data) { return ({
                url: "/".concat(resourceName, "/bulk/"),
                method: 'DELETE',
                body: data
            }); },
        }),
        deleteUser: builder.mutation({
            query: function (id) { return ({
                url: "/".concat(resourceName, "/").concat(id),
                method: 'DELETE',
            }); },
        }),
        getUser: builder.query({
            query: function (userId) { return "/user/".concat(userId, "?populateMedia[]=profilePicture&populate[]=roles"); },
        }),
        updateUserProfile: builder.mutation({
            query: function (_a) {
                var data = _a.data;
                return ({
                    url: "/user/profile",
                    method: 'POST',
                    body: data,
                });
            },
        }),
    }); },
});
export var useUpdateProfileMutation = userApi.useUpdateProfileMutation, useLazyUpdateSessionQuery = userApi.useLazyUpdateSessionQuery, useLazyCheckIfPermissionExistsQuery = userApi.useLazyCheckIfPermissionExistsQuery, useUpdatePasswordMutation = userApi.useUpdatePasswordMutation, useUploadAvatarMutation = userApi.useUploadAvatarMutation, 
// useUpdateUserMutation,
useDeleteUserMutation = userApi.useDeleteUserMutation, useGetusersQuery = userApi.useGetusersQuery, useLazyGetusersQuery = userApi.useLazyGetusersQuery, useGetusersByIdQuery = userApi.useGetusersByIdQuery, useCreateuserroleMutation = userApi.useCreateuserroleMutation, useCreateuserrolebulkMutation = userApi.useCreateuserrolebulkMutation, useDeleteMultipleUsersMutation = userApi.useDeleteMultipleUsersMutation, useLazyGetusersByIdQuery = userApi.useLazyGetusersByIdQuery, useRegisterUserMutation = userApi.useRegisterUserMutation, useUpdateSessionQuery = userApi.useUpdateSessionQuery, useGetUserQuery = userApi.useGetUserQuery, useLazyGetUserQuery = userApi.useLazyGetUserQuery, useUpdateUserProfileMutation = userApi.useUpdateUserProfileMutation;
