// @solidxai/core-ui/dist/components/core/common/SolidGenericImport/SolidImportTransaction.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Button } from 'primereact/button';
import styles from './SolidImport.module.css';
import { useCreateImportSyncMutation, useLazyGetImportMappingInfoQuery, usePatchUpdateImportTransactionMutation } from '../../../../redux/api/importTransactionApi';
import React, { useEffect, useRef, useState } from 'react';
import { Dropdown } from 'primereact/dropdown';
import { Toast } from 'primereact/toast';
import { ERROR_MESSAGES } from '../../../../constants/error-messages';
import showToast from "../../../../helpers/showToast";
export var SolidImportTransaction = function (_a) {
    // console.log("get transaction id", transactionId);
    var _b, _c, _d, _e, _f;
    var setImportStatusResult = _a.setImportStatusResult, transactionId = _a.transactionId, setImportStep = _a.setImportStep;
    var toast = useRef(null);
    var _g = useLazyGetImportMappingInfoQuery(), trigger = _g[0], _h = _g[1], mappingInfo = _h.data, isLoading = _h.isLoading, isError = _h.isError;
    var patchUpdateImportTransaction = usePatchUpdateImportTransactionMutation()[0];
    var _j = useCreateImportSyncMutation(), createImportSync = _j[0], isImporting = _j[1].isLoading;
    var _k = useState({}), fieldMapping = _k[0], setFieldMapping = _k[1];
    var _l = useState([]), visibleHeaders = _l[0], setVisibleHeaders = _l[1];
    useEffect(function () {
        if (transactionId) {
            trigger({ id: transactionId });
        }
    }, [transactionId]);
    useEffect(function () {
        var _a, _b;
        var defaultMapping = {};
        var headers = [];
        (_b = (_a = mappingInfo === null || mappingInfo === void 0 ? void 0 : mappingInfo.data) === null || _a === void 0 ? void 0 : _a.sampleImportedRecordInfo) === null || _b === void 0 ? void 0 : _b.forEach(function (sample) {
            defaultMapping[sample.cellHeader] = sample.defaultMappedFieldName || "";
            headers.push(sample.cellHeader);
        });
        setFieldMapping(defaultMapping);
        setVisibleHeaders(headers);
    }, [mappingInfo === null || mappingInfo === void 0 ? void 0 : mappingInfo.data]);
    var dropdownOptions = (_b = mappingInfo === null || mappingInfo === void 0 ? void 0 : mappingInfo.data) === null || _b === void 0 ? void 0 : _b.importableFields.map(function (field) { return ({
        label: field.displayName,
        value: field.name,
    }); });
    var handleChange = function (cellHeader, selectedField) {
        setFieldMapping(function (prev) {
            var _a;
            return (__assign(__assign({}, prev), (_a = {}, _a[cellHeader] = selectedField, _a)));
        });
    };
    var handleRemoveRow = function (cellHeader) {
        setVisibleHeaders(function (prev) { return prev.filter(function (header) { return header !== cellHeader; }); });
        setFieldMapping(function (prev) {
            var updated = __assign({}, prev);
            delete updated[cellHeader];
            return updated;
        });
    };
    var handleImportTransaction = function () { return __awaiter(void 0, void 0, void 0, function () {
        var mappingArray, patchData, patchResult, importResult, importError_1, error_1, errorMessage;
        var _a, _b, _c, _d;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0:
                    _e.trys.push([0, 6, , 7]);
                    mappingArray = Object.entries(fieldMapping).map(function (_a) {
                        var header = _a[0], fieldName = _a[1];
                        return ({
                            header: header,
                            fieldName: fieldName,
                        });
                    });
                    patchData = {
                        status: "mapping_created",
                        mapping: JSON.stringify(mappingArray),
                    };
                    return [4 /*yield*/, patchUpdateImportTransaction({
                            id: transactionId,
                            data: patchData,
                        }).unwrap()];
                case 1:
                    patchResult = _e.sent();
                    if (!((patchResult === null || patchResult === void 0 ? void 0 : patchResult.statusCode) === 200)) return [3 /*break*/, 5];
                    _e.label = 2;
                case 2:
                    _e.trys.push([2, 4, , 5]);
                    return [4 /*yield*/, createImportSync({ id: transactionId }).unwrap()];
                case 3:
                    importResult = _e.sent();
                    if (((_a = importResult === null || importResult === void 0 ? void 0 : importResult.data) === null || _a === void 0 ? void 0 : _a.status) === "import_succeeded") {
                        setImportStatusResult(importResult);
                        setImportStep(4);
                    }
                    if (((_b = importResult === null || importResult === void 0 ? void 0 : importResult.data) === null || _b === void 0 ? void 0 : _b.status) === "import_failed") {
                        setImportStatusResult(importResult);
                        setImportStep(4);
                    }
                    return [3 /*break*/, 5];
                case 4:
                    importError_1 = _e.sent();
                    showToast(toast, "error", ERROR_MESSAGES.IMPORT_ERROR, (_c = importError_1 === null || importError_1 === void 0 ? void 0 : importError_1.data) === null || _c === void 0 ? void 0 : _c.error);
                    return [3 /*break*/, 5];
                case 5: return [3 /*break*/, 7];
                case 6:
                    error_1 = _e.sent();
                    errorMessage = ((_d = error_1 === null || error_1 === void 0 ? void 0 : error_1.data) === null || _d === void 0 ? void 0 : _d.error) || ERROR_MESSAGES.SOMETHING_WRONG;
                    showToast(toast, "error", ERROR_MESSAGES.ERROR, errorMessage);
                    return [3 /*break*/, 7];
                case 7: return [2 /*return*/];
            }
        });
    }); };
    var sampleRecords = (_d = (_c = mappingInfo === null || mappingInfo === void 0 ? void 0 : mappingInfo.data) === null || _c === void 0 ? void 0 : _c.sampleImportedRecordInfo) !== null && _d !== void 0 ? _d : [];
    return (_jsxs("div", { children: [_jsx(Toast, { ref: toast }), _jsx("div", { className: styles.SolidImportContextWrapper, children: _jsxs("div", { className: 'grid m-0 relative', children: [_jsxs("div", { className: "col-6 font-bold p-3 ".concat(styles.ImportTableHeader), children: ["File Column", _jsx("div", { className: styles.TransactionsHeaderDivider })] }), _jsx("div", { className: "col-6 font-bold p-3 ".concat(styles.ImportTableHeader), children: "SolidX Field" }), sampleRecords.length > 0 ? (sampleRecords === null || sampleRecords === void 0 ? void 0 : sampleRecords.filter(function (sample) { return visibleHeaders.includes(sample.cellHeader); }).map(function (sample) {
                            var fieldMeta = mappingInfo.data.importableFields.find(function (f) { return f.name === fieldMapping[sample.cellHeader]; });
                            var isRequired = fieldMeta === null || fieldMeta === void 0 ? void 0 : fieldMeta.required;
                            // {sampleRecords.length > 0 ? (
                            //    sampleRecords    
                            //     ?.filter((sample: any) => visibleHeaders.includes(sample.cellHeader))
                            //     ?.sort((a: any, b: any) => {
                            //         const aRequired = mappingInfo.data.importableFields.find((f: any) => f.name === fieldMapping[a.cellHeader])?.required;
                            //         const bRequired = mappingInfo.data.importableFields.find((f: any) => f.name === fieldMapping[b.cellHeader])?.required;
                            //         return (bRequired ? 1 : 0) - (aRequired ? 1 : 0); // true comes before false
                            //     })
                            //     .map((sample: any) => {
                            //         const fieldMeta = mappingInfo.data.importableFields.find(
                            //             (f: any) => f.name === fieldMapping[sample.cellHeader]
                            //         );
                            //         const isRequired = fieldMeta?.required;
                            return (_jsxs(React.Fragment, { children: [_jsxs("div", { className: "col-5 md:col-6 py-2 pl-3 pr-2 md:px-3 flex flex-column justify-content-center", style: { borderBottom: "1px solid var(--gray-100)" }, children: [_jsx("div", { className: "font-medium text-primary", children: sample.cellHeader }), _jsx("div", { className: "text-sm", style: { color: "var(--solid-grey-500)" }, children: sample.cellValue || "" })] }), _jsx("div", { className: "col-7 md:col-6 py-0", style: { borderBottom: "1px solid var(--gray-100)" }, children: _jsxs("div", { className: "flex w-full justify-between align-items-center gap-2  py-2 md:px-3  ".concat(fieldMapping[sample.cellHeader] ? "col-11 sm:col-12" : "col-12", " "), children: [_jsx(Dropdown, { value: fieldMapping[sample.cellHeader], options: dropdownOptions, onChange: function (e) { return handleChange(sample.cellHeader, e.value); }, className: "w-full p-inputtext-sm", placeholder: "Select field" }), !isRequired ? (_jsx("span", { onClick: function () { return handleRemoveRow(sample.cellHeader); }, title: "Remove this row", style: { cursor: 'pointer' }, children: _jsxs("svg", { xmlns: "http://www.w3.org/2000/svg", width: "16", height: "16", viewBox: "0 0 16 16", fill: "none", children: [_jsx("rect", { width: "16", height: "16", rx: "8", fill: "#F0F0F0" }), _jsx("path", { d: "M5.6 11L5 10.4L7.4 8L5 5.6L5.6 5L8 7.4L10.4 5L11 5.6L8.6 8L11 10.4L10.4 11L8 8.6L5.6 11Z", fill: "#4B4D52" })] }) }))
                                                    :
                                                        _jsx("span", { style: { height: 16, width: 16 } })] }) })] }, sample.cellHeader));
                        })) : (_jsxs("div", { className: 'col-12 flex flex-column align-items-center', children: [_jsx("h4", { className: 'text-center px-2', children: "No Sample Imported Record Info Found" }), _jsx("p", { children: "Please Add Records" })] }))] }) }), _jsx("div", { className: 'mt-3 flex align-items-center gap-3', children: _jsx(Button, { label: 'Import', size: 'small', onClick: handleImportTransaction, loading: isImporting, disabled: isImporting || ((_f = (_e = mappingInfo === null || mappingInfo === void 0 ? void 0 : mappingInfo.data) === null || _e === void 0 ? void 0 : _e.sampleImportedRecordInfo) === null || _f === void 0 ? void 0 : _f.length) === 0 }) })] }));
};
