// @solidxai/core-ui/dist/components/core/common/SolidGenericImport/SolidImportTransactionStatus.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { ERROR_MESSAGES } from '../../../../constants/error-messages';
import { getSession } from "../../../../adapters/auth/index";
import { Button } from 'primereact/button';
import { Dialog } from 'primereact/dialog';
import { useState } from 'react';
import { env } from "../../../../adapters/env";
export var SolidImportTransactionStatus = function (_a) {
    var importStatusResult = _a.importStatusResult, transactionId = _a.transactionId, setOpenImportDialog = _a.setOpenImportDialog, handleFetchUpdatedRecords = _a.handleFetchUpdatedRecords;
    var _b = useState(false), showPartialDialog = _b[0], setShowPartialDialog = _b[1];
    var handleDownloadFailedRecords = function (transactionId) { return __awaiter(void 0, void 0, void 0, function () {
        var session, token, response, blob, url, a, contentDisposition, match, error_1;
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _c.trys.push([0, 4, , 5]);
                    return [4 /*yield*/, getSession()];
                case 1:
                    session = _c.sent();
                    token = ((_a = session === null || session === void 0 ? void 0 : session.user) === null || _a === void 0 ? void 0 : _a.accessToken) || "";
                    if (!token) {
                        throw new Error(ERROR_MESSAGES.NO_AUTH_TOKEN_FOUND);
                    }
                    return [4 /*yield*/, fetch("".concat(env("API_URL"), "/api/import-transaction/").concat(transactionId, "/export-failed-import-records"), {
                            method: 'GET',
                            headers: {
                                Authorization: "Bearer ".concat(token),
                            },
                        })];
                case 2:
                    response = _c.sent();
                    if (!response.ok)
                        throw new Error(ERROR_MESSAGES.FAILED_TO_DOWNLOAD);
                    return [4 /*yield*/, response.blob()];
                case 3:
                    blob = _c.sent();
                    url = window.URL.createObjectURL(blob);
                    a = document.createElement("a");
                    a.href = url;
                    contentDisposition = response.headers.get("Content-Disposition");
                    match = contentDisposition === null || contentDisposition === void 0 ? void 0 : contentDisposition.match(/filename="(.+)"/);
                    a.download = (_b = match === null || match === void 0 ? void 0 : match[1]) !== null && _b !== void 0 ? _b : "failed-records.xlsx"; // fallback if not found
                    document.body.appendChild(a);
                    a.click();
                    a.remove();
                    window.URL.revokeObjectURL(url);
                    return [3 /*break*/, 5];
                case 4:
                    error_1 = _c.sent();
                    console.error(ERROR_MESSAGES.DOWNLOAD_FAILED, error_1);
                    return [3 /*break*/, 5];
                case 5: return [2 /*return*/];
            }
        });
    }); };
    var _c = (importStatusResult === null || importStatusResult === void 0 ? void 0 : importStatusResult.data) || {}, status = _c.status, _d = _c.failedCount, failedCount = _d === void 0 ? 0 : _d, _e = _c.importedCount, importedCount = _e === void 0 ? 0 : _e;
    var handleSuccessSyncImport = function () {
        setOpenImportDialog(false);
        setTimeout(function () {
            handleFetchUpdatedRecords();
        }, 150);
    };
    var handlePartialDialogConfirm = function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, handleDownloadFailedRecords(transactionId)];
                case 1:
                    _a.sent();
                    setOpenImportDialog(false);
                    setTimeout(function () {
                        handleFetchUpdatedRecords();
                    }, 150);
                    return [2 /*return*/];
            }
        });
    }); };
    var getStatusTag = function () {
        if (status === "import_succeeded") {
            return _jsx(Button, { type: 'button', size: 'small', icon: "pi pi-check-circle", severity: "success", label: "".concat(importedCount, " row").concat(importedCount > 1 ? 's' : '', " imported successfully") });
        }
        if (status === "import_failed" && importedCount === 0) {
            return _jsx(Button, { type: 'button', size: 'small', icon: "pi pi-times-circle", severity: "danger", label: "Import Failed" });
        }
        if (status === "import_failed" && importedCount > 0) {
            return _jsx(Button, { type: 'button', size: 'small', icon: "pi pi-exclamation-triangle", severity: "warning", label: "Completed with Some Errors" });
        }
        return _jsx("div", { children: "Import Status" });
    };
    return (_jsxs("div", { children: [_jsxs("div", { style: { borderRadius: 6, border: '1px solid var(--primary-light-color)' }, children: [_jsx("div", { className: 'p-3', style: { borderBottom: '1px solid var(--primary-light-color)', pointerEvents: 'none' }, children: getStatusTag() }), _jsxs("div", { className: 'font-bold', style: { color: 'var(--solid-primary-black)' }, children: [_jsxs("div", { className: 'p-3 flex justify-content-between', children: [_jsx("span", { children: "No. of Successful Rows" }), " ", _jsx("span", { children: importedCount !== null && importedCount !== void 0 ? importedCount : 0 })] }), _jsx("div", { className: 'px-3', children: _jsx("div", { style: { borderTop: '1px dashed var(--primary-light-color)' } }) }), _jsxs("div", { className: 'p-3 flex justify-content-between', children: [_jsx("span", { children: "No. of Failed Rows" }), " ", _jsx("span", { children: failedCount !== null && failedCount !== void 0 ? failedCount : 0 })] })] })] }), _jsxs("div", { className: 'mt-3 flex align-items-center gap-3', children: [importedCount > 0 &&
                        _jsx(Button, { label: 'View Imported Records', size: 'small', onClick: function () {
                                if (status === "import_failed" && importedCount > 0) {
                                    setShowPartialDialog(true);
                                }
                                else {
                                    handleSuccessSyncImport();
                                }
                            } }), status === "import_failed" && transactionId &&
                        _jsx(Button, { label: "Download Failed Records", size: 'small', icon: "pi pi-download", outlined: true, severity: 'secondary', onClick: function () { return handleDownloadFailedRecords(transactionId); } })] }), _jsx(Dialog, { header: "View Imported Records", visible: showPartialDialog, onHide: function () { return setShowPartialDialog(false); }, style: {
                    width: '50vh'
                }, footer: _jsxs("div", { className: "flex justify-content-start gap-2", children: [_jsx(Button, { label: "Yes", size: 'small', icon: "pi pi-download", onClick: handlePartialDialogConfirm }), _jsx(Button, { label: "No", size: 'small', icon: "pi pi-times", severity: "secondary", outlined: true, onClick: function () {
                                handleSuccessSyncImport();
                            } })] }), modal: true, children: _jsx("p", { children: "Do you want to download failed records before viewing the successful imports?" }) })] }));
};
