// @solidxai/core-ui/dist/components/core/chatter/SolidChatter.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useState, useEffect } from 'react';
import { SolidChatterHeader } from './SolidChatterHeader';
import { SolidChatterDateDivider } from './SolidChatterDateDivider';
import { SolidChatterMessageBox } from './SolidChatterMessageBox';
import { useLazyGetchatterMessageQuery } from '../../../redux/api/solidChatterMessageApi';
import qs from "qs";
import { ERROR_MESSAGES } from '../../../constants/error-messages';
import { permissionExpression } from '../../../helpers/permissions';
export var SolidChatter = function (_a) {
    var modelSingularName = _a.modelSingularName, id = _a.id, refreshChatterMessage = _a.refreshChatterMessage, setRefreshChatterMessage = _a.setRefreshChatterMessage, _b = _a.actionsAllowed, actionsAllowed = _b === void 0 ? [] : _b;
    var _c = useState('email-message'), activeTab = _c[0], setActiveTab = _c[1];
    var _d = useState(null), visibleBox = _d[0], setVisibleBox = _d[1];
    var _e = useState([]), messages = _e[0], setMessages = _e[1];
    var _f = useState(25), limit = _f[0], setLimit = _f[1];
    var _g = useState(0), totalRecords = _g[0], setTotalRecords = _g[1];
    var _h = useState({
        name: '',
        startDate: null,
        endDate: null
    }), filters = _h[0], setFilters = _h[1];
    var _j = useLazyGetchatterMessageQuery(), getchatterMessage = _j[0], isChatterLoading = _j[1].isLoading;
    useEffect(function () {
        if (refreshChatterMessage) {
            if (id !== 'new') {
                fetchData();
                setRefreshChatterMessage(false);
            }
        }
    }, [refreshChatterMessage]);
    useEffect(function () {
        if (id !== 'new') {
            fetchData();
        }
    }, [filters, limit]);
    var handleTabClick = function (tab) {
        setActiveTab(tab);
        setVisibleBox(function (prev) { return (prev === tab ? null : tab); });
    };
    var formatDate = function (dateString) {
        var date = new Date(dateString);
        var today = new Date();
        var yesterday = new Date(today);
        yesterday.setDate(yesterday.getDate() - 1);
        if (date.toDateString() === today.toDateString()) {
            return 'Today';
        }
        else if (date.toDateString() === yesterday.toDateString()) {
            return 'Yesterday';
        }
        else {
            return date.toLocaleDateString('en-US', {
                year: 'numeric',
                month: 'long',
                day: 'numeric'
            });
        }
    };
    var fetchData = function () { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, queryString, response, processedMessages, error_1;
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _c.trys.push([0, 2, , 3]);
                    queryData = {
                        populate: ['user', 'chatterMessageDetails'],
                        limit: limit
                    };
                    if (filters.name) {
                        queryData.filters = queryData.filters || {};
                        queryData.filters['user'] = { fullName: { $containsi: filters.name } };
                    }
                    if (filters.startDate && filters.endDate) {
                        queryData.filters = queryData.filters || {};
                        queryData.filters.createdAt = {
                            $between: [filters.startDate.toISOString(), filters.endDate.toISOString()]
                        };
                    }
                    else if (filters.startDate) {
                        queryData.filters = queryData.filters || {};
                        queryData.filters.createdAt = { $gte: filters.startDate.toISOString() };
                    }
                    else if (filters.endDate) {
                        queryData.filters = queryData.filters || {};
                        queryData.filters.createdAt = { $lte: filters.endDate.toISOString() };
                    }
                    queryString = qs.stringify(queryData, {
                        encodeValuesOnly: true,
                    });
                    return [4 /*yield*/, getchatterMessage({
                            entityId: id,
                            entityName: modelSingularName,
                            qs: queryString
                        }).unwrap()];
                case 1:
                    response = _c.sent();
                    processedMessages = response.data.records.map(function (msg) {
                        var _a, _b, _c;
                        if (msg.messageType === 'custom') {
                            // Custom message
                            return {
                                id: msg.id,
                                user: ((_a = msg.user) === null || _a === void 0 ? void 0 : _a.fullName) || "System",
                                messageType: "custom",
                                message: msg.messageBody,
                                time: new Date(msg.createdAt).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
                                createdAt: msg.createdAt,
                                date: formatDate(msg.createdAt),
                                media: msg._media,
                                messageSubType: msg.messageSubType,
                                modelDisplayName: msg.modelDisplayName,
                                modelUserKey: msg.modelUserKey
                            };
                        }
                        else {
                            // Audit message
                            var auditRecord = ((_b = msg.chatterMessageDetails) === null || _b === void 0 ? void 0 : _b.map(function (detail) { return ({
                                field: detail.fieldName,
                                fieldDisplayName: detail.fieldDisplayName,
                                previous: detail.oldValueDisplay || detail.oldValue || 'None',
                                current: detail.newValueDisplay || detail.newValue
                            }); })) || [];
                            return {
                                id: msg.id,
                                user: ((_c = msg.user) === null || _c === void 0 ? void 0 : _c.fullName) || "System",
                                messageType: "audit",
                                time: new Date(msg.createdAt).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' }),
                                auditRecord: auditRecord,
                                createdAt: msg.createdAt,
                                date: formatDate(msg.createdAt),
                                messageSubType: msg.messageSubType,
                                modelDisplayName: msg.modelDisplayName,
                                modelUserKey: msg.modelUserKey,
                                message: msg.messageBody,
                            };
                        }
                    });
                    setMessages(processedMessages);
                    setTotalRecords(((_b = (_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.meta) === null || _b === void 0 ? void 0 : _b.totalRecords) || 0);
                    return [3 /*break*/, 3];
                case 2:
                    error_1 = _c.sent();
                    console.error(ERROR_MESSAGES.FETCHING_MESSAGE, error_1);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    var handleFilterChange = function (newFilters) {
        setFilters({
            name: newFilters.name,
            startDate: newFilters.startDate,
            endDate: newFilters.endDate
        });
        setLimit(25);
    };
    var handleLoadMore = function () {
        setLimit(function (prevLimit) { return prevLimit + 25; });
    };
    return (_jsxs("div", { className: 'h-full', children: [_jsx(SolidChatterHeader, { id: id, refetch: fetchData, modelSingularName: modelSingularName, activeTab: activeTab, handleTabClick: handleTabClick, visibleBox: visibleBox, onFilterChange: handleFilterChange }), _jsx("div", { className: 'p-3', style: {
                    overflowY: 'scroll',
                    height: visibleBox === 'email-message'
                        ? 'calc(100vh - 196px)'
                        : visibleBox === 'log'
                            ? 'calc(100vh - 172px)'
                            : 'calc(100vh - 65px)',
                }, children: isChatterLoading ? (_jsx("div", { className: 'flex align-items-center justify-content-center h-full font-medium', children: "Loading..." })) : messages.length === 0 ? (_jsx("div", { className: 'flex align-items-center justify-content-center h-full font-medium', children: "No Data Available" })) : (_jsxs(_Fragment, { children: [actionsAllowed.includes("".concat(permissionExpression('chatterMessage', 'findMany'))) ?
                            (messages.map(function (message, index) {
                                var showDateDivider = index === 0 || message.date !== messages[index - 1].date;
                                return (_jsxs("div", { children: [showDateDivider && _jsx(SolidChatterDateDivider, { date: message.date }), _jsx(SolidChatterMessageBox, { user: message.user, messageType: message.messageType, message: message.message, time: message.time, auditRecord: message.auditRecord, media: message.media, messageSubType: message.messageSubType, modelDisplayName: message.modelDisplayName, modelUserKey: message.modelUserKey })] }, message.id));
                            }))
                            : _jsx("div", { className: "p-3", children: _jsxs("div", { className: 'p-3 border-round text-red-500 bg-red-50 flex align-items-center gap-3', children: [_jsx("i", { className: "pi pi-exclamation-triangle" }), "You do not have permission to view these messages."] }) }), totalRecords > messages.length && (_jsx("div", { className: 'flex justify-content-center mt-3 mb-3', children: _jsx("span", { onClick: handleLoadMore, className: 'cursor-pointer text-primary font-medium px-4 py-2', style: {
                                    userSelect: 'none',
                                    borderRadius: '6px',
                                    transition: 'all 0.2s ease',
                                    display: 'inline-block'
                                }, onMouseEnter: function (e) {
                                    e.currentTarget.style.backgroundColor = 'rgba(59, 130, 246, 0.1)';
                                    e.currentTarget.style.transform = 'translateY(-1px)';
                                }, onMouseLeave: function (e) {
                                    e.currentTarget.style.backgroundColor = 'transparent';
                                    e.currentTarget.style.transform = 'translateY(0)';
                                }, children: "Load More..." }) }))] })) })] }));
};
