// @solidxai/core-ui/dist/components/core/chatter/SolidChatterHeader.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Button } from 'primereact/button';
import { Calendar } from 'primereact/calendar';
import { Dialog } from 'primereact/dialog';
import { AutoComplete } from 'primereact/autocomplete';
import { useEffect, useState } from 'react';
import styles from './chatter.module.css';
import { SolidMessageComposer } from './SolidMessageComposer';
import { useLazyGetusersQuery } from '../../../redux/api/userApi';
import { ERROR_MESSAGES } from '../../../constants/error-messages';
export var SolidChatterHeader = function (props) {
    var activeTab = props.activeTab, visibleBox = props.visibleBox, handleTabClick = props.handleTabClick, modelSingularName = props.modelSingularName, refetch = props.refetch, id = props.id, onFilterChange = props.onFilterChange;
    var _a = useState(false), showFilterDialog = _a[0], setShowFilterDialog = _a[1];
    var _b = useState({
        name: '',
        startDate: null,
        endDate: null
    }), filters = _b[0], setFilters = _b[1];
    var _c = useState(false), hasActiveFilters = _c[0], setHasActiveFilters = _c[1];
    var _d = useState([]), suggestions = _d[0], setSuggestions = _d[1];
    var getUsers = useLazyGetusersQuery()[0];
    useEffect(function () {
        var isActive = filters.name !== '' || filters.startDate !== null || filters.endDate !== null;
        setHasActiveFilters(isActive);
    }, [filters]);
    var handleFilterClick = function () {
        setShowFilterDialog(true);
    };
    var handleApplyFilters = function () {
        if (onFilterChange) {
            onFilterChange(filters);
        }
        setShowFilterDialog(false);
    };
    var handleClearFilters = function () {
        var clearedFilters = {
            name: '',
            startDate: null,
            endDate: null
        };
        setFilters(clearedFilters);
        if (onFilterChange) {
            onFilterChange(clearedFilters);
        }
        setShowFilterDialog(false);
    };
    var searchUsers = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var response, error_1;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _b.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, getUsers("filters[fullName][$containsi]=".concat(event.query)).unwrap()];
                case 1:
                    response = _b.sent();
                    setSuggestions(((_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.records) || []);
                    return [3 /*break*/, 3];
                case 2:
                    error_1 = _b.sent();
                    console.error(ERROR_MESSAGES.FETCHING_USER, error_1);
                    setSuggestions([]);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    var itemTemplate = function (item) {
        return (_jsx("div", { className: "flex align-items-center", children: _jsx("div", { children: item.fullName }) }));
    };
    return (_jsxs("div", { className: styles.chatterTitle, children: [_jsxs("div", { className: 'flex justify-content-between align-items-center', children: [_jsx("div", { className: "form-wrapper-title ", children: "Activity" }), _jsxs("div", { className: 'flex align-items-center gap-2', children: [_jsx(Button, __assign({ label: "Log Note", size: "small", type: "button", onClick: function () { return handleTabClick('log'); } }, (activeTab !== 'log' && {
                                severity: 'secondary',
                                outlined: true,
                            }))), _jsx(Button, { icon: "pi pi-filter-fill", text: true, size: 'small', severity: "secondary", "aria-label": "Filter", style: {
                                    width: 20,
                                    color: hasActiveFilters ? '#722ED1' : undefined
                                }, onClick: handleFilterClick }), _jsx(Button, { icon: "pi pi-refresh", text: true, size: 'small', severity: "secondary", "aria-label": "Refresh", style: {
                                    width: 20
                                }, onClick: refetch })] })] }), visibleBox &&
                _jsxs("div", { className: 'mt-4', children: [visibleBox === "email-message" &&
                            _jsx(SolidMessageComposer, { id: id, refetch: refetch, modelSingularName: modelSingularName, type: "email" }), visibleBox === "log" &&
                            _jsx(SolidMessageComposer, { id: id, refetch: refetch, modelSingularName: modelSingularName })] }), _jsx(Dialog, { header: "Filter Messages", visible: showFilterDialog, style: { width: 500 }, onHide: function () { return setShowFilterDialog(false); }, className: 'solid-chatter-filter-main', children: _jsxs("div", { className: "p-fluid", children: [_jsxs("div", { className: "flex flex-column gap-2", children: [_jsx("label", { htmlFor: "name", children: "Name:" }), _jsx(AutoComplete, { id: "fullName", value: filters.name, suggestions: suggestions, completeMethod: searchUsers, field: "fullName", onChange: function (e) { return setFilters(function (prev) { return (__assign(__assign({}, prev), { name: typeof e.value === 'object' ? e.value.fullName : e.value })); }); }, placeholder: "Filter By User", itemTemplate: itemTemplate, className: "w-full" })] }), _jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { htmlFor: "dateRange", children: "Date Range" }), _jsxs("div", { className: "flex gap-2", children: [_jsx("div", { className: "flex-1", children: _jsx(Calendar, { id: "startDate", value: filters.startDate, onChange: function (e) { return setFilters(function (prev) { return (__assign(__assign({}, prev), { startDate: e.value })); }); }, placeholder: "Start Date", showIcon: true, dateFormat: "dd/mm/yy", className: "w-full" }) }), _jsx("div", { className: "flex-1", children: _jsx(Calendar, { id: "endDate", value: filters.endDate, onChange: function (e) { return setFilters(function (prev) { return (__assign(__assign({}, prev), { endDate: e.value })); }); }, placeholder: "End Date", showIcon: true, dateFormat: "dd/mm/yy", className: "w-full" }) })] })] }), _jsxs("div", { className: "flex justify-content-end gap-2 mt-4", children: [_jsx(Button, { label: "Clear", severity: "secondary", outlined: true, onClick: handleClearFilters }), _jsx(Button, { label: "Apply", onClick: handleApplyFilters })] })] }) })] }));
};
