// @solidxai/core-ui/dist/redux/features/settingsSlice.js
import { createSlice } from "@reduxjs/toolkit";
const initialState = {
    authSettings: {},
    solidSettings: {},
};
export const settingsSlice = createSlice({
    name: "settingsSlice",
    initialState,
    reducers: {
        // // -------- AUTH SETTINGS --------
        // setAuthSettings: (
        //   state,
        //   action: PayloadAction<Record<string, any>>
        // ) => {
        //   state.authSettings = action.payload;
        // },
        // updateAuthSetting: (
        //   state,
        //   action: PayloadAction<{ key: string; value: any }>
        // ) => {
        //   state.authSettings[action.payload.key] = action.payload.value;
        // },
        // -------- SOLID SETTINGS --------
        setSolidSettings: (state, action) => {
            state.solidSettings = action.payload;
        },
        updateSolidSetting: (state, action) => {
            state.solidSettings[action.payload.key] = action.payload.value;
        },
        // -------- RESET --------
        resetSettings: () => initialState,
    },
});
export default settingsSlice.reducer;
export const { 
// setAuthSettings,
// updateAuthSetting,
setSolidSettings, updateSolidSetting, resetSettings, } = settingsSlice.actions;
