// @solidxai/core-ui/dist/helpers/eventBus.js
var EventBus = /** @class */ (function () {
    function EventBus() {
        Object.defineProperty(this, "listeners", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new Map()
        });
    }
    Object.defineProperty(EventBus.prototype, "on", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (event, handler) {
            var _this = this;
            if (!this.listeners.has(event)) {
                this.listeners.set(event, new Set());
            }
            this.listeners.get(event).add(handler);
            return function () { return _this.off(event, handler); };
        }
    });
    Object.defineProperty(EventBus.prototype, "off", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (event, handler) {
            var set = this.listeners.get(event);
            if (!set)
                return;
            set.delete(handler);
            if (set.size === 0)
                this.listeners.delete(event);
        }
    });
    Object.defineProperty(EventBus.prototype, "emit", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (event, payload) {
            var set = this.listeners.get(event);
            if (!set)
                return;
            for (var _i = 0, _a = Array.from(set); _i < _a.length; _i++) {
                var handler = _a[_i];
                try {
                    handler(payload);
                }
                catch (_b) {
                    // swallow to avoid breaking other listeners
                }
            }
        }
    });
    return EventBus;
}());
export var eventBus = new EventBus();
export var AppEvents = {
    SessionUpdated: "session:updated",
    SessionCleared: "session:cleared",
    GlobalError: "error:global",
};
