// @solidxai/core-ui/dist/helpers/helpers.js
import { camelCase, kebabCase, snakeCase } from "lodash";
import moment from "moment";
import pluralize from "pluralize";
import qs from "qs";
export var calculateDaysOfStay = function (checkInDate, checkOutDate) {
    var startDate = moment(checkInDate);
    var endDate = moment(checkOutDate);
    return endDate.diff(startDate, "days") + 1;
};
export var addCommasToAmount = function (amount) {
    amount = amount.toString();
    var parts = amount.split(".");
    parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",");
    return parts.join(".");
};
export var getSingularAndPlural = function (displayName) {
    // Convert display name to kebab-case (lowercase with hyphens)
    var toKebabCase = kebabCase(displayName);
    var toSnakeCase = snakeCase(displayName);
    var toCamelCase = camelCase(displayName);
    // Get the plural form using pluralize
    var toPluralKebabCase = pluralize(toKebabCase);
    var toPluralCamelCase = pluralize(toCamelCase);
    return { toKebabCase: toKebabCase, toSnakeCase: toSnakeCase, toPluralKebabCase: toPluralKebabCase, toPluralCamelCase: toPluralCamelCase, toCamelCase: toCamelCase };
};
function cleanForQuery(values) {
    // remove files, blobs, data: URIs, null/undefined
    return JSON.parse(JSON.stringify(values, function (k, v) {
        if (v === null || v === undefined)
            return undefined;
        if (v instanceof File || v instanceof Blob)
            return undefined;
        if (typeof v === "string" && v.startsWith("data:"))
            return undefined;
        return v;
    }));
}
export function formikValuestoQueryString(values) {
    var filtered = cleanForQuery(values);
    return qs.stringify(filtered, {
        encode: true,
        arrayFormat: "brackets",
        allowDots: true,
        skipNulls: true,
    });
}
