// @solidxai/core-ui/dist/helpers/hydrateRelationRules.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { FilterRuleType } from "../components/core/common/FilterComponent";
import { kebabCase } from "change-case";
import { getSession } from "../adapters/auth/index";
import { env } from "../adapters/env";
import { solidGet } from "../http/solidHttp";
var API_URL = env("NEXT_PUBLIC_BACKEND_API_URL");
export function hydrateRelationRules(rootRules, viewData) {
    var _a, _b, _c, _d, _e, _f;
    return __awaiter(this, void 0, void 0, function () {
        var relationMap, walk, buckets, session, token, _loop_1, _i, buckets_1, bucket;
        return __generator(this, function (_g) {
            switch (_g.label) {
                case 0:
                    relationMap = new Map();
                    walk = function (rule) {
                        var _a, _b;
                        if (rule.type === FilterRuleType.RULE && rule.fieldName) {
                            var meta = (_b = (_a = viewData === null || viewData === void 0 ? void 0 : viewData.data) === null || _a === void 0 ? void 0 : _a.solidFieldsMetadata) === null || _b === void 0 ? void 0 : _b[rule.fieldName];
                            if ((meta === null || meta === void 0 ? void 0 : meta.type) === "relation" &&
                                meta.relationType === "many-to-many" &&
                                Array.isArray(rule.value) &&
                                typeof rule.value[0] === "number") {
                                if (!relationMap.has(rule.fieldName)) {
                                    relationMap.set(rule.fieldName, {
                                        meta: meta,
                                        rules: [],
                                        ids: new Set()
                                    });
                                }
                                var bucket_1 = relationMap.get(rule.fieldName);
                                rule.value.forEach(function (id) { return bucket_1.ids.add(id); });
                                bucket_1.rules.push(rule);
                            }
                        }
                        if (Array.isArray(rule.children)) {
                            rule.children.forEach(walk);
                        }
                    };
                    rootRules.forEach(walk);
                    buckets = Array.from(relationMap.values());
                    return [4 /*yield*/, getSession()];
                case 1:
                    session = _g.sent();
                    token = (_a = session === null || session === void 0 ? void 0 : session.user) === null || _a === void 0 ? void 0 : _a.accessToken;
                    if (!token) {
                        console.error("hydrateRelationRules: Not authenticated");
                        return [2 /*return*/, rootRules];
                    }
                    _loop_1 = function (bucket) {
                        var meta, rules, ids, kebabEntityName, resp, records, labelField, idToLabel;
                        return __generator(this, function (_h) {
                            switch (_h.label) {
                                case 0:
                                    meta = bucket.meta, rules = bucket.rules, ids = bucket.ids;
                                    if (!meta.relationCoModelSingularName ||
                                        !((_c = (_b = meta.relationModel) === null || _b === void 0 ? void 0 : _b.userKeyField) === null || _c === void 0 ? void 0 : _c.name)) {
                                        return [2 /*return*/, "continue"];
                                    }
                                    kebabEntityName = kebabCase(meta.relationCoModelSingularName);
                                    return [4 /*yield*/, solidGet("".concat(API_URL, "/api/").concat(kebabEntityName), {
                                            params: {
                                                offset: 0,
                                                limit: 100,
                                                filters: {
                                                    id: { $in: Array.from(ids) },
                                                },
                                            },
                                            headers: {
                                                Authorization: "Bearer ".concat(token),
                                            },
                                        })];
                                case 1:
                                    resp = _h.sent();
                                    records = (_f = (_e = (_d = resp === null || resp === void 0 ? void 0 : resp.data) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.records) !== null && _f !== void 0 ? _f : [];
                                    labelField = meta.relationModel.userKeyField.name;
                                    idToLabel = new Map(records.map(function (item) { return [
                                        item.id,
                                        item[labelField]
                                    ]; }));
                                    // 3️⃣ Replace numeric IDs with { label, value }
                                    rules.forEach(function (rule) {
                                        rule.value = rule.value.map(function (id) {
                                            var _a;
                                            return ({
                                                label: (_a = idToLabel.get(id)) !== null && _a !== void 0 ? _a : String(id),
                                                value: id
                                            });
                                        });
                                    });
                                    return [2 /*return*/];
                            }
                        });
                    };
                    _i = 0, buckets_1 = buckets;
                    _g.label = 2;
                case 2:
                    if (!(_i < buckets_1.length)) return [3 /*break*/, 5];
                    bucket = buckets_1[_i];
                    return [5 /*yield**/, _loop_1(bucket)];
                case 3:
                    _g.sent();
                    _g.label = 4;
                case 4:
                    _i++;
                    return [3 /*break*/, 2];
                case 5: return [2 /*return*/, rootRules];
            }
        });
    });
}
