// @solidxai/core-ui/dist/helpers/logger.js
import { env } from "../adapters/env";
var isDev = env("SOLIDX_ENV", "prod") === "dev";
function shouldLog(level) {
    if (!isDev && level === "debug")
        return false;
    return true;
}
export var logger = {
    debug: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (!shouldLog("debug"))
            return;
        // eslint-disable-next-line no-console
        console.debug.apply(console, args);
    },
    info: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (!shouldLog("info"))
            return;
        // eslint-disable-next-line no-console
        console.info.apply(console, args);
    },
    warn: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (!shouldLog("warn"))
            return;
        // eslint-disable-next-line no-console
        console.warn.apply(console, args);
    },
    error: function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (!shouldLog("error"))
            return;
        // eslint-disable-next-line no-console
        console.error.apply(console, args);
    },
};
