// @solidxai/core-ui/dist/helpers/ToastContainer.js
import { jsx as _jsx } from "react/jsx-runtime";
import { ERROR_MESSAGES } from '../constants/error-messages';
import { Toast } from 'primereact/toast';
import { useRef } from 'react';
var toastRef = null;
export var ToastContainer = function () {
    toastRef = useRef(null);
    return _jsx(Toast, { ref: toastRef });
};
function isFetchBaseQueryErrorWithErrorResponse(error) {
    return error && typeof error === 'object' && 'data' in error && 'message' in error.data;
}
var extractErrorMessages = function (errorObj) {
    return Object.values(errorObj);
};
export var handleError = function (errorToast) {
    var _a;
    var errorMessage = errorToast ? errorToast : [ERROR_MESSAGES.ERROR_OCCURED];
    if (typeof errorMessage == 'object') {
        if (isFetchBaseQueryErrorWithErrorResponse(errorToast)) {
            errorMessage = errorToast.data.message;
        }
    }
    // if (isFetchBaseQueryErrorWithErrorResponse(errorToast)) {
    //     errorMessage = errorToast.data.message;
    // } else {
    //     errorMessage = ['Something went wrong'];
    // }
    (_a = toastRef === null || toastRef === void 0 ? void 0 : toastRef.current) === null || _a === void 0 ? void 0 : _a.show({
        severity: 'error',
        summary: ERROR_MESSAGES.ERROR,
        detail: errorMessage,
        life: 3000,
        //@ts-ignore
        content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: Array.isArray(errorMessage) ? (errorMessage.map(function (message, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: message.trim() }) }, index)); })) : (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: errorMessage === null || errorMessage === void 0 ? void 0 : errorMessage.trim() }) })) })); },
    });
};
export var handleSuccess = function (successMessage) {
    var _a;
    (_a = toastRef === null || toastRef === void 0 ? void 0 : toastRef.current) === null || _a === void 0 ? void 0 : _a.show({
        severity: 'error',
        summary: ERROR_MESSAGES.IS_SUCCESS,
        detail: successMessage,
        life: 3000,
        //@ts-ignore
        content: function (props) { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: Array.isArray(successMessage) ? (successMessage.map(function (message, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: message.trim() }) }, index)); })) : (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: successMessage === null || successMessage === void 0 ? void 0 : successMessage.trim() }) })) })); },
    });
};
