// @solidxai/core-ui/dist/components/core/list/SolidVarInputsFilterElement.js
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { Button } from "primereact/button";
import { Calendar } from "primereact/calendar";
import { InputNumber } from "primereact/inputnumber";
import { InputText } from "primereact/inputtext";
import { useEffect, useState } from "react";
import { SolidManyToOneFilterElement } from "./SolidManyToOneFilterElement";
import { SolidSelectionDynamicFilterElement } from "./SolidSelectionDynamicFilterElement";
import { SolidSelectionStaticFilterElement } from "./SolidSelectionStaticFilterElement";
export var InputTypes;
(function (InputTypes) {
    InputTypes["Date"] = "Date";
    InputTypes["DateTime"] = "DateTime";
    InputTypes["Time"] = "Time";
    InputTypes["Numeric"] = "Numeric";
    InputTypes["Text"] = "Text";
    InputTypes["SelectionStatic"] = "SelectionStatic";
    InputTypes["RelationManyToOne"] = "RelationManyToOne";
    InputTypes["SelectionDynamic"] = "SelectionDynamic";
})(InputTypes || (InputTypes = {}));
// Based on numberOfInputs map the input filed and hide add and delete 
export var SolidVarInputsFilterElement = function (_a) {
    var values = _a.values, onChange = _a.onChange, _b = _a.inputType, inputType = _b === void 0 ? InputTypes.Text : _b, _c = _a.numberOfInputs, numberOfInputs = _c === void 0 ? null : _c, solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata;
    if (!values) {
        values = numberOfInputs && numberOfInputs > 0 ? Array(numberOfInputs).fill('') : [''];
        // values = [''];
    }
    else {
        if (values[0] == '') {
            values = numberOfInputs && numberOfInputs > 0 ? Array(numberOfInputs).fill('') : [''];
        }
        else {
            values = values;
        }
    }
    // TODO: Ideally values will be an array, so we can spread them here instead of making a nested array.
    var _d = useState(__spreadArray([], values, true)), inputs = _d[0], setInputs = _d[1];
    useEffect(function () {
        setInputs(__spreadArray([], values, true));
    }, [numberOfInputs]);
    var updateInputs = function (index, value) {
        var updatedSpecification = inputs.map(function (item, i) {
            return i === index ? value : item;
        });
        setInputs(updatedSpecification);
    };
    var addInput = function () {
        setInputs(__spreadArray(__spreadArray([], inputs, true), [''], false));
    };
    var deleteInput = function (index) {
        if (inputs.length > 1) {
            var updatedRows = inputs.filter(function (_, rowIndex) { return rowIndex !== index; });
            setInputs(updatedRows);
        }
        else {
        }
    };
    useEffect(function () {
        onChange(inputs);
    }, [inputs]);
    return (_jsx("div", { className: "grid formgrid", children: _jsx("div", { className: "col-12", children: _jsx("div", { className: "flex flex-column gap-2", children: inputs && inputs.map(function (value, index) { return (_jsxs("div", { className: "flex align-items-center gap-2", children: [inputType === InputTypes.Text &&
                            _jsx(InputText, { value: value, onChange: function (e) { return updateInputs(index, e.target.value); }, placeholder: "", className: "p-inputtext-sm small-input w-full" }), inputType === InputTypes.Numeric &&
                            _jsx(InputNumber, { value: value, onChange: function (e) { return updateInputs(index, e.value); }, placeholder: "", className: "p-inputtext-sm small-input w-full" }), inputType === InputTypes.Date &&
                            _jsx(Calendar, { value: value, onChange: function (e) { return updateInputs(index, e.target.value); }, dateFormat: "mm/dd/yy", placeholder: "mm/dd/yyyy", mask: "99/99/9999" }), inputType === InputTypes.DateTime &&
                            _jsx(Calendar, { value: value, onChange: function (e) { return updateInputs(index, e.target.value); }, dateFormat: "mm/dd/yy", placeholder: "mm/dd/yyyy hh:mm", mask: "99/99/9999 99:99", showTime: true, hourFormat: "24" }), inputType === InputTypes.Time &&
                            _jsx(Calendar, { value: value, onChange: function (e) { return updateInputs(index, e.target.value); }, placeholder: "hh:mm", mask: "99:99", timeOnly: true, hourFormat: "24" }), inputType === InputTypes.RelationManyToOne &&
                            _jsx(SolidManyToOneFilterElement, { value: value, index: index, updateInputs: updateInputs, fieldMetadata: fieldMetadata }), inputType === InputTypes.SelectionDynamic &&
                            _jsx(SolidSelectionDynamicFilterElement, { value: value, index: index, updateInputs: updateInputs, fieldMetadata: fieldMetadata }), inputType === InputTypes.SelectionStatic &&
                            _jsx(SolidSelectionStaticFilterElement, { value: value, index: index, updateInputs: updateInputs, fieldMetadata: fieldMetadata }), numberOfInputs === null &&
                            _jsxs(_Fragment, { children: [_jsx(Button, { icon: "pi pi-plus", size: "small", className: "small-button", onClick: addInput, type: "button" }), _jsx(Button, { icon: "pi pi-trash", size: "small", className: "small-button", onClick: function () { return deleteInput(index); }, outlined: true, severity: "danger", type: "button" })] })] }, index)); }) }) }) }));
};
