// @solidxai/core-ui/dist/components/auth/SolidLogin.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { Form, Formik } from "formik";
import { signIn } from "../../adapters/auth/index";
import Link from "../common/Link";
import { useRouter } from "../../hooks/useRouter";
import { Button } from "primereact/button";
import { Divider } from "primereact/divider";
import { InputText } from "primereact/inputtext";
import { Message } from "primereact/message";
import { Password } from "primereact/password";
import { TabPanel, TabView } from 'primereact/tabview';
import { Toast } from "primereact/toast";
import { useEffect, useRef, useState } from "react";
import * as Yup from "yup";
import { SocialMediaLogin } from "../common/SocialMediaLogin";
import { useInitateLoginMutation } from "../../redux/api/authApi";
import Image from "../common/Image";
import SolidLogo from '../../resources/images/SolidXLogo.svg';
import { formatTimeLeft } from "../../helpers/resendOtpHelper";
import { ERROR_MESSAGES } from "../../constants/error-messages";
import { RadioButton } from "primereact/radiobutton";
import { useLazyGetAuthSettingsQuery } from "../../redux/api/solidSettingsApi";
import { env } from "../../adapters/env";
import showToast from "../../helpers/showToast";
var SolidLogin = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k;
    var signInValidatorLabel = _a.signInValidatorLabel, signInValidatorPlaceholder = _a.signInValidatorPlaceholder;
    var _l = useLazyGetAuthSettingsQuery(), trigger = _l[0], solidSettingsData = _l[1].data;
    useEffect(function () {
        trigger(""); // Fetch settings on mount
    }, [trigger]);
    var initiateLogin = useInitateLoginMutation()[0];
    var _m = useState(0), activeIndex = _m[0], setActiveIndex = _m[1];
    useEffect(function () {
        sessionStorage.removeItem("app-mounted");
    }, []);
    var toast = useRef(null);
    var router = useRouter();
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var emailOrUsernameRegex = /^[a-zA-Z0-9._]{3,30}$/;
    var PasswordLogin = function () {
        return (_jsx(Formik, { initialValues: {
                identifier: "",
                // email: initialEmail,
                password: "",
                // rememberMe: rememberMe,
            }, enableReinitialize: false, validationSchema: Yup.object({
                identifier: Yup.string()
                    .required(ERROR_MESSAGES.FIELD_REUQIRED("Email or Username"))
                    .test("email-or-username", ERROR_MESSAGES.FIELD_INVALID("email or username"), function (value) {
                    // if (!value) return false;
                    var isEmail = Yup.string().email().isValidSync(value);
                    var isUsername = emailOrUsernameRegex.test(value);
                    return isEmail || isUsername;
                }),
                password: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED("Password")),
            }), onSubmit: function (values, _a) {
                var setSubmitting = _a.setSubmitting, setErrors = _a.setErrors;
                return __awaiter(void 0, void 0, void 0, function () {
                    var response, redirectUrl, error_1;
                    var _b;
                    return __generator(this, function (_c) {
                        switch (_c.label) {
                            case 0:
                                _c.trys.push([0, 2, 3, 4]);
                                return [4 /*yield*/, signIn("credentials", {
                                        redirect: false,
                                        identifier: values.identifier,
                                        password: values.password,
                                    })];
                            case 1:
                                response = _c.sent();
                                if (response === null || response === void 0 ? void 0 : response.error) {
                                    showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, response.error);
                                    setErrors({
                                        identifier: ERROR_MESSAGES.INVALID_CREDENTIALS,
                                        password: ERROR_MESSAGES.INVALID_CREDENTIALS,
                                    });
                                }
                                else {
                                    showToast(toast, "success", ERROR_MESSAGES.LOGIN_SUCCESS, ERROR_MESSAGES.DASHBOARD_REDIRECTING);
                                    redirectUrl = env("NEXT_PUBLIC_LOGIN_REDIRECT_URL") || "/admin";
                                    router.push(redirectUrl);
                                }
                                return [3 /*break*/, 4];
                            case 2:
                                error_1 = _c.sent();
                                showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, (error_1 === null || error_1 === void 0 ? void 0 : error_1.data) ? (_b = error_1 === null || error_1 === void 0 ? void 0 : error_1.data) === null || _b === void 0 ? void 0 : _b.message : ERROR_MESSAGES.SOMETHING_WRONG);
                                return [3 /*break*/, 4];
                            case 3:
                                setSubmitting(false); // Re-enable the button after submission
                                return [7 /*endfinally*/];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            }, children: function (formik) {
                var _a, _b, _c, _d;
                return (_jsxs(Form, { children: [_jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { htmlFor: "email", className: "solid-auth-input-label", children: signInValidatorLabel ? signInValidatorLabel : "Username or Email" }), _jsx(InputText, { id: "identifier", name: "identifier", placeholder: signInValidatorPlaceholder ? signInValidatorPlaceholder : "Email or Username", onChange: formik.handleChange, value: formik.values.identifier, invalid: !!formik.errors.identifier, onBlur: formik.handleBlur }), isFormFieldValid(formik, "identifier") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_b = (_a = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _a === void 0 ? void 0 : _a.identifier) === null || _b === void 0 ? void 0 : _b.toString() })] }), _jsxs("div", { className: "flex flex-column gap-1 mt-4", style: {}, children: [_jsx("label", { htmlFor: "password", className: "solid-auth-input-label", children: "Password" }), _jsx(Password, { id: "password", name: "password", value: formik.values.password, onChange: formik.handleChange, toggleMask: true, invalid: !!formik.errors.password, inputClassName: "w-full", feedback: false }), isFormFieldValid(formik, "password") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_d = (_c = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _c === void 0 ? void 0 : _c.password) === null || _d === void 0 ? void 0 : _d.toString() })] }), _jsxs("div", { className: "mt-4 flex align-items-center justify-content-between", children: [_jsx("div", {}), _jsx(Link, { href: "/auth/initiate-forgot-password", className: "solid-auth-input-label font-bold", children: "Forgot Password?" })] }), _jsx("div", { className: "mt-4", children: _jsx(Button, { className: "w-full font-light auth-submit-button", label: "Sign In", disabled: formik.isSubmitting, loading: formik.isSubmitting }) })] }));
            } }));
    };
    var PasswordLessLogin = function () {
        var _a;
        var validationType = ((_a = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _a === void 0 ? void 0 : _a.passwordlessRegistrationValidateWhat) || "email";
        var _b = useState("email"), selectedAuthMethod = _b[0], setSelectedAuthMethod = _b[1];
        var getFieldConfig = function () {
            if (validationType === "transactional") {
                if (selectedAuthMethod === "mobile") {
                    return {
                        label: "Mobile Number / Username",
                        placeholder: "Enter your mobile number or username",
                        type: "mobile",
                        // validationSchema: Yup.string()
                        //     .matches(/^[0-9]{10}$/, ERROR_MESSAGES.FIELD_INVALID('mobile number'))
                        //     .required(ERROR_MESSAGES.FIELD_REUQIRED('Mobile number'))
                    };
                }
                else {
                    return {
                        label: "Email Address / Username",
                        placeholder: "Enter your email or username",
                        type: "email",
                        // validationSchema: Yup.string()
                        //     .email(ERROR_MESSAGES.FIELD_INVALID('email address'))
                        //     .required(ERROR_MESSAGES.FIELD_REUQIRED('Email'))
                    };
                }
            }
            switch (validationType) {
                case "mobile":
                    return {
                        label: signInValidatorLabel || "Mobile Number / Username",
                        placeholder: signInValidatorPlaceholder || "Enter your mobile number / username",
                        type: "mobile",
                        // validationSchema: Yup.string()
                        //     .matches(/^[0-9]{10}$/, ERROR_MESSAGES.FIELD_INVALID('mobile number'))
                        //     .required(ERROR_MESSAGES.FIELD_REUQIRED('Mobile number'))
                    };
                case "email":
                default:
                    return {
                        label: signInValidatorLabel || "Email / Username",
                        placeholder: signInValidatorPlaceholder || "Email ID / username",
                        type: "email",
                        // validationSchema: Yup.string()
                        //     .email(ERROR_MESSAGES.FIELD_INVALID('email address'))
                        //     .required(ERROR_MESSAGES.FIELD_REUQIRED('Email'))
                    };
            }
        };
        var fieldConfig = getFieldConfig();
        return (_jsx(Formik, { initialValues: {
                identifier: "",
            }, 
            // validationSchema={
            //     fieldConfig.validationSchema 
            //     ? Yup.object({
            //           identifier: fieldConfig.validationSchema
            //       })
            //     : Yup.object({
            //           identifier: Yup.string().required("required"),
            //       })
            //   }
            validationSchema: Yup.object({ identifier: Yup.string().required("required") }), enableReinitialize: false, onSubmit: function (values, _a) {
                var setSubmitting = _a.setSubmitting, setErrors = _a.setErrors;
                return __awaiter(void 0, void 0, void 0, function () {
                    var RESEND_OTP_KEY, RESEND_OTP_TIMER_MIN, RESEND_OTP_TIMER, authType, payload, storedTimeStr, now, lastSent, elapsed, remaining, formatted, response, identifier, err_1;
                    var _b, _c;
                    return __generator(this, function (_d) {
                        switch (_d.label) {
                            case 0:
                                _d.trys.push([0, 2, 3, 4]);
                                RESEND_OTP_KEY = "resendOtpLogin_".concat(values.identifier);
                                RESEND_OTP_TIMER_MIN = parseFloat(env("NEXT_PUBLIC_RESEND_OTP_TIMER") || '0.5');
                                RESEND_OTP_TIMER = Math.round(RESEND_OTP_TIMER_MIN * 60);
                                authType = validationType === "transactional" ? selectedAuthMethod : fieldConfig.type;
                                payload = {
                                    type: authType,
                                    identifier: values.identifier,
                                };
                                storedTimeStr = localStorage.getItem(RESEND_OTP_KEY);
                                now = Date.now();
                                if (storedTimeStr) {
                                    lastSent = parseInt(storedTimeStr, 10);
                                    elapsed = Math.floor((now - lastSent) / 1000);
                                    remaining = RESEND_OTP_TIMER - elapsed;
                                    if (remaining > 0) {
                                        formatted = formatTimeLeft(remaining);
                                        showToast(toast, "error", ERROR_MESSAGES.PLEASE_WAIT, ERROR_MESSAGES.OPT_FORMAT(formatted));
                                        setSubmitting(false);
                                        return [2 /*return*/]; //  Prevent request
                                    }
                                }
                                return [4 /*yield*/, initiateLogin(payload).unwrap()];
                            case 1:
                                response = _d.sent();
                                if ((response === null || response === void 0 ? void 0 : response.statusCode) === 200) {
                                    showToast(toast, "success", ERROR_MESSAGES.OPT_RESEND, (_b = response === null || response === void 0 ? void 0 : response.data) === null || _b === void 0 ? void 0 : _b.message);
                                    identifier = values.identifier;
                                    localStorage.setItem("resendOtpLogin_".concat(identifier), Date.now().toString());
                                    router.push("/auth/initiate-login?identifier=".concat(encodeURIComponent(identifier), "&type=").concat(authType));
                                }
                                else {
                                    showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, response.error);
                                }
                                return [3 /*break*/, 4];
                            case 2:
                                err_1 = _d.sent();
                                showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, (err_1 === null || err_1 === void 0 ? void 0 : err_1.data) ? (_c = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _c === void 0 ? void 0 : _c.message : ERROR_MESSAGES.SOMETHING_WRONG);
                                setErrors({
                                    identifier: "Invalid Credentials",
                                });
                                return [3 /*break*/, 4];
                            case 3:
                                setSubmitting(false);
                                return [7 /*endfinally*/];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            }, children: function (formik) {
                var _a, _b;
                return (_jsxs(Form, { children: [validationType === "transactional" && (_jsxs("div", { className: "flex flex-column gap-3 mt-3", children: [_jsx("label", { className: "solid-auth-input-label", children: "Select Authentication Method" }), _jsxs("div", { className: "flex gap-4", children: [_jsxs("div", { className: "flex align-items-center", children: [_jsx(RadioButton, { inputId: "auth-email", name: "authMethod", value: "email", checked: selectedAuthMethod === "email", onChange: function (e) {
                                                        setSelectedAuthMethod("email");
                                                        formik.setFieldValue("identifier", "");
                                                    }, className: "mr-2" }), _jsx("label", { htmlFor: "auth-email", className: "solid-auth-input-label cursor-pointer", children: "Email / Username" })] }), _jsxs("div", { className: "flex align-items-center", children: [_jsx(RadioButton, { inputId: "auth-mobile", name: "authMethod", value: "mobile", checked: selectedAuthMethod === "mobile", onChange: function (e) {
                                                        setSelectedAuthMethod("mobile");
                                                        formik.setFieldValue("identifier", "");
                                                    }, className: "mr-2" }), _jsx("label", { htmlFor: "auth-mobile", className: "solid-auth-input-label cursor-pointer", children: "Mobile / Username" })] })] })] })), _jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { htmlFor: "identifier", className: "solid-auth-input-label", children: fieldConfig.label }), _jsx(InputText, { id: "identifier", name: "identifier", placeholder: fieldConfig.placeholder, onChange: formik.handleChange, value: formik.values.identifier, invalid: !!formik.errors.identifier, onBlur: formik.handleBlur }), isFormFieldValid(formik, "identifier") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_b = (_a = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _a === void 0 ? void 0 : _a.identifier) === null || _b === void 0 ? void 0 : _b.toString() })] }), _jsx("div", { className: "mt-4", children: _jsx(Button, { className: "w-full font-light auth-submit-button", label: "Sign In", disabled: formik.isSubmitting, loading: formik.isSubmitting }) })] }));
            } }));
    };
    var AuthTabs = function (_a) {
        var passwordBasedAuth = _a.passwordBasedAuth, passwordLessAuth = _a.passwordLessAuth;
        if (passwordBasedAuth && passwordLessAuth) {
            return (_jsxs(TabView, { className: "solid-auth-tabview", activeIndex: activeIndex, onTabChange: function (e) { return setActiveIndex(e.index); }, children: [_jsx(TabPanel, { header: "With Password", children: _jsx(PasswordLogin, {}) }), _jsx(TabPanel, { header: "Without Password", children: _jsx(PasswordLessLogin, {}) })] }));
        }
        else if (passwordBasedAuth) {
            return _jsx(PasswordLogin, {});
        }
        else if (passwordLessAuth) {
            return _jsx(PasswordLessLogin, {});
        }
        else {
            return _jsx("p", { children: "No authentication method available" });
        }
    };
    return (_jsxs("div", { className: "", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "auth-container ".concat(((_b = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _b === void 0 ? void 0 : _b.authPagesLayout) === 'center' ? 'center' : 'side'), children: [((_c = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _c === void 0 ? void 0 : _c.authPagesLayout) === 'center' &&
                        _jsx("div", { className: "flex justify-content-center", children: _jsx("div", { className: "solid-logo flex align-items-center ".concat((_d = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _d === void 0 ? void 0 : _d.appLogoPosition), children: _jsx(Image, { alt: "solid logo", src: ((_e = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _e === void 0 ? void 0 : _e.appLogo) || SolidLogo, className: "relative", fill: true }) }) }), _jsx("h2", { className: "solid-auth-title ".concat(((_f = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _f === void 0 ? void 0 : _f.authPagesLayout) === 'center' ? 'text-center mt-2 md:mt-4' : 'text-left'), children: "Sign In To Your Account" }), _jsx(AuthTabs, { passwordBasedAuth: (_g = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _g === void 0 ? void 0 : _g.passwordBasedAuth, passwordLessAuth: (_h = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _h === void 0 ? void 0 : _h.passwordLessAuth }), ((_j = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _j === void 0 ? void 0 : _j.iamGoogleOAuthEnabled) &&
                        _jsxs(_Fragment, { children: [_jsx(Divider, { align: "center", children: _jsx("div", { className: "inline-flex align-items-center", children: "OR" }) }), _jsx(SocialMediaLogin, {})] })] }), ((_k = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _k === void 0 ? void 0 : _k.allowPublicRegistration) && _jsx("div", { className: "mt-3 md:mt-5", children: _jsxs("div", { className: "text-sm text-center text-400 secondary-dark-color", children: ["Don\u2019t have an account ? ", _jsx(Link, { className: "font-bold", href: "/auth/register", children: "Sign Up" })] }) })] }));
};
export default SolidLogin;
