// @solidxai/core-ui/dist/components/auth/SolidResetPassword.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useConfirmForgotPasswordMutation } from "../../redux/api/authApi";
import { useFormik } from "formik";
import Image from "../common/Image";
import { useRouter } from "../../hooks/useRouter";
import { useSearchParams } from "../../hooks/useSearchParams";
import { Button } from "primereact/button";
import { Message } from "primereact/message";
import { Password } from "primereact/password";
import { Toast } from "primereact/toast";
import { classNames } from "primereact/utils";
import { useEffect, useRef } from "react";
import * as Yup from "yup";
import SolidLogo from '../../resources/images/SolidXLogo.svg';
import { ERROR_MESSAGES } from "../../constants/error-messages";
import { useLazyGetAuthSettingsQuery } from "../../redux/api/solidSettingsApi";
import { env } from "../../adapters/env";
import showToast from "../../helpers/showToast";
var SolidResetPassword = function () {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    var searchParams = useSearchParams();
    var verificationToken = searchParams.get('token');
    // const decodedUsername = searchParams.get('username');
    // const username = decodedUsername ? decodeURIComponent(decodedUsername) : '';
    var _k = useLazyGetAuthSettingsQuery(), trigger = _k[0], solidSettingsData = _k[1].data;
    useEffect(function () {
        trigger(""); // Fetch settings on mount
    }, [trigger]);
    var toast = useRef(null);
    var router = useRouter();
    var confirmForgotPassword = useConfirmForgotPasswordMutation()[0];
    var envPasswordRegex = env("NEXT_PUBLIC_PASSWORD_REGEX");
    var envPasswordHelperText = env("NEXT_PUBLIC_PASSWORD_COMPLEXITY_DESC");
    var passwordRegex = null;
    try {
        if (envPasswordRegex) {
            var unescaped = JSON.parse("\"".concat(envPasswordRegex, "\""));
            passwordRegex = new RegExp(unescaped);
        }
    }
    catch (error) {
        console.error(ERROR_MESSAGES.INVALID_PASSWORD_REGX, error);
    }
    var newPasswordValidation = passwordRegex
        ? Yup.string()
            .matches(passwordRegex, ERROR_MESSAGES.PASSWORD_DO_NOT_MEET)
            .required(ERROR_MESSAGES.FIELD_REUQIRED('New password'))
        : Yup.string().min(6, ERROR_MESSAGES.PASSWORD_CHARACTER(6))
            .required('New password is required');
    var validationSchema = Yup.object({
        password: newPasswordValidation,
        confirmPassword: Yup.string().oneOf([Yup.ref('password')], ERROR_MESSAGES.MUST_MATCH).required(ERROR_MESSAGES.FIELD_REUQIRED('Confirm password')),
    });
    var formik = useFormik({
        initialValues: {
            // username: username,
            password: "",
            confirmPassword: "",
            verificationToken: verificationToken,
        },
        validationSchema: validationSchema,
        onSubmit: function (values) { return __awaiter(void 0, void 0, void 0, function () {
            var payload, response, err_1;
            var _a, _b, _c, _d, _e, _f;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        _g.trys.push([0, 2, , 3]);
                        payload = {
                            // username: values.username,
                            password: values.password,
                            verificationToken: values.verificationToken,
                        };
                        return [4 /*yield*/, confirmForgotPassword(payload).unwrap()];
                    case 1:
                        response = _g.sent();
                        if ((response === null || response === void 0 ? void 0 : response.statusCode) === 200) {
                            showToast(toast, "success", ERROR_MESSAGES.FIELD_UPDATE('Password'), ERROR_MESSAGES.FIELD_UPDATE_SUCCESSFULLY('Password'));
                            router.push('/auth/login');
                        }
                        else
                            (showToast(toast, "error", ERROR_MESSAGES.ERROR, response.error));
                        return [3 /*break*/, 3];
                    case 2:
                        err_1 = _g.sent();
                        showToast(toast, "error", (_a = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _a === void 0 ? void 0 : _a.message, ((_c = (_b = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _b === void 0 ? void 0 : _b.data) === null || _c === void 0 ? void 0 : _c.message) ? (_e = (_d = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.message : (_f = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _f === void 0 ? void 0 : _f.message);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        }); },
    });
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    return (_jsxs(_Fragment, { children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "auth-container ".concat(((_a = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _a === void 0 ? void 0 : _a.authPagesLayout) === 'center' ? 'center' : 'side'), children: [((_b = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _b === void 0 ? void 0 : _b.authPagesLayout) === 'center' &&
                        _jsx("div", { className: "flex justify-content-center", children: _jsx("div", { className: "solid-logo flex align-items-center ".concat((_c = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _c === void 0 ? void 0 : _c.appLogoPosition), children: _jsx(Image, { alt: "solid logo", src: ((_d = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _d === void 0 ? void 0 : _d.appLogo) || SolidLogo, className: "relative", fill: true }) }) }), _jsx("h2", { className: "solid-auth-title ".concat(((_e = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _e === void 0 ? void 0 : _e.authPagesLayout) === 'center' ? 'text-center' : 'text-left'), children: "Create New Password" }), _jsxs("form", { onSubmit: formik.handleSubmit, children: [_jsxs("div", { className: "flex flex-column gap-2", children: [_jsx("label", { htmlFor: "password", className: "solid-auth-input-label", children: "New Password" }), _jsx(Password, { id: "password", name: "password", placeholder: "***************", value: formik.values.password, onChange: formik.handleChange, onBlur: formik.handleBlur, toggleMask: true, className: classNames("", {
                                            "p-invalid": isFormFieldValid(formik, "password"),
                                        }), inputClassName: "w-full", feedback: false }), isFormFieldValid(formik, "password") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_g = (_f = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _f === void 0 ? void 0 : _f.password) === null || _g === void 0 ? void 0 : _g.toString() })] }), _jsxs("div", { className: "flex flex-column gap-1 mt-4", style: {}, children: [_jsx("label", { htmlFor: "password", className: "solid-auth-input-label", children: "Confirm Password" }), _jsx(Password, { id: "confirmPassword", name: "confirmPassword", placeholder: "***************", value: formik.values.confirmPassword, onChange: formik.handleChange, onBlur: formik.handleBlur, toggleMask: true, className: classNames("", {
                                            "p-invalid": isFormFieldValid(formik, "confirmPassword"),
                                        }), inputClassName: "w-full", feedback: false }), isFormFieldValid(formik, "confirmPassword") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_j = (_h = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _h === void 0 ? void 0 : _h.confirmPassword) === null || _j === void 0 ? void 0 : _j.toString() })] }), envPasswordHelperText && (_jsx("div", { className: "mt-4 text-sm", children: _jsx("div", { className: "grid", children: envPasswordHelperText
                                        .split('\\n')
                                        .map(function (text, idx) { return (_jsx("div", { className: "col-6 pt-0", children: _jsxs("div", { className: 'flex gap-2', children: [_jsx("span", { children: "\u2022" }), _jsx("span", { children: text })] }) }, idx)); }) }) })), _jsxs("div", { className: "mt-4", children: [_jsx(Button, { className: "w-full font-light auth-submit-button", label: "Reset Password", disabled: formik.isSubmitting, loading: formik.isSubmitting }), _jsx(Button, { type: "button", label: "Back", className: "w-full auth-back-button text-center mt-1", link: true, onClick: function () { return (window.location.href = '/auth/login'); } })] })] })] })] }));
};
export default SolidResetPassword;
