// @solidxai/core-ui/dist/components/auth/SolidRegister.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useInitateRegisterMutation, useRegisterMutation } from "../../redux/api/authApi";
import { Form, Formik } from "formik";
import Link from "../common/Link";
import { useRouter } from "../../hooks/useRouter";
import { Button } from "primereact/button";
import { Divider } from "primereact/divider";
import { InputText } from "primereact/inputtext";
import { Message } from "primereact/message";
import { Password } from "primereact/password";
import { TabPanel, TabView } from "primereact/tabview";
import { Toast } from "primereact/toast";
import { useEffect, useRef, useState } from "react";
import * as Yup from "yup";
import { SocialMediaLogin } from "../common/SocialMediaLogin";
import Image from "../common/Image";
import SolidLogo from '../../resources/images/SolidXLogo.svg';
import { formatTimeLeft } from "../../helpers/resendOtpHelper";
import { ProgressSpinner } from "primereact/progressspinner";
import { ERROR_MESSAGES } from "../../constants/error-messages";
import { useLazyGetAuthSettingsQuery } from "../../redux/api/solidSettingsApi";
import { env } from "../../adapters/env";
import showToast from "../../helpers/showToast";
var SolidRegister = function () {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j;
    var envPasswordHelperText = env("NEXT_PUBLIC_PASSWORD_COMPLEXITY_DESC");
    var _k = useState(0), activeIndex = _k[0], setActiveIndex = _k[1];
    var _l = useLazyGetAuthSettingsQuery(), trigger = _l[0], solidSettingsData = _l[1].data;
    useEffect(function () {
        trigger(""); // Fetch settings on mount
    }, [trigger]);
    var _m = useState(false), showOverlay = _m[0], setShowOverlay = _m[1];
    var toast = useRef(null);
    var router = useRouter();
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var _o = useRegisterMutation(), register = _o[0], _p = _o[1], isLoading = _p.isLoading, error = _p.error, isSuccess = _p.isSuccess, data = _p.data;
    var initiateRegister = useInitateRegisterMutation()[0];
    var showError = function () {
        var _a, _b, _c, _d, _e;
        if (error) {
            if ("data" in error) {
                var apiError = error;
                // @ts-ignore
                var errorMessages = Array.isArray((_a = apiError.data) === null || _a === void 0 ? void 0 : _a.message) ? (_b = apiError.data) === null || _b === void 0 ? void 0 : _b.message : [(_c = apiError.data) === null || _c === void 0 ? void 0 : _c.message];
                (_d = toast.current) === null || _d === void 0 ? void 0 : _d.show({
                    severity: "error",
                    summary: ERROR_MESSAGES.ERROR,
                    detail: errorMessages.join(", "),
                    sticky: true
                });
            }
            else {
                var serializedError = error;
                (_e = toast.current) === null || _e === void 0 ? void 0 : _e.show({
                    severity: "error",
                    summary: ERROR_MESSAGES.ERROR,
                    detail: serializedError.message || ERROR_MESSAGES.ERROR_OCCURED,
                    sticky: true
                });
            }
        }
    };
    useEffect(function () {
        if (error) {
            showError();
        }
    }, [error]);
    // useEffect(() => {
    //     if (isSuccess) {
    //             router.replace("/auth/login");
    //     }
    // }, [isSuccess])
    var PasswordSignup = function (_a) {
        var showNameFieldsForRegistration = _a.showNameFieldsForRegistration;
        console.log("showNameFieldsForRegistration", showNameFieldsForRegistration);
        return (_jsx(Formik, { initialValues: {
                username: "",
                email: "",
                password: "",
                firstName: "",
                lastName: ""
            }, validationSchema: Yup.object({
                username: showNameFieldsForRegistration
                    ? Yup.string().notRequired()
                    : Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED('User Name')),
                firstName: showNameFieldsForRegistration
                    ? Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED('First Name'))
                    : Yup.string().notRequired(),
                lastName: showNameFieldsForRegistration
                    ? Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED('Last Name '))
                    : Yup.string().notRequired(),
                email: Yup.string()
                    .email(ERROR_MESSAGES.FIELD_INVALID('email address'))
                    .required(ERROR_MESSAGES.FIELD_REUQIRED('Email')),
                password: Yup.string()
                    .required(ERROR_MESSAGES.FIELD_REUQIRED('Password'))
                    .min(8, ERROR_MESSAGES.PASSWORD_CHARACTER(8))
                    .matches(/[a-z]/, ERROR_MESSAGES.PASSWORD_CONTAIN('lowercase'))
                    .matches(/[A-Z]/, ERROR_MESSAGES.PASSWORD_CONTAIN('uppercase'))
                    .matches(/\d/, ERROR_MESSAGES.PASSWORD_CONTAIN('one', 'number'))
                    .matches(/[@$!%*?&#^(){}[\]|\\/~`+=<>:;'"_,.-]/, ERROR_MESSAGES.PASSWORD_CONTAIN('special', 'character')),
            }), onSubmit: function (values, _a) {
                var setSubmitting = _a.setSubmitting;
                return __awaiter(void 0, void 0, void 0, function () {
                    var userData, fullName, username, response, err_1;
                    var _b, _c;
                    return __generator(this, function (_d) {
                        switch (_d.label) {
                            case 0:
                                _d.trys.push([0, 2, 3, 4]);
                                userData = {
                                    email: values.email,
                                    password: values.password,
                                };
                                if (showNameFieldsForRegistration) {
                                    fullName = "".concat(values.firstName || "", " ").concat(values.lastName || "").trim();
                                    username = "".concat(values.firstName || "").concat(values.lastName || "").trim().toLowerCase();
                                    console.log("fullName, username", fullName, username);
                                    userData = __assign(__assign({}, userData), { fullName: fullName, username: username });
                                }
                                else {
                                    userData = __assign(__assign({}, userData), { username: values.username });
                                }
                                return [4 /*yield*/, register(userData).unwrap()];
                            case 1:
                                response = _d.sent();
                                if ((response === null || response === void 0 ? void 0 : response.statusCode) === 200) {
                                    showToast(toast, "success", ERROR_MESSAGES.USER_REGISTER, (_b = response === null || response === void 0 ? void 0 : response.data) === null || _b === void 0 ? void 0 : _b.message);
                                    setShowOverlay(true);
                                    setTimeout(function () {
                                        router.push("/auth/login");
                                    }, 3000);
                                }
                                else {
                                    showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, response.error);
                                }
                                return [3 /*break*/, 4];
                            case 2:
                                err_1 = _d.sent();
                                showToast(toast, "error", ERROR_MESSAGES.EMAIL_ALREADY_TAKEN, (err_1 === null || err_1 === void 0 ? void 0 : err_1.data) ? (_c = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _c === void 0 ? void 0 : _c.message : ERROR_MESSAGES.SOMETHING_WRONG);
                                return [3 /*break*/, 4];
                            case 3:
                                setSubmitting(false);
                                return [7 /*endfinally*/];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            }, children: function (formik) {
                var _a, _b, _c, _d, _e;
                return (_jsxs(Form, { children: [showNameFieldsForRegistration ? (_jsx(_Fragment, { children: _jsxs("div", { className: "flex gap-2 mt-3", children: [_jsxs("div", { className: "flex flex-column w-full gap-2", children: [_jsx("label", { className: "solid-auth-input-label", children: "First Name" }), _jsx(InputText, { id: "firstName", name: "firstName", placeholder: "First Name", onChange: formik.handleChange, value: formik.values.firstName, invalid: !!formik.errors.firstName }), isFormFieldValid(formik, "firstName") && (_jsx(Message, { severity: "error", text: (_a = formik.errors.firstName) === null || _a === void 0 ? void 0 : _a.toString() }))] }), _jsxs("div", { className: "flex flex-column w-full gap-2", children: [_jsx("label", { className: "solid-auth-input-label", children: "Last Name" }), _jsx(InputText, { id: "lastName", name: "lastName", placeholder: "Last Name", onChange: formik.handleChange, value: formik.values.lastName, invalid: !!formik.errors.lastName }), isFormFieldValid(formik, "lastName") && (_jsx(Message, { severity: "error", text: (_b = formik.errors.lastName) === null || _b === void 0 ? void 0 : _b.toString() }))] })] }) })) : (_jsx(_Fragment, { children: _jsxs("div", { children: [_jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { className: "solid-auth-input-label", children: "Username" }), _jsx(InputText, { id: "username", name: "username", placeholder: "username", onChange: formik.handleChange, value: formik.values.username, invalid: !!formik.errors.username })] }), isFormFieldValid(formik, "username") &&
                                        _jsx(Message, { severity: "error", text: (_c = formik.errors.username) === null || _c === void 0 ? void 0 : _c.toString() })] }) })), _jsxs("div", { children: [_jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { className: "solid-auth-input-label", children: "Email" }), _jsx(InputText, { id: "email", name: "email", placeholder: "Yourgmail@123.com", onChange: formik.handleChange, value: formik.values.email, invalid: !!formik.errors.email })] }), isFormFieldValid(formik, "email") &&
                                    _jsx(Message, { severity: "error", text: (_d = formik.errors.email) === null || _d === void 0 ? void 0 : _d.toString() })] }), _jsxs("div", { children: [_jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { className: "solid-auth-input-label", children: "Password" }), _jsx(Password, { id: "password", name: "password", value: formik.values.password, onChange: formik.handleChange, toggleMask: true, inputClassName: "w-full", feedback: false, invalid: !!formik.errors.password })] }), isFormFieldValid(formik, "password") &&
                                    _jsx(Message, { severity: "error", text: (_e = formik.errors.password) === null || _e === void 0 ? void 0 : _e.toString() })] }), _jsx("div", { className: "mt-4", children: _jsx(Button, { className: "w-full font-light auth-submit-button", label: "Sign Up", disabled: formik.isSubmitting, loading: formik.isSubmitting }) })] }));
            } }));
    };
    var PasswordLessSignup = function () {
        return (_jsx(Formik, { initialValues: {
                username: "",
                email: "",
            }, validationSchema: Yup.object({
                username: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED('"User Name')),
                email: Yup.string()
                    .email(ERROR_MESSAGES.FIELD_INVALID('email address'))
                    .required(ERROR_MESSAGES.FIELD_REUQIRED('Email')),
            }), onSubmit: function (values, _a) {
                var setSubmitting = _a.setSubmitting;
                return __awaiter(void 0, void 0, void 0, function () {
                    var RESEND_OTP_KEY, RESEND_OTP_TIMER_MIN, RESEND_OTP_TIMER, payload, storedTimeStr, now, lastSent, elapsed, remaining, formatted, response, email, err_2;
                    var _b, _c;
                    return __generator(this, function (_d) {
                        switch (_d.label) {
                            case 0:
                                _d.trys.push([0, 2, 3, 4]);
                                RESEND_OTP_KEY = "resendOtpRegister_".concat(values.email);
                                RESEND_OTP_TIMER_MIN = parseFloat(env("NEXT_PUBLIC_RESEND_OTP_TIMER") || '0.5');
                                RESEND_OTP_TIMER = Math.round(RESEND_OTP_TIMER_MIN * 60);
                                payload = {
                                    username: values.username,
                                    email: values.email,
                                    validationSources: ["email"]
                                };
                                storedTimeStr = localStorage.getItem(RESEND_OTP_KEY);
                                now = Date.now();
                                if (storedTimeStr) {
                                    lastSent = parseInt(storedTimeStr, 10);
                                    elapsed = Math.floor((now - lastSent) / 1000);
                                    remaining = RESEND_OTP_TIMER - elapsed;
                                    if (remaining > 0) {
                                        formatted = formatTimeLeft(remaining);
                                        showToast(toast, "error", ERROR_MESSAGES.PLEASE_WAIT, ERROR_MESSAGES.OPT_FORMAT(formatted));
                                        setSubmitting(false);
                                        return [2 /*return*/]; //  Prevent request
                                    }
                                }
                                return [4 /*yield*/, initiateRegister(payload).unwrap()];
                            case 1:
                                response = _d.sent();
                                if ((response === null || response === void 0 ? void 0 : response.statusCode) === 200) {
                                    showToast(toast, "success", ERROR_MESSAGES.OPT_SEND, (_b = response === null || response === void 0 ? void 0 : response.data) === null || _b === void 0 ? void 0 : _b.message);
                                    email = values.email;
                                    localStorage.setItem("resendOtpRegister_".concat(email), Date.now().toString());
                                    router.push("/auth/initiate-register?email=".concat(email, "&username=").concat(values.username));
                                }
                                else {
                                    showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, response.error);
                                }
                                return [3 /*break*/, 4];
                            case 2:
                                err_2 = _d.sent();
                                showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, (err_2 === null || err_2 === void 0 ? void 0 : err_2.data) ? (_c = err_2 === null || err_2 === void 0 ? void 0 : err_2.data) === null || _c === void 0 ? void 0 : _c.message : ERROR_MESSAGES.SOMETHING_WRONG);
                                return [3 /*break*/, 4];
                            case 3:
                                setSubmitting(false);
                                return [7 /*endfinally*/];
                            case 4: return [2 /*return*/];
                        }
                    });
                });
            }, children: function (formik) {
                var _a, _b, _c, _d;
                return (_jsxs(Form, { children: [_jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { htmlFor: "email", className: "solid-auth-input-label", children: "Username" }), _jsx(InputText, { id: "username", name: "username", placeholder: "username", onChange: formik.handleChange, value: formik.values.username, invalid: !!formik.errors.username }), isFormFieldValid(formik, "username") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_b = (_a = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _a === void 0 ? void 0 : _a.username) === null || _b === void 0 ? void 0 : _b.toString() })] }), _jsxs("div", { className: "flex flex-column gap-2 mt-3", children: [_jsx("label", { htmlFor: "email", className: "solid-auth-input-label", children: "Email" }), _jsx(InputText, { id: "email", name: "email", placeholder: "Yourgmail@123.com", onChange: formik.handleChange, value: formik.values.email, invalid: !!formik.errors.email }), isFormFieldValid(formik, "email") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_d = (_c = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _c === void 0 ? void 0 : _c.email) === null || _d === void 0 ? void 0 : _d.toString() })] }), _jsx("div", { className: "mt-4", children: _jsx(Button, { className: "w-full font-light auth-submit-button", label: "Sign Up", disabled: formik.isSubmitting, loading: formik.isSubmitting }) })] }));
            } }));
    };
    var AuthTabs = function (_a) {
        var passwordBasedAuth = _a.passwordBasedAuth, passwordLessAuth = _a.passwordLessAuth, showNameFieldsForRegistration = _a.showNameFieldsForRegistration;
        if (passwordBasedAuth && passwordLessAuth) {
            return (_jsxs(TabView, { className: "solid-auth-tabview", activeIndex: activeIndex, onTabChange: function (e) { return setActiveIndex(e.index); }, children: [_jsx(TabPanel, { header: "With Password", children: _jsx(PasswordSignup, { showNameFieldsForRegistration: showNameFieldsForRegistration }) }), _jsx(TabPanel, { header: "Without Password", children: _jsx(PasswordLessSignup, {}) })] }));
        }
        else if (passwordBasedAuth) {
            return _jsx(PasswordSignup, { showNameFieldsForRegistration: showNameFieldsForRegistration });
        }
        else if (passwordLessAuth) {
            return _jsx(PasswordLessSignup, {});
        }
        else {
            return _jsx("p", { children: "No authentication method available" });
        }
    };
    return (_jsxs("div", { className: "", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "auth-container position-relative ".concat(((_a = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _a === void 0 ? void 0 : _a.authPagesLayout) === 'center' ? 'center' : 'side'), children: [showOverlay && (_jsx("div", { className: "absolute top-0 left-0 w-full h-full flex align-items-center justify-content-center register-success-popup", children: _jsxs("div", { className: "inline-flex flex-column align-items-center justify-content-center text-center", children: [_jsx(ProgressSpinner, { style: { width: '50px', height: '50px' }, strokeWidth: "4" }), _jsxs("p", { className: "mt-3 text-lg font-medium text-700", children: ["Registration successful,", _jsx("br", {}), "you will be redirected..."] })] }) })), ((_b = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _b === void 0 ? void 0 : _b.authPagesLayout) === 'center' &&
                        _jsx("div", { className: "flex justify-content-center", children: _jsx("div", { className: "solid-logo flex align-items-center ".concat((_c = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _c === void 0 ? void 0 : _c.appLogoPosition), children: _jsx(Image, { alt: "solid logo", src: ((_d = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _d === void 0 ? void 0 : _d.appLogo) || SolidLogo, className: "relative", fill: true }) }) }), _jsx("h2", { className: "solid-auth-title ".concat(((_e = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _e === void 0 ? void 0 : _e.authPagesLayout) === 'center' ? 'text-center mt-2 md:mt-4' : 'text-left'), children: "Sign Up" }), _jsx(AuthTabs, { passwordBasedAuth: (_f = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _f === void 0 ? void 0 : _f.passwordBasedAuth, passwordLessAuth: (_g = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _g === void 0 ? void 0 : _g.passwordLessAuth, showNameFieldsForRegistration: (_h = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _h === void 0 ? void 0 : _h.showNameFieldsForRegistration }), ((_j = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _j === void 0 ? void 0 : _j.iamGoogleOAuthEnabled) &&
                        _jsxs(_Fragment, { children: [_jsx(Divider, { align: "center", children: _jsx("div", { className: "inline-flex align-items-center", children: "or" }) }), _jsx(SocialMediaLogin, {})] })] }), _jsx("div", { className: "text-center mt-3 md:mt-4", children: _jsxs("div", { className: "text-sm text-400 secondary-dark-color", children: ["Already have an account ? ", _jsx(Link, { className: "font-bold", href: "/auth/login", children: "Sign In" })] }) })] }));
};
export default SolidRegister;
