// @solidxai/core-ui/dist/components/auth/SolidForgotPassword.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useInitiateChangePasswordMutation } from "../../redux/api/authApi";
import { useFormik } from "formik";
import { useRouter } from "../../hooks/useRouter";
import { Button } from "primereact/button";
import { InputText } from "primereact/inputtext";
import { Message } from "primereact/message";
import { Toast } from "primereact/toast";
import { useEffect, useRef } from "react";
import * as Yup from "yup";
import Image from "../common/Image";
import SolidLogo from '../../resources/images/SolidXLogo.svg';
import { ERROR_MESSAGES } from "../../constants/error-messages";
import { useLazyGetAuthSettingsQuery } from "../../redux/api/solidSettingsApi";
import showToast from "../../helpers/showToast";
var SolidForgotPassword = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h;
    var signInValidatorLabel = _a.signInValidatorLabel, signInValidatorPlaceholder = _a.signInValidatorPlaceholder;
    var _j = useLazyGetAuthSettingsQuery(), trigger = _j[0], solidSettingsData = _j[1].data;
    useEffect(function () {
        trigger(""); // Fetch settings on mount
    }, [trigger]);
    var toast = useRef(null);
    var router = useRouter();
    var initiateChangePassword = useInitiateChangePasswordMutation()[0];
    var validationSchema = Yup.object({
        email: Yup.string()
            .email(ERROR_MESSAGES.FIELD_INAVLID_FORMAT('email'))
            .required(ERROR_MESSAGES.FIELD_REUQIRED('Email')),
    });
    function maskEmail(email) {
        var _a = email.split('@'), localPart = _a[0], domain = _a[1];
        var maskedLocal = localPart.slice(0, 3) + '*'.repeat(localPart.length - 3);
        var maskedDomain = domain.slice(-8).padStart(domain.length, '*');
        return "".concat(maskedLocal, "@").concat(maskedDomain);
    }
    var formik = useFormik({
        initialValues: {
            email: "",
        },
        validationSchema: validationSchema,
        onSubmit: function (values) { return __awaiter(void 0, void 0, void 0, function () {
            var payload, response, email, maskedEmail, err_1;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        payload = {
                            email: values.email,
                        };
                        return [4 /*yield*/, initiateChangePassword(payload).unwrap()];
                    case 1:
                        response = _b.sent();
                        if ((response === null || response === void 0 ? void 0 : response.statusCode) === 200) {
                            email = values === null || values === void 0 ? void 0 : values.email;
                            maskedEmail = maskEmail(email);
                            router.push("/auth/initiate-forgot-password-thank-you?email=".concat(maskedEmail));
                        }
                        else
                            (showToast(toast, "error", ERROR_MESSAGES.ERROR, response.error));
                        return [3 /*break*/, 3];
                    case 2:
                        err_1 = _b.sent();
                        showToast(toast, "error", ERROR_MESSAGES.ERROR, (err_1 === null || err_1 === void 0 ? void 0 : err_1.data) ? (_a = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _a === void 0 ? void 0 : _a.message : ERROR_MESSAGES.SOMETHING_WRONG);
                        return [3 /*break*/, 3];
                    case 3: return [2 /*return*/];
                }
            });
        }); },
    });
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    return (_jsxs(_Fragment, { children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "auth-container ".concat(((_b = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _b === void 0 ? void 0 : _b.authPagesLayout) === 'center' ? 'center' : 'side'), children: [((_c = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _c === void 0 ? void 0 : _c.authPagesLayout) === 'center' &&
                        _jsx("div", { className: "flex justify-content-center", children: _jsx("div", { className: "solid-logo flex align-items-center ".concat((_d = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _d === void 0 ? void 0 : _d.appLogoPosition), children: _jsx(Image, { alt: "solid logo", src: ((_e = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _e === void 0 ? void 0 : _e.appLogo) || SolidLogo, className: "relative", fill: true }) }) }), _jsx("h2", { className: "solid-auth-title ".concat(((_f = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _f === void 0 ? void 0 : _f.authPagesLayout) === 'center' ? 'text-center' : 'text-left'), children: "Forgot Password" }), _jsxs("form", { onSubmit: formik.handleSubmit, children: [_jsxs("div", { className: "flex flex-column gap-2", children: [_jsx("label", { htmlFor: "email", className: "solid-auth-input-label", children: signInValidatorLabel ? signInValidatorLabel : "Username or Email" }), _jsx(InputText, { id: "email", name: "email", placeholder: signInValidatorPlaceholder ? signInValidatorPlaceholder : "Email ID", value: formik.values.email, onChange: formik.handleChange, onBlur: formik.handleBlur }), isFormFieldValid(formik, "email") && _jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_h = (_g = formik === null || formik === void 0 ? void 0 : formik.errors) === null || _g === void 0 ? void 0 : _g.email) === null || _h === void 0 ? void 0 : _h.toString() })] }), _jsxs("div", { className: "mt-4", children: [_jsx(Button, { className: "w-full font-light auth-submit-button", label: "Continue", disabled: formik.isSubmitting, loading: formik.isSubmitting }), _jsx(Button, { type: "button", label: "Back", className: "w-full auth-back-button text-center mt-1", link: true, onClick: function () { return (window.location.href = '/auth/login'); } })] })] })] })] }));
};
export default SolidForgotPassword;
