// @solidxai/core-ui/dist/components/auth/SolidInitiateRegisterOtp.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useConfirmOtpRegisterMutation, useInitateRegisterMutation } from "../../redux/api/authApi";
import { Form, Formik } from "formik";
import Image from "../common/Image";
import { useRouter } from "../../hooks/useRouter";
import { useSearchParams } from "../../hooks/useSearchParams";
import { Button } from "primereact/button";
import { InputOtp } from "primereact/inputotp";
import { Message } from "primereact/message";
import { Toast } from "primereact/toast";
import { useEffect, useRef, useState } from "react";
import * as Yup from "yup";
import SolidLogo from '../../resources/images/SolidXLogo.svg';
import { ERROR_MESSAGES } from "../../constants/error-messages";
import { useLazyGetAuthSettingsQuery } from "../../redux/api/solidSettingsApi";
import { env } from "../../adapters/env";
import showToast from "../../helpers/showToast";
var SolidInitiateRegisterOtp = function () {
    var _a, _b, _c, _d, _e;
    var searchParams = useSearchParams();
    var tempEmail = searchParams.get('email');
    var email = tempEmail ? decodeURIComponent(tempEmail) : '';
    var RESEND_OTP_KEY = "resendOtpRegister_".concat(email);
    var RESEND_OTP_TIMER_MIN = parseFloat(env("NEXT_PUBLIC_RESEND_OTP_TIMER") || '0.5');
    var RESEND_OTP_TIMER = Math.round(RESEND_OTP_TIMER_MIN * 60);
    var username = searchParams.get('username') || '';
    var _f = useLazyGetAuthSettingsQuery(), trigger = _f[0], solidSettingsData = _f[1].data;
    useEffect(function () {
        trigger(""); // Fetch settings on mount
    }, [trigger]);
    var initiateResendOTP = useInitateRegisterMutation()[0];
    var initiateOtpRegister = useConfirmOtpRegisterMutation()[0];
    var toast = useRef(null);
    var router = useRouter();
    var _g = useState(RESEND_OTP_TIMER), timeLeft = _g[0], setTimeLeft = _g[1];
    var _h = useState(false), resendEnabled = _h[0], setResendEnabled = _h[1];
    useEffect(function () {
        // Set timer if not already set (e.g., after login)
        var storedTime = localStorage.getItem(RESEND_OTP_KEY);
        if (!storedTime) {
            localStorage.setItem(RESEND_OTP_KEY, Date.now().toString());
        }
        var lastSent = storedTime ? parseInt(storedTime, 10) : Date.now();
        var elapsed = Math.floor((Date.now() - lastSent) / 1000);
        var remaining = RESEND_OTP_TIMER - elapsed;
        if (remaining > 0) {
            setTimeLeft(remaining);
            setResendEnabled(false);
        }
        else {
            setTimeLeft(0);
            setResendEnabled(true);
        }
    }, [email]);
    useEffect(function () {
        if (resendEnabled || timeLeft <= 0)
            return;
        var interval = setInterval(function () {
            setTimeLeft(function (prev) {
                if (prev <= 1) {
                    clearInterval(interval);
                    setResendEnabled(true);
                    return 0;
                }
                return prev - 1;
            });
        }, 1000);
        return function () { return clearInterval(interval); };
    }, [resendEnabled, timeLeft]);
    var validationSchema = Yup.object({
        otp: Yup.string()
            .matches(/^\d{6}$/, ERROR_MESSAGES.OTP_CHARACTER(6))
            .required(ERROR_MESSAGES.FIELD_REUQIRED('OTP')),
    });
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var handleResendOtp = function () { return __awaiter(void 0, void 0, void 0, function () {
        var payload, response, err_1;
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _c.trys.push([0, 2, , 3]);
                    payload = {
                        username: username,
                        email: email,
                        validationSources: ["email"]
                    };
                    return [4 /*yield*/, initiateResendOTP(payload).unwrap()];
                case 1:
                    response = _c.sent();
                    if ((response === null || response === void 0 ? void 0 : response.statusCode) === 200) {
                        showToast(toast, "success", ERROR_MESSAGES.OPT_RESEND, (_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.message);
                        localStorage.setItem(RESEND_OTP_KEY, Date.now().toString());
                        setTimeLeft(RESEND_OTP_TIMER);
                        setResendEnabled(false);
                    }
                    else {
                        showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, response.error);
                    }
                    return [3 /*break*/, 3];
                case 2:
                    err_1 = _c.sent();
                    showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, ((_b = err_1 === null || err_1 === void 0 ? void 0 : err_1.data) === null || _b === void 0 ? void 0 : _b.message) || ERROR_MESSAGES.SOMETHING_WRONG);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    return (_jsxs(_Fragment, { children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "auth-container ".concat(((_a = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _a === void 0 ? void 0 : _a.authPagesLayout) === 'center' ? 'center' : 'side'), style: { minWidth: 480 }, children: [((_b = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _b === void 0 ? void 0 : _b.authPagesLayout) === 'center' &&
                        _jsx("div", { className: "flex justify-content-center", children: _jsx("div", { className: "solid-logo flex align-items-center ".concat((_c = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _c === void 0 ? void 0 : _c.appLogoPosition), children: _jsx(Image, { alt: "solid logo", src: ((_d = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _d === void 0 ? void 0 : _d.appLogo) || SolidLogo, className: "relative", fill: true }) }) }), _jsx("h2", { className: "solid-auth-title ".concat(((_e = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _e === void 0 ? void 0 : _e.authPagesLayout) === 'center' ? 'text-center' : 'text-left'), children: "OTP Verification" }), _jsx("p", { className: "solid-auth-subtitle text-sm", children: "Please enter the OTP sent to your email to complete verification" }), _jsx(_Fragment, { children: _jsx(Formik, { initialValues: {
                                otp: "",
                            }, validationSchema: validationSchema, onSubmit: function (values, _a) {
                                var setSubmitting = _a.setSubmitting, setErrors = _a.setErrors;
                                return __awaiter(void 0, void 0, void 0, function () {
                                    var payload, response, err_2;
                                    var _b;
                                    return __generator(this, function (_c) {
                                        switch (_c.label) {
                                            case 0:
                                                _c.trys.push([0, 2, 3, 4]);
                                                payload = {
                                                    type: "email",
                                                    identifier: email,
                                                    otp: values.otp
                                                };
                                                return [4 /*yield*/, initiateOtpRegister(payload).unwrap()];
                                            case 1:
                                                response = _c.sent();
                                                if ((response === null || response === void 0 ? void 0 : response.statusCode) === 200) {
                                                    localStorage.removeItem("resendOtpRegister_".concat(email));
                                                    showToast(toast, "success", ERROR_MESSAGES.LOGIN_SUCCESSFULLY, "Login");
                                                    router.push("/auth/login");
                                                }
                                                else {
                                                    showToast(toast, "error", ERROR_MESSAGES.INAVLID_OTP, response.error);
                                                    setErrors({
                                                        otp: ERROR_MESSAGES.INAVLID_OTP,
                                                    });
                                                }
                                                return [3 /*break*/, 4];
                                            case 2:
                                                err_2 = _c.sent();
                                                showToast(toast, "error", ERROR_MESSAGES.LOGIN_ERROR, (err_2 === null || err_2 === void 0 ? void 0 : err_2.data) ? (_b = err_2 === null || err_2 === void 0 ? void 0 : err_2.data) === null || _b === void 0 ? void 0 : _b.message : ERROR_MESSAGES.SOMETHING_WRONG);
                                                return [3 /*break*/, 4];
                                            case 3:
                                                setSubmitting(false);
                                                return [7 /*endfinally*/];
                                            case 4: return [2 /*return*/];
                                        }
                                    });
                                });
                            }, children: function (formik) {
                                var _a;
                                return (_jsxs(Form, { children: [_jsxs("div", { className: "flex flex-column gap-2", children: [_jsx("label", { htmlFor: "otp", className: "solid-auth-input-label", children: "Enter OTP" }), _jsx(InputOtp, { value: formik.values.otp, onChange: function (e) { return formik.setFieldValue("otp", e.value); }, length: 6, style: { width: '100%' }, invalid: !!formik.errors.otp }), isFormFieldValid(formik, "otp") && (_jsx(Message, { className: "text-red-500 text-sm", severity: "error", text: (_a = formik.errors.otp) === null || _a === void 0 ? void 0 : _a.toString() })), _jsxs("div", { className: "flex align-items-center justify-content-between", children: [_jsx(Button, { type: "button", icon: 'pi pi-refresh', iconPos: "left", link: true, label: "Resend Code", className: "px-0 text-sm font-normal", onClick: handleResendOtp, disabled: !resendEnabled }), _jsx("p", { className: "m-0 text-sm text-color", children: resendEnabled
                                                                ? "You can resend now"
                                                                : "Time left: ".concat(Math.floor(timeLeft / 60)
                                                                    .toString()
                                                                    .padStart(2, "0"), ":").concat((timeLeft % 60).toString().padStart(2, "0")) })] })] }), _jsxs("div", { className: "mt-4", children: [_jsx(Button, { type: "submit", className: "w-full font-light auth-submit-button", label: "Verify", disabled: formik.isSubmitting, loading: formik.isSubmitting }), _jsx(Button, { type: "button", label: "Back", className: "w-full auth-back-button text-center mt-1", link: true, onClick: function () { return (window.location.href = '/auth/login'); } })] })] }));
                            } }) })] })] }));
};
export default SolidInitiateRegisterOtp;
