// @solidxai/core-ui/dist/components/layout/context/layoutcontext.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { useState, createContext } from 'react';
export var LayoutContext = createContext({});
export var LayoutProvider = function (_a) {
    var children = _a.children;
    var _b = useState({
        inputStyle: 'outlined',
        colorScheme: 'light',
        theme: 'solid-light-purple',
        scale: 14,
        authLayout: 'Center'
    }), layoutConfig = _b[0], setLayoutConfig = _b[1];
    var _c = useState({
        overlayMenuActive: false,
        profileSidebarVisible: false,
        configSidebarVisible: false,
        staticMenuMobileActive: false,
        menuHoverActive: false
    }), layoutState = _c[0], setLayoutState = _c[1];
    var onMenuToggle = function () {
        if (isDesktop()) {
            setLayoutState(function (prevLayoutState) { return (__assign(__assign({}, prevLayoutState), { staticMenuDesktopInactive: !prevLayoutState.staticMenuDesktopInactive })); });
        }
        else {
            setLayoutState(function (prevLayoutState) { return (__assign(__assign({}, prevLayoutState), { staticMenuMobileActive: !prevLayoutState.staticMenuMobileActive })); });
        }
    };
    var showProfileSidebar = function () {
        setLayoutState(function (prevLayoutState) { return (__assign(__assign({}, prevLayoutState), { profileSidebarVisible: !prevLayoutState.profileSidebarVisible })); });
    };
    var isDesktop = function () {
        return window.innerWidth > 991;
    };
    var value = {
        layoutConfig: layoutConfig,
        setLayoutConfig: setLayoutConfig,
        layoutState: layoutState,
        setLayoutState: setLayoutState,
        onMenuToggle: onMenuToggle,
        showProfileSidebar: showProfileSidebar
    };
    return _jsx(LayoutContext.Provider, { value: value, children: children });
};
