// @solidxai/core-ui/dist/components/core/form/fields/widgets/SolidS3FileViewerWidget.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useEffect, useRef, useState } from "react";
import { Button } from "primereact/button";
import { Dialog } from "primereact/dialog";
import { useResolveS3UrlMutation } from "../../../../../redux/api/fieldApi";
import Viewer from "viewerjs";
import "viewerjs/dist/viewer.css";
/**
 * SolidS3FileViewerWidget (PrimeReact version)
 */
export var SolidS3FileViewerWidget = function (_a) {
    var _b, _c;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var value = formik.values[fieldLayoutInfo.attrs.name];
    var fileType = (_c = fieldLayoutInfo.attrs.fileType) === null || _c === void 0 ? void 0 : _c.toLowerCase();
    var downloadAllowed = fieldLayoutInfo.attrs.downloadAllowed !== false;
    var bucketName = fieldLayoutInfo.attrs.bucketName;
    var mediaStorageProviderUserKey = fieldLayoutInfo.attrs.mediaStorageProviderUserKey;
    var isPrivate = fieldLayoutInfo.attrs.isPrivate ? fieldLayoutInfo.attrs.isPrivate : "false";
    var imageRef = useRef(null);
    var viewerRef = useRef(null);
    var _d = useState(""), previewUrl = _d[0], setPreviewUrl = _d[1];
    var _e = useState(false), isLoading = _e[0], setIsLoading = _e[1];
    var _f = useState(false), open = _f[0], setOpen = _f[1];
    var _g = useState(false), shouldShowViewer = _g[0], setShouldShowViewer = _g[1];
    var resolveS3Url = useResolveS3UrlMutation()[0];
    var fetchS3Url = function () { return __awaiter(void 0, void 0, void 0, function () {
        var result, e_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    console.log("fetch url called");
                    setIsLoading(true);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, resolveS3Url({
                            modelName: fieldContext.modelName,
                            fieldName: fieldContext.fieldMetadata.name,
                            s3Key: value,
                            fileType: fileType,
                            bucketName: bucketName,
                            mediaStorageProviderUserKey: mediaStorageProviderUserKey,
                            isPrivate: isPrivate
                        }).unwrap()];
                case 2:
                    result = _a.sent();
                    setIsLoading(false);
                    if (result.statusCode == "200") {
                        console.log("fetch url success", result.data.url);
                        return [2 /*return*/, result.data.url];
                    }
                    return [3 /*break*/, 4];
                case 3:
                    e_1 = _a.sent();
                    console.error("Failed to resolve S3 URL:", e_1);
                    setIsLoading(false);
                    return [2 /*return*/, null];
                case 4: return [2 /*return*/];
            }
        });
    }); };
    var handleDownload = function () { return __awaiter(void 0, void 0, void 0, function () {
        var url, a;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    if (isLoading)
                        return [2 /*return*/];
                    return [4 /*yield*/, fetchS3Url()];
                case 1:
                    url = _a.sent();
                    if (!url)
                        return [2 /*return*/];
                    a = document.createElement("a");
                    a.href = url;
                    a.download = (value === null || value === void 0 ? void 0 : value.split("/").pop()) || "file";
                    a.target = "_blank";
                    a.rel = "noopener noreferrer";
                    a.click();
                    return [2 /*return*/];
            }
        });
    }); };
    var handleView = function () { return __awaiter(void 0, void 0, void 0, function () {
        var url;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    console.log("isLoading in view", isLoading);
                    if (isLoading)
                        return [2 /*return*/];
                    return [4 /*yield*/, fetchS3Url()];
                case 1:
                    url = _a.sent();
                    console.log("url after fetch success", url);
                    if (!url)
                        return [2 /*return*/];
                    setPreviewUrl(url);
                    if (isImage) {
                        // Trigger viewer to show after state update
                        setShouldShowViewer(true);
                    }
                    else {
                        setOpen(true);
                    }
                    return [2 /*return*/];
            }
        });
    }); };
    var isImage = ["jpeg", "jpg", "png", "gif", "webp"].includes(fileType);
    var isPDF = fileType === "pdf";
    var isDownloadOnly = ["xlsx", "xls", "csv", "doc", "docx"].includes(fileType);
    // 🔹 Initialize Viewer.js once image exists
    useEffect(function () {
        if (imageRef.current && previewUrl && isImage) {
            // Destroy existing viewer if any
            if (viewerRef.current) {
                viewerRef.current.destroy();
            }
            // Create new viewer instance
            viewerRef.current = new Viewer(imageRef.current, {
                toolbar: {
                    zoomIn: 1,
                    zoomOut: 1,
                    rotateLeft: 1,
                    rotateRight: 1,
                    reset: 1,
                },
                navbar: false,
                title: false,
                transition: true,
                movable: true,
                scalable: true,
                rotatable: true,
                zoomable: true,
                zIndex: 9999,
                // Add hidden event to reset state
                hidden: function () {
                    setShouldShowViewer(false);
                }
            });
            console.log("Viewer initialized");
        }
        return function () {
            if (viewerRef.current) {
                viewerRef.current.destroy();
                viewerRef.current = null;
            }
        };
    }, [previewUrl, isImage]);
    // 🔹 Show viewer when shouldShowViewer becomes true
    useEffect(function () {
        if (shouldShowViewer && viewerRef.current) {
            console.log("Showing viewer");
            viewerRef.current.show();
        }
    }, [shouldShowViewer]);
    return (_jsxs("div", { className: "mt-2 flex flex-col gap-2", children: [_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel }), value ? (_jsxs("div", { className: "flex gap-3 items-center", children: [(isImage || isPDF) && (_jsx(Button, { icon: "pi pi-eye", type: "button", style: { minWidth: 66 }, loading: isLoading, tooltip: value, disabled: isLoading, label: "View ".concat(fileType), size: "small", iconPos: "left", onClick: handleView })), downloadAllowed && (_jsx(Button, { size: "small", type: "button", icon: "pi pi-download", tooltip: "Download ".concat(value === null || value === void 0 ? void 0 : value.split("/").pop()), loading: isLoading, onClick: handleDownload, disabled: isLoading, className: 'solid-icon-button' }))] })) : (_jsx("p", { className: "text-sm text-muted-foreground", children: "No file uploaded" })), isImage && previewUrl && (_jsx("img", { ref: imageRef, src: previewUrl, alt: value, style: {
                    position: "absolute",
                    visibility: "hidden",
                    width: "1px",
                    height: "1px"
                } })), _jsx(Dialog, { header: value, visible: open, modal: true, style: { width: "80vw", maxHeight: "90vh" }, onHide: function () { return setOpen(false); }, headerClassName: 'p-1 form-wrapper-title', contentClassName: 'p-0', contentStyle: { borderRadius: 6 }, children: previewUrl && isPDF && (_jsx("div", { style: {
                        width: "100%",
                        height: "75vh",
                        overflow: "hidden",
                    }, children: _jsx("iframe", { src: previewUrl, style: { width: "100%", height: "100%", border: "none" } }) })) })] }));
};
