// @solidxai/core-ui/dist/components/core/form/fields/widgets/SolidShortTextFieldAvatarWidget.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
export var SolidShortTextFieldAvatarWidget = function (_a) {
    var _b;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var value = formik.values[fieldLayoutInfo.attrs.name];
    var getInitials = function (value) {
        if (value) {
            var names = value === null || value === void 0 ? void 0 : value.trim().split(' ');
            var initials_1 = names.length === 1
                ? names[0][0]
                : names[0][0] + names[names.length - 1][0];
            return initials_1.toUpperCase();
        }
        else {
            return "";
        }
    };
    var getColorFromInitials = function (initials) {
        var hash = 0;
        for (var i = 0; i < initials.length; i++) {
            hash = initials.charCodeAt(i) + ((hash << 5) - hash);
        }
        var hue = Math.abs(hash) % 360;
        return "hsl(".concat(hue, ", 60%, 60%)"); // nice pastel color
    };
    var initials = getInitials(value);
    var bgColor = getColorFromInitials(initials);
    return (_jsxs("div", { className: "mt-2 flex-column gap-2", children: [_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel }), _jsx("div", { children: value &&
                    _jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '0.5rem' }, children: [_jsx("div", { style: {
                                    width: 20,
                                    height: 20,
                                    borderRadius: '50%',
                                    backgroundColor: bgColor,
                                    color: 'white',
                                    display: 'flex',
                                    alignItems: 'center',
                                    justifyContent: 'center',
                                    fontWeight: 600,
                                    fontSize: 12,
                                }, children: initials }), _jsx("span", { children: value })] }) })] }));
};
