// @solidxai/core-ui/dist/components/core/module/ModuleListViewData.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { CreateButton } from "../../../components/common/CreateButton";
import { handleError, handleSuccess } from "../../../helpers/ToastContainer";
import { useDeleteMultiplemodulesMutation, useGenerateCodeFormoduleMutation, useLazyGetmodulesQuery, useRefreshPermissionsMutation } from "../../../redux/api/moduleApi";
import Link from "../../common/Link";
import { FilterMatchMode } from "primereact/api";
import { Button } from "primereact/button";
import { Column } from "primereact/column";
import { DataTable, } from "primereact/datatable";
import { Dialog } from "primereact/dialog";
import { Toast } from "primereact/toast";
import qs from "qs";
import { useEffect, useRef, useState } from "react";
export var ModuleListViewData = function () {
    var toast = useRef(null);
    var _a = useState([]), moduleMetadata = _a[0], setModuleMetadata = _a[1];
    var _b = useState({
        displayName: { value: null, matchMode: FilterMatchMode.STARTS_WITH },
    }), filters = _b[0], setFilters = _b[1];
    // Lazy loading datatable
    var _c = useState(0), first = _c[0], setFirst = _c[1];
    var _d = useState(25), rows = _d[0], setRows = _d[1];
    var _e = useState(0), totalRecords = _e[0], setTotalRecords = _e[1];
    var _f = useState(""), sortField = _f[0], setSortField = _f[1];
    var _g = useState(0), sortOrder = _g[0], setSortOrder = _g[1];
    var _h = useState([]), selectedMenus = _h[0], setSelectedMenus = _h[1];
    var _j = useState(true), loading = _j[0], setLoading = _j[1];
    var _k = useState(false), isDialogVisible = _k[0], setDialogVisible = _k[1];
    var _l = useState(false), isGenerateCodeVisible = _l[0], setGenerateCodeVisible = _l[1];
    var _m = useState(), generateCodeForModule = _m[0], setGenerateCodeForModule = _m[1];
    var _o = useLazyGetmodulesQuery(), triggerGetModule = _o[0], _p = _o[1], module = _p.data, isLoading = _p.isLoading, error = _p.error;
    var _q = useGenerateCodeFormoduleMutation(), generateCode = _q[0], _r = _q[1], isGenerateCodeUpdating = _r.isLoading, isGenerateCodeSuceess = _r.isSuccess, isGenerateCodeError = _r.isError, generateCodeError = _r.error, generateCodeData = _r.data;
    var _s = useRefreshPermissionsMutation(), refreshPermission = _s[0], _t = _s[1], isRefreshPermissionUpdating = _t.isLoading, isRefreshPermissionSuceess = _t.isSuccess, isRefreshPermissionError = _t.isError, refreshPermissionError = _t.error, refreshPermissionData = _t.data;
    var _u = useDeleteMultiplemodulesMutation(), deleteManyModules = _u[0], _v = _u[1], isModuleDeleted = _v.isLoading, isDeleteModuleSuceess = _v.isSuccess, isModuleDeleteError = _v.isError, ModuleDeleteError = _v.error, DeletedModule = _v.data;
    useEffect(function () {
        if (module) {
            setModuleMetadata(module === null || module === void 0 ? void 0 : module.records);
            setTotalRecords(module === null || module === void 0 ? void 0 : module.meta.totalRecords);
            setLoading(false);
        }
    }, [module]);
    useEffect(function () {
        var queryData = {
            offset: 0,
            limit: 25,
            sort: ["id:desc"],
        };
        var queryString = qs.stringify(queryData, {
            encodeValuesOnly: true
        });
        triggerGetModule(queryString);
        setSelectedMenus([]);
    }, [isDeleteModuleSuceess]);
    var onPageChange = function (event) {
        setFirst(event.first);
        setRows(event.rows);
        setQueryString(event.first, event.rows, sortField, sortOrder, filters);
    };
    var onSort = function (event) {
        var sortField = event.sortField, sortOrder = event.sortOrder;
        var validSortOrder = sortOrder === 1 || sortOrder === -1 ? sortOrder : 0;
        setSortField(sortField);
        setSortOrder(validSortOrder);
        setFirst(0);
        setQueryString(0, rows, sortField, sortOrder === 1 || sortOrder === -1 ? sortOrder : 0, filters);
    };
    var onSelectionChange = function (event) {
        var value = event.value;
        setSelectedMenus(value);
    };
    // Filter fuctions
    var setQueryString = function (offset, row, field, order, filterQuery) { return __awaiter(void 0, void 0, void 0, function () {
        var formattedFilters, queryData, queryString;
        return __generator(this, function (_a) {
            formattedFilters = {};
            Object.keys(filterQuery).forEach(function (field) {
                var _a, _b;
                var filterValue = filterQuery[field].value;
                var matchMode = filterQuery[field].matchMode;
                if (filterValue !== null && filterValue !== undefined) {
                    var operator = matchMode;
                    switch (matchMode) {
                        case FilterMatchMode.CONTAINS:
                            operator = "$containsi";
                            break;
                        case FilterMatchMode.STARTS_WITH:
                            operator = "$startsWithi";
                            break;
                        case FilterMatchMode.EQUALS:
                            operator = "$eqi";
                            break;
                        case FilterMatchMode.NOT_CONTAINS:
                            operator = "$notContainsi";
                            break;
                        case FilterMatchMode.NOT_EQUALS:
                            operator = "$nei";
                            break;
                        case FilterMatchMode.ENDS_WITH:
                            operator = "$endsWithi";
                            break;
                        default:
                            operator = matchMode;
                    }
                    if (field.includes(".")) {
                        var fieldParts = field.split(".");
                        if (!formattedFilters[fieldParts[0]]) {
                            formattedFilters[fieldParts[0]] = {};
                        }
                        formattedFilters[fieldParts[0]][fieldParts[1]] = (_a = {},
                            _a[operator] = filterValue,
                            _a);
                    }
                    else {
                        formattedFilters[field] = (_b = {}, _b[operator] = filterValue, _b);
                    }
                }
            });
            queryData = {
                offset: offset !== null && offset !== void 0 ? offset : first,
                limit: row !== null && row !== void 0 ? row : rows,
                filters: formattedFilters,
            };
            if (field) {
                queryData.sort = [
                    "".concat(field, ":").concat(order == 0 ? null : order == 1 ? "asc" : "desc"),
                ];
            }
            queryString = qs.stringify(queryData, {
                encodeValuesOnly: true,
            });
            triggerGetModule(queryString);
            return [2 /*return*/];
        });
    }); };
    var onFilter = function (e) {
        setFilters(e.filters);
        setQueryString(0, rows, sortField, sortOrder === 1 || sortOrder === -1 ? sortOrder : 0, e.filters);
    };
    var detailsBodyTemplate = function (moduleMetadata) {
        return (_jsxs(Link, { href: "".concat(moduleMetadata.id), rel: "noopener noreferrer", className: "text-sm font-bold p-0", style: { color: "#12415D" }, children: [moduleMetadata.isSystem === true &&
                    _jsx("i", { className: "pi pi-eye", style: { fontSize: "1rem" } }), moduleMetadata.isSystem === false &&
                    _jsx("i", { className: "pi pi-pencil", style: { fontSize: "1rem" } })] }));
    };
    var generateCodeBodyTemplate = function (moduleMetadata) {
        return (_jsx("a", { onClick: function () {
                setGenerateCodeForModule(moduleMetadata.id);
                setGenerateCodeVisible(true);
            }, rel: "noopener noreferrer", className: "text-sm font-bold p-0", style: { color: "#12415D" }, children: moduleMetadata.isSystem === false &&
                _jsx("i", { className: "pi pi-cog", style: { fontSize: "1rem" } }) }));
    };
    var deleteBulk = function () {
        var deleteList = [];
        selectedMenus.forEach(function (element) {
            deleteList.push(element.id);
        });
        deleteManyModules(deleteList);
        setDialogVisible(false);
    };
    var onDeleteClose = function () {
        setDialogVisible(false);
        setSelectedMenus([]);
    };
    var generateCodeHandler = function () { return __awaiter(void 0, void 0, void 0, function () {
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, generateCode({ id: generateCodeForModule })];
                case 1:
                    _a.sent();
                    setGenerateCodeVisible(false);
                    return [2 /*return*/];
            }
        });
    }); };
    var onGenerateCodeClose = function () {
        setGenerateCodeVisible(false);
        setSelectedMenus([]);
    };
    var isFetchBaseQueryErrorWithErrorResponse = function (error) {
        return error && typeof error === 'object' && 'data' in error && 'message' in error.data;
    };
    // const handleError = (errorToast: any) => {
    //   let errorMessage: any = ['An error occurred'];
    //   if (isFetchBaseQueryErrorWithErrorResponse(errorToast)) {
    //     errorMessage = errorToast.data.message;
    //   } else {
    //     errorMessage = ['Something went wrong'];
    //   }
    //   toast?.current?.show({
    //     severity: 'error',
    //     summary: 'Error',
    //     detail: errorMessage,
    //     life: 3000,
    //     //@ts-ignore
    //     content: (props) => (
    //       <div className="flex flex-column align-items-left" style={{ flex: "1" }}>
    //         {Array.isArray(errorMessage) ? (
    //           errorMessage.map((message, index) => (
    //             <div className="flex align-items-center gap-2" key={index}>
    //               <span className="font-bold text-900">{message.trim()}</span>
    //             </div>
    //           ))
    //         ) : (
    //           <div className="flex align-items-center gap-2">
    //             <span className="font-bold text-900">{errorMessage?.trim()}</span>
    //           </div>
    //         )}
    //       </div>
    //     ),
    //   });
    // };
    useEffect(function () {
        if (isGenerateCodeError) {
            handleError(generateCodeError);
        }
    }, [isGenerateCodeError]);
    useEffect(function () {
        if (isGenerateCodeSuceess) {
            handleSuccess(["Code Generated Successfully"]);
        }
    }, [isGenerateCodeSuceess]);
    var handleRefreshPermission = function () {
        refreshPermission("");
    };
    return (_jsxs("div", { className: "", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "flex gap-3 mb-4", children: [_jsx(CreateButton, {}), _jsx(Button, { className: 'small-button', onClick: handleRefreshPermission, children: "Refresh Permissions" }), selectedMenus.length > 0 && _jsx(Button, { type: "button", label: "Delete", size: "small", onClick: function () { return setDialogVisible(true); }, className: "small-button", severity: "danger" })] }), _jsx("style", { children: "\n      .p-datatable .p-datatable-loading-overlay {\n      background-color: rgba(0, 0, 0, 0.0);\n      }\n      " }), _jsxs(DataTable, { value: moduleMetadata, lazy: true, size: "small", paginator: true, rows: rows, rowsPerPageOptions: [10, 25, 50], dataKey: "id", filters: filters, emptyMessage: "No Modules found", onFilter: onFilter, filterDisplay: "row", totalRecords: totalRecords, first: first, onPage: onPageChange, onSort: function (e) { return onSort(e); }, sortField: sortField, sortOrder: sortOrder === 1 || sortOrder === -1 ? sortOrder : 0, loading: loading || isLoading, loadingIcon: "pi pi-spinner", selection: selectedMenus, onSelectionChange: onSelectionChange, selectionMode: "multiple", removableSort: true, children: [_jsx(Column, { selectionMode: "multiple", headerStyle: { width: "3em" } }), _jsx(Column, { field: "id", header: "Id", className: "text-sm", sortable: true, headerClassName: "table-header-fs" }), _jsx(Column, { field: "displayName", header: "Display Name", className: "text-sm", sortable: true, filter: true, filterPlaceholder: "Search model by name", style: { minWidth: "12rem" }, headerClassName: "table-header-fs" }), _jsx(Column, { header: "Edit", body: detailsBodyTemplate }), _jsx(Column, { header: "Code", body: generateCodeBodyTemplate })] }), _jsx(Dialog, { visible: isDialogVisible, header: "Confirm Delete", modal: true, className: "solid-confirm-dialog", footer: function () { return (_jsxs("div", { className: "flex justify-content-center", children: [_jsx(Button, { label: "Yes", icon: "pi pi-check", className: 'small-button', severity: "danger", autoFocus: true, onClick: deleteBulk }), _jsx(Button, { label: "No", icon: "pi pi-times", className: 'small-button', onClick: onDeleteClose })] })); }, onHide: function () { return setDialogVisible(false); }, children: _jsx("p", { children: "Are you sure you want to delete the selected Modules?" }) }), _jsx(Dialog, { visible: isGenerateCodeVisible, header: "Generate Code", headerClassName: "text-center", modal: true, className: "solid-confirm-dialog", footer: function () { return (_jsxs("div", { className: "flex justify-content-center", children: [_jsx(Button, { label: "Yes", icon: "pi pi-check", className: 'small-button', severity: "danger", autoFocus: true, onClick: generateCodeHandler }), _jsx(Button, { label: "No", icon: "pi pi-times", className: 'small-button', onClick: onGenerateCodeClose })] })); }, onHide: function () { return setGenerateCodeVisible(false); }, children: _jsx("p", { className: "text-center", children: "Proceed with module code generation? Existing files will be overwritten." }) })] }));
};
