// @solidxai/core-ui/dist/components/core/dashboard/SolidDashboardVariable.js
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { AutoComplete } from "primereact/autocomplete";
import { Calendar } from "primereact/calendar";
import { useEffect, useState } from "react";
import styles from './SolidDashboard.module.css';
import { useGetDashboardVariableSelectionDynamicValuesQuery } from "../../../redux/api/dashboardApi";
import { ProgressSpinner } from "primereact/progressspinner";
import { Button } from "primereact/button";
export var DateVariableFilterComponent = function (_a) {
    var setFilters = _a.setFilters, clearSignal = _a.clearSignal, dashboardVariable = _a.dashboardVariable;
    // Initialize the default dates state
    // If the dashboardVariable has a defaultOperator as Between and a defaultValue as two dates, we can set those as the initial values, otherwise null
    var defaultDatesString = dashboardVariable.defaultOperator === '$between' && dashboardVariable.defaultValue;
    var defaultDatesArray = JSON.parse(defaultDatesString || '[]');
    // Map the defaultDatesArray to Date objects
    var defaultDates = defaultDatesArray.length === 2 ? [
        new Date(defaultDatesArray[0]),
        new Date(defaultDatesArray[1])
    ] : null;
    var _b = useState(defaultDates), dates = _b[0], setDates = _b[1];
    useEffect(function () {
        setDates(null); // reset UI
    }, [clearSignal]);
    return (_jsxs("div", { className: "flex align-items-center ".concat(styles.SolidDashboardDateRangeFilterWrapper), children: [_jsx(Calendar, { value: dates, onChange: function (e) {
                    console.log("Calendar changed values are: ");
                    console.log(e.value);
                    var newDates = e.value;
                    setDates(newDates);
                    if (newDates && newDates.length === 2 && newDates[0] && newDates[1]) {
                        var filter_1 = {
                            variableName: dashboardVariable.variableName,
                            // @ts-ignore
                            operator: '$between',
                            value: [
                                newDates[0].toISOString().split('T')[0],
                                newDates[1].toISOString().split('T')[0]
                            ]
                        };
                        setFilters(function (prev) {
                            // Remove any existing dashboardDate filter and replace with the new one
                            // const filtered = prev.filter(f => f.variableName !== 'dashboardDate');
                            return [filter_1];
                        });
                    }
                }, selectionMode: "range", readOnlyInput: true, hideOnRangeSelection: true, inputStyle: { height: 36.38 }, className: styles.SolidDashboardDateRangeFilter }), _jsx("div", { className: "px-2", children: _jsx("i", { className: "pi pi-calendar opacity-50" }) })] }));
};
export var SelectionDynamicVariableFilterComponent = function (_a) {
    var setFilters = _a.setFilters, clearSignal = _a.clearSignal, dashboardVariable = _a.dashboardVariable;
    // Initialize the selection dynamic values state
    // Pick the values from defaultValue if present, for default operator $in
    var defaultDynamicValuesString = dashboardVariable.defaultOperator === '$in' && dashboardVariable.defaultValue;
    var defaultDynamicValues = JSON.parse(defaultDynamicValuesString || '[]');
    // Selection Dynamic Values
    var _b = useState(defaultDynamicValues), selectionDynamicValues = _b[0], setSelectionDynamicValues = _b[1];
    var _c = useState([]), filteredItems = _c[0], setFilteredItems = _c[1];
    // Using rtk query to fetch dynamic values based on variableId
    var variableId = dashboardVariable.id;
    var queryString = "variableId=".concat(variableId);
    var _d = useGetDashboardVariableSelectionDynamicValuesQuery(queryString), dynamicValues = _d.data, isLoading = _d.isLoading, isError = _d.isError;
    var allItems = dynamicValues ? dynamicValues : [];
    // Dynamic search
    var searchDynamic = function (event) {
        var query = event.query.toLowerCase();
        var filtered = allItems.map(function (item) { return item.label; }).filter(function (item) {
            return item.toLowerCase().includes(query);
        });
        setFilteredItems(filtered);
    };
    // Set the filters whenever selectionDynamicValues change
    useEffect(function () {
        console.log("Selection Dynamic Values changed:", selectionDynamicValues);
        if (selectionDynamicValues.length > 0) {
            var filter_2 = {
                variableName: dashboardVariable.variableName,
                // @ts-ignore
                operator: '$in',
                value: selectionDynamicValues
            };
            setFilters(function (prev) {
                // Remove any existing filter for this variable and replace with the new one
                var filtered = prev.filter(function (f) { return f.variableName !== dashboardVariable.variableName; });
                return __spreadArray(__spreadArray([], filtered, true), [filter_2], false);
            });
        }
        // else {
        //   // If no values selected, remove the filter for this variable
        //   setFilters(prev => prev.filter(f => f.variableName !== dashboardVariable.variableName));
        // }
    }, [selectionDynamicValues]);
    useEffect(function () {
        setSelectionDynamicValues([]);
    }, [clearSignal]);
    return (_jsxs(_Fragment, { children: [isLoading && _jsx(ProgressSpinner, {}), isError && _jsx("div", { children: "Error loading values" }), !isLoading && !isError &&
                _jsx(AutoComplete, { value: selectionDynamicValues, suggestions: filteredItems, completeMethod: searchDynamic, onChange: function (e) {
                        console.log('AutoComplete changed values are:');
                        console.log(e);
                        setSelectionDynamicValues(e.value);
                    }, multiple: true, dropdown: true, placeholder: dashboardVariable.variableName, className: "solid-standard-autocomplete", style: { minHeight: 38 } })] }));
};
export var SelectionStaticVariableFilterComponent = function (_a) {
    var setFilters = _a.setFilters, clearSignal = _a.clearSignal, dashboardVariable = _a.dashboardVariable;
    // Initialize the selection static values state
    // Pick the values from defaultValue if present, for default operator $in
    var defaultStaticValuesString = dashboardVariable.defaultOperator === '$in' && dashboardVariable.defaultValue;
    var defaultStaticValues = JSON.parse(defaultStaticValuesString || '[]');
    // Selection Static Values
    var _b = useState(defaultStaticValues), selectionStaticValues = _b[0], setSelectionStaticValues = _b[1];
    var _c = useState([]), filteredStaticItems = _c[0], setFilteredStaticItems = _c[1];
    var staticValues = dashboardVariable.selectionStaticValues || [];
    // The values are in the format val:label, we need to extract the labels
    var staticValueItems = staticValues.map(function (val) { return ({ value: val.split(':')[0], label: val.split(':')[1] }); });
    // Static search
    var searchStatic = function (event) {
        var query = event.query.toLowerCase();
        var filtered = staticValueItems.map(function (item) { return item.label; }).filter(function (item) {
            return item.toLowerCase().includes(query);
        });
        setFilteredStaticItems(filtered);
    };
    // Set the filters whenever selectionStaticValues change
    useEffect(function () {
        // console.log(`Selection Static Values changed:`, selectionStaticValues);
        if (selectionStaticValues.length > 0) {
            var filter_3 = {
                variableName: dashboardVariable.variableName,
                // @ts-ignore
                operator: '$in',
                value: selectionStaticValues
            };
            setFilters(function (prev) {
                // Remove any existing filter for this variable and replace with the new one
                var filtered = prev.filter(function (f) { return f.variableName !== dashboardVariable.variableName; });
                return __spreadArray(__spreadArray([], filtered, true), [filter_3], false);
            });
        }
        // else {
        //   // If no values selected, remove the filter for this variable
        //   setFilters(prev => prev.filter(f => f.variableName !== dashboardVariable.variableName));
        // }
    }, [selectionStaticValues]);
    useEffect(function () {
        setSelectionStaticValues([]);
    }, [clearSignal]);
    return (_jsx(AutoComplete, { value: selectionStaticValues, suggestions: filteredStaticItems, completeMethod: searchStatic, onChange: function (e) { return setSelectionStaticValues(e.value); }, multiple: true, dropdown: true, placeholder: dashboardVariable.variableName, className: "solid-standard-autocomplete", style: { minHeight: 38 } }));
};
var SolidDashboardVariable = function (_a) {
    var dashboardVariables = _a.dashboardVariables, filters = _a.filters, setFilters = _a.setFilters;
    var _b = useState(0), clearSignal = _b[0], setClearSignal = _b[1];
    var dashboardVariableComponents = dashboardVariables.map(function (dashboardVariable, index) {
        switch (dashboardVariable.variableType) {
            case 'date':
                return _jsx(DateVariableFilterComponent, { setFilters: setFilters, clearSignal: clearSignal, dashboardVariable: dashboardVariable }, index);
            case 'selectionStatic':
                return _jsx(SelectionStaticVariableFilterComponent, { setFilters: setFilters, clearSignal: clearSignal, dashboardVariable: dashboardVariable }, index);
            case 'selectionDynamic':
                return _jsx(SelectionDynamicVariableFilterComponent, { setFilters: setFilters, clearSignal: clearSignal, dashboardVariable: dashboardVariable }, index);
            default:
                return null;
        }
    });
    var clearAllFilters = function () {
        setFilters([]);
        setClearSignal(function (prev) { return prev + 1; }); // triggers children to reset
    };
    // TODO [HP]: Currently this is static, we need this to be dynamic how we are invoking setFilters below has to be fully dynamic...
    return (_jsxs("div", { className: "flex align-items-center gap-3", children: [dashboardVariableComponents, filters.length > 0 && (_jsx(Button, { onClick: clearAllFilters, size: "small", outlined: true, 
                // severity="info"
                icon: "pi pi-filter-slash" }))] }));
};
export default SolidDashboardVariable;
