// @solidxai/core-ui/dist/components/core/dashboard/SolidDashboard.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useGetDashboardQuery } from '../../../redux/api/dashboardApi';
import { Button } from 'primereact/button';
import { Tooltip } from "primereact/tooltip";
import qs from 'qs';
import { useEffect, useState } from 'react';
import { SolidXAIIcon } from '../solid-ai/SolidXAIIcon';
import styles from './SolidDashboard.module.css';
import SolidDashboardBody from './SolidDashboardBody';
import SolidDashboardVariable from './SolidDashboardVariable';
import { SolidAiMainWrapper } from '../solid-ai/SolidAiMainWrapper';
import { SolidDashboardFilterRequired } from './SolidDashboardFilterRequired';
import { SolidDashboardLoading } from './SolidDashboardLoading';
import { SolidDashboardRenderError } from './SolidDashboardRenderError';
import { useDispatch, useSelector } from "react-redux";
import { showNavbar, toggleNavbar } from "../../../redux/features/navbarSlice";
import SolidDashboardNotAvailable from './SolidDashboardNotAvailable';
import { useLazyGetMcpUrlQuery, useLazyGetSolidSettingsQuery } from '../../../redux/api/solidSettingsApi';
export var DashboardVariableType;
(function (DashboardVariableType) {
    DashboardVariableType["DATE"] = "date";
    DashboardVariableType["SELECTION_STATIC"] = "selectionStatic";
    DashboardVariableType["SELECTION_DYNAMIC"] = "selectionDynamic";
})(DashboardVariableType || (DashboardVariableType = {}));
var SOURCE_TYPE;
(function (SOURCE_TYPE) {
    SOURCE_TYPE["SQL"] = "sql";
    SOURCE_TYPE["PROVIDER"] = "provider";
})(SOURCE_TYPE || (SOURCE_TYPE = {}));
function handleDashboardData(data, setDashboardVariables, setQuestions) {
    var records = data.records, meta = data.meta;
    if (records && records.length > 0) {
        // Set the layout options for the dashboard body
        var dashboardData = records[0]; // Assuming we want the first dashboard
        // Set the dashboard variables
        setDashboardVariables(dashboardData.dashboardVariables || []);
        // Set the dashboard questions
        setQuestions(dashboardData.questions);
    }
}
function getQueryParams(moduleName, dashboardId, dashboardName) {
    var filters = {
        module: {
            name: {
                $eq: moduleName
            }
        }
    };
    if (dashboardId !== undefined) {
        filters.id = { $eq: dashboardId };
    }
    else if (dashboardName !== undefined) {
        filters.name = { $eq: dashboardName };
    }
    var query = {
        filters: filters,
        populate: ['dashboardVariables', 'questions']
    };
    var urlQuery = qs.stringify(query, {
        encodeValuesOnly: true,
    });
    return urlQuery;
}
// Render the dashboard body only if:
// 1. There are dashboard questions
// AND
// (
// 1. There are dashboard variables and all dashboard variable filter rules have been applied i.e (all dashboard variables have been selected)
//. OR
// 2. There are no dashboard variables
// )
// 
function isRenderDashboardBody(questions, dashboardVariables, filters) {
    if (questions.length === 0) {
        return false;
    }
    if (dashboardVariables.length === 0) {
        return true;
    }
    // Check if all dashboard variables have corresponding filters applied
    var allVariablesFiltered = dashboardVariables.every(function (variable) {
        return filters.some(function (filter) { return filter.variableName === variable.variableName; });
    });
    return allVariablesFiltered;
}
var SolidDashboard = function (params) {
    var _a, _b, _c, _d, _e;
    var _f = useGetDashboardQuery(getQueryParams(params.moduleName, params.dashboardId, params.dashboardName)), data = _f.data, isLoading = _f.isLoading, error = _f.error; // FIXME : error handling should be done properly
    // Define a state called layoutOption and pass it after destructing the widgetOptions and dashboardOptions from layoutOption
    // TODO [HP]: Shouldn't the type of this state variable be something different? Why are we muddling this with layout but calling it body props? 
    // TODO [HP]: Body props should be clearly made up of Gridstack layout options, the questions that make up the body & the filter[] which is an array of SqlExpressions
    // TODO [HP]: This is fully CONFUSED
    // const [layoutOption, setLayoutOption] = useState<SolidDashboardBodyProps>({
    //   filters: [],
    //   questions: [],
    // });
    // TODO [HP]: replace dashboardVariableFilterRules with filters everywhere...
    // const [dashboardVariableFilterRules, setDashboardVariableFilterRules] = useState<ISolidDashboardVariableFilterRule[]>([]);
    var dispatch = useDispatch();
    var visibleNavbar = useSelector(function (state) { var _a; return (_a = state.navbarState) === null || _a === void 0 ? void 0 : _a.visibleNavbar; });
    var _g = useState([]), filters = _g[0], setFilters = _g[1];
    var _h = useState(false), isOpenSolidXAiPanel = _h[0], setIsOpenSolidXAiPanel = _h[1];
    var _j = useState(380), chatterWidth = _j[0], setChatterWidth = _j[1];
    var _k = useState(false), isResizing = _k[0], setIsResizing = _k[1];
    var _l = useState([]), questions = _l[0], setQuestions = _l[1];
    var _m = useState([]), dashboardVariables = _m[0], setDashboardVariables = _m[1];
    useEffect(function () {
        // Invoke the dashboard api to fetch the dashboard data
        // console.log('Dashboard Data testing:', isLoading, data, error);
        if (!isLoading && data) {
            // Assuming data contains the layout options
            handleDashboardData(data, setDashboardVariables, setQuestions);
        }
    }, [isLoading, data]);
    useEffect(function () {
        var storedOpen = localStorage.getItem('d_solidxai_open');
        var storedWidth = localStorage.getItem('d_solidxai_width');
        if (storedOpen !== null) {
            setIsOpenSolidXAiPanel(storedOpen === 'true');
        }
        if (storedWidth !== null) {
            var width = parseInt(storedWidth, 10);
            if (!isNaN(width)) {
                setChatterWidth(width);
            }
        }
    }, []);
    useEffect(function () {
        if (isResizing) {
            var handleMouseMove_1 = function (e) {
                var newWidth = window.innerWidth - e.clientX;
                var clampedWidth = Math.max(280, Math.min(newWidth, 700));
                setChatterWidth(clampedWidth);
                localStorage.setItem('d_solidxai_width', clampedWidth.toString());
            };
            var handleMouseUp_1 = function () {
                setIsResizing(false);
            };
            window.addEventListener('mousemove', handleMouseMove_1);
            window.addEventListener('mouseup', handleMouseUp_1);
            return function () {
                window.removeEventListener('mousemove', handleMouseMove_1);
                window.removeEventListener('mouseup', handleMouseUp_1);
            };
        }
    }, [isResizing]);
    var handleOpen = function () {
        setIsOpenSolidXAiPanel(true);
        localStorage.setItem('d_solidxai_open', 'true');
    };
    var handleClose = function () {
        setIsOpenSolidXAiPanel(false);
        localStorage.setItem('d_solidxai_open', 'false');
    };
    var toggleBothSidebars = function () {
        if (visibleNavbar) {
            dispatch(toggleNavbar()); // close both
        }
        else {
            dispatch(showNavbar()); // open both
        }
    };
    var _o = useState(null), mcpUrl = _o[0], setMcpUrl = _o[1];
    var getMcpUrl = useLazyGetMcpUrlQuery()[0];
    var _p = useLazyGetSolidSettingsQuery(), trigger = _p[0], solidSettingsData = _p[1].data;
    useEffect(function () {
        trigger(""); // Fetch settings on mount
    }, []);
    useEffect(function () {
        var _a, _b;
        if (((_a = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _a === void 0 ? void 0 : _a.mcpEnabled) && ((_b = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _b === void 0 ? void 0 : _b.mcpServerUrl)) {
            enableSolidXAiPanel();
        }
    }, [solidSettingsData]);
    var enableSolidXAiPanel = function () { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, queryString, response, error_1;
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _c.trys.push([0, 2, , 3]);
                    queryData = {
                        showHeader: "true",
                        inListView: "true"
                    };
                    queryString = qs.stringify(__assign({}, queryData), { encodeValuesOnly: true });
                    return [4 /*yield*/, getMcpUrl(queryString).unwrap()];
                case 1:
                    response = _c.sent();
                    console.log("response", response);
                    if (response && ((_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.mcpUrl)) {
                        setMcpUrl((_b = response === null || response === void 0 ? void 0 : response.data) === null || _b === void 0 ? void 0 : _b.mcpUrl);
                    }
                    return [3 /*break*/, 3];
                case 2:
                    error_1 = _c.sent();
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    return (_jsxs("div", { className: "h-screen surface-0 flex", children: [_jsxs("div", { className: "h-full flex-grow-1 ".concat(styles.SolidDashboardPageContentWrapper), children: [isLoading && _jsx(SolidDashboardLoading, {}), error && _jsx(SolidDashboardRenderError, {}), !isLoading && !error && data && data.records.length === 0 && (_jsx(SolidDashboardNotAvailable, {})), !isLoading && !error && data && data.records.length > 0 && (_jsxs(_Fragment, { children: [_jsxs("div", { className: "page-header", style: { borderBottom: '1px solid var(--primary-light-color)' }, children: [_jsxs("div", { className: 'flex align-items-center gap-2', children: [_jsx("div", { className: "apps-icon block md:hidden cursor-pointer", onClick: toggleBothSidebars, children: _jsx("i", { className: "pi pi-th-large" }) }), _jsxs("p", { className: "view-title solid-text-wrapper flex align-items-center gap-1 ".concat(styles.SolidDashboardTitle), children: [((_a = data === null || data === void 0 ? void 0 : data.records[0]) === null || _a === void 0 ? void 0 : _a.displayName) ? (_b = data === null || data === void 0 ? void 0 : data.records[0]) === null || _b === void 0 ? void 0 : _b.displayName : (_c = data === null || data === void 0 ? void 0 : data.records[0]) === null || _c === void 0 ? void 0 : _c.name, ((_d = data === null || data === void 0 ? void 0 : data.records[0]) === null || _d === void 0 ? void 0 : _d.description) &&
                                                        _jsxs(_Fragment, { children: [_jsx(Tooltip, { className: 'solid-field-tooltip', target: ".solid-field-tooltip-icon" }), _jsx("i", { className: "pi pi-info-circle solid-field-tooltip-icon", "data-pr-tooltip": (_e = data === null || data === void 0 ? void 0 : data.records[0]) === null || _e === void 0 ? void 0 : _e.description, "data-pr-position": 'right' })] })] })] }), dashboardVariables && dashboardVariables.length > 0 && _jsx(SolidDashboardVariable, { dashboardVariables: dashboardVariables, filters: filters, setFilters: setFilters })] }), !isRenderDashboardBody(questions, dashboardVariables, filters) && _jsx(SolidDashboardFilterRequired, {}), isRenderDashboardBody(questions, dashboardVariables, filters) && _jsx(SolidDashboardBody, { questions: questions, filters: filters })] }))] }), mcpUrl && (_jsxs("div", { className: "chatter-section ".concat(isOpenSolidXAiPanel === false ? 'collapsed' : 'open'), style: { width: chatterWidth }, children: [isOpenSolidXAiPanel && (_jsx("div", { style: {
                            width: 5,
                            cursor: 'col-resize',
                            position: 'absolute',
                            left: 0,
                            top: 0,
                            bottom: 0,
                            height: '100%',
                            zIndex: 9,
                        }, onMouseDown: function () { return setIsResizing(true); } })), isOpenSolidXAiPanel &&
                        _jsx(Button, { icon: "pi pi-angle-double-right", size: "small", text: true, className: "chatter-collapse-btn", style: { width: 30, height: 30, aspectRatio: '1/1' }, onClick: handleClose }), isOpenSolidXAiPanel === false ?
                        _jsxs("div", { className: "flex flex-column gap-2 justify-content-center p-2", children: [_jsx("div", { className: "chatter-collapsed-content", onClick: handleOpen, children: _jsxs("div", { className: "flex gap-2", children: [" ", _jsx(SolidXAIIcon, {}), " SolidX AI "] }) }), _jsx(Button, { icon: "pi pi-chevron-left", size: "small", className: "px-0", style: { width: 30 }, onClick: handleOpen })] })
                        :
                            _jsx(SolidAiMainWrapper, { mcpUrl: mcpUrl })] }))] }));
};
export default SolidDashboard;
