// @solidxai/core-ui/dist/components/common/AutoCompleteField.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx } from "react/jsx-runtime";
import { AutoComplete } from "primereact/autocomplete";
import { useEffect, useState } from "react";
import { getVirtualScrollerOptions } from "../../helpers/autoCompleteVirtualScroll";
export var AutoCompleteField = function (_a) {
    var multiple = _a.multiple, isFormFieldValid = _a.isFormFieldValid, formik = _a.formik, fieldName = _a.fieldName, searchData = _a.searchData, existingData = _a.existingData, existingDataTitle = _a.existingDataTitle, existingDataId = _a.existingDataId, onBlur = _a.onBlur;
    var _b = useState(existingData), selectedItem = _b[0], setSelectedItem = _b[1];
    var _c = useState([]), filteredItem = _c[0], setFilteredItem = _c[1];
    useEffect(function () {
        if (existingData) {
            setSelectedItem(existingData);
            // formik.setFieldValue(fieldName, existingData);
        }
    }, [existingData]);
    var searchItems = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var data;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, searchData(event)];
                case 1:
                    data = _a.sent();
                    setFilteredItem(data);
                    return [2 /*return*/];
            }
        });
    }); };
    return (_jsx(AutoComplete, { value: selectedItem, multiple: multiple ? multiple : false, suggestions: filteredItem, invalid: isFormFieldValid(formik, fieldName), completeMethod: searchItems, 
        // virtualScrollerOptions={{ itemSize: 38 }}
        virtualScrollerOptions: getVirtualScrollerOptions({
            itemsLength: filteredItem.length,
        }), className: "solid-standard-autocomplete w-full", 
        // style={{
        //     maxHeight: 39.67
        // }}
        field: "label", dropdown: true, onBlur: onBlur, onChange: function (e) {
            var data = {};
            if (typeof (e.value) == "string") {
                data = e.value;
            }
            else {
                if (multiple === true) {
                    data = e.value.map(function (m) {
                        return ({
                            label: m.label,
                            name: m.label,
                            value: m.value,
                            id: m.value,
                        });
                    });
                }
                else {
                    data = {
                        label: e.value.label,
                        name: e.value.label,
                        value: e.value.value,
                        id: e.value.value,
                    };
                }
            }
            setSelectedItem(data);
            formik.setFieldValue(fieldName, data);
        } }));
};
