// @solidxai/core-ui/dist/components/common/MarkdownViewer.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';
import { Prism as SyntaxHighlighter } from 'react-syntax-highlighter';
import { oneDark } from 'react-syntax-highlighter/dist/esm/styles/prism';
// Custom CodeBlock renderer
var CodeBlock = function (_a) {
    var _b;
    var inline = _a.inline, className = _a.className, children = _a.children, props = __rest(_a, ["inline", "className", "children"]);
    var codeString = String(children).replace(/\n$/, '');
    var isFenced = !inline && /language-(\w+)/.test(className || '');
    var match = /language-(\w+)/.exec(className || '');
    var language = (_b = match === null || match === void 0 ? void 0 : match[1]) !== null && _b !== void 0 ? _b : '';
    if (inline || !isFenced) {
        // Inline code or non-fenced code
        return (_jsx("code", { className: "bg-gray-100 px-1 py-0.5 rounded text-sm", children: children }));
    }
    // Fenced code block
    return (_jsxs("div", { className: "relative group my-4", children: [_jsx("button", { className: "absolute right-2 top-2 px-2 py-1 text-xs rounded bg-white text-black transition", style: { cursor: 'pointer', right: 0 }, onClick: function () { return navigator.clipboard.writeText(codeString); }, children: "Copy" }), _jsx(SyntaxHighlighter, __assign({ language: language, style: oneDark, PreTag: "div", customStyle: {
                    borderRadius: '0.5rem',
                    padding: '1rem',
                    fontSize: '0.85rem',
                    background: '#282c34',
                } }, props, { children: codeString }))] }));
};
var MarkdownViewer = function (_a) {
    var data = _a.data;
    return (_jsx(ReactMarkdown, { remarkPlugins: [remarkGfm], className: "markdown-view prose prose-invert max-w-none", components: {
            code: CodeBlock,
            table: function (props) { return (_jsx("div", { className: "overflow-x-auto my-4", children: _jsx("table", { className: "min-w-full border border-gray-700 text-sm", children: props.children }) })); },
            th: function (props) { return (_jsx("th", { className: "border border-gray-700 px-3 py-1 font-semibold text-left bg-gray-800", style: { color: "#fff" }, children: props.children })); },
            td: function (props) { return (_jsx("td", { className: "border border-gray-700 px-3 py-1 align-top", children: props.children })); },
        }, children: data }));
};
export default MarkdownViewer;
