// @solidxai/core-ui/dist/components/common/SolidModuleHome.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useSession } from "../../hooks/useSession";
import styles from './solidModuleHome.module.css';
import { Button } from 'primereact/button';
import { DocsSvg } from '../Svg/DocsSvg';
import { SettingsSvg } from '../Svg/SettingsSvg';
import { DevDocs } from '../Svg/DevDocs';
import { HomePageModuleSvg } from '../Svg/HomePageModuleSvg';
import Link from "./Link";
import { useState } from 'react';
import { SolidAccountSettings } from '../core/common/SolidAccountSettings/SolidAccountSettings';
import { useDispatch, useSelector } from "react-redux";
import { showNavbar, toggleNavbar } from "../../redux/features/navbarSlice";
export var SolidModuleHome = function (_a) {
    var _b;
    var _c = _a.moduleName, moduleName = _c === void 0 ? "Dashboard" : _c;
    var _d = useSession(), session = _d.data, status = _d.status;
    var dispatch = useDispatch();
    var visibleNavbar = useSelector(function (state) { var _a; return (_a = state.navbarState) === null || _a === void 0 ? void 0 : _a.visibleNavbar; });
    //@ts-ignore
    var user = (_b = session === null || session === void 0 ? void 0 : session.user) === null || _b === void 0 ? void 0 : _b.username;
    var _e = useState(false), showAccountSettings = _e[0], setShowAccountSettings = _e[1];
    var toggleBothSidebars = function () {
        if (visibleNavbar) {
            dispatch(toggleNavbar()); // close both
        }
        else {
            dispatch(showNavbar()); // open both
        }
    };
    return (_jsxs("div", { className: "h-screen surface-0 overflow-y-auto", children: [_jsx("div", { className: "page-header", style: { borderBottom: '1px solid var(--primary-light-color)' }, children: _jsxs("div", { className: 'flex align-items-center gap-2', children: [_jsx("div", { className: "apps-icon block md:hidden cursor-pointer", onClick: toggleBothSidebars, children: _jsx("i", { className: "pi pi-th-large" }) }), _jsx("p", { className: "m-0 view-title solid-text-wrapper", children: moduleName })] }) }), _jsxs("div", { className: "p-4 md:p-5 flex flex-column gap-4", children: [_jsxs("div", { className: styles.solidModuleWelcomeSection, children: [_jsxs("div", { className: 'flex flex-column gap-2 pr-3 lg:pr-0', children: [_jsxs("h3", { className: styles.solidModuleHomeTitle1, children: ["Welcome, ", status === "loading" ? "." : user] }), _jsxs("p", { className: 'm-0 w-35rem text-sl', children: ["This is the home page for the ", moduleName, " module. You can find documentation and settings below."] })] }), _jsx("div", { className: styles.homeSvg, children: _jsx(HomePageModuleSvg, {}) })] }), _jsx("div", { className: 'grid', children: [{
                                icon: _jsx(DocsSvg, {}),
                                title: 'Admin Docs',
                                description: 'Explore the administrator documentation to manage and configure your system',
                                class: styles.dashboardCardOne,
                                url: 'https://docs.solidxai.com/docs/admin-docs/'
                            }, {
                                icon: _jsx(DevDocs, {}),
                                title: 'Dev Docs',
                                description: 'Dive into the developer documentation to build, extend, and integrate with the platform',
                                class: styles.dashboardCardTwo,
                                url: 'https://docs.solidxai.com/docs/developer-docs/'
                            }, {
                                icon: _jsx(SettingsSvg, {}),
                                title: 'Settings',
                                description: 'Access and update your account settings, preferences, and profile information',
                                class: styles.dashboardCardThree,
                                type: 'popup'
                            }].map(function (card, i) { return (_jsx("div", { className: 'col-12 md:col-6 xl:col-4', children: _jsxs("div", { className: "".concat(styles.solidModuleHomeColorCards, " ").concat(card.class), children: [_jsxs("div", { children: [_jsx("div", { className: styles.solidModuleHomeCardSvgWrapper, children: card.icon }), _jsx("h3", { className: "".concat(styles.solidModuleHomeTitle1, " mt-4"), children: card.title }), _jsx("p", { className: 'mb-0 mt-2 mr-2 text-xs', children: card.description })] }), _jsx("div", { children: card.url ? (_jsx(Link, { href: card.url, target: "_blank", rel: "noopener noreferrer", children: _jsx(Button, { outlined: true, label: 'Explore', size: 'small', icon: "pi pi-arrow-right", iconPos: 'right', className: "font-bold", style: { background: 'var(--solid-module-home-add-button-bg)' } }) })) : (_jsx(Button, { outlined: true, label: 'Open', size: 'small', icon: "pi pi-cog", iconPos: 'right', className: "font-bold", style: { background: 'var(--solid-module-home-add-button-bg)' }, onClick: function () {
                                                if (card.type === 'popup') {
                                                    setShowAccountSettings(true);
                                                }
                                            } })) })] }) }, i)); }) })] }), showAccountSettings &&
                _jsx(SolidAccountSettings, { showProfileSettingsDialog: showAccountSettings, setShowProfileSettingsDialog: setShowAccountSettings })] }));
};
