// @solidxai/core-ui/dist/components/common/SolidFormStepper.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { OverlayPanel } from 'primereact/overlaypanel';
import { useEffect, useRef, useState } from 'react';
import { createSolidEntityApi } from '../../redux/api/solidEntityApi';
import { useFormik } from 'formik';
import { Toast } from 'primereact/toast';
import { useSearchParams } from "../../hooks/useSearchParams";
import { ERROR_MESSAGES } from '../../constants/error-messages';
import showToast from "../../helpers/showToast";
export var SolidFormStepper = function (props) {
    var _a;
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
    var solidFormViewMetaData = props.solidFormViewMetaData, modelName = props.modelName, initialEntityData = props.initialEntityData, id = props.id, solidWorkflowFieldValue = props.solidWorkflowFieldValue, setSolidWorkflowFieldValue = props.setSolidWorkflowFieldValue, onStepperUpdate = props.onStepperUpdate;
    var toast = useRef(null);
    var formStepperOverlay = useRef(null);
    var containerRef = useRef(null);
    var leftFormStepperOverlay = useRef(null);
    var searchParams = useSearchParams();
    var viewMode = searchParams.get('viewMode');
    var solidFormViewWorkflowData = (_b = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _b === void 0 ? void 0 : _b.solidFormViewWorkflowData;
    var solidWorkflowField = (_f = (_e = (_d = (_c = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _c === void 0 ? void 0 : _c.solidView) === null || _d === void 0 ? void 0 : _d.layout) === null || _e === void 0 ? void 0 : _e.attrs) === null || _f === void 0 ? void 0 : _f.workflowField;
    var solidWorkflowFieldEnabled = (_k = (_j = (_h = (_g = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _g === void 0 ? void 0 : _g.solidView) === null || _h === void 0 ? void 0 : _h.layout) === null || _j === void 0 ? void 0 : _j.attrs) === null || _k === void 0 ? void 0 : _k.workflowFieldUpdateEnabled;
    var defaultWorkflowFieldValue = (_o = (_m = (_l = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _l === void 0 ? void 0 : _l.solidFieldsMetadata) === null || _m === void 0 ? void 0 : _m[solidWorkflowField]) === null || _o === void 0 ? void 0 : _o.defaultValue;
    var defaultWorkflowFieldDisplayName = (_r = (_q = (_p = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _p === void 0 ? void 0 : _p.solidFieldsMetadata) === null || _q === void 0 ? void 0 : _q[solidWorkflowField]) === null || _r === void 0 ? void 0 : _r.displayName;
    var activeStep = (_s = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _s === void 0 ? void 0 : _s.solidFormViewWorkflowData[0].value;
    var _u = useState(""), solidWorkflowFieldKey = _u[0], setSolidWorkflowFieldKey = _u[1];
    var _v = useState(1), visibleStepsCount = _v[0], setVisibleStepsCount = _v[1];
    // Dynamic responsive logic
    useEffect(function () {
        var calculateVisibleSteps = function () {
            if (!containerRef.current || !solidFormViewWorkflowData || solidFormViewWorkflowData.length === 0)
                return;
            var containerWidth = containerRef.current.offsetWidth;
            var overflowButtonWidth = 50;
            var arrowWidth = 30; // Width for arrow between buttons
            var spacing = 0; // No gap needed as arrows connect buttons
            var tempContainer = document.createElement('div');
            tempContainer.style.visibility = 'hidden';
            tempContainer.style.position = 'absolute';
            document.body.appendChild(tempContainer);
            var buttonWidths = [];
            solidFormViewWorkflowData.forEach(function (step) {
                var tempButton = document.createElement('button');
                tempButton.className = 'arrow-step-button';
                tempButton.textContent = step.label;
                tempButton.style.padding = '0.5rem 1.5rem';
                tempButton.style.minWidth = '100px';
                tempContainer.appendChild(tempButton);
                buttonWidths.push(Math.max(100, tempButton.offsetWidth));
            });
            document.body.removeChild(tempContainer);
            var count = 0;
            var totalWidth = 0;
            for (var i = 0; i < buttonWidths.length; i++) {
                var buttonWidth = buttonWidths[i];
                var arrowSpace = i > 0 ? arrowWidth : 0;
                var needsOverflow = i < buttonWidths.length - 1;
                var requiredWidth = totalWidth + buttonWidth + arrowSpace + (needsOverflow ? overflowButtonWidth : 0);
                if (requiredWidth <= containerWidth) {
                    totalWidth += buttonWidth + arrowSpace;
                    count++;
                }
                else {
                    break;
                }
            }
            count = Math.max(1, count);
            setVisibleStepsCount(count);
        };
        calculateVisibleSteps();
        var resizeObserver = new ResizeObserver(calculateVisibleSteps);
        if (containerRef.current) {
            resizeObserver.observe(containerRef.current);
        }
        var timeoutId = setTimeout(calculateVisibleSteps, 100);
        return function () {
            resizeObserver.disconnect();
            clearTimeout(timeoutId);
        };
    }, [solidFormViewWorkflowData]);
    useEffect(function () {
        if (!solidWorkflowField)
            return;
        setSolidWorkflowFieldKey(solidWorkflowField);
        setSolidWorkflowFieldValue(function () {
            var _a, _b, _c, _d;
            if ((initialEntityData === null || initialEntityData === void 0 ? void 0 : initialEntityData[solidWorkflowField]) !== undefined) {
                if (((_c = (_b = (_a = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _a === void 0 ? void 0 : _a.solidFieldsMetadata) === null || _b === void 0 ? void 0 : _b[solidWorkflowField]) === null || _c === void 0 ? void 0 : _c.type) === "relation") {
                    return ((_d = initialEntityData[solidWorkflowField]) === null || _d === void 0 ? void 0 : _d.id) || initialEntityData[solidWorkflowField];
                }
                else {
                    return initialEntityData[solidWorkflowField];
                }
            }
            else if (defaultWorkflowFieldValue !== undefined) {
                return defaultWorkflowFieldValue;
            }
            else {
                return activeStep;
            }
        });
    }, [solidWorkflowField, initialEntityData, defaultWorkflowFieldValue, activeStep]);
    var formik = useFormik({
        enableReinitialize: true,
        initialValues: (_a = {
                id: +id
            },
            _a[solidWorkflowFieldKey] = solidWorkflowFieldValue || "",
            _a),
        onSubmit: function (values) {
            handleStepChange(values);
        }
    });
    var entityApi = createSolidEntityApi(modelName);
    var usePatchUpdateSolidEntityMutation = entityApi.usePatchUpdateSolidEntityMutation;
    var _w = usePatchUpdateSolidEntityMutation(), updateStepper = _w[0], _x = _w[1], isStepperUpdateSuccessfull = _x.isSuccess, isStepperUpdateError = _x.isError, stepperUpdateError = _x.error;
    var handleStepChange = function (values) { return __awaiter(void 0, void 0, void 0, function () {
        var result, error_1;
        var _a;
        var _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _c.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, updateStepper({ id: values.id, data: (_a = {}, _a[solidWorkflowFieldKey] = values[solidWorkflowFieldKey], _a) }).unwrap()];
                case 1:
                    result = _c.sent();
                    if ((result === null || result === void 0 ? void 0 : result.statusCode) === 200) {
                        showToast(toast, "success", ERROR_MESSAGES.FIELD_UPDATE(defaultWorkflowFieldDisplayName), ERROR_MESSAGES.FIELD_UPDATE_SUCCESSFULLY(defaultWorkflowFieldDisplayName));
                        if ((_b = result === null || result === void 0 ? void 0 : result.data) === null || _b === void 0 ? void 0 : _b[solidWorkflowFieldKey]) {
                            setSolidWorkflowFieldValue(result.data[solidWorkflowFieldKey]);
                        }
                        if (onStepperUpdate) {
                            onStepperUpdate();
                        }
                    }
                    return [3 /*break*/, 3];
                case 2:
                    error_1 = _c.sent();
                    console.error(ERROR_MESSAGES.UPDATING_STEPPER, error_1);
                    showToast(toast, "error", ERROR_MESSAGES.UPDATE_FAILED, ERROR_MESSAGES.FAILED_UPDATE_FROM);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    var handleButtonClick = function (stepValue) {
        if (solidWorkflowFieldEnabled === false || id === "new" || viewMode === "view") {
            return;
        }
        else {
            formik.setFieldValue(solidWorkflowFieldKey, stepValue);
            formik.handleSubmit();
        }
    };
    var activeIndex = (_t = solidFormViewWorkflowData === null || solidFormViewWorkflowData === void 0 ? void 0 : solidFormViewWorkflowData.findIndex(function (step) { return step.value === solidWorkflowFieldValue; })) !== null && _t !== void 0 ? _t : -1;
    var visibleSteps = [];
    var previousSteps = [];
    var nextSteps = [];
    var isSingleVisibleStep = visibleStepsCount === 1;
    if (solidFormViewWorkflowData && solidFormViewWorkflowData.length > 0) {
        if (activeIndex === -1) {
            visibleSteps = solidFormViewWorkflowData.slice(0, visibleStepsCount);
            nextSteps = solidFormViewWorkflowData.slice(visibleStepsCount);
        }
        else if (visibleStepsCount === 1) {
            visibleSteps = [solidFormViewWorkflowData[activeIndex]];
            previousSteps = solidFormViewWorkflowData.slice(0, activeIndex);
            nextSteps = solidFormViewWorkflowData.slice(activeIndex + 1);
        }
        else {
            var totalSteps = solidFormViewWorkflowData.length;
            if (activeIndex < visibleStepsCount) {
                visibleSteps = solidFormViewWorkflowData.slice(0, visibleStepsCount);
                nextSteps = solidFormViewWorkflowData.slice(visibleStepsCount);
            }
            else {
                var slotsAvailable = visibleStepsCount - 1;
                var afterActive = Math.min(Math.floor(slotsAvailable / 2), totalSteps - activeIndex - 1);
                var beforeActive = slotsAvailable - afterActive;
                var startIndex = Math.max(0, activeIndex - beforeActive);
                var endIndex = Math.min(totalSteps, activeIndex + afterActive + 1);
                previousSteps = solidFormViewWorkflowData.slice(0, startIndex);
                visibleSteps = solidFormViewWorkflowData.slice(startIndex, endIndex);
                nextSteps = solidFormViewWorkflowData.slice(endIndex);
            }
        }
    }
    var hasPreviousSteps = previousSteps.length > 0 && !isSingleVisibleStep;
    var hasNextSteps = nextSteps.length > 0 || (isSingleVisibleStep && previousSteps.length > 0);
    return (_jsxs(_Fragment, { children: [_jsx(Toast, { ref: toast }), _jsxs("div", { ref: containerRef, className: 'arrow-stepper-container', children: [hasPreviousSteps && (_jsxs("div", { style: { display: 'flex', alignItems: 'center' }, children: [_jsx("button", { type: 'button', className: "overflow-button overlow-left-button", onClick: function (e) {
                                    // @ts-ignore
                                    leftFormStepperOverlay.current.toggle(e);
                                }, children: _jsx("i", { className: "pi pi-ellipsis-h" }) }), _jsx(OverlayPanel, { ref: leftFormStepperOverlay, className: "solid-custom-overlay solid-form-stepper-overlay", children: _jsx("div", { style: { display: 'flex', flexDirection: 'column', gap: '0.25rem', padding: '0.5rem' }, children: previousSteps.map(function (step, index) {
                                        var stepIndex = index;
                                        var isStepActive = stepIndex === activeIndex;
                                        var isStepBeforeActive = stepIndex < activeIndex;
                                        return (_jsx("button", { type: 'button', className: "overlay-step-button ".concat(isStepActive ? 'active' : '', " ").concat(isStepBeforeActive ? 'completed' : ''), onClick: function () {
                                                handleButtonClick(step.value);
                                                // @ts-ignore
                                                leftFormStepperOverlay.current.hide();
                                            }, children: step.label }, stepIndex));
                                    }) }) })] })), visibleSteps.map(function (step) {
                        var stepIndex = solidFormViewWorkflowData.findIndex(function (s) { return s.value === step.value; });
                        var isActive = stepIndex === activeIndex;
                        var isBeforeActive = stepIndex < activeIndex;
                        var isSingleButton = visibleSteps.length === 1;
                        return (_jsx("button", { type: "button", className: "arrow-step-button ".concat(isSingleButton ? 'single-button' : '', " ").concat(isActive
                                ? 'arrow-step-active'
                                : isBeforeActive
                                    ? 'arrow-step-completed'
                                    : 'arrow-step-future'), onClick: function () { return handleButtonClick(step.value); }, title: step.label, children: _jsx("span", { className: "step-text", children: step.label }) }, step.value));
                    }), hasNextSteps && (_jsxs("div", { style: { display: 'flex', alignItems: 'center' }, children: [_jsx("button", { type: 'button', className: "overflow-button overflow-right-button", onClick: function (e) {
                                    // @ts-ignore
                                    formStepperOverlay.current.toggle(e);
                                }, children: _jsx("i", { className: "pi pi-ellipsis-h" }) }), _jsx(OverlayPanel, { ref: formStepperOverlay, className: "solid-custom-overlay solid-form-stepper-overlay", children: _jsx("div", { style: { display: 'flex', flexDirection: 'column', gap: '0.25rem', padding: '0.5rem' }, children: (isSingleVisibleStep
                                        ? __spreadArray(__spreadArray([], previousSteps, true), nextSteps, true) : nextSteps).map(function (step) {
                                        var stepIndex = solidFormViewWorkflowData.findIndex(function (s) { return s.value === step.value; });
                                        var isStepActive = stepIndex === activeIndex;
                                        var isStepBeforeActive = stepIndex < activeIndex;
                                        return (_jsx("button", { type: 'button', className: "overlay-step-button ".concat(isStepActive ? 'active' : '', " ").concat(isStepBeforeActive ? 'completed' : ''), onClick: function () {
                                                handleButtonClick(step.value);
                                                // @ts-ignore
                                                formStepperOverlay.current.hide();
                                            }, children: step.label }, stepIndex));
                                    }) }) })] }))] })] }));
};
