// @solidxai/core-ui/dist/components/core/form/fields/SolidShortTextField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { InputText } from "primereact/inputtext";
import { Message } from "primereact/message";
import * as Yup from 'yup';
import { getExtensionComponent } from "../../../../helpers/registry";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { Password } from "primereact/password";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
var SolidShortTextField = /** @class */ (function () {
    function SolidShortTextField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidShortTextField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var _a, _b, _c;
            var fieldLayoutInfo = this.fieldContext.field;
            var editWidget = (_a = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.editWidget;
            if (editWidget === "PseudoRelationManyToOneFormWidget") {
                if (value === null || value === void 0 ? void 0 : value.solidManyToOneValue) {
                    formData.append((_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.name, value.solidManyToOneValue);
                }
            }
            else {
                if (value !== undefined && value !== null) {
                    formData.append((_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.name, value);
                }
            }
        }
    });
    Object.defineProperty(SolidShortTextField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b, _c;
            var fieldName = this.fieldContext.field.attrs.name;
            var fieldDefaultValue = (_b = (_a = this.fieldContext) === null || _a === void 0 ? void 0 : _a.fieldMetadata) === null || _b === void 0 ? void 0 : _b.defaultValue;
            var fieldLayoutInfo = this.fieldContext.field;
            var editWidget = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.editWidget;
            if (this.fieldContext.parentData && this.fieldContext.parentData[fieldName]) {
                var parentDataForKey = this.fieldContext.parentData[fieldName];
                if (editWidget === "PseudoRelationManyToOneFormWidget") {
                    if (parentDataForKey) {
                        return { solidManyToOneLabel: this.fieldContext.parentData[fieldName], solidManyToOneValue: this.fieldContext.parentData[fieldName] };
                    }
                }
                else {
                    if (parentDataForKey && typeof parentDataForKey !== 'object') {
                        return this.fieldContext.parentData[fieldName];
                    }
                }
            }
            var existingValue = this.fieldContext.data[fieldName];
            if (editWidget === "PseudoRelationManyToOneFormWidget" && this.fieldContext.data[fieldName]) {
                existingValue = { solidManyToOneLabel: this.fieldContext.data[fieldName], solidManyToOneValue: this.fieldContext.data[fieldName] };
            }
            return existingValue !== undefined && existingValue !== null ? existingValue : fieldDefaultValue || '';
        }
    });
    Object.defineProperty(SolidShortTextField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b, _c, _d, _e;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_b = (_a = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
            var editWidget = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.editWidget;
            if (editWidget === "PseudoRelationManyToOneFormWidget") {
                var schema = Yup.mixed();
                // Custom validation for relation field
                if (fieldMetadata.required) {
                    schema = schema.test(ERROR_MESSAGES.REQUIRED_REALTION, ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel), function (value) {
                        // Handle empty values
                        if (!value)
                            return false;
                        // If it's an object with solidManyToOneValue, check if it's valid
                        if (typeof value === 'object' && value !== null && value.solidManyToOneValue) {
                            return true;
                        }
                        // If it's a string (user typed but didn't select), it's invalid for required field
                        if (typeof value === 'string') {
                            return false;
                        }
                        return false;
                    });
                }
                // Add validation to ensure valid selection
                schema = schema.test(ERROR_MESSAGES.VALIDATE_SELECTION, ERROR_MESSAGES.SELECT_VALID_FROM_DROPDOWN(fieldLabel), function (value) {
                    // If not required and empty, it's valid
                    if (!fieldMetadata.required && (!value || value === '')) {
                        return true;
                    }
                    // If it's an object with solidManyToOneValue, it's a valid selection
                    if (typeof value === 'object' && value !== null && value.solidManyToOneValue) {
                        return true;
                    }
                    // If it's a string (user typed but didn't select), it's invalid
                    if (typeof value === 'string' && value.trim() !== '') {
                        return false;
                    }
                    // Empty value for non-required field
                    return !fieldMetadata.required;
                });
                return schema;
            }
            else {
                var schema = Yup.string();
                var fieldMetadata_1 = this.fieldContext.fieldMetadata;
                var fieldLayoutInfo_1 = this.fieldContext.field;
                var fieldLabel_1 = (_d = fieldLayoutInfo_1.attrs.label) !== null && _d !== void 0 ? _d : fieldMetadata_1.displayName;
                // 1. required  
                // 1. required
                if (fieldMetadata_1.required) {
                    schema = schema.required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel_1));
                }
                else {
                    schema = schema.nullable(); // Allow null when not required
                }
                // 2. length (min/max)
                if (fieldMetadata_1.min && fieldMetadata_1.min > 0) {
                    schema = schema.min(fieldMetadata_1.min, ERROR_MESSAGES.FIELD_MINIMUM_CHARACTER(fieldLabel_1, fieldMetadata_1.min));
                }
                if (fieldMetadata_1.max && fieldMetadata_1.max > 0) {
                    schema = schema.max(fieldMetadata_1.max, ERROR_MESSAGES.FIELD_MAXIMUM_CHARACTER(fieldLabel_1, fieldMetadata_1.max));
                }
                // 3. regular expression
                if (fieldMetadata_1.regexPattern) {
                    var regexPatternNotMatchingErrorMsg = (_e = fieldMetadata_1.regexPatternNotMatchingErrorMsg) !== null && _e !== void 0 ? _e : ERROR_MESSAGES.FIELD_INVALID_DATA(fieldLabel_1);
                    schema = schema.matches(fieldMetadata_1.regexPattern, regexPatternNotMatchingErrorMsg);
                }
                return schema;
            }
        }
    });
    Object.defineProperty(SolidShortTextField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var solidFormViewMetaData = this.fieldContext.solidFormViewMetaData;
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultShortTextFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultShortTextFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" &&
                            _jsx(_Fragment, { children: editWidget &&
                                    this.renderExtensionRenderMode(editWidget, formik) })] }) }));
        }
    });
    Object.defineProperty(SolidShortTextField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidShortTextField;
}());
export { SolidShortTextField };
export var DefaultShortTextFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var includeWrapper = ((_c = fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.includeWrapper) || 'yes';
    var fieldLabel = (_d = fieldLayoutInfo.attrs.label) !== null && _d !== void 0 ? _d : fieldMetadata.displayName;
    var fieldDescription = (_e = fieldLayoutInfo.attrs.description) !== null && _e !== void 0 ? _e : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var isPrimaryKey = fieldMetadata.isPrimaryKey || false;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var fieldDisabled = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.disabled;
    var fieldReadonly = (_h = fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.readonly;
    var formDisabled = (_l = (_k = (_j = solidFormViewMetaData.data.solidView) === null || _j === void 0 ? void 0 : _j.layout) === null || _k === void 0 ? void 0 : _k.attrs) === null || _l === void 0 ? void 0 : _l.disabled;
    var formReadonly = (_p = (_o = (_m = solidFormViewMetaData.data.solidView) === null || _m === void 0 ? void 0 : _m.layout) === null || _o === void 0 ? void 0 : _o.attrs) === null || _p === void 0 ? void 0 : _p.readonly;
    return (_jsxs(_Fragment, { children: [includeWrapper === 'yes' &&
                _jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                                    _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx(InputText, { readOnly: formReadonly || fieldReadonly || readOnlyPermission || isPrimaryKey, disabled: formDisabled || fieldDisabled || isPrimaryKey, id: fieldLayoutInfo.attrs.name, name: fieldMetadata.name, "aria-describedby": "".concat(fieldLayoutInfo.attrs.name, "-help"), 
                                    // onChange={formik.handleChange}
                                    onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, onBlur: function (e) { return fieldContext.onBlur(e, 'onFieldBlur'); }, value: formik.values[fieldLayoutInfo.attrs.name] || '' })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_q = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _q === void 0 ? void 0 : _q.toString() }) }))] }), includeWrapper === 'no' &&
                _jsxs(_Fragment, { children: [showFieldLabel != false &&
                            _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx(InputText, { readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled, id: fieldLayoutInfo.attrs.name, name: fieldMetadata.name, "aria-describedby": "".concat(fieldLayoutInfo.attrs.name, "-help"), 
                            // onChange={formik.handleChange}
                            onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, onBlur: function (e) { return fieldContext.onBlur(e, 'onFieldBlur'); }, value: formik.values[fieldLayoutInfo.attrs.name] || '' })] })] }));
};
export var DefaultShortTextFormViewWidget = function (_a) {
    var _b, _c;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    return (_jsxs("div", { className: "mt-2 flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: formik.values[fieldLayoutInfo.attrs.name] && typeof formik.values[fieldLayoutInfo.attrs.name] !== "object" && typeof formik.values[fieldLayoutInfo.attrs.name] == "string" && formik.values[fieldLayoutInfo.attrs.name] })] }));
};
// Masked ShortText - LIST
export var MaskedShortTextListViewWidget = function (_a) {
    var rowData = _a.rowData, solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column;
    var colVal = rowData[column.attrs.name];
    // Mask it (show same number of * as characters, or fixed length if you want)
    var maskedValue = colVal ? '*'.repeat(colVal.length) : '';
    return (_jsx("span", { children: maskedValue }));
};
// Masked ShortText - VIEW
export var MaskedShortTextFormViewWidget = function (_a) {
    var _b, _c;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    // Get actual value
    var rawValue = formik.values[fieldLayoutInfo.attrs.name] || '';
    // Mask it (show same number of * as characters, or fixed length if you want)
    var maskedValue = rawValue ? '*'.repeat(rawValue.length) : '';
    return (_jsxs("div", { className: "mt-2 flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: maskedValue })] }));
};
// Masked ShortText - EDIT
export var MaskedShortTextFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var includeWrapper = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.includeWrapper) || 'yes';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var isPrimaryKey = fieldMetadata.isPrimaryKey || false;
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var fieldDisabled = (_e = fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.disabled;
    var fieldReadonly = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.readonly;
    var formDisabled = (_j = (_h = (_g = solidFormViewMetaData.data.solidView) === null || _g === void 0 ? void 0 : _g.layout) === null || _h === void 0 ? void 0 : _h.attrs) === null || _j === void 0 ? void 0 : _j.disabled;
    var formReadonly = (_m = (_l = (_k = solidFormViewMetaData.data.solidView) === null || _k === void 0 ? void 0 : _k.layout) === null || _l === void 0 ? void 0 : _l.attrs) === null || _m === void 0 ? void 0 : _m.readonly;
    var passwordField = (_jsx(Password, { toggleMask: true, feedback: false, readOnly: formReadonly || fieldReadonly || readOnlyPermission || isPrimaryKey, disabled: formDisabled || fieldDisabled || isPrimaryKey, id: fieldLayoutInfo.attrs.name, name: fieldMetadata.name, "aria-describedby": "".concat(fieldLayoutInfo.attrs.name, "-help"), onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, onBlur: function (e) { return fieldContext.onBlur(e, 'onFieldBlur'); }, value: formik.values[fieldLayoutInfo.attrs.name] || '', className: "w-full" }));
    return (_jsxs(_Fragment, { children: [includeWrapper === 'yes' && (_jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel !== false && (_jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] })), passwordField] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_o = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _o === void 0 ? void 0 : _o.toString() }) }))] })), includeWrapper === 'no' && (_jsxs(_Fragment, { children: [showFieldLabel !== false && (_jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] })), passwordField] }))] }));
};
