// @solidxai/core-ui/dist/components/core/form/fields/SolidIntegerField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { createElement as _createElement } from "react";
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { InputNumber } from "primereact/inputnumber";
import { Message } from "primereact/message";
import * as Yup from 'yup';
import { getExtensionComponent } from "../../../../helpers/registry";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { Range } from "react-range";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
var SolidIntegerField = /** @class */ (function () {
    function SolidIntegerField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidIntegerField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            if (value !== undefined && value !== null) {
                formData.append(fieldLayoutInfo.attrs.name, value);
            }
        }
    });
    Object.defineProperty(SolidIntegerField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b;
            var fieldName = this.fieldContext.field.attrs.name;
            var fieldDefaultValue = (_b = (_a = this.fieldContext) === null || _a === void 0 ? void 0 : _a.fieldMetadata) === null || _b === void 0 ? void 0 : _b.defaultValue;
            if (this.fieldContext.parentData && this.fieldContext.parentData[fieldName]) {
                var parentDataForKey = this.fieldContext.parentData[fieldName];
                if (parentDataForKey && typeof parentDataForKey !== 'object') {
                    return this.fieldContext.parentData[fieldName];
                }
            }
            var existingValue = this.fieldContext.data[fieldName];
            return existingValue !== undefined && existingValue !== null ? existingValue : fieldDefaultValue || '';
        }
    });
    Object.defineProperty(SolidIntegerField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            var schema = Yup.number();
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // 1. required 
            if (fieldMetadata.required) {
                schema = schema.required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
            }
            else {
                schema = schema.nullable(); // Allow null when not required
            }
            // 2. length (min/max)
            if (fieldMetadata.min && fieldMetadata.min > 0) {
                schema = schema.min(fieldMetadata.min, ERROR_MESSAGES.FIELD_MINIMUM_CHARACTER(fieldLabel, fieldMetadata.min));
            }
            if (fieldMetadata.max && fieldMetadata.max > 0) {
                schema = schema.max(fieldMetadata.max, ERROR_MESSAGES.FIELD_MAXIMUM_CHARACTER(fieldLabel, fieldMetadata.max));
            }
            return schema;
        }
    });
    Object.defineProperty(SolidIntegerField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a;
            var fieldLayoutInfo = this.fieldContext.field;
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultIntegerFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultIntegerFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" && (_jsx(_Fragment, { children: editWidget &&
                                this.renderExtensionRenderMode(editWidget, formik) }))] }) }));
        }
    });
    Object.defineProperty(SolidIntegerField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidIntegerField;
}());
export { SolidIntegerField };
export var DefaultIntegerFormViewWidget = function (_a) {
    var _b, _c;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    return (_jsxs("div", { className: "mt-2 flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: formik.values[fieldLayoutInfo.attrs.name] && typeof formik.values[fieldLayoutInfo.attrs.name] !== "object" && formik.values[fieldLayoutInfo.attrs.name] })] }));
};
export var DefaultIntegerFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var fieldDisabled = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.disabled;
    var fieldReadonly = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.readonly;
    var formDisabled = (_k = (_j = (_h = solidFormViewMetaData.data.solidView) === null || _h === void 0 ? void 0 : _h.layout) === null || _j === void 0 ? void 0 : _j.attrs) === null || _k === void 0 ? void 0 : _k.disabled;
    var formReadonly = (_o = (_m = (_l = solidFormViewMetaData.data.solidView) === null || _l === void 0 ? void 0 : _l.layout) === null || _m === void 0 ? void 0 : _m.attrs) === null || _o === void 0 ? void 0 : _o.readonly;
    return (_jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx(InputNumber, { readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled, id: fieldLayoutInfo.attrs.name, "aria-describedby": "".concat(fieldLayoutInfo.attrs.name, "-help"), onChange: function (e) {
                            formik.setFieldValue(fieldLayoutInfo.attrs.name, e.value);
                        }, value: formik.values[fieldLayoutInfo.attrs.name] || '' })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_p = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _p === void 0 ? void 0 : _p.toString() }) }))] }));
};
export var SolidIntegerSliderStyleFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var fieldDisabled = (_d = fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.disabled;
    var fieldReadonly = (_e = fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.readonly;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var formDisabled = (_h = (_g = (_f = solidFormViewMetaData.data.solidView) === null || _f === void 0 ? void 0 : _f.layout) === null || _g === void 0 ? void 0 : _g.attrs) === null || _h === void 0 ? void 0 : _h.disabled;
    var formReadonly = (_l = (_k = (_j = solidFormViewMetaData.data.solidView) === null || _j === void 0 ? void 0 : _j.layout) === null || _k === void 0 ? void 0 : _k.attrs) === null || _l === void 0 ? void 0 : _l.readonly;
    var min = fieldMetadata.min || 0;
    var max = fieldMetadata.max || 5;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var fieldName = fieldLayoutInfo.attrs.name;
    var currentValue = Number((_m = formik.values[fieldName]) !== null && _m !== void 0 ? _m : min);
    return (_jsxs("div", { className: "w-full", style: { height: '60px' }, children: [showFieldLabel != false && (_jsxs("div", { className: "font-medium mb-2", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" })] })), _jsxs("div", { className: "relative h-12", children: [_jsx(Range, { step: 1, min: min, max: max, values: [currentValue], onChange: function (values) {
                            formik.setFieldValue(fieldName, values[0]);
                        }, renderTrack: function (_a) {
                            var props = _a.props, children = _a.children;
                            var percent = ((currentValue - min) / (max - min)) * 100;
                            return (_jsxs("div", __assign({}, props, { style: __assign(__assign({}, props.style), { height: "10px", width: "100%", borderRadius: "8px", backgroundColor: "var(--primary-light-color)", position: "relative" }), children: [_jsx("div", { style: {
                                            position: "absolute",
                                            height: "100%",
                                            width: "".concat(percent, "%"),
                                            backgroundColor: "var(--primary-color)",
                                            borderRadius: "5px",
                                            top: 0,
                                            left: 0,
                                        } }), children] })));
                        }, renderThumb: function (_a) {
                            var props = _a.props;
                            return (_createElement("div", __assign({}, props, { key: props.key, style: __assign(__assign({}, props.style), { height: "18px", width: "18px", border: "4px solid var(--surface-0)", borderRadius: '50%', backgroundColor: "var(--primary-color)" }) })));
                        } }), _jsx("div", { className: "flex align-item-center justify-content-between mt-2", children: Array.from({ length: max - min + 1 }, function (_, i) {
                            var num = i + min;
                            return (_jsx("span", { className: "text-sm", children: num === 0 ? '' : num }, num));
                        }) }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-2", children: _jsx(Message, { severity: "error", text: (_o = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _o === void 0 ? void 0 : _o.toString() }) }))] })] }));
};
