// @solidxai/core-ui/dist/components/core/form/fields/SolidTimeField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Calendar } from "primereact/calendar";
import { Message } from "primereact/message";
import { useRef } from "react";
import * as Yup from 'yup';
import { getExtensionComponent } from "../../../../helpers/registry";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
import { DateFieldViewComponent } from '../../../../components/core/common/DateFieldViewComponent';
// Converts multiple time formats into a JavaScript Date object
function parseTimeStringToDate(timeStr) {
    if (!timeStr)
        return null;
    // CASE 1: HH:mm:ss
    if (typeof timeStr === "string" && /^\d{2}:\d{2}:\d{2}$/.test(timeStr)) {
        var _a = timeStr.split(":").map(Number), h = _a[0], m = _a[1], s = _a[2];
        var d = new Date();
        d.setHours(h, m, s, 0);
        return d;
    }
    // CASE 2: timestamp or timestamptz
    try {
        var d = new Date(timeStr); // ← AUTO UTC → LOCAL conversion
        if (!isNaN(d.getTime()))
            return d;
    }
    catch (e) {
        console.error("".concat(ERROR_MESSAGES.FIELD_INAVLID_FORMAT('Date'), ":,").concat(e));
    }
    return null;
}
// Formats a Date object to a HH:mm:ss string for display
function formatTime(date) {
    if (!date)
        return "";
    return date.toLocaleTimeString();
}
var SolidTimeField = /** @class */ (function () {
    function SolidTimeField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidTimeField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            if (value instanceof Date) {
                // Use local date with selected time
                var now = new Date();
                var localDate = new Date(now.getFullYear(), now.getMonth(), now.getDate(), value.getHours(), value.getMinutes(), value.getSeconds());
                formData.append(fieldLayoutInfo.attrs.name, localDate.toISOString());
            }
            else if (value) {
                formData.append(fieldLayoutInfo.attrs.name, value);
            }
        }
    });
    Object.defineProperty(SolidTimeField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b;
            var fieldName = this.fieldContext.field.attrs.name;
            var fieldDefaultValue = (_b = (_a = this.fieldContext) === null || _a === void 0 ? void 0 : _a.fieldMetadata) === null || _b === void 0 ? void 0 : _b.defaultValue;
            var existingValue = this.fieldContext.data[fieldName];
            if (existingValue) {
                if (existingValue instanceof Date) {
                    return existingValue;
                }
                if (typeof existingValue === "string") {
                    var parsed = parseTimeStringToDate(existingValue);
                    if (parsed)
                        return parsed;
                }
            }
            return fieldDefaultValue ? new Date(fieldDefaultValue) : null;
        }
    });
    Object.defineProperty(SolidTimeField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            var schema = Yup.date().nullable();
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // 1. required 
            if (fieldMetadata.required) {
                schema = schema.required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
            }
            return schema;
        }
    });
    Object.defineProperty(SolidTimeField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultTimeFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultTimeFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" &&
                            _jsx(_Fragment, { children: editWidget &&
                                    this.renderExtensionRenderMode(editWidget, formik) })] }) }));
        }
    });
    Object.defineProperty(SolidTimeField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidTimeField;
}());
export { SolidTimeField };
export var DefaultTimeFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var calendarRef = useRef(null); // Reference for the Calendar component
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var fieldDisabled = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.disabled;
    var formDisabled = (_j = (_h = (_g = solidFormViewMetaData.data.solidView) === null || _g === void 0 ? void 0 : _g.layout) === null || _h === void 0 ? void 0 : _h.attrs) === null || _j === void 0 ? void 0 : _j.disabled;
    var fieldValue = formik.values[fieldLayoutInfo.attrs.name];
    return (_jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx(Calendar, { disabled: formDisabled || fieldDisabled || readOnlyPermission, ref: calendarRef, id: fieldLayoutInfo.attrs.name, "aria-describedby": "".concat(fieldLayoutInfo.attrs.name, "-help"), 
                        // onChange={formik.handleChange}
                        onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, 
                        //@ts-ignore
                        // value={formik.values[fieldLayoutInfo.attrs.name] ? formik.values[fieldLayoutInfo.attrs.name] : Date()}
                        value: fieldValue instanceof Date ? fieldValue : typeof fieldValue === "string" ? parseTimeStringToDate(fieldValue) : null, 
                        // dateFormat="mm/dd/yy"
                        // placeholder="mm/dd/yyyy hh:mm"
                        hideOnDateTimeSelect: true, timeOnly: true, showTime: true, className: "", hourFormat: "24" })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_k = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _k === void 0 ? void 0 : _k.toString() }) }))] }));
};
export var DefaultTimeFormViewWidget = function (_a) {
    var _b, _c, _d;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldName = fieldLayoutInfo.attrs.name;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    var rawValue = formik.values[fieldName];
    var format = (_d = fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.format;
    return (_jsxs("div", { className: "mt-2 flex flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: _jsx(DateFieldViewComponent, { value: rawValue, format: format, fallback: "-" }) })] }));
};
