// @solidxai/core-ui/dist/components/core/form/fields/SolidDateField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Calendar } from "primereact/calendar";
import { Message } from "primereact/message";
import { useEffect, useRef, useState } from "react";
import * as Yup from 'yup';
import { getExtensionComponent } from "../../../../helpers/registry";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
import { StatusIcon } from '../../../../components/core/extension/solid-core/CustomIcon/StatusIcon';
import { DateFieldViewComponent } from '../../../../components/core/common/DateFieldViewComponent';
var SolidDateField = /** @class */ (function () {
    function SolidDateField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidDateField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            if (value instanceof Date && !isNaN(value.getTime())) {
                var localDate = new Date(value);
                var utcDate = new Date(Date.UTC(localDate.getFullYear(), localDate.getMonth(), localDate.getDate()));
                // "2025-06-10T00:00:00.000Z"
                var isoString = utcDate.toISOString();
                formData.append(fieldLayoutInfo.attrs.name, isoString.split('T')[0]);
            }
            else if (value) {
                formData.append(fieldLayoutInfo.attrs.name, value);
            }
        }
    });
    Object.defineProperty(SolidDateField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b;
            var fieldName = this.fieldContext.field.attrs.name;
            var fieldDefaultValue = (_b = (_a = this.fieldContext) === null || _a === void 0 ? void 0 : _a.fieldMetadata) === null || _b === void 0 ? void 0 : _b.defaultValue;
            var existingValue = this.fieldContext.data[fieldName];
            return existingValue !== undefined && existingValue !== null ? existingValue : fieldDefaultValue || '';
        }
    });
    Object.defineProperty(SolidDateField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            var schema = Yup.date();
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // 1. required 
            if (fieldMetadata.required) {
                schema = schema.required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
            }
            else {
                schema = schema.nullable();
            }
            return schema;
        }
    });
    Object.defineProperty(SolidDateField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a, _b;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultDateFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultDateFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" && (_jsx(_Fragment, { children: editWidget &&
                                this.renderExtensionRenderMode(editWidget, formik) }))] }) }));
        }
    });
    Object.defineProperty(SolidDateField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidDateField;
}());
export { SolidDateField };
export var DefaultDateFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var _l = useState(false), overlayVisible = _l[0], setOverlayVisible = _l[1];
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var calendarRef = useRef(null); // Reference for the Calendar component
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var fieldDisabled = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.disabled;
    // const fieldReadonly = fieldLayoutInfo.attrs?.readonly;
    var formDisabled = (_j = (_h = (_g = solidFormViewMetaData.data.solidView) === null || _g === void 0 ? void 0 : _g.layout) === null || _h === void 0 ? void 0 : _h.attrs) === null || _j === void 0 ? void 0 : _j.disabled;
    useEffect(function () {
        var handleDocumentClick = function (event) {
            var _a, _b;
            var inputElement = (_a = calendarRef.current) === null || _a === void 0 ? void 0 : _a.getInput();
            var overlayElement = (_b = calendarRef.current) === null || _b === void 0 ? void 0 : _b.getOverlay();
            if (overlayVisible && inputElement && !inputElement.contains(event.target) && overlayElement && !overlayElement.contains(event.target)) {
                setOverlayVisible(false);
            }
        };
        document.addEventListener('mousedown', handleDocumentClick);
        return function () {
            document.removeEventListener('mousedown', handleDocumentClick);
        };
    }, [overlayVisible]);
    var handleInputClick = function () {
        setOverlayVisible(true);
    };
    return (_jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx("div", { onClick: handleInputClick, id: fieldLayoutInfo.attrs.name, children: _jsx(Calendar, { disabled: formDisabled || fieldDisabled || readOnlyPermission, ref: calendarRef, "aria-describedby": "".concat(fieldLayoutInfo.attrs.name, "-help"), name: fieldMetadata.name, 
                            // onChange={formik.handleChange}
                            onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, 
                            //@ts-ignore
                            value: formik.values[fieldLayoutInfo.attrs.name] ? new Date(formik.values[fieldLayoutInfo.attrs.name]) : Date(), 
                            // dateFormat="mm/dd/yy"
                            // placeholder="mm/dd/yyyy hh:mm"
                            mask: "99/99/9999", hideOnDateTimeSelect: true, appendTo: "self", className: "", onFocus: function () { return setOverlayVisible(true); }, visible: overlayVisible, onVisibleChange: function (e) {
                                console.log("Overlay visibility changed:", e.visible);
                                setOverlayVisible(e.visible);
                            }, onBlur: function (e) {
                                var _a, _b;
                                if ((_b = (_a = calendarRef.current) === null || _a === void 0 ? void 0 : _a.getOverlay()) === null || _b === void 0 ? void 0 : _b.contains(e.relatedTarget)) {
                                    return;
                                }
                                setOverlayVisible(false);
                            }, onKeyDown: function (e) {
                                if (e.key === "Tab") {
                                    setOverlayVisible(false);
                                }
                            } }) })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_k = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _k === void 0 ? void 0 : _k.toString() }) }))] }));
};
export var DefaultDateFormViewWidget = function (_a) {
    var _b, _c, _d;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldName = fieldLayoutInfo.attrs.name;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    var rawValue = formik.values[fieldName];
    var format = (_d = fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.format;
    return (_jsxs("div", { className: "mt-2 flex flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: _jsx(DateFieldViewComponent, { value: rawValue, format: format, fallback: "-" }) })] }));
};
export var PublishedStatusListViewWidget = function (_a) {
    var rowData = _a.rowData, solidListViewMetaData = _a.solidListViewMetaData, fieldMetadata = _a.fieldMetadata, column = _a.column;
    var colVal = rowData[column.attrs.name]; // publishedAt value
    var isPublished = !!colVal;
    return (_jsx(StatusIcon, { isPublished: isPublished }));
};
