// @solidxai/core-ui/dist/components/core/form/fields/SolidSelectionStaticField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { AutoComplete } from "primereact/autocomplete";
import { Message } from "primereact/message";
import { useState } from "react";
import * as Yup from 'yup';
import { getExtensionComponent } from "../../../../helpers/registry";
import { RadioButton } from "primereact/radiobutton";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { SelectButton } from "primereact/selectbutton";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
var SolidSelectionStaticField = /** @class */ (function () {
    function SolidSelectionStaticField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidSelectionStaticField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
            if (isMultiSelect && Array.isArray(value)) {
                formData.append(fieldLayoutInfo.attrs.name, JSON.stringify(value.map(function (v) { return v.value; })));
            }
            else if (value) {
                formData.append(fieldLayoutInfo.attrs.name, value.value);
            }
            // if (value) {
            //     formData.append(fieldLayoutInfo.attrs.name, value.value);
            // }
        }
    });
    Object.defineProperty(SolidSelectionStaticField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            // Get field name and metadata
            var fieldName = this.fieldContext.field.attrs.name;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldDefaultValue = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.defaultValue;
            var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
            // Get existing value from form data
            var existingValue = this.fieldContext.data[fieldName];
            // Function to get display value based on selectionStaticValues
            // const getDisplayValue = (value: string | null): string | null => {
            //     if (!value) return null;
            //     for (const item of fieldMetadata.selectionStaticValues) {
            //         const [lhs, rhs] = item.split(':');
            //         if (lhs === value) {
            //             return rhs;
            //         }
            //     }
            //     return null;
            // };
            // Function to get display value based on selectionStaticValues
            var getDisplayValue = function (value) {
                var match = fieldMetadata.selectionStaticValues.find(function (item) { return item.startsWith(value + ':'); });
                return match ? match.split(':')[1] : value;
            };
            // Determine the final value to use (existing value or default value)
            var finalValue = (_a = existingValue !== null && existingValue !== void 0 ? existingValue : fieldDefaultValue) !== null && _a !== void 0 ? _a : '';
            if (isMultiSelect) {
                var values = [];
                if (Array.isArray(finalValue)) {
                    values = finalValue;
                }
                else {
                    try {
                        var parsed = JSON.parse(finalValue);
                        if (Array.isArray(parsed))
                            values = parsed;
                    }
                    catch (_b) { }
                }
                return values.map(function (val) { return ({
                    label: getDisplayValue(val),
                    value: val
                }); });
            }
            // Get display value for the final value
            // const displayValue = getDisplayValue(finalValue);
            // return { label: displayValue ?? '', value: finalValue };
            return { label: getDisplayValue(finalValue), value: finalValue };
        }
    });
    Object.defineProperty(SolidSelectionStaticField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
            // let schema = Yup.object({
            //     value: Yup.string().required(`${fieldLabel} is required.`)
            // });
            var schema;
            if (isMultiSelect) {
                // For multi-select, create array schema
                if (fieldMetadata.required) {
                    schema = Yup.array()
                        .of(Yup.object({
                        value: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel))
                    }))
                        .min(1, ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel))
                        .required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
                }
                else {
                    schema = Yup.array()
                        .of(Yup.object({
                        value: Yup.string()
                    }))
                        .nullable(); // Allow null/undefined for non-required fields
                }
            }
            else {
                // For single select, create object schema
                if (fieldMetadata.required) {
                    schema = Yup.object({
                        value: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel))
                    }).required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
                }
                else {
                    schema = Yup.object({
                        value: Yup.string()
                    }).nullable(); // Allow null/undefined for non-required fields
                }
            }
            return schema;
        }
    });
    Object.defineProperty(SolidSelectionStaticField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var isFormFieldValid = function (formik, fieldName) { return formik.errors[fieldName]; };
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultSelectionStaticAutocompleteFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultSelectionStaticFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" &&
                            _jsx(_Fragment, { children: editWidget &&
                                    this.renderExtensionRenderMode(editWidget, formik) })] }) }));
        }
    });
    Object.defineProperty(SolidSelectionStaticField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidSelectionStaticField;
}());
export { SolidSelectionStaticField };
export var DefaultSelectionStaticAutocompleteFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var fieldDisabled = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.disabled;
    var fieldReadonly = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.readonly;
    var formDisabled = (_k = (_j = (_h = solidFormViewMetaData.data.solidView) === null || _h === void 0 ? void 0 : _h.layout) === null || _j === void 0 ? void 0 : _j.attrs) === null || _k === void 0 ? void 0 : _k.disabled;
    var formReadonly = (_o = (_m = (_l = solidFormViewMetaData.data.solidView) === null || _l === void 0 ? void 0 : _l.layout) === null || _m === void 0 ? void 0 : _m.attrs) === null || _o === void 0 ? void 0 : _o.readonly;
    var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
    var _v = useState([]), selectionStaticItems = _v[0], setSelectionStaticItems = _v[1];
    var selectionStaticSearch = function (event) {
        var selectionStaticData = fieldMetadata.selectionStaticValues.map(function (i) {
            return {
                label: i.split(":")[1],
                value: i.split(":")[0]
            };
        });
        var suggestionData = selectionStaticData.filter(function (t) { return t.value.toLowerCase().startsWith(event.query.toLowerCase()); });
        setSelectionStaticItems(suggestionData);
    };
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    return (_jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label font-medium", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx(AutoComplete, __assign({ multiple: isMultiSelect, readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled }, formik.getFieldProps(fieldLayoutInfo.attrs.name), { id: fieldLayoutInfo.attrs.name, name: fieldLayoutInfo.attrs.name, field: "label", 
                        // value={formik.values[fieldLayoutInfo.attrs.name] || null}
                        value: formik.values[fieldLayoutInfo.attrs.name] || (isMultiSelect ? [] : null), dropdown: true, suggestions: selectionStaticItems, completeMethod: selectionStaticSearch, 
                        // onChange={(e) => updateInputs(index, e.value)} />
                        // onChange={formik.handleChange}
                        onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, className: "solid-standard-autocomplete" }))] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: 
                    // formik?.errors[fieldLayoutInfo.attrs.name]?.toString()
                    typeof formik.errors[(_p = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _p === void 0 ? void 0 : _p.name] === 'object'
                        ? (_s = (_r = formik.errors[(_q = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _q === void 0 ? void 0 : _q.name]) === null || _r === void 0 ? void 0 : _r.value) === null || _s === void 0 ? void 0 : _s.toString()
                        : (_u = formik.errors[(_t = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _t === void 0 ? void 0 : _t.name]) === null || _u === void 0 ? void 0 : _u.toString() }) }))] }));
};
export var SolidSelectionStaticRadioFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var fieldDisabled = (_e = fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.disabled;
    var fieldReadonly = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.readonly;
    var formDisabled = (_j = (_h = (_g = fieldContext.solidFormViewMetaData.data.solidView) === null || _g === void 0 ? void 0 : _g.layout) === null || _h === void 0 ? void 0 : _h.attrs) === null || _j === void 0 ? void 0 : _j.disabled;
    var formReadonly = (_m = (_l = (_k = fieldContext.solidFormViewMetaData.data.solidView) === null || _k === void 0 ? void 0 : _k.layout) === null || _l === void 0 ? void 0 : _l.attrs) === null || _m === void 0 ? void 0 : _m.readonly;
    var fieldName = fieldLayoutInfo.attrs.name;
    var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
    // Convert selectionStaticValues to usable radio options
    var radioOptions = fieldMetadata.selectionStaticValues.map(function (i) {
        var _a = i.split(":"), value = _a[0], label = _a[1];
        return { label: label, value: value };
    });
    var isFormFieldValid = function (formik, fieldName) {
        return formik.errors[fieldName];
    };
    if (isMultiSelect) {
        return (_jsx("div", { className: className, children: _jsx(Message, { severity: "error", text: "This render mode is not supported for multi select." }) }));
    }
    return (_jsx("div", { className: className, children: _jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel !== false && (_jsxs("label", { htmlFor: fieldName, className: "form-field-label font-medium", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] })), _jsx("div", { className: "flex flex-wrap gap-3", children: radioOptions.map(function (option) {
                                var _a;
                                return (_jsxs("div", { className: "flex items-center", children: [_jsx(RadioButton, { id: "".concat(fieldName, "-").concat(option.value), name: fieldName, value: { label: option.label, value: option.value }, checked: ((_a = formik.values[fieldName]) === null || _a === void 0 ? void 0 : _a.value) === option.value, 
                                            // onChange={(e) => formik.setFieldValue(fieldName, e.value)} 
                                            // onChange={(e) => fieldContext.onChange(e, 'onFieldChange')}
                                            // onChange={(e) =>
                                            //     formik.setFieldValue(fieldName, { label: option.label, value: option.value })
                                            // }
                                            onChange: function (e) {
                                                formik.setFieldValue(fieldName, { label: option.label, value: option.value });
                                                fieldContext.onChange(e, 'onFieldChange');
                                            }, disabled: formReadonly || fieldReadonly || readOnlyPermission || formDisabled || fieldDisabled, className: "mr-2" }, option.value), _jsx("label", { htmlFor: "".concat(fieldName, "-").concat(option.value), className: "cursor-pointer", children: option.label })] }, option.value));
                            }) })] }), isFormFieldValid(formik, fieldName) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_o = formik === null || formik === void 0 ? void 0 : formik.errors[fieldName]) === null || _o === void 0 ? void 0 : _o.toString() }) }))] }) }));
};
export var SolidSelectionStaticSelectButtonFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var fieldDisabled = (_e = fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.disabled;
    var fieldReadonly = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.readonly;
    var formDisabled = (_j = (_h = (_g = fieldContext.solidFormViewMetaData.data.solidView) === null || _g === void 0 ? void 0 : _g.layout) === null || _h === void 0 ? void 0 : _h.attrs) === null || _j === void 0 ? void 0 : _j.disabled;
    var formReadonly = (_m = (_l = (_k = fieldContext.solidFormViewMetaData.data.solidView) === null || _k === void 0 ? void 0 : _k.layout) === null || _l === void 0 ? void 0 : _l.attrs) === null || _m === void 0 ? void 0 : _m.readonly;
    var fieldName = fieldLayoutInfo.attrs.name;
    var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
    var isFormFieldValid = function (formik, fieldName) {
        return formik.errors[fieldName];
    };
    if (isMultiSelect) {
        return (_jsx("div", { className: className, children: _jsx(Message, { severity: "error", text: "This render mode is not supported for multi select." }) }));
    }
    var options = fieldMetadata.selectionStaticValues.map(function (i) {
        var _a = i.split(":"), value = _a[0], label = _a[1];
        return { label: label, value: value };
    });
    var currentValue = formik.values[fieldName];
    // SelectButton needs only primitive "yes" or "no" to show active correctly
    var buttonValue = typeof currentValue === "object" ? currentValue.value : currentValue || null;
    console.log("selection static options", options);
    return (_jsx("div", { className: className, children: _jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel !== false && (_jsxs("label", { htmlFor: fieldName, className: "form-field-label font-medium", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" })] })), _jsx(SelectButton, { id: fieldName, name: fieldName, value: buttonValue, options: options, optionLabel: "label", onChange: function (e) {
                                // Always store object in Formik so validation works
                                var selectedOption = options.find(function (opt) { return opt.value === e.value; });
                                formik.setFieldValue(fieldName, selectedOption);
                            }, disabled: formReadonly ||
                                fieldReadonly ||
                                readOnlyPermission ||
                                formDisabled ||
                                fieldDisabled })] }), isFormFieldValid(formik, fieldName) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_o = formik.errors[fieldName]) === null || _o === void 0 ? void 0 : _o.toString() }) }))] }) }));
};
export var DefaultSelectionStaticFormViewWidget = function (_a) {
    var _b, _c;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var value = formik.values[fieldLayoutInfo.attrs.name];
    var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    return (
    // <div className="mt-2 flex-column gap-2">
    //     <p className="m-0 form-field-label font-medium">{fieldLabel}</p>
    //     <p className="m-0">{value && value.label && value.label}</p>
    // </div>
    _jsxs("div", { className: "mt-2 flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: isMultiSelect
                    ? Array.isArray(value)
                        ? value.map(function (v) { return v === null || v === void 0 ? void 0 : v.label; }).filter(Boolean).join(', ')
                        : ''
                    : (value === null || value === void 0 ? void 0 : value.label) || '' })] }));
};
