// @solidxai/core-ui/dist/components/core/form/fields/SolidBooleanField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { getExtensionComponent } from "../../../../helpers/registry";
import { Message } from "primereact/message";
import { SelectButton } from "primereact/selectbutton";
import { classNames } from "primereact/utils";
import { useEffect, useState } from "react";
import * as Yup from 'yup';
import { Checkbox } from "primereact/checkbox";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { InputSwitch } from "primereact/inputswitch";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
var SolidBooleanField = /** @class */ (function () {
    function SolidBooleanField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidBooleanField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            formData.append(fieldLayoutInfo.attrs.name, value === "true" || value === true ? "true" : "");
        }
    });
    Object.defineProperty(SolidBooleanField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b;
            var fieldName = this.fieldContext.field.attrs.name;
            var fieldDefaultValue = (_b = (_a = this.fieldContext) === null || _a === void 0 ? void 0 : _a.fieldMetadata) === null || _b === void 0 ? void 0 : _b.defaultValue;
            var existingValue = this.fieldContext.data[fieldName];
            // return existingValue !== undefined && existingValue !== null ? existingValue : fieldDefaultValue || '';
            // Ensure the value is always a string "true" or "false"
            // const result = existingValue
            //     ? (existingValue === true || existingValue === "true" ? "true" : "false")
            //     : (fieldDefaultValue === true || fieldDefaultValue === "true" ? "true" : "false");
            // return result;
            if (existingValue !== undefined && existingValue !== null) {
                return existingValue === true || existingValue === "true";
            }
            return fieldDefaultValue === true || fieldDefaultValue === "true";
        }
    });
    Object.defineProperty(SolidBooleanField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a;
            var schema = Yup.boolean();
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // 1. required 
            if (fieldMetadata.required) {
                schema = schema.required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
            }
            else {
                schema = schema.nullable(); // Allow null when not required
            }
            // // 2. length (min/max)
            // if (fieldMetadata.min && fieldMetadata.min > 0) {
            //     schema = schema.min(fieldMetadata.min, `${fieldLabel} should be at-least ${fieldMetadata.min} characters long.`);
            // }
            // if (fieldMetadata.max && fieldMetadata.max > 0) {
            //     schema = schema.max(fieldMetadata.max, `${fieldLabel} should not be more than ${fieldMetadata.max} characters long.`);
            // }
            // 3. regular expression
            //*********Regex doesnt make sense in boolean right? */
            // if (fieldMetadata.regexPattern) {
            //     const regexPatternNotMatchingErrorMsg = fieldMetadata.regexPatternNotMatchingErrorMsg ?? `${fieldLabel} has invalid data.`
            //     schema = schema.matches(fieldMetadata.regexPattern, regexPatternNotMatchingErrorMsg);
            // }
            return schema;
        }
    });
    Object.defineProperty(SolidBooleanField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a, _b;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
            useEffect(function () { formik.setFieldValue(fieldLayoutInfo.attrs.name, "false"); }, []);
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'booleanCheckbox';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultBooleanFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" &&
                            (_jsx(_Fragment, { children: this.renderExtensionRenderMode(editWidget, formik) }))] }) }));
        }
    });
    Object.defineProperty(SolidBooleanField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidBooleanField;
}());
export { SolidBooleanField };
export var DefaultBooleanFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var trueLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.trueLabel;
    var falseLabel = (_f = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.falseLabel;
    var readOnlyPermission = fieldContext.readOnly;
    // const [booleanOptions, setBooleanOptions] = useState<string[]>(["false", "true"]);
    var _r = useState([
        { label: "False", value: "false" },
        { label: "True", value: "true" },
    ]), booleanOptions = _r[0], setBooleanOptions = _r[1];
    // let booleanOptions = ["false", "true"];
    useEffect(function () {
        if (trueLabel || falseLabel) {
            setBooleanOptions([
                {
                    label: falseLabel !== null && falseLabel !== void 0 ? falseLabel : "False",
                    value: "false",
                },
                {
                    label: trueLabel !== null && trueLabel !== void 0 ? trueLabel : "True",
                    value: "true",
                },
            ]);
        }
    }, [trueLabel, falseLabel]);
    // useEffect(() => { formik.setFieldValue(fieldLayoutInfo.attrs.name, "false") }, [])
    useEffect(function () {
        var name = fieldLayoutInfo.attrs.name;
        if (formik.values[name] === undefined || formik.values[name] === null) {
            formik.setFieldValue(name, "false");
        }
    }, []);
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var fieldDisabled = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.disabled;
    var fieldReadonly = (_h = fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.readonly;
    var formDisabled = (_l = (_k = (_j = solidFormViewMetaData.data.solidView) === null || _j === void 0 ? void 0 : _j.layout) === null || _k === void 0 ? void 0 : _k.attrs) === null || _l === void 0 ? void 0 : _l.disabled;
    var formReadonly = (_p = (_o = (_m = solidFormViewMetaData.data.solidView) === null || _m === void 0 ? void 0 : _m.layout) === null || _o === void 0 ? void 0 : _o.attrs) === null || _p === void 0 ? void 0 : _p.readonly;
    return (_jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label font-medium", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx(SelectButton, { readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled, optionLabel: "label", optionValue: "value", id: fieldLayoutInfo.attrs.name, "aria-describedby": "".concat(fieldLayoutInfo.attrs.name, "-help"), onChange: function (e) { formik.setFieldValue(fieldLayoutInfo.attrs.name, e.value); console.log("value is", e.value); }, value: formik.values[fieldLayoutInfo.attrs.name] ? formik.values[fieldLayoutInfo.attrs.name].toString() : "false", options: booleanOptions, className: classNames("", {
                            "p-invalid": isFormFieldValid(formik, "defaultValue"),
                        }) })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_q = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _q === void 0 ? void 0 : _q.toString() }) }))] }));
};
export var SolidBooleanCheckboxStyleFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || "field col-12";
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var checkboxLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.checkboxLabel;
    var readOnlyPermission = fieldContext.readOnly;
    // Set default value to false on mount
    useEffect(function () {
        if (formik.values[fieldLayoutInfo.attrs.name] === undefined) {
            console.log("Setting default value:", false);
            formik.setFieldValue(fieldLayoutInfo.attrs.name, false);
        }
    }, []);
    var handleChange = function (e) {
        var _a;
        var newValue = e.checked; // This returns `true` or `false`
        console.log("".concat(fieldLayoutInfo.attrs.name, ", new value:"), newValue);
        formik.setFieldValue(fieldLayoutInfo.attrs.name, newValue === true ? 'true' : 'false');
        formik.setTouched(__assign(__assign({}, formik.touched), (_a = {}, _a[fieldLayoutInfo.attrs.name] = true, _a))); // Ensure Formik registers the change
        // ✅ Check if Formik updated the value correctly
        setTimeout(function () {
            console.log("Formik values after update:", formik.values);
        }, 0);
    };
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var fieldDisabled = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.disabled;
    var fieldReadonly = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.readonly;
    var formDisabled = (_k = (_j = (_h = solidFormViewMetaData.data.solidView) === null || _h === void 0 ? void 0 : _h.layout) === null || _j === void 0 ? void 0 : _j.attrs) === null || _k === void 0 ? void 0 : _k.disabled;
    var formReadonly = (_o = (_m = (_l = solidFormViewMetaData.data.solidView) === null || _l === void 0 ? void 0 : _l.layout) === null || _m === void 0 ? void 0 : _m.attrs) === null || _o === void 0 ? void 0 : _o.readonly;
    return (_jsx("div", { className: className, children: _jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel !== false && (_jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label font-medium", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] })), _jsxs("div", { className: "flex align-items-center", children: [_jsx(Checkbox, { id: fieldLayoutInfo.attrs.name, checked: formik.values[fieldLayoutInfo.attrs.name] === true, 
                                    // onChange={handleChange}
                                    onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, disabled: formDisabled || fieldDisabled, readOnly: formReadonly || fieldReadonly || readOnlyPermission, className: classNames("", {
                                        "p-invalid": isFormFieldValid(formik, fieldLayoutInfo.attrs.name),
                                    }) }), checkboxLabel &&
                                    _jsx("span", { className: "ml-2", children: checkboxLabel || "Yes" })] })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_p = formik.errors[fieldLayoutInfo.attrs.name]) === null || _p === void 0 ? void 0 : _p.toString() }) }))] }) }));
};
export var SolidBooleanSwitchStyleFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || "field col-12";
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    // Set default value to false on mount
    useEffect(function () {
        if (formik.values[fieldLayoutInfo.attrs.name] === undefined) {
            console.log("Setting default value:", false);
            formik.setFieldValue(fieldLayoutInfo.attrs.name, false);
        }
    }, []);
    var handleChange = function (e) {
        var _a;
        var newValue = e.value;
        console.log("".concat(fieldLayoutInfo.attrs.name, " switch clicked, new value:"), newValue);
        // formik.setFieldValue(fieldLayoutInfo.attrs.name, newValue === true ? true : false);
        // formik.setTouched({ ...formik.touched, [fieldLayoutInfo.attrs.name]: true }); // Ensure Formik registers the change
        // // ✅ Check if Formik updated the value correctly
        // setTimeout(() => {
        //     console.log("Formik values after update:", formik.values);
        // }, 0);
        // 1: BYPASS Formik's validation by directly mutating the values object
        // This is necessary because the validation schema was rejecting boolean values
        // Direct assignment skips all validation logic that was causing the issue
        formik.values[fieldLayoutInfo.attrs.name] = newValue;
        // 2: Force Formik to recognize the change and trigger re-renders
        // setValues() with the second parameter as 'false' means:
        // - false = skip validation (avoid the schema conflict)
        // - This triggers Formik's internal change detection and component re-renders
        formik.setValues(__assign({}, formik.values), false);
        // 3: Mark the field as "touched" to indicate user interaction
        // This is important for validation state and form submission logic
        // The 'false' parameter means don't validate immediately
        formik.setTouched(__assign(__assign({}, formik.touched), (_a = {}, _a[fieldLayoutInfo.attrs.name] = true, _a)), false);
        fieldContext.onChange(e, 'onFieldChange');
    };
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var fieldDisabled = (_e = fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.disabled;
    var fieldReadonly = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.readonly;
    var formDisabled = (_j = (_h = (_g = solidFormViewMetaData.data.solidView) === null || _g === void 0 ? void 0 : _g.layout) === null || _h === void 0 ? void 0 : _h.attrs) === null || _j === void 0 ? void 0 : _j.disabled;
    var formReadonly = (_m = (_l = (_k = solidFormViewMetaData.data.solidView) === null || _k === void 0 ? void 0 : _k.layout) === null || _l === void 0 ? void 0 : _l.attrs) === null || _m === void 0 ? void 0 : _m.readonly;
    return (_jsx("div", { className: className, children: _jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel !== false && (_jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label font-medium", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] })), _jsxs("div", { className: "flex align-items-center", children: [_jsx(InputSwitch, { id: fieldLayoutInfo.attrs.name, name: fieldLayoutInfo.attrs.name, checked: formik.values[fieldLayoutInfo.attrs.name] === true || formik.values[fieldLayoutInfo.attrs.name] === "true", onChange: handleChange, 
                                    // onChange={(e) => {fieldContext.onChange(e, 'onFieldChange')}
                                    disabled: formDisabled || fieldDisabled, readOnly: formReadonly || fieldReadonly || readOnlyPermission, className: classNames("", {
                                        "p-invalid": isFormFieldValid(formik, fieldLayoutInfo.attrs.name),
                                    }) }), _jsx("span", { className: "ml-2", children: fieldLabel || "Yes" })] })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_o = formik.errors[fieldLayoutInfo.attrs.name]) === null || _o === void 0 ? void 0 : _o.toString() }) }))] }) }));
};
export var DefaultBooleanFormViewWidget = function (_a) {
    var _b, _c, _d, _e;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    // const trueLabel = fieldLayoutInfo?.attrs?.trueLabel;
    // const falseLabel = fieldLayoutInfo?.attrs?.falseLabel;
    var _f = useState("true"), trueLabel = _f[0], setTrueLabel = _f[1];
    var _g = useState("false"), falseLabel = _g[0], setFalseLabel = _g[1];
    useEffect(function () {
        var _a, _b, _c, _d;
        if ((_a = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.trueLabel) {
            setTrueLabel((_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.trueLabel);
        }
        if ((_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.falseLabel) {
            setFalseLabel((_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.falseLabel);
        }
    }, [(_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.falseLabel, (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.trueLabel]);
    return (_jsxs("div", { className: "mt-2 flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: formik.values[fieldLayoutInfo.attrs.name] === true || formik.values[fieldLayoutInfo.attrs.name] === "true" ? trueLabel : falseLabel })] }));
};
