// @solidxai/core-ui/dist/components/core/form/fields/SolidPasswordField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Message } from "primereact/message";
import * as Yup from 'yup';
import { Password } from "primereact/password";
import { getExtensionComponent } from "../../../../helpers/registry";
import { useState } from "react";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { useFormik } from "formik";
import { Dialog } from "primereact/dialog";
import { Button } from "primereact/button";
import { usePathname } from "../../../../hooks/usePathname";
import { updatePasswordField } from "../../../../helpers/updatePasswordField";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
var SolidPasswordField = /** @class */ (function () {
    function SolidPasswordField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidPasswordField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var _a, _b, _c;
            var fieldName = (_c = (_b = (_a = this.fieldContext) === null || _a === void 0 ? void 0 : _a.field) === null || _b === void 0 ? void 0 : _b.attrs) === null || _c === void 0 ? void 0 : _c.name;
            if (value !== undefined && value !== null && value !== '') {
                formData.append(fieldName, value);
            }
        }
    });
    Object.defineProperty(SolidPasswordField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b;
            var fieldName = this.fieldContext.field.attrs.name;
            var fieldDefaultValue = (_b = (_a = this.fieldContext) === null || _a === void 0 ? void 0 : _a.fieldMetadata) === null || _b === void 0 ? void 0 : _b.defaultValue;
            var existingValue = this.fieldContext.data[fieldName];
            // return existingValue !== undefined && existingValue !== null ? existingValue : fieldDefaultValue || '';
            return existingValue !== undefined && existingValue !== null ? undefined : fieldDefaultValue || '';
        }
    });
    Object.defineProperty(SolidPasswordField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b;
            var schema = Yup.string();
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldName = fieldLayoutInfo.attrs.name;
            var confirmFieldName = "".concat(fieldName, "Confirm");
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // 1. required 
            // 1. required
            if (fieldMetadata.required) {
                schema = schema.required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel));
            }
            else {
                schema = schema.nullable(); // Allow null when not required
            }
            // 2. length (min/max)
            if (fieldMetadata.min && fieldMetadata.min > 0) {
                schema = schema.min(fieldMetadata.min, ERROR_MESSAGES.FIELD_MINIMUM_CHARACTER(fieldLabel, fieldMetadata.min));
            }
            if (fieldMetadata.max && fieldMetadata.max > 0) {
                schema = schema.max(fieldMetadata.max, ERROR_MESSAGES.FIELD_MAXIMUM_CHARACTER(fieldLabel, fieldMetadata.max));
            }
            // 3. regular expression
            if (fieldMetadata.regexPattern) {
                var regexPatternNotMatchingErrorMsg = (_b = fieldMetadata.regexPatternNotMatchingErrorMsg) !== null && _b !== void 0 ? _b : ERROR_MESSAGES.FIELD_INVALID_DATA(fieldLabel);
                schema = schema.matches(fieldMetadata.regexPattern, regexPatternNotMatchingErrorMsg);
            }
            //check password and confirm password match if password have value
            schema = schema.test('passwords-match', ERROR_MESSAGES.FIELD_MUST_MATCH(fieldLabel), function (value) {
                var _a = this, path = _a.path, parent = _a.parent;
                var confirmValue = parent[confirmFieldName];
                //  Edit mode / untouched → SKIP
                if (!value && !confirmValue) {
                    return true;
                }
                //  Only one filled → ERROR
                if (value || confirmValue) {
                    return value === confirmValue;
                }
                return true; // If password is empty, don't validate match
            });
            return schema;
        }
    });
    Object.defineProperty(SolidPasswordField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a, _b;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            var createWidget = (_b = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.createWidget;
            var pathname = usePathname();
            var isCreateForm = pathname.endsWith('/new');
            var data = this.fieldContext.data || {};
            var isFieldDataEmpty = Object.keys(data).length === 0;
            if (!createWidget) {
                createWidget = 'DefaultPasswordFormCreateWidget';
            }
            if (!editWidget) {
                editWidget = 'DefaultPasswordFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultPasswordFormViewWidget'; // add of this need to createWidget
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            _jsx(_Fragment, {})
                        // this.renderExtensionRenderMode(viewWidget, formik)
                        , viewMode === "edit" && (!isCreateForm && !isFieldDataEmpty) &&
                            _jsx(_Fragment, { children: editWidget &&
                                    this.renderExtensionRenderMode(editWidget, formik) }), viewMode === "edit" && (isCreateForm || isFieldDataEmpty) &&
                            _jsx(_Fragment, { children: createWidget &&
                                    this.renderExtensionRenderMode(createWidget, formik) })] }) }));
        }
    });
    Object.defineProperty(SolidPasswordField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidPasswordField;
}());
export { SolidPasswordField };
export var DefaultPasswordFormViewWidget = function (_a) {
    var _b, _c;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    var _d = useState(false), isText = _d[0], setIsText = _d[1];
    return (_jsxs("div", { className: "mt-2 flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsxs("div", { className: "flex align-items-center gap-4", children: [_jsx("p", { className: "m-0", children: isText ? formik.values[fieldLayoutInfo.attrs.name] : "••••••••" }), _jsx("i", { className: "pi ".concat(isText ? 'pi-eye' : 'pi-eye-slash'), onClick: function () { return setIsText(!isText); }, style: { cursor: 'pointer' } })] })] }));
};
//  Adding formCreateWidget
export var DefaultPasswordFormCreateWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var fieldDisabled = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.disabled;
    var fieldReadonly = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.readonly;
    var formDisabled = (_k = (_j = (_h = solidFormViewMetaData.data.solidView) === null || _h === void 0 ? void 0 : _h.layout) === null || _j === void 0 ? void 0 : _j.attrs) === null || _k === void 0 ? void 0 : _k.disabled;
    var formReadonly = (_o = (_m = (_l = solidFormViewMetaData.data.solidView) === null || _l === void 0 ? void 0 : _l.layout) === null || _m === void 0 ? void 0 : _m.attrs) === null || _o === void 0 ? void 0 : _o.readonly;
    return (_jsx("div", { className: "relative ", children: _jsxs("div", { className: "password-field-component", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                            _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx(Password, { id: fieldLayoutInfo.attrs.name, name: fieldMetadata.name, value: formik.values[fieldLayoutInfo.attrs.name] || '', onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, onBlur: function (e) { return fieldContext.onBlur(e, 'onFieldBlur'); }, readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled, toggleMask: true, autoComplete: "new-password" })] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx(Message, { severity: "error", text: (_p = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _p === void 0 ? void 0 : _p.toString() })), _jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [_jsxs("label", { htmlFor: "".concat(fieldLayoutInfo.attrs.name, "Confirm"), className: "form-field-label", children: ["Confirm ", fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" })] }), _jsx(Password, { id: "".concat(fieldLayoutInfo.attrs.name, "Confirm"), name: "".concat(fieldLayoutInfo.attrs.name, "Confirm"), value: formik.values["".concat(fieldLayoutInfo.attrs.name, "Confirm")] || '', onChange: function (e) {
                                formik.setFieldValue("".concat(fieldLayoutInfo.attrs.name, "Confirm"), e.target.value);
                            }, onBlur: function (e) {
                                formik.setFieldTouched("".concat(fieldLayoutInfo.attrs.name, "Confirm"), true);
                            }, readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled, toggleMask: true, autoComplete: "new-password" }), isFormFieldValid(formik, "".concat(fieldLayoutInfo.attrs.name, "Confirm")) && (_jsx(Message, { severity: "error", text: (_q = formik === null || formik === void 0 ? void 0 : formik.errors["".concat(fieldLayoutInfo.attrs.name, "Confirm")]) === null || _q === void 0 ? void 0 : _q.toString() }))] })] }) }));
};
// Adding formEditWidget custom
export var DefaultPasswordFormEditWidget = function (_a) {
    var _b, _c;
    var _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_d = fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.className) || 'field col-12';
    var fieldLabel = (_e = fieldLayoutInfo.attrs.label) !== null && _e !== void 0 ? _e : fieldMetadata.displayName;
    var fieldDescription = (_f = fieldLayoutInfo.attrs.description) !== null && _f !== void 0 ? _f : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_g = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var fieldDisabled = (_h = fieldLayoutInfo.attrs) === null || _h === void 0 ? void 0 : _h.disabled;
    var fieldReadonly = (_j = fieldLayoutInfo.attrs) === null || _j === void 0 ? void 0 : _j.readonly;
    var formDisabled = (_m = (_l = (_k = solidFormViewMetaData.data.solidView) === null || _k === void 0 ? void 0 : _k.layout) === null || _l === void 0 ? void 0 : _l.attrs) === null || _m === void 0 ? void 0 : _m.disabled;
    var formReadonly = (_q = (_p = (_o = solidFormViewMetaData.data.solidView) === null || _o === void 0 ? void 0 : _o.layout) === null || _p === void 0 ? void 0 : _p.attrs) === null || _q === void 0 ? void 0 : _q.readonly;
    var url = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.modelName.replace(/([a-z0-9])([A-Z])/g, '$1-$2').replace(/([A-Z])([A-Z][a-z])/g, '$1-$2').toLowerCase();
    var id = (_r = fieldContext === null || fieldContext === void 0 ? void 0 : fieldContext.data) === null || _r === void 0 ? void 0 : _r.id;
    var isFormFieldValid = function (formik, fieldName) {
        return formik.touched[fieldName] && formik.errors[fieldName];
    };
    var _u = useState(false), visible = _u[0], setVisible = _u[1];
    var fieldName = fieldLayoutInfo.attrs.name;
    var confirmFieldName = "".concat(fieldName, "Confirm");
    var modalFormik = useFormik({
        initialValues: (_b = {},
            _b[fieldName] = '',
            _b[confirmFieldName] = '',
            _b),
        validationSchema: Yup.object((_c = {},
            _c[fieldName] = Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel)),
            _c[confirmFieldName] = Yup.string()
                .required("Confirm ".concat(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel)))
                .oneOf([Yup.ref(fieldName)], ERROR_MESSAGES.FIELD_MUST_MATCH(fieldLabel)),
            _c)),
        onSubmit: function (values, _a) {
            var resetForm = _a.resetForm;
            return __awaiter(void 0, void 0, void 0, function () {
                var err_1;
                return __generator(this, function (_b) {
                    switch (_b.label) {
                        case 0:
                            _b.trys.push([0, 2, , 3]);
                            return [4 /*yield*/, updatePasswordField({
                                    url: url,
                                    id: id,
                                    fieldName: fieldName,
                                    fieldValue: values[fieldName],
                                    confirmFieldName: confirmFieldName,
                                    confirmFieldValue: values[confirmFieldName],
                                })];
                        case 1:
                            _b.sent();
                            // formik.setFieldValue(fieldName, values[fieldName]);
                            resetForm();
                            setVisible(false);
                            return [3 /*break*/, 3];
                        case 2:
                            err_1 = _b.sent();
                            console.error(err_1);
                            return [3 /*break*/, 3];
                        case 3: return [2 /*return*/];
                    }
                });
            });
        },
    });
    return (_jsxs("div", { className: className, children: [showFieldLabel !== false && (_jsxs("label", { htmlFor: fieldName, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] })), _jsx(Button, { type: "button", label: "Change Password", icon: "pi pi-lock", onClick: function () { return setVisible(true); }, className: "mt-2", disabled: formDisabled || fieldDisabled || readOnlyPermission || fieldReadonly }), _jsx(Dialog, { header: "Change ".concat(fieldLabel), visible: visible, onHide: function () { return setVisible(false); }, style: { width: '30vw' }, className: "solid-confirm-dialog", children: _jsxs("form", { onSubmit: modalFormik.handleSubmit, className: "p-fluid", children: [_jsxs("div", { className: "field", children: [_jsxs("label", { htmlFor: fieldName, children: ["New ", fieldLabel] }), _jsx(Password, { id: fieldName, name: fieldName, value: modalFormik.values[fieldName], onChange: modalFormik.handleChange, onBlur: modalFormik.handleBlur, toggleMask: true, feedback: false, autoComplete: "new-password" }), isFormFieldValid(modalFormik, fieldName) && (_jsx(Message, { severity: "error", text: (_s = modalFormik.errors[fieldName]) === null || _s === void 0 ? void 0 : _s.toString() }))] }), _jsxs("div", { className: "field mt-4", children: [_jsxs("label", { htmlFor: confirmFieldName, children: ["Confirm ", fieldLabel] }), _jsx(Password, { id: confirmFieldName, name: confirmFieldName, value: modalFormik.values[confirmFieldName], onChange: modalFormik.handleChange, onBlur: modalFormik.handleBlur, toggleMask: true, feedback: false, autoComplete: "new-password" }), isFormFieldValid(modalFormik, confirmFieldName) && (_jsx(Message, { severity: "error", text: (_t = modalFormik.errors[confirmFieldName]) === null || _t === void 0 ? void 0 : _t.toString() }))] }), _jsx("div", { className: "mt-5", children: _jsx(Button, { label: "Update Password", icon: "pi pi-check", type: "submit", className: "w-full" }) })] }) })] }));
};
