// @solidxai/core-ui/dist/components/core/form/fields/SolidMediaSingleField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { DropzonePlaceholder } from "../../../../components/common/DropzonePlaceholder";
import { useDeleteMediaMutation } from "../../../../redux/api/mediaApi";
import { Button } from "primereact/button";
import { Dialog } from "primereact/dialog";
import { Message } from "primereact/message";
import { useEffect, useState } from "react";
import { useDropzone } from "react-dropzone";
import * as Yup from 'yup';
import { FileReaderExt } from "../../../../components/common/FileReaderExt";
import getAcceptedFileTypes from "../../../../helpers/getAcceptedFileTypes";
import { downloadMediaFile } from "../../../../helpers/downloadMediaFile";
import { getExtensionComponent } from "../../../../helpers/registry";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
var SolidMediaSingleField = /** @class */ (function () {
    function SolidMediaSingleField(fieldContext, setLightboxUrls, setOpenLightbox) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "setLightboxUrls", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "setOpenLightbox", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
        this.setLightboxUrls = setLightboxUrls;
        this.setOpenLightbox = setOpenLightbox;
    }
    Object.defineProperty(SolidMediaSingleField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            if (value instanceof File) {
                formData.append(fieldLayoutInfo.attrs.name, value);
            }
        }
    });
    Object.defineProperty(SolidMediaSingleField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var mediaUrls = this.fieldContext.data && this.fieldContext.data._media && this.fieldContext.data._media[this.fieldContext.field.attrs.name][0];
            return mediaUrls;
        }
    });
    Object.defineProperty(SolidMediaSingleField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            // let schema: Yup.StringSchema = Yup.object();
            var _a;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // Dynamically determine the base type based on the 'required' flag
            // Define base schema
            var schema;
            if (fieldMetadata.required) {
                // For required fields: disallow null and undefined
                schema = Yup.mixed()
                    .required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel))
                    .test(ERROR_MESSAGES.FILE_OBJECT, ERROR_MESSAGES.MUST_BE_FILE_OBJECT(fieldLabel), function (value) {
                    return value instanceof File || typeof value === "object";
                } // Validate File or object
                );
            }
            else {
                // For optional fields: allow null or undefined
                schema = Yup.mixed()
                    .nullable() // Allow null explicitly
                    .test(ERROR_MESSAGES.FILE_OBJECT, ERROR_MESSAGES.MUST_BE_FILE_OBJECT(fieldLabel), function (value) {
                    return value === null || // Allow null
                        value === undefined || // Allow undefined
                        value instanceof File || // Allow File
                        value === 'false' ||
                        typeof value === "object";
                } // Allow object
                );
            }
            return schema;
        }
    });
    Object.defineProperty(SolidMediaSingleField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultMediaSingleFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultMediaSingleFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" &&
                            _jsx(_Fragment, { children: editWidget &&
                                    this.renderExtensionRenderMode(editWidget, formik) })] }) }));
        }
    });
    Object.defineProperty(SolidMediaSingleField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
                setLightboxUrls: this.setLightboxUrls,
                setOpenLightbox: this.setOpenLightbox
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidMediaSingleField;
}());
export { SolidMediaSingleField };
export var DefaultMediaSingleFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f;
    var formik = _a.formik, fieldContext = _a.fieldContext, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly ? fieldContext.readOnly : fieldLayoutInfo.attrs.readonly;
    var viewMode = fieldContext.viewMode;
    var _g = useState(false), isDeleteImageDialogVisible = _g[0], setDeleteImageDialogVisible = _g[1];
    var _h = useState(), imageToBeDeletedData = _h[0], setImageToBeDeletedData = _h[1];
    var _j = useState(null), fileDetails = _j[0], setFileDetails = _j[1];
    var _k = useState(false), isReplaceImageDialogVisible = _k[0], setReplaceImageDialogVisible = _k[1];
    var _l = useState(null), newFileToUpload = _l[0], setNewFileToUpload = _l[1];
    var _m = useState(null), fileSizeError = _m[0], setFileSizeError = _m[1];
    var formatFileSize = function (size) {
        return size >= 1024 * 1024
            ? "".concat((size / (1024 * 1024)).toFixed(1), " MB")
            : "".concat((size / 1024).toFixed(1), " KB");
    };
    var _o = useDeleteMediaMutation(), deleteMedia = _o[0], _p = _o[1], isMediaDeleted = _p.isLoading, isDeleteMediaSuceess = _p.isSuccess, isMediaDeleteError = _p.isError, mediaDeleteError = _p.error, DeletedMedia = _p.data;
    var handleCancelUpload = function (e) {
        e.stopPropagation();
        if (imageToBeDeletedData) {
            deleteMedia(imageToBeDeletedData);
        }
        e.stopPropagation();
        setFileDetails(null);
        formik.setFieldValue(fieldLayoutInfo.attrs.name, null);
        setDeleteImageDialogVisible(false);
    };
    var handleDropImage = function (acceptedFiles) {
        var file = acceptedFiles[0];
        if (!file)
            return;
        setFileSizeError(null);
        if (fileDetails) {
            // If a file is already uploaded, show the confirmation dialog
            setNewFileToUpload(file);
            setReplaceImageDialogVisible(true);
        }
        else {
            // If no file is present, proceed with upload
            uploadFile(file);
        }
    };
    var uploadFile = function (file) {
        setFileDetails({
            name: file.name,
            type: file.type,
            fileUrl: URL.createObjectURL(file),
            fileSize: file.size
        });
        var reader = new FileReader();
        reader.readAsDataURL(file);
        formik.setFieldValue(fieldLayoutInfo.attrs.name, file);
    };
    var handleReplaceFile = function () {
        // Delete the existing file first
        if (imageToBeDeletedData) {
            deleteMedia(imageToBeDeletedData);
        }
        setFileDetails(null);
        formik.setFieldValue(fieldLayoutInfo.attrs.name, null);
        // Proceed with uploading new file
        if (newFileToUpload) {
            uploadFile(newFileToUpload);
            setNewFileToUpload(null);
        }
        setReplaceImageDialogVisible(false);
    };
    useEffect(function () {
        var fieldValue = formik === null || formik === void 0 ? void 0 : formik.values[fieldLayoutInfo.attrs.name];
        if (fieldValue && typeof fieldValue === "object") {
            var fileUrl = "";
            var fileName = "Unknown File";
            var fileSize = 0;
            if (fieldValue instanceof File) {
                fileUrl = URL.createObjectURL(fieldValue);
                fileName = fieldValue.name;
                fileSize = fieldValue.size;
            }
            else if (fieldValue._full_url) {
                fileUrl = fieldValue._full_url;
                fileName = fieldValue.originalFileName;
                fileSize = fieldValue.fileSize;
            }
            setFileDetails({
                name: fileName,
                type: fieldValue.mimeType ? fieldValue.mimeType : fieldValue.type,
                fileUrl: fileUrl,
                fileSize: fileSize
            });
            // Set file ID for delete operation
            setImageToBeDeletedData(fieldValue.id);
            // Ensure formik has the correct value
            formik.setFieldValue(fieldLayoutInfo.attrs.name, fieldValue);
        }
    }, [formik.values, fieldLayoutInfo.attrs.name]);
    var _q = useDropzone({
        onDrop: handleDropImage,
        onDropRejected: function (fileRejections) {
            var _a;
            var rejection = fileRejections[0];
            var sizeError = rejection.errors.find(function (err) { return err.code === 'file-too-large'; });
            if (sizeError) {
                setFileSizeError(ERROR_MESSAGES.FILE_TOO_LAREG(fieldMetadata.mediaMaxSizeKb));
            }
            else {
                setFileSizeError(((_a = rejection.errors[0]) === null || _a === void 0 ? void 0 : _a.message) || ERROR_MESSAGES.FILE_NOT_ACCEPT);
            }
        },
        accept: getAcceptedFileTypes(fieldMetadata.mediaTypes),
        maxSize: fieldMetadata.mediaMaxSizeKb * 1024,
    }), getRootProps = _q.getRootProps, getInputProps = _q.getInputProps, isDragActive = _q.isDragActive;
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var DynamicWidget = getExtensionComponent("SolidFormFieldViewMediaSingleWidget");
    var widgetProps = {
        formik: formik,
        fieldContext: fieldContext,
        setLightboxUrls: setLightboxUrls,
        setOpenLightbox: setOpenLightbox
    };
    var handleFileView = function (url) {
        var _a;
        var downloadOnlyExt = [
            "txt", "zip", "rar",
            "doc", "docx",
            "xls", "xlsx",
            "ppt", "pptx"
        ];
        var fileUrl = (url === null || url === void 0 ? void 0 : url.fileUrl) || "";
        var cleanUrl = fileUrl.split("?")[0];
        var ext = (_a = cleanUrl.split(".").pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
        if (ext && downloadOnlyExt.includes(ext)) {
            var link = document.createElement('a');
            link.href = url.fileUrl;
            link.download = ''; // or specify a file name like 'file.pdf'
            link.target = '_blank';
            document.body.appendChild(link);
            link.click();
            document.body.removeChild(link);
        }
        else {
            setLightboxUrls === null || setLightboxUrls === void 0 ? void 0 : setLightboxUrls([
                { src: url.fileUrl, downloadUrl: url.fileUrl },
            ]);
            setOpenLightbox === null || setOpenLightbox === void 0 ? void 0 : setOpenLightbox(true);
        }
    };
    return (_jsxs("div", { style: readOnlyPermission === true ? { filter: 'opacity(50%)', pointerEvents: 'none' } : {}, children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4 relative", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, fieldMetadata.required && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsxs("div", { className: "relative", children: [_jsxs("div", __assign({}, getRootProps(), { className: "solid-dropzone-wrapper", children: [_jsx("input", __assign({}, getInputProps())), _jsx(DropzonePlaceholder, { mediaTypes: fieldMetadata.mediaTypes, mediaMaxSizeKb: fieldMetadata.mediaMaxSizeKb })] })), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: (_f = formik === null || formik === void 0 ? void 0 : formik.errors[fieldLayoutInfo.attrs.name]) === null || _f === void 0 ? void 0 : _f.toString() }) }))] }), fileSizeError &&
                        _jsx(Message, { severity: "error", text: fileSizeError === null || fileSizeError === void 0 ? void 0 : fileSizeError.toString() }), fileDetails && (_jsx("div", { className: "solid-file-upload-wrapper mt-4", children: _jsxs("div", { className: "flex align-items-center md:gap-2", children: [_jsx(FileReaderExt, { fileDetails: fileDetails }), _jsxs("div", { className: "w-full flex flex-column gap-1", children: [_jsxs("div", { className: "flex align-items-start justify-content-between", children: [_jsx("p", { className: "font-normal w-9 text-primary m-0 solid-img-text-wrapper", style: { cursor: 'pointer' }, onClick: function () { return handleFileView(fileDetails); }, children: fileDetails.name }), _jsxs("div", { className: "flex align-items-center md:gap-2", children: [_jsx("div", { children: _jsx(Button, { type: "button", text: true, icon: "pi pi-download", size: "small", severity: "secondary", 
                                                                // className="p-2"
                                                                style: {
                                                                    height: 16,
                                                                    width: 16
                                                                }, onClick: function () { return downloadMediaFile(fileDetails === null || fileDetails === void 0 ? void 0 : fileDetails.fileUrl, fileDetails === null || fileDetails === void 0 ? void 0 : fileDetails.name); } }) }), _jsx("div", { children: _jsx(Button, { type: "button", text: true, icon: "pi pi-times", size: "small", severity: "secondary", 
                                                                // className="p-2"
                                                                style: {
                                                                    height: 16,
                                                                    width: 16
                                                                }, onClick: function () { return setDeleteImageDialogVisible(true); } }) })] })] }), _jsx("div", { className: "flex align-items-center gap-2 text-sm", children: fileDetails && formatFileSize(fileDetails.fileSize) })] })] }) }))] }), _jsx(Dialog, { visible: isDeleteImageDialogVisible, header: "Confirm Delete", modal: true, className: "solid-confirm-dialog", footer: function () { return (_jsxs("div", { className: "flex justify-content-center", children: [_jsx(Button, { type: "button", label: "Yes", icon: "pi pi-check", className: 'small-button', severity: "danger", autoFocus: true, onClick: handleCancelUpload }), _jsx(Button, { type: "button", label: "No", icon: "pi pi-times", className: 'small-button', onClick: function () { return setDeleteImageDialogVisible(false); } })] })); }, onHide: function () { return setDeleteImageDialogVisible(false); }, children: _jsx("p", { children: "Are you sure you want to delete media?" }) }), _jsx(Dialog, { visible: isReplaceImageDialogVisible, header: "Replace Image", modal: true, className: "solid-confirm-dialog", footer: function () { return (_jsxs("div", { className: "flex justify-content-center", children: [_jsx(Button, { type: "button", label: "Yes, Replace", icon: "pi pi-check", className: 'small-button', severity: "danger", onClick: handleReplaceFile }), _jsx(Button, { type: "button", label: "Cancel", icon: "pi pi-times", className: 'small-button', onClick: function () { return setReplaceImageDialogVisible(false); } })] })); }, onHide: function () { return setReplaceImageDialogVisible(false); }, children: _jsx("p", { children: "An media is already uploaded. Do you want to delete it and upload a new one?" }) })] }));
};
export var DefaultMediaSingleFormViewWidget = function (_a) {
    var _b, _c, _d;
    var formik = _a.formik, fieldContext = _a.fieldContext, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox;
    var _e = useState(null), fileDetails = _e[0], setFileDetails = _e[1];
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var showFieldLabel = (_d = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _d === void 0 ? void 0 : _d.showLabel;
    // useEffect(() => { formik.setFieldValue(fieldLayoutInfo.attrs.name, "false") }, [])
    var isFormFieldValid = function (formik, fieldName) { return formik.touched[fieldName] && formik.errors[fieldName]; };
    var formatFileSize = function (size) {
        return size >= 1024 * 1024
            ? "".concat((size / (1024 * 1024)).toFixed(1), " MB")
            : "".concat((size / 1024).toFixed(1), " KB");
    };
    useEffect(function () {
        var fieldValue = formik === null || formik === void 0 ? void 0 : formik.values[fieldLayoutInfo.attrs.name];
        if (fieldValue && typeof fieldValue === "object") {
            var fileUrl = "";
            var fileName = "Unknown File";
            var fileSize = 0;
            if (fieldValue instanceof File) {
                fileUrl = URL.createObjectURL(fieldValue);
                fileName = fieldValue.name;
                fileSize = fieldValue.size;
            }
            else if (fieldValue._full_url) {
                fileUrl = fieldValue._full_url;
                fileName = fieldValue.originalFileName;
                fileSize = fieldValue.fileSize;
            }
            setFileDetails({
                name: fileName,
                type: fieldValue.mimeType ? fieldValue.mimeType : fieldValue.type,
                fileUrl: fileUrl,
                fileSize: fileSize
            });
            // Ensure formik has the correct value
            formik.setFieldValue(fieldLayoutInfo.attrs.name, fieldValue);
        }
    }, [formik.values, fieldLayoutInfo.attrs.name]);
    var handleFileView = function (url) {
        var _a;
        var downloadOnlyExt = [
            "txt", "zip", "rar",
            "doc", "docx",
            "xls", "xlsx",
            "ppt", "pptx"
        ];
        var fileUrl = (url === null || url === void 0 ? void 0 : url.fileUrl) || "";
        var cleanUrl = fileUrl.split("?")[0];
        var ext = (_a = cleanUrl.split(".").pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
        if (ext && downloadOnlyExt.includes(ext)) {
            var link = document.createElement('a');
            link.href = url.fileUrl;
            link.download = ''; // or specify a file name like 'file.pdf'
            link.target = '_blank';
            document.body.appendChild(link);
            link.click();
            document.body.removeChild(link);
        }
        else {
            setLightboxUrls === null || setLightboxUrls === void 0 ? void 0 : setLightboxUrls([
                { src: url.fileUrl, downloadUrl: url.fileUrl },
            ]);
            setOpenLightbox === null || setOpenLightbox === void 0 ? void 0 : setOpenLightbox(true);
        }
    };
    return (_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4 relative", children: [showFieldLabel != false &&
                _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label font-medium", children: [fieldLabel, _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), fileDetails && (_jsx("div", { className: "solid-file-view-wrapper mt-4", children: _jsxs("div", { className: "flex align-items-center md:gap-2", children: [_jsx(FileReaderExt, { fileDetails: fileDetails }), _jsxs("div", { className: "w-full flex flex-column gap-1", children: [_jsxs("div", { className: "flex align-items-start justify-content-between", children: [_jsx("p", { className: "font-normal w-9 text-primary m-0 solid-img-text-wrapper", style: { cursor: 'pointer' }, onClick: function () { return handleFileView(fileDetails); }, children: fileDetails.name }), _jsx("div", { className: "flex align-items-center md:gap-2", children: _jsx("div", { children: _jsx(Button, { type: "button", text: true, icon: "pi pi-download", size: "small", severity: "secondary", 
                                                    // className="p-2"
                                                    style: {
                                                        height: 16,
                                                        width: 16
                                                    }, onClick: function () { return downloadMediaFile(fileDetails === null || fileDetails === void 0 ? void 0 : fileDetails.fileUrl, fileDetails === null || fileDetails === void 0 ? void 0 : fileDetails.name); } }) }) })] }), _jsx("div", { className: "flex align-items-center gap-2 text-sm", children: fileDetails && formatFileSize(fileDetails.fileSize) })] })] }) }))] }));
};
