// @solidxai/core-ui/dist/components/core/form/fields/SolidSelectionDynamicField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useLazyGetSelectionDynamicValuesQuery } from "../../../../redux/api/fieldApi";
import { AutoComplete } from "primereact/autocomplete";
import { Message } from "primereact/message";
import qs from "qs";
import { useState } from "react";
import * as Yup from 'yup';
import { getExtensionComponent } from "../../../../helpers/registry";
import { SolidFieldTooltip } from "../../../../components/common/SolidFieldTooltip";
import { formikValuestoQueryString } from "../../../../helpers/helpers";
import { ERROR_MESSAGES } from "../../../../constants/error-messages";
var SolidSelectionDynamicField = /** @class */ (function () {
    function SolidSelectionDynamicField(fieldContext) {
        Object.defineProperty(this, "fieldContext", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fieldContext = fieldContext;
    }
    Object.defineProperty(SolidSelectionDynamicField.prototype, "initialValue", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            // TODO: Use the field metadata to re-create the object in the valid format 
            // {label: '', value: ''}
            var optionValue = this.fieldContext.data[this.fieldContext.field.attrs.name];
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
            // isMultiSelect: fieldMetaData ? fieldMetaData?.isMultiSelect : false,
            // return { label: optionValue || '', value: optionValue || '' };
            var stripBracketsAndQuotes = function (str) {
                return str.replace(/^\[?"?/, '').replace(/"?\]?$/, '').trim();
            };
            var cleanValue = function (val) {
                var _a, _b;
                if (typeof val === 'string') {
                    var cleaned = stripBracketsAndQuotes(val);
                    return { label: cleaned, value: cleaned };
                }
                else if (val && typeof val === 'object') {
                    var label = stripBracketsAndQuotes(String((_a = val.label) !== null && _a !== void 0 ? _a : val));
                    var value = stripBracketsAndQuotes(String((_b = val.value) !== null && _b !== void 0 ? _b : val));
                    return { label: label, value: value };
                }
                else {
                    var strVal = stripBracketsAndQuotes(String(val));
                    return { label: strVal, value: strVal };
                }
            };
            if (isMultiSelect) {
                if (Array.isArray(optionValue)) {
                    return optionValue.map(cleanValue);
                }
                else if (typeof optionValue === 'string') {
                    try {
                        var parsed = JSON.parse(optionValue); // try to parse as JSON array
                        if (Array.isArray(parsed)) {
                            return parsed.map(cleanValue);
                        }
                    }
                    catch (_a) {
                        // fallback to comma-split only if not valid JSON
                        var splitValues = optionValue.split(',').map(function (v) { return v.trim(); }).filter(Boolean);
                        return splitValues.map(cleanValue);
                    }
                }
                return [];
            }
            else {
                return cleanValue(optionValue || '');
            }
        }
    });
    Object.defineProperty(SolidSelectionDynamicField.prototype, "updateFormData", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (value, formData) {
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
            var stripQuotes = function (str) { return str.replace(/^"+|"+$/g, '').replace(/^'+|'+$/g, ''); };
            if (value) {
                if (isMultiSelect) {
                    var selectedValues = value.map(function (v) {
                        var cleanedValue = stripQuotes(v.value); // still fine
                        return cleanedValue;
                    });
                    var jsonString = JSON.stringify(selectedValues); // ✅ Proper JSON array
                    formData.append(fieldLayoutInfo.attrs.name, jsonString);
                }
                else {
                    var cleanedValue = stripQuotes(value.value);
                    formData.append(fieldLayoutInfo.attrs.name, cleanedValue);
                }
            }
            // if (value) {
            //     formData.append(fieldLayoutInfo.attrs.name, value.value);
            // }
        }
    });
    Object.defineProperty(SolidSelectionDynamicField.prototype, "validationSchema", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function () {
            var _a, _b, _c;
            var schema = Yup.object();
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var fieldLabel = (_a = fieldLayoutInfo.attrs.label) !== null && _a !== void 0 ? _a : fieldMetadata.displayName;
            // 1. required 
            // if (fieldMetadata.required) {
            //     schema = schema.required(`${fieldLabel} is required.`);
            // }
            // return schema;
            var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
            var isRequired = (_c = (_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.required) !== null && _c !== void 0 ? _c : fieldMetadata.required;
            if (!isRequired) {
                return Yup.mixed();
            }
            if (isMultiSelect) {
                return Yup.array()
                    .min(1, ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel))
                    .of(Yup.object().shape({ label: Yup.string(), value: Yup.string() }));
            }
            return Yup.object().shape({
                label: Yup.string(),
                value: Yup.string().required(ERROR_MESSAGES.FIELD_REUQIRED(fieldLabel)),
            });
        }
    });
    Object.defineProperty(SolidSelectionDynamicField.prototype, "render", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (formik) {
            var _a;
            var fieldMetadata = this.fieldContext.fieldMetadata;
            var fieldLayoutInfo = this.fieldContext.field;
            var className = ((_a = fieldLayoutInfo.attrs) === null || _a === void 0 ? void 0 : _a.className) || 'field col-12';
            // const isFormFieldValid = (formik: any, fieldName: string) => formik.touched[fieldName] && formik.errors[fieldName];
            var isFormFieldValid = function (formik, fieldName) { return formik.errors[fieldName]; };
            var viewWidget = fieldLayoutInfo.attrs.viewWidget;
            var editWidget = fieldLayoutInfo.attrs.editWidget;
            if (!editWidget) {
                editWidget = 'DefaultSelectionDynamicFormEditWidget';
            }
            if (!viewWidget) {
                viewWidget = 'DefaultSelectionDynamicFormViewWidget';
            }
            var viewMode = this.fieldContext.viewMode;
            return (_jsx(_Fragment, { children: _jsxs("div", { className: className, children: [viewMode === "view" &&
                            this.renderExtensionRenderMode(viewWidget, formik), viewMode === "edit" &&
                            _jsx(_Fragment, { children: editWidget &&
                                    this.renderExtensionRenderMode(editWidget, formik) })] }) }));
        }
    });
    Object.defineProperty(SolidSelectionDynamicField.prototype, "renderExtensionRenderMode", {
        enumerable: false,
        configurable: true,
        writable: true,
        value: function (widget, formik) {
            var DynamicWidget = getExtensionComponent(widget);
            var widgetProps = {
                formik: formik,
                fieldContext: this.fieldContext,
            };
            return (_jsx(_Fragment, { children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
        }
    });
    return SolidSelectionDynamicField;
}());
export { SolidSelectionDynamicField };
export var DefaultSelectionDynamicFormEditWidget = function (_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var className = ((_b = fieldLayoutInfo.attrs) === null || _b === void 0 ? void 0 : _b.className) || 'field col-12';
    var fieldLabel = (_c = fieldLayoutInfo.attrs.label) !== null && _c !== void 0 ? _c : fieldMetadata.displayName;
    var fieldDescription = (_d = fieldLayoutInfo.attrs.description) !== null && _d !== void 0 ? _d : fieldMetadata.description;
    var solidFormViewMetaData = fieldContext.solidFormViewMetaData;
    var showFieldLabel = (_e = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _e === void 0 ? void 0 : _e.showLabel;
    var readOnlyPermission = fieldContext.readOnly;
    var fieldDisabled = (_f = fieldLayoutInfo.attrs) === null || _f === void 0 ? void 0 : _f.disabled;
    var fieldReadonly = (_g = fieldLayoutInfo.attrs) === null || _g === void 0 ? void 0 : _g.readonly;
    var formDisabled = (_k = (_j = (_h = solidFormViewMetaData.data.solidView) === null || _h === void 0 ? void 0 : _h.layout) === null || _j === void 0 ? void 0 : _j.attrs) === null || _k === void 0 ? void 0 : _k.disabled;
    var formReadonly = (_o = (_m = (_l = solidFormViewMetaData.data.solidView) === null || _l === void 0 ? void 0 : _l.layout) === null || _m === void 0 ? void 0 : _m.attrs) === null || _o === void 0 ? void 0 : _o.readonly;
    var whereClause = fieldLayoutInfo.attrs.whereClause;
    var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
    // selection dynamic specific code. 
    var triggerGetSelectionDynamicValues = useLazyGetSelectionDynamicValuesQuery()[0];
    var _x = useState([]), selectionDynamicItems = _x[0], setSelectionDynamicItems = _x[1];
    var isFormFieldValid = function (formik, fieldName) {
        return (formik.touched[fieldName] || formik.submitCount > 0) && !!formik.errors[fieldName];
    };
    var isRequired = (_q = (_p = fieldLayoutInfo.attrs) === null || _p === void 0 ? void 0 : _p.required) !== null && _q !== void 0 ? _q : fieldMetadata.required;
    var selectionDynamicSearch = function (event) { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, sdQs, sdResponse, items, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    queryData = {
                        offset: 0,
                        limit: 10,
                        query: event.query,
                        fieldId: fieldMetadata.id,
                        formValues: formikValuestoQueryString(formik.values),
                    };
                    if (whereClause) {
                        queryData.query = whereClause;
                    }
                    sdQs = qs.stringify(queryData, {
                        encodeValuesOnly: true,
                        // encoder: (str, defaultEncoder, charset, type) => {
                        //     if (type === 'key' || type === 'value') {
                        //         if (str === queryData.query) return str;
                        //     }
                        //     return defaultEncoder(str);
                        // }
                    });
                    return [4 /*yield*/, triggerGetSelectionDynamicValues(sdQs).unwrap()];
                case 1:
                    sdResponse = _a.sent();
                    items = Array.isArray(sdResponse === null || sdResponse === void 0 ? void 0 : sdResponse.data) ? sdResponse.data : [];
                    // TODO: if no data found then can we show no matching "entities", where entities can be replaced with the model plural name,
                    // const sdData = sdResponse.data.data;
                    // @ts-ignore
                    setSelectionDynamicItems(__spreadArray([], items, true));
                    return [3 /*break*/, 3];
                case 2:
                    err_1 = _a.sent();
                    setSelectionDynamicItems([]);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    return (_jsxs("div", { className: "relative", children: [_jsxs("div", { className: "flex flex-column gap-2 mt-1 sm:mt-2 md:mt-3 lg:mt-4", children: [showFieldLabel != false &&
                        _jsxs("label", { htmlFor: fieldLayoutInfo.attrs.name, className: "form-field-label", children: [fieldLabel, isRequired && _jsx("span", { className: "text-red-500", children: " *" }), _jsx(SolidFieldTooltip, { fieldContext: fieldContext })] }), _jsx(AutoComplete, __assign({ multiple: isMultiSelect, readOnly: formReadonly || fieldReadonly || readOnlyPermission, disabled: formDisabled || fieldDisabled }, formik.getFieldProps(fieldLayoutInfo.attrs.name), { id: fieldLayoutInfo.attrs.name, field: "label", 
                        // value={formik.values[fieldLayoutInfo.attrs.name] || null}
                        value: formik.values[fieldLayoutInfo.attrs.name] || (isMultiSelect ? [] : null), dropdown: true, suggestions: selectionDynamicItems, completeMethod: selectionDynamicSearch, 
                        // onChange={(e) => updateInputs(index, e.value)} />
                        // onChange={formik.handleChange}
                        onChange: function (e) { return fieldContext.onChange(e, 'onFieldChange'); }, className: "solid-standard-autocomplete", emptyMessage: "No records found" }))] }), isFormFieldValid(formik, fieldLayoutInfo.attrs.name) && (_jsx("div", { className: "absolute mt-1", children: _jsx(Message, { severity: "error", text: 
                    // formik?.errors[fieldLayoutInfo.attrs.name]?.toString()
                    typeof formik.errors[(_r = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _r === void 0 ? void 0 : _r.name] === 'object'
                        ? (_u = (_t = formik.errors[(_s = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _s === void 0 ? void 0 : _s.name]) === null || _t === void 0 ? void 0 : _t.value) === null || _u === void 0 ? void 0 : _u.toString()
                        : (_w = formik.errors[(_v = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _v === void 0 ? void 0 : _v.name]) === null || _w === void 0 ? void 0 : _w.toString() }) }))] }));
};
export var DefaultSelectionDynamicFormViewWidget = function (_a) {
    var _b, _c;
    var formik = _a.formik, fieldContext = _a.fieldContext;
    var fieldMetadata = fieldContext.fieldMetadata;
    var fieldLayoutInfo = fieldContext.field;
    var fieldLabel = (_b = fieldLayoutInfo.attrs.label) !== null && _b !== void 0 ? _b : fieldMetadata.displayName;
    var showFieldLabel = (_c = fieldLayoutInfo === null || fieldLayoutInfo === void 0 ? void 0 : fieldLayoutInfo.attrs) === null || _c === void 0 ? void 0 : _c.showLabel;
    var value = formik.values[fieldLayoutInfo.attrs.name];
    var isMultiSelect = fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.isMultiSelect;
    var values = [];
    if (isMultiSelect) {
        if (typeof value === 'string') {
            try {
                var parsed = JSON.parse(value);
                if (Array.isArray(parsed)) {
                    values = parsed;
                }
            }
            catch (_d) {
                // Fallback if not a JSON string (or invalid JSON)
                values = value.split(',').map(function (v) { return v.trim(); });
            }
        }
        else if (Array.isArray(value)) {
            values = value.map(function (v) { return (typeof v === 'object' && v.label ? v.label : String(v)); });
        }
    }
    return (
    // <div className="mt-2 flex-column gap-2">
    //     <p className="m-0 form-field-label font-medium">{fieldLabel}</p>
    //     <p className="m-0 solid-custom-selection-dynamic-pill">{value && value.label && value.label}</p>
    // </div>
    _jsxs("div", { className: "mt-2 flex-column gap-2", children: [showFieldLabel !== false && (_jsx("p", { className: "m-0 form-field-label font-medium", children: fieldLabel })), _jsx("p", { className: "m-0", children: isMultiSelect ? (values.length > 0 ? (_jsx("span", { children: values.join(', ') }) // ✅ Join with commas
                ) : (_jsx("span", { className: "text-gray-500", children: "No selection" }))) : (value && (value === null || value === void 0 ? void 0 : value.label) ? value.label : _jsx("span", { className: "text-gray-500", children: "No selection" })) })] }));
};
