// @solidxai/core-ui/dist/components/core/form/SolidFormUserViewLayout.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useFormik } from "formik";
import { Button } from "primereact/button";
import { useRef } from "react";
import { createSolidEntityApi } from "../../../redux/api/solidEntityApi";
import { javascript } from "@codemirror/lang-javascript";
import { oneDark } from "@codemirror/theme-one-dark";
import CodeMirror, { EditorView } from "@uiw/react-codemirror";
import { Toast } from "primereact/toast";
import { ERROR_MESSAGES } from "../../../constants/error-messages";
import showToast from "../../../helpers/showToast";
export var SolidFormUserViewLayout = function (_a) {
    var _b;
    var solidFormViewMetaData = _a.solidFormViewMetaData, setLayoutDialogVisible = _a.setLayoutDialogVisible;
    var toast = useRef(null);
    var entityApi = createSolidEntityApi("userViewMetadata");
    var useUpsertSolidEntityMutation = entityApi.useUpsertSolidEntityMutation;
    var upsertUserView = useUpsertSolidEntityMutation()[0];
    if (!((_b = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _b === void 0 ? void 0 : _b.solidView))
        return null;
    var solidView = solidFormViewMetaData.data.solidView;
    var formik = useFormik({
        initialValues: {
            layoutString: JSON.stringify(solidView.layout, null, 2),
        },
        onSubmit: function (values) { return __awaiter(void 0, void 0, void 0, function () {
            var parsedLayout, response, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        parsedLayout = JSON.parse(values.layoutString);
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 4, , 5]);
                        if (!solidView.id) return [3 /*break*/, 3];
                        return [4 /*yield*/, upsertUserView({
                                viewMetadataId: solidView.id,
                                layout: JSON.stringify(parsedLayout)
                            }).unwrap()];
                    case 2:
                        response = _a.sent();
                        if (response.statusCode === 200) {
                            showToast(toast, "success", ERROR_MESSAGES.LAYOUT, ERROR_MESSAGES.FORM_LAYOUT_UPDATE);
                            setLayoutDialogVisible(false);
                            window.location.reload();
                        }
                        _a.label = 3;
                    case 3: return [3 /*break*/, 5];
                    case 4:
                        error_1 = _a.sent();
                        console.error(ERROR_MESSAGES.UPDATE_FAILED, error_1);
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        }); },
    });
    return (_jsxs(_Fragment, { children: [_jsx(Toast, { ref: toast }), _jsxs("form", { onSubmit: formik.handleSubmit, children: [_jsx(CodeMirror, { value: formik.values.layoutString, height: "500px", theme: oneDark, style: { fontSize: '10px' }, extensions: [javascript(), EditorView.lineWrapping], onChange: function (value) {
                            formik.setFieldValue("layoutString", value);
                        } }), _jsxs("div", { className: "pt-3 flex gap-2", children: [_jsx(Button, { type: "submit", label: "Apply", size: "small" }), _jsx(Button, { type: "button", outlined: true, label: "Cancel", size: "small", onClick: function () { return setLayoutDialogVisible(false); } })] })] })] }));
};
