// @solidxai/core-ui/dist/components/core/form/SolidFormView.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useSession } from '../../../hooks/useSession';
import { permissionExpression } from "../../../helpers/permissions";
import { createSolidEntityApi } from "../../../redux/api/solidEntityApi";
import { useGetSolidViewLayoutQuery } from "../../../redux/api/solidViewApi";
import { useLazyCheckIfPermissionExistsQuery } from "../../../redux/api/userApi";
import { useFormik } from "formik";
import { usePathname } from "../../../hooks/usePathname";
import { useRouter } from "../../../hooks/useRouter";
import { useSearchParams } from "../../../hooks/useSearchParams";
import "primeflex/primeflex.css";
import { Button } from "primereact/button";
import { Dialog } from "primereact/dialog";
import { TabPanel, TabView } from "primereact/tabview";
import { Toast } from "primereact/toast";
import qs from "qs";
import React, { useEffect, useMemo, useRef, useState } from "react";
import * as Yup from "yup";
import { SolidBooleanField } from "./fields/SolidBooleanField";
import { SolidDateField } from "./fields/SolidDateField";
import { SolidDateTimeField } from "./fields/SolidDateTimeField";
import { SolidDecimalField } from "./fields/SolidDecimalField";
import { SolidIntegerField } from "./fields/SolidIntegerField";
import { SolidJsonField } from "./fields/SolidJsonField";
import { SolidLongTextField } from "./fields/SolidLongTextField";
import { SolidMediaMultipleField } from "./fields/SolidMediaMultipleField";
import { SolidMediaSingleField } from "./fields/SolidMediaSingleField";
import { SolidRelationField } from "./fields/SolidRelationField";
import { SolidRichTextField } from "./fields/SolidRichTextField";
import { SolidSelectionDynamicField } from "./fields/SolidSelectionDynamicField";
import { SolidSelectionStaticField } from "./fields/SolidSelectionStaticField";
import { SolidShortTextField } from "./fields/SolidShortTextField";
import { SolidTimeField } from "./fields/SolidTimeField";
import { getExtensionComponent, getExtensionFunction } from "../../../helpers/registry";
import { SolidPasswordField } from "./fields/SolidPasswordField";
import { SolidEmailField } from "./fields/SolidEmailField";
import { Panel } from "primereact/panel";
import { SolidFormUserViewLayout } from "./SolidFormUserViewLayout";
import Lightbox from "yet-another-react-lightbox";
import Counter from "yet-another-react-lightbox/plugins/counter";
import Download from "yet-another-react-lightbox/plugins/download";
import Video from "yet-another-react-lightbox/plugins/video";
import "yet-another-react-lightbox/styles.css";
import "yet-another-react-lightbox/plugins/counter.css";
import { SolidFormActionHeader } from "./SolidFormActionHeader";
import { SolidFormViewShimmerLoading } from "./SolidFormViewShimmerLoading";
import { hasAnyRole } from "../../../helpers/rolesHelper";
import SolidChatterLocaleTabView from "../locales/SolidChatterLocaleTabView";
import { ConfirmDialog } from "primereact/confirmdialog";
import { SolidXAIIcon } from "../solid-ai/SolidXAIIcon";
import { ERROR_MESSAGES } from "../../../constants/error-messages";
import { useLazyGetMcpUrlQuery, useLazyGetSolidSettingsQuery } from "../../../redux/api/solidSettingsApi";
import { SolidFormFooter } from "./SolidFormFooter";
import { normalizeSolidFormActionPath } from "../../../helpers/routePaths";
import showToast from "../../../helpers/showToast";
var getLayoutFields = function (node) {
    var fields = [];
    if (node.type === "field") {
        fields.push(node);
    }
    if (node.children && Array.isArray(node.children)) {
        for (var _i = 0, _a = node.children; _i < _a.length; _i++) {
            var child = _a[_i];
            fields = fields.concat(getLayoutFields(child));
        }
    }
    return fields;
};
export var getLayoutFieldsAsObject = function (layout) {
    var allFields = layout.flatMap(getLayoutFields);
    return allFields.reduce(function (result, field) {
        if (field.attrs.name) {
            result[field.attrs.name] = __assign({}, field);
        }
        return result;
    }, {});
};
export var getActualFieldMetadata = function (key, solidFieldsMetadata) {
    if (solidFieldsMetadata[key]) {
        return solidFieldsMetadata[key];
    }
    if (key.endsWith("Confirm")) {
        var baseKey = key.slice(0, -"Confirm".length); // Remove "Confirm"
        if (solidFieldsMetadata[baseKey]) {
            return solidFieldsMetadata[baseKey];
        }
    }
    return null; // or handle fallback
};
var fieldFactory = function (type, fieldContext, setLightboxUrls, setOpenLightbox) {
    if (type === 'shortText') {
        return new SolidShortTextField(fieldContext);
    }
    if (type === 'longText') {
        return new SolidLongTextField(fieldContext);
    }
    if (type === 'int' || type === 'bigint') {
        return new SolidIntegerField(fieldContext);
    }
    if (type === 'decimal' || type === 'float') {
        return new SolidDecimalField(fieldContext);
    }
    if (type === 'boolean') {
        return new SolidBooleanField(fieldContext);
    }
    if (type === 'richText') {
        return new SolidRichTextField(fieldContext);
    }
    if (type === 'date') {
        return new SolidDateField(fieldContext);
    }
    if (type === 'datetime') {
        return new SolidDateTimeField(fieldContext);
    }
    if (type === 'time') {
        return new SolidTimeField(fieldContext);
    }
    if (type === 'json') {
        return new SolidJsonField(fieldContext);
    }
    if (type === 'selectionStatic') {
        return new SolidSelectionStaticField(fieldContext);
    }
    if (type === 'selectionDynamic') {
        return new SolidSelectionDynamicField(fieldContext);
    }
    if (type === 'relation') {
        return new SolidRelationField(fieldContext);
    }
    if (type === 'mediaSingle') {
        return new SolidMediaSingleField(fieldContext, setLightboxUrls, setOpenLightbox);
    }
    if (type === 'mediaMultiple') {
        return new SolidMediaMultipleField(fieldContext, setLightboxUrls, setOpenLightbox);
    }
    if (type === 'password') {
        return new SolidPasswordField(fieldContext);
    }
    if (type === 'email') {
        return new SolidEmailField(fieldContext);
    }
    return null;
};
// solidFieldsMetadata={solidFieldsMetadata} solidView={solidView}
var SolidField = function (_a) {
    var formik = _a.formik, field = _a.field, fieldMetadata = _a.fieldMetadata, initialEntityData = _a.initialEntityData, solidFormViewMetaData = _a.solidFormViewMetaData, modelName = _a.modelName, readOnly = _a.readOnly, viewMode = _a.viewMode, onChange = _a.onChange, onBlur = _a.onBlur, parentFieldName = _a.parentFieldName, parentData = _a.parentData, setLightboxUrls = _a.setLightboxUrls, setOpenLightbox = _a.setOpenLightbox, onEmbeddedFormSave = _a.onEmbeddedFormSave;
    var fieldContext = {
        // field metadata - coming from the field-metadata table.
        fieldMetadata: fieldMetadata,
        // field layout - coming from view.layout
        field: field,
        // initial data
        data: initialEntityData,
        // complete form view metadata - this includes layout of the whole form & metadata about all fields in the corresponding model.
        solidFormViewMetaData: solidFormViewMetaData,
        modelName: modelName,
        readOnly: readOnly,
        viewMode: viewMode,
        onChange: onChange,
        onBlur: onBlur
    };
    if (parentData) {
        fieldContext.parentData = parentData;
    }
    if (parentFieldName) {
        fieldContext.parentFieldName = parentFieldName;
    }
    if (onEmbeddedFormSave) {
        fieldContext.onEmbeddedFormSave = onEmbeddedFormSave;
    }
    var solidField = fieldFactory(fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.type, fieldContext, setLightboxUrls, setOpenLightbox);
    return solidField === null || solidField === void 0 ? void 0 : solidField.render(formik);
};
var SolidGroup = function (_a) {
    var children = _a.children, attrs = _a.attrs;
    var className = attrs.className;
    return (_jsxs("div", { className: className, children: [attrs.label && _jsx("p", { children: attrs.label }), _jsx("div", { className: "grid", children: children })] })
    // <div className={className}>
    //     <div className="s_group">
    //         <fieldset>
    //             {attrs.label && <p className="s_group_heading">{attrs.label}</p>}
    //             <div className="grid">{children}</div>
    //         </fieldset>
    //     </div>
    // </div>
    // <div className="formgrid grid">
    //     {children}
    // </div>
    );
};
var SolidRow = function (_a) {
    var children = _a.children, attrs = _a.attrs;
    var className = attrs.className;
    return (
    // <div className={`row ${className}`}>
    //     <div className="s_group">
    //         <fieldset>
    //             {attrs.label && <p className="s_group_heading">{attrs.label}</p>}
    //             <div className="grid">{children}</div>
    //         </fieldset>
    //     </div>
    // </div>
    _jsxs("div", { className: "row ".concat(className), children: [attrs.label && _jsx("p", { children: attrs.label }), _jsx("div", { className: "grid", children: children })] })
    // <div>{children}</div>
    );
};
var SolidColumn = function (_a) {
    var children = _a.children, attrs = _a.attrs;
    var className = attrs.className;
    return (
    // first fieldset ui
    // <div className={`${className}`}>
    //     <div className="s_group">
    //         <fieldset>
    //             {attrs.label && <p className="s_group_heading">{attrs.label}</p>}
    //             <div className="grid">{children}</div>
    //         </fieldset>
    //     </div>
    // </div>
    //second fieldset ui
    // <div className={`${className}`}>
    //     {attrs.label && <p>{attrs.label}</p>}
    //     <div className="grid">{children}</div>
    // </div>
    //figma fieldset ui
    attrs.label ?
        _jsx("div", { className: "".concat(className), children: _jsx(Panel, { header: attrs.label, className: "solid-column-panel", children: _jsx("div", { className: "grid", children: children }) }) })
        :
            _jsx("div", { className: "".concat(className), children: _jsx("div", { className: "grid", children: children }) }));
};
var SolidSheet = function (_a) {
    var children = _a.children;
    return (_jsx("div", { className: "p-fluid p-grid", children: children }));
};
var SolidNotebook = function (_a) {
    var children = _a.children, activeTab = _a.activeTab, embeded = _a.embeded;
    // const childrenArray = children;
    // const childrenArray = React.Children.toArray(children).filter(Boolean);
    var childrenArray = React.Children.toArray(children).filter(function (child) { return !!child; });
    var router = useRouter();
    var pathname = usePathname();
    var searchParams = useSearchParams();
    // Local state to manage active tab in embedded context
    var _b = useState(activeTab), localActiveTab = _b[0], setLocalActiveTab = _b[1];
    var activeIndex = useMemo(function () {
        return childrenArray.findIndex(function (child) {
            return child.key === (embeded ? localActiveTab : activeTab);
        });
    }, [childrenArray, activeTab, localActiveTab, embeded]);
    var handleTabChange = function (e) {
        var selectedChild = childrenArray[e.index];
        var newTabLabel = selectedChild === null || selectedChild === void 0 ? void 0 : selectedChild.key;
        if (newTabLabel) {
            if (!embeded) {
                var queryParams = new URLSearchParams(searchParams.toString());
                queryParams.set('activeTab', newTabLabel);
                router.push("".concat(pathname, "?").concat(queryParams.toString()));
            }
            else {
                // Update the active tab state locally for embedded view
                setLocalActiveTab(newTabLabel);
            }
        }
    };
    return (_jsx("div", { className: "solid-tab-view w-full", children: _jsx(TabView, { activeIndex: activeIndex >= 0 ? activeIndex : 0, onTabChange: handleTabChange, children: childrenArray }) }));
};
var SolidDynamicWidget = function (_a) {
    var widgetName = _a.widgetName, formik = _a.formik, field = _a.field, solidFormViewMetaData = _a.solidFormViewMetaData, solidFormViewData = _a.solidFormViewData;
    var solidView = solidFormViewMetaData.data.solidView;
    var solidFieldsMetadata = solidFormViewMetaData.data.solidFieldsMetadata;
    var DynamicWidget = getExtensionComponent(widgetName);
    var widgetProps = {
        formData: formik.values,
        field: field,
        fieldsMetadata: solidFieldsMetadata,
        viewMetadata: solidView,
        formViewData: solidFormViewData
    };
    return (_jsx("div", { className: "solid-tab-view w-full", children: DynamicWidget && _jsx(DynamicWidget, __assign({}, widgetProps)) }));
};
var SolidPage = function (_a) {
    var attrs = _a.attrs, children = _a.children, key = _a.key, formik = _a.formik, fields = _a.fields;
    var fieldsName = fields.map(function (f) { return f.attrs.name; });
    var errorCount = formik.submitCount > 0 ? fieldsName.filter(function (name) { return !!formik.errors[name]; }).length : 0;
    var label = (_jsxs("span", { style: { color: errorCount > 0 ? 'red' : 'inherit' }, children: [attrs.label, errorCount > 0 && " (".concat(errorCount, ")")] }));
    return (_jsx(TabPanel, { header: label, children: _jsx("div", { className: "p-fluid", children: children }) }, key));
};
// Original code...
// const addLevelToGroups = (layout: any, level = 1) => {
//     return layout.map((element: any) => {
//         if (element.type === "group") {
//             // Add the level to the group's attrs
//             element.attrs = {
//                 ...element.attrs,
//                 level: level,  // Add level information to the attrs
//                 // className: level === 1 ? 'col-12' : 'col-6',  // Assign className based on level
//             };
//         } else {
//             element.children = addLevelToGroups(element.children, level);
//         }
//         // If the element has children, recursively apply this logic
//         // if (element.children) {
//         //     element.children = addLevelToGroups(element.children, level + 1);
//         // }
//         return element;
//     });
// };
// Fix for immutable objects. 
// const addLevelToGroups = (layout: any, level = 1): any[] => {
//     return layout.map((element: any) => {
//         // Create a new object for immutability
//         const updatedElement = { ...element };
//         if (updatedElement.type === "group") {
//             // Add level to attrs, creating a new object for immutability
//             updatedElement.attrs = {
//                 ...updatedElement.attrs,
//                 level: level, // Add level information
//             };
//         }
//         // If the element has children, recursively apply this logic
//         if (updatedElement.children) {
//             updatedElement.children = addLevelToGroups(updatedElement.children, level + 1);
//         }
//         return updatedElement;
//     });
// };
var SolidFormView = function (params) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
    var _r = useSession(), session = _r.data, status = _r.status;
    var user = session === null || session === void 0 ? void 0 : session.user;
    var pathname = usePathname();
    var router = useRouter();
    var toast = useRef(null);
    var searchParams = useSearchParams();
    var _s = useState(false), confirmVisible = _s[0], setConfirmVisible = _s[1];
    var confirmResolveRef = useRef();
    var _t = useState(false), redirectToList = _t[0], setRedirectToList = _t[1];
    var _u = useState('en'), selectedLocale = _u[0], setSelectedLocale = _u[1];
    var _v = useState(null), defaultEntityLocaleId = _v[0], setDefaultEntityLocaleId = _v[1];
    var _w = useState(false), isDeleteDialogVisible = _w[0], setDeleteDialogVisible = _w[1];
    var _x = useState(false), isLayoutDialogVisible = _x[0], setLayoutDialogVisible = _x[1];
    var _y = useState(null), published = _y[0], setPublished = _y[1];
    var _z = useState([]), actionsAllowed = _z[0], setActionsAllowed = _z[1];
    var _0 = useState(params.embeded === true ? "edit" : "view"), viewMode = _0[0], setViewMode = _0[1];
    var _1 = useState(false), createMode = _1[0], setCreateMode = _1[1];
    var _2 = useState(false), openLightbox = _2[0], setOpenLightbox = _2[1];
    var _3 = useState([]), lightboxUrls = _3[0], setLightboxUrls = _3[1];
    var _4 = useState(false), isShowChatter = _4[0], setShowChatter = _4[1];
    var _5 = useState(380), chatterLocaleWidth = _5[0], setChatterLocaleWidth = _5[1]; // default width
    var _6 = useState(false), isResizingChatterLocale = _6[0], setIsResizingChatterLocale = _6[1];
    var _7 = useState(false), isSubmitting = _7[0], setIsSubmitting = _7[1];
    var _8 = useState(""), solidWorkflowFieldValue = _8[0], setSolidWorkflowFieldValue = _8[1];
    var _9 = useState(1), defaultTabViewOptionIndex = _9[0], setDefaultTabViewOptionIndex = _9[1];
    var errorFields = [];
    var triggerCheckIfPermissionExists = useLazyCheckIfPermissionExistsQuery()[0];
    var _10 = useState(null), mcpUrl = _10[0], setMcpUrl = _10[1];
    var getMcpUrl = useLazyGetMcpUrlQuery()[0];
    // when rendering the form view we will optionally get action params...
    // these we can bubble up in the event that is being raised onFormLayoutLoad, onFormDataLoad & onFormLoad
    var actionName = searchParams.get('actionName');
    var actionType = searchParams.get('actionType');
    var actionContext = searchParams.get('actionContext');
    var _11 = useLazyGetSolidSettingsQuery(), trigger = _11[0], solidSettingsData = _11[1].data;
    useEffect(function () {
        trigger(""); // Fetch settings on mount
    }, []);
    useEffect(function () {
        var _a, _b;
        if (((_a = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _a === void 0 ? void 0 : _a.mcpEnabled) && ((_b = solidSettingsData === null || solidSettingsData === void 0 ? void 0 : solidSettingsData.data) === null || _b === void 0 ? void 0 : _b.mcpServerUrl)) {
            enableSolidXAiPanel();
        }
    }, [solidSettingsData]);
    var enableSolidXAiPanel = function () { return __awaiter(void 0, void 0, void 0, function () {
        var queryData, queryString, response, error_1;
        var _a, _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _c.trys.push([0, 2, , 3]);
                    queryData = {
                        showHeader: "false",
                        inListView: "false"
                    };
                    queryString = qs.stringify(__assign({}, queryData), { encodeValuesOnly: true });
                    return [4 /*yield*/, getMcpUrl(queryString).unwrap()];
                case 1:
                    response = _c.sent();
                    console.log("response", response);
                    if (response && ((_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.mcpUrl)) {
                        setMcpUrl((_b = response === null || response === void 0 ? void 0 : response.data) === null || _b === void 0 ? void 0 : _b.mcpUrl);
                    }
                    return [3 /*break*/, 3];
                case 2:
                    error_1 = _c.sent();
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/];
            }
        });
    }); };
    var op = useRef(null);
    useEffect(function () {
        var stored = localStorage.getItem('chatter_locale_width');
        if (stored) {
            setChatterLocaleWidth(parseInt(stored, 10));
        }
    }, []);
    useEffect(function () {
        var handleMouseMove = function (e) {
            if (!isResizingChatterLocale)
                return;
            var newWidth = window.innerWidth - e.clientX;
            var clampedWidth = Math.max(280, Math.min(newWidth, 700));
            setChatterLocaleWidth(clampedWidth);
            localStorage.setItem('chatter_locale_width', clampedWidth.toString());
        };
        var handleMouseUp = function () {
            setIsResizingChatterLocale(false);
        };
        if (isResizingChatterLocale) {
            window.addEventListener('mousemove', handleMouseMove);
            window.addEventListener('mouseup', handleMouseUp);
        }
        return function () {
            window.removeEventListener('mousemove', handleMouseMove);
            window.removeEventListener('mouseup', handleMouseUp);
        };
    }, [isResizingChatterLocale]);
    useEffect(function () {
        var mode = searchParams.get('viewMode');
        var locale = searchParams.get('locale');
        var defaultEntityLocaleIdn = searchParams.get('defaultEntityLocaleId');
        if (params.id === 'new' && !locale) {
            setViewMode('edit');
            setCreateMode(true);
            return;
        }
        if (locale) {
            setSelectedLocale(locale);
        }
        if (defaultEntityLocaleIdn) {
            setDefaultEntityLocaleId(defaultEntityLocaleIdn);
        }
        // Set the viewMode based on the URL
        if (mode === 'view' || mode === 'edit') {
            setViewMode(mode);
        }
        else {
            setViewMode('view'); // Default to 'view' if no valid mode is provided
        }
    }, [searchParams, params.id]);
    // function that updates view mode 
    var updateViewMode = function (newMode) {
        setViewMode(newMode);
        var params = new URLSearchParams(searchParams.toString());
        params.set("viewMode", newMode);
        router.push("".concat(pathname, "?").concat(params.toString()));
    };
    useEffect(function () {
        var fetchPermissions = function () { return __awaiter(void 0, void 0, void 0, function () {
            var permissionNames, queryData, queryString, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!params.modelName) return [3 /*break*/, 2];
                        permissionNames = [
                            permissionExpression(params.modelName, 'create'),
                            permissionExpression(params.modelName, 'delete'),
                            permissionExpression(params.modelName, 'update'),
                            permissionExpression(params.modelName, 'findOne'),
                            permissionExpression(params.modelName, 'publish'),
                            permissionExpression(params.modelName, 'unpublish'),
                            permissionExpression('chatterMessage', 'findMany')
                        ];
                        queryData = {
                            permissionNames: permissionNames
                        };
                        queryString = qs.stringify(queryData, {
                            encodeValuesOnly: true
                        });
                        return [4 /*yield*/, triggerCheckIfPermissionExists(queryString)];
                    case 1:
                        response = _a.sent();
                        setActionsAllowed(response.data.data);
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        }); };
        fetchPermissions();
    }, [params.modelName]);
    // Create the RTK slices for this entitor (const id of fieldValue) {
    //     if (!isInt(id)) {
    //         errors.push({ field: this.fieldMetadata.name, error: `Invalid ids in ${ commandFieldName } ` });
    //     }
    // }y
    var entityApi = createSolidEntityApi(params.modelName);
    var useCreateSolidEntityMutation = entityApi.useCreateSolidEntityMutation, useDeleteSolidEntityMutation = entityApi.useDeleteSolidEntityMutation, useGetSolidEntityByIdQuery = entityApi.useGetSolidEntityByIdQuery, useUpdateSolidEntityMutation = entityApi.useUpdateSolidEntityMutation, usePatchUpdateSolidEntityMutation = entityApi.usePatchUpdateSolidEntityMutation, usePublishSolidEntityMutation = entityApi.usePublishSolidEntityMutation, useUnpublishSolidEntityMutation = entityApi.useUnpublishSolidEntityMutation;
    var _12 = useCreateSolidEntityMutation(), createEntity = _12[0], _13 = _12[1], isEntityCreateSuccess = _13.isSuccess, isEntityCreateError = _13.isError, entityCreateError = _13.error;
    var _14 = useUpdateSolidEntityMutation(), updateEntity = _14[0], _15 = _14[1], isEntityUpdateSuceess = _15.isSuccess, isEntityUpdateError = _15.isError, entityUpdateError = _15.error;
    var _16 = useDeleteSolidEntityMutation(), deleteEntity = _16[0], _17 = _16[1], isEntityDeleteSuceess = _17.isSuccess, isEntityDeleteError = _17.isError, entityDeleteError = _17.error;
    var _18 = usePatchUpdateSolidEntityMutation(), patchEntity = _18[0], _19 = _18[1], isEntityPatchSuceess = _19.isSuccess, isEntityPatchError = _19.isError, entityPatchError = _19.error;
    var _20 = usePublishSolidEntityMutation(), publishSolidEntity = _20[0], _21 = _20[1], isEntityPublishedSuccess = _21.isSuccess, isEntityPublishedError = _21.isError, entityPublishedError = _21.error;
    var _22 = useUnpublishSolidEntityMutation(), unpublishSolidEntity = _22[0], _23 = _22[1], isEntityUnpublishedSuccess = _23.isSuccess, isEntityUnpublishedError = _23.isError, entityUnpublishedError = _23.error;
    // - - - - - - - - - - - -- - - - - - - - - - - - METADATA here
    // Get the form view layout & metadata first. 
    var formViewMetaDataQs = qs.stringify(__assign(__assign({}, params), { viewType: 'form', defaultEntityLocaleId: defaultEntityLocaleId }), {
        encodeValuesOnly: true,
    });
    var _24 = useState({}), formViewMetaData = _24[0], setFormViewMetaData = _24[1];
    var _25 = useState(null), formViewLayout = _25[0], setFormViewLayout = _25[1];
    var _26 = useGetSolidViewLayoutQuery(formViewMetaDataQs), solidFormViewMetaData = _26.data, solidFormViewMetaDataIsLoading = _26.isLoading;
    var _27 = useState(true), refreshChatterMessage = _27[0], setRefreshChatterMessage = _27[1];
    useEffect(function () {
        if (isEntityCreateSuccess == true ||
            isEntityUpdateSuceess == true ||
            isEntityDeleteSuceess == true ||
            isEntityPatchSuceess == true ||
            isEntityPublishedSuccess == true ||
            isEntityUnpublishedSuccess == true) {
            setRefreshChatterMessage(true);
            if (params.embeded == true && params.onEmbeddedFormSave) {
                params.onEmbeddedFormSave();
            }
            // Close The pop in case the form is used in embeded form
            if (params.embeded == true) {
                params.handlePopupClose();
            }
            if (redirectToList === true) {
                if (params.redirectToPath) {
                    router.push(params.redirectToPath);
                    window.location.reload();
                }
                else {
                    var segments = pathname.split('/').filter(Boolean); // Split and filter empty segments
                    var newPath = '/' + segments.slice(0, -2).join('/') + '/list'; // Remove last segment and add "/all"
                    router.push(newPath);
                }
            }
        }
    }, [isEntityCreateSuccess, isEntityUpdateSuceess, isEntityDeleteSuceess, isEntityPatchSuceess, isEntityPublishedSuccess, isEntityUnpublishedSuccess]);
    useEffect(function () {
        var _a, _b, _c, _d, _e;
        if ((_c = (_b = (_a = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _a === void 0 ? void 0 : _a.solidView) === null || _b === void 0 ? void 0 : _b.model) === null || _c === void 0 ? void 0 : _c.internationalisation) {
            setDefaultTabViewOptionIndex(0);
            var matchedLocale = (_e = (_d = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _d === void 0 ? void 0 : _d.applicableLocales) === null || _e === void 0 ? void 0 : _e.find(function (x) { return x.isDefault === 'yes'; });
            //this is to attach default locale when adding data in popup view where relations exists
            if (!selectedLocale && matchedLocale && !searchParams.get('locale')) {
                setSelectedLocale(matchedLocale.locale);
            }
        }
    }, [params.modelName, solidFormViewMetaData]);
    function isFetchBaseQueryErrorWithErrorResponse(error) {
        return error && typeof error === 'object' && 'data' in error && 'message' in error.data;
    }
    useEffect(function () {
        var handleError = function (errorToast) {
            var _a;
            var errorMessage = [ERROR_MESSAGES.ERROR_OCCURED];
            if (isFetchBaseQueryErrorWithErrorResponse(errorToast)) {
                errorMessage = errorToast.data.message;
            }
            else {
                errorMessage = [ERROR_MESSAGES.SOMETHING_WRONG];
            }
            (_a = toast.current) === null || _a === void 0 ? void 0 : _a.show({
                severity: 'error',
                summary: 'Error',
                detail: errorMessage,
                sticky: true,
                //@ts-ignore
                content: function () { return (_jsx("div", { className: "flex flex-column align-items-left", style: { flex: "1" }, children: Array.isArray(errorMessage) ? (errorMessage.map(function (message, index) { return (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: message.trim() }) }, index)); })) : (_jsx("div", { className: "flex align-items-center gap-2", children: _jsx("span", { className: "font-bold text-900", children: errorMessage === null || errorMessage === void 0 ? void 0 : errorMessage.trim() }) })) })); },
            });
        };
        // Check and handle errors from each API operation
        if (isEntityCreateError) {
            handleError(entityCreateError);
        }
        else if (isEntityDeleteError) {
            handleError(entityDeleteError);
        }
        else if (isEntityUpdateError) {
            handleError(entityUpdateError);
        }
        else if (isEntityPatchError) {
            handleError(entityPatchError);
        }
        else if (isEntityPublishedError) {
            handleError(entityPublishedError);
        }
        else if (isEntityUnpublishedError) {
            handleError(entityUnpublishedError);
        }
    }, [
        isEntityCreateError,
        isEntityDeleteError,
        isEntityUpdateError,
        isEntityPatchError,
        isEntityPublishedError,
        isEntityUnpublishedError
    ]);
    var confirmDialogWithPromise = function () {
        return new Promise(function (resolve) {
            confirmResolveRef.current = resolve;
            setConfirmVisible(true);
        });
    };
    var onFormikSubmit = function (values) { return __awaiter(void 0, void 0, void 0, function () {
        var solidView, solidFieldsMetadata, layoutFieldsObj, formData_1, solidWorkflowField, result, baseFormPath, queryParams, result, err_1;
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t;
        return __generator(this, function (_u) {
            switch (_u.label) {
                case 0:
                    solidView = solidFormViewMetaData.data.solidView;
                    solidFieldsMetadata = solidFormViewMetaData.data.solidFieldsMetadata;
                    layoutFieldsObj = getLayoutFieldsAsObject([formViewLayout]);
                    setIsSubmitting(true);
                    _u.label = 1;
                case 1:
                    _u.trys.push([1, 7, 8, 9]);
                    formData_1 = new FormData();
                    // Iterate through the keys in the values object
                    Object.entries(values).forEach(function (_a) {
                        var key = _a[0], value = _a[1];
                        var fieldMetadata = solidFieldsMetadata[key];
                        //  const fieldMetadata = getActualFieldMetadata(key, solidFieldsMetadata);
                        var fieldContext = {
                            fieldMetadata: fieldMetadata,
                            field: layoutFieldsObj[key],
                            data: initialEntityData,
                            solidFormViewMetaData: solidFormViewMetaData,
                            modelName: params.modelName
                        };
                        var solidField = fieldFactory(fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.type, fieldContext);
                        // Append each field to the FormData
                        if (value !== undefined && value !== null && solidField) {
                            solidField.updateFormData(value, formData_1);
                        }
                        if (value !== undefined && value !== null && key.endsWith("Confirm")) {
                            formData_1.append(key, String(value));
                        }
                    });
                    solidWorkflowField = (_d = (_c = (_b = (_a = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _a === void 0 ? void 0 : _a.solidView) === null || _b === void 0 ? void 0 : _b.layout) === null || _c === void 0 ? void 0 : _c.attrs) === null || _d === void 0 ? void 0 : _d.workflowField;
                    if (params.id !== "new") {
                        if ((_e = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _e === void 0 ? void 0 : _e.solidFormViewWorkflowData) {
                            if (((_h = (_g = (_f = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _f === void 0 ? void 0 : _f.solidFieldsMetadata) === null || _g === void 0 ? void 0 : _g[solidWorkflowField]) === null || _h === void 0 ? void 0 : _h.type) === "selectionStatic") {
                                formData_1.append(solidWorkflowField, solidWorkflowFieldValue);
                            }
                            if (((_l = (_k = (_j = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _j === void 0 ? void 0 : _j.solidFieldsMetadata) === null || _k === void 0 ? void 0 : _k[solidWorkflowField]) === null || _l === void 0 ? void 0 : _l.type) === "many-to-one") {
                                formData_1.append("".concat(solidWorkflowField, "Id"), solidWorkflowFieldValue);
                            }
                        }
                    }
                    if ((_p = (_o = (_m = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _m === void 0 ? void 0 : _m.solidView) === null || _o === void 0 ? void 0 : _o.model) === null || _p === void 0 ? void 0 : _p.internationalisation) {
                        if (selectedLocale && !formData_1.has('localeName')) {
                            formData_1.append('localeName', selectedLocale);
                        }
                        if (defaultEntityLocaleId) {
                            formData_1.append('defaultEntityLocaleId', defaultEntityLocaleId.toString());
                        }
                    }
                    if ((_s = (_r = (_q = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _q === void 0 ? void 0 : _q.solidView) === null || _r === void 0 ? void 0 : _r.model) === null || _s === void 0 ? void 0 : _s.draftPublishWorkflow) {
                        if (published) {
                            formData_1.append('publishedAt', published);
                        }
                    }
                    if (!(params.inlineCreateAutoSave === true)) return [3 /*break*/, 2];
                    params.customCreateHandler(formData_1);
                    return [3 /*break*/, 6];
                case 2:
                    if (!(params.id === 'new')) return [3 /*break*/, 4];
                    return [4 /*yield*/, createEntity(formData_1).unwrap()];
                case 3:
                    result = _u.sent();
                    showToast(toast, "success", ERROR_MESSAGES.FORM_SAVED, ERROR_MESSAGES.FORM_SAVED_SUCCESSFULLY);
                    // if (!params.embeded && result?.data?.id) {
                    //     const newPathname = pathname.replace(/new$/, result.data.id);
                    //     const params = new URLSearchParams(searchParams.toString());
                    //     params.set("viewMode", "view");
                    //     const updatedUrl = `${newPathname}?${params.toString()}`;
                    //     await router.push(updatedUrl, { scroll: false });
                    //     setViewMode("view")
                    // }
                    if (!params.embeded) {
                        baseFormPath = normalizeSolidFormActionPath(pathname, "form");
                        queryParams = new URLSearchParams(searchParams.toString());
                        queryParams.set("viewMode", "view");
                        router.push("".concat(baseFormPath, "/").concat((_t = result === null || result === void 0 ? void 0 : result.data) === null || _t === void 0 ? void 0 : _t.id, "?").concat(queryParams.toString()));
                        setViewMode("view");
                    }
                    return [2 /*return*/, result];
                case 4: return [4 /*yield*/, updateEntity({ id: +params.id, data: formData_1 }).unwrap()];
                case 5:
                    result = _u.sent();
                    // const result = await updateEntity({ id: +params.id, data: formData }).unwrap();
                    if (!params.embeded) {
                        showToast(toast, "success", ERROR_MESSAGES.FORM_UPDATE, ERROR_MESSAGES.FORM_UPDATE_SUCCESSFULLY);
                        if ((result === null || result === void 0 ? void 0 : result.statusCode) === 200) {
                            updateViewMode("view");
                        }
                    }
                    return [2 /*return*/, result];
                case 6: return [3 /*break*/, 9];
                case 7:
                    err_1 = _u.sent();
                    console.error(ERROR_MESSAGES.ENTITY_FAILED, err_1);
                    return [3 /*break*/, 9];
                case 8:
                    setIsSubmitting(false);
                    return [7 /*endfinally*/];
                case 9: return [2 /*return*/];
            }
        });
    }); };
    var showFieldError = function () {
        if ((errorFields === null || errorFields === void 0 ? void 0 : errorFields.length) === 0)
            return;
        errorFields.forEach(function (error) {
            var _a;
            (_a = toast === null || toast === void 0 ? void 0 : toast.current) === null || _a === void 0 ? void 0 : _a.show({
                severity: "error",
                summary: "Metadata Error",
                detail: error,
                life: 3000,
            });
        });
        // errorFields.length = 0;
    };
    useEffect(function () {
        if ((errorFields === null || errorFields === void 0 ? void 0 : errorFields.length) > 0) {
            showFieldError();
        }
    }, [errorFields]);
    // - - - - - - - - - - - -- - - - - - - - - - - - DATA here
    // Fetch the actual data here. 
    // This is the initial value of this form, will come from an API call in the case of edit. 
    var layoutFields = [];
    var toPopulate = [];
    var toPopulateMedia = [];
    if (solidFormViewMetaData && formViewLayout) {
        var solidView = solidFormViewMetaData.data.solidView;
        var solidFieldsMetadata = solidFormViewMetaData.data.solidFieldsMetadata;
        layoutFields = [formViewLayout].flatMap(getLayoutFields);
        for (var i = 0; i < (layoutFields === null || layoutFields === void 0 ? void 0 : layoutFields.length); i++) {
            var formLayoutField = layoutFields[i];
            var fieldMetadata = solidFieldsMetadata[formLayoutField.attrs.name];
            if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.type) === 'relation') {
                toPopulate.push(fieldMetadata.name);
            }
            if ((fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.type) === 'mediaSingle' || (fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.type) === 'mediaMultiple') {
                toPopulateMedia.push(fieldMetadata.name);
            }
        }
        if (((_a = formViewLayout.attrs) === null || _a === void 0 ? void 0 : _a.workflowField) && ((_b = solidFieldsMetadata[formViewLayout.attrs.workflowField]) === null || _b === void 0 ? void 0 : _b.type) === 'relation' && ((_c = solidFieldsMetadata[formViewLayout.attrs.workflowField]) === null || _c === void 0 ? void 0 : _c.relationType) === 'many-to-one') {
            toPopulate.push(solidFieldsMetadata[formViewLayout.attrs.workflowField].name);
        }
    }
    // TODO: Possible optimisation here, we are firing 2 queries to load the form view data object. 
    // once without populate & populateMedia and then again once after that.
    var formViewDataQs = qs.stringify({ populate: toPopulate, populateMedia: toPopulateMedia }, {
        encodeValuesOnly: true,
    });
    var _28 = useState({}), initialEntityData = _28[0], setInitialEntityData = _28[1];
    var _29 = useGetSolidEntityByIdQuery({ id: params.id, qs: formViewDataQs }, {
        skip: params.id === 'new'
    }), solidFormViewData = _29.data, solidFormViewDataIsLoading = _29.isLoading, refetchSolidFormViewData = _29.refetch;
    useEffect(function () {
        if (params.id !== 'new') {
            refetchSolidFormViewData();
        }
    }, [formViewDataQs]);
    useEffect(function () {
        var _a, _b, _c;
        if (solidFormViewMetaData) {
            if (params.customLayout) {
                setFormViewLayout(params.customLayout);
            }
            else {
                setFormViewLayout((_b = (_a = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _a === void 0 ? void 0 : _a.solidView) === null || _b === void 0 ? void 0 : _b.layout);
            }
            setPublished((_c = solidFormViewData === null || solidFormViewData === void 0 ? void 0 : solidFormViewData.data) === null || _c === void 0 ? void 0 : _c.publishedAt);
            setFormViewMetaData(solidFormViewMetaData);
        }
    }, [solidFormViewMetaData]);
    // useEffect(() => {
    //     const handleOnFormLayoutLoadEvent = async () => {
    //         if (solidFormViewMetaData) {
    //             // let formLayout = solidFormViewMetaData;
    //             // let customLayout = params?.customLayout;
    //             const onFormLayoutLoadHandlerExtensionFunction = solidFormViewMetaData?.data?.solidView?.layout?.onFormLayoutLoad;
    //             // let dynamicExtensionFunction = null;
    //             let formLayout = solidFormViewMetaData?.data?.solidView?.layout;
    //             if (params.customLayout) {
    //                 formLayout = params.customLayout;
    //             }
    //             const event: SolidLoadForm = {
    //                 parentData: params?.parentData,
    //                 fieldsMetadata: solidFormViewMetaData,
    //                 formData: solidFormViewData?.data,
    //                 type: 'onFormLayoutLoad',
    //                 viewMetadata: solidFormViewMetaData?.data?.solidView,
    //                 formViewLayout: formLayout,
    //                 queryParams: {
    //                     actionName,
    //                     actionType,
    //                     actionContext
    //                 }
    //             }
    //             if (onFormLayoutLoadHandlerExtensionFunction) {
    //                 const dynamicExtensionFunction = getExtensionFunction(onFormLayoutLoadHandlerExtensionFunction);
    //                 if (dynamicExtensionFunction) {
    //                     try {
    //                         const updatedFormLayout: SolidUiEventResponse = await dynamicExtensionFunction(event);
    //                         if (updatedFormLayout && updatedFormLayout?.layoutChanged && updatedFormLayout?.newLayout) {
    //                             setFormViewLayout(updatedFormLayout.newLayout);
    //                             // const newFormLayout = {
    //                             //     ...formLayout,
    //                             //     data: {
    //                             //         ...formLayout.data,
    //                             //         solidView: {
    //                             //             ...formLayout.data.solidView,
    //                             //             layout: updatedFormLayout.newLayout
    //                             //         }
    //                             //     }
    //                             // };
    //                             // formLayout = newFormLayout;
    //                             // customLayout = updatedFormLayout.newLayout;
    //                         }
    //                     } catch (error) {
    //                         console.error(ERROR_MESSAGES.DYNAMIC_FUNCTION_ERROR, error);
    //                     }
    //                 }
    //             }
    //             // setFormViewMetaData(formLayout);
    //             // if (params.customLayout) {
    //             //     setFormViewLayout(customLayout);
    //             // } else {
    //             //     setFormViewLayout(formLayout.data.solidView.layout);
    //             // }
    //         }
    //     };
    //     const handleOnFormDataLoadEvent = async () => {
    //         const onFormDataLoadHandlerExtensionFunction = solidFormViewMetaData?.data?.solidView?.layout?.onFormDataLoad;
    //         // let dynamicExtensionFunction = null;
    //         let formViewData = solidFormViewData?.data;
    //         let formLayout = solidFormViewMetaData?.data?.solidView?.layout;
    //         if (params.customLayout) {
    //             formLayout = params.customLayout;
    //         }
    //         const event: SolidLoadForm = {
    //             fieldsMetadata: solidFormViewMetaData,
    //             formData: solidFormViewData?.data,
    //             type: "onFormDataLoad",
    //             viewMetadata: solidFormViewMetaData?.data?.solidView,
    //             formViewLayout: formLayout,
    //             queryParams: {
    //                 actionName,
    //                 actionType,
    //                 actionContext
    //             }
    //         };
    //         if (onFormDataLoadHandlerExtensionFunction) {
    //             const dynamicExtensionFunction = getExtensionFunction(onFormDataLoadHandlerExtensionFunction);
    //             if (dynamicExtensionFunction) {
    //                 const updatedFormData: SolidUiEventResponse = await dynamicExtensionFunction(event);
    //                 if (updatedFormData && updatedFormData?.dataChanged && updatedFormData?.newFormData) {
    //                     formViewData = updatedFormData.newFormData;
    //                 }
    //             }
    //             if (formViewData) {
    //                 setInitialEntityData(formViewData);
    //             }
    //         }
    //     };
    //     const handleOnFormLoadEvent = async () => {
    //         const onFormLoadHandlerExtensionFunction = solidFormViewMetaData?.data?.solidView?.layout?.onFormLoad;
    //         // let dynamicExtensionFunction = null;
    //         let localFormViewMetadata = solidFormViewMetaData;
    //         // let customLayout = params?.customLayout;
    //         let formViewData = solidFormViewData?.data;
    //         let formLayout = solidFormViewMetaData?.data?.solidView?.layout;
    //         if (params.customLayout) {
    //             formLayout = params.customLayout;
    //         }
    //         const event: SolidLoadForm = {
    //             parentData: params?.parentData,
    //             fieldsMetadata: solidFormViewMetaData,
    //             formData: solidFormViewData?.data,
    //             type: 'onFormLoad',
    //             viewMetadata: solidFormViewMetaData?.data?.solidView,
    //             formViewLayout: formViewLayout,
    //             queryParams: {
    //                 actionName,
    //                 actionType,
    //                 actionContext
    //             }
    //         };
    //         if (onFormLoadHandlerExtensionFunction) {
    //             const dynamicExtensionFunction = getExtensionFunction(onFormLoadHandlerExtensionFunction);
    //             if (dynamicExtensionFunction) {
    //                 try {
    //                     const result: SolidUiEventResponse = await dynamicExtensionFunction(event);
    //                     if (result && result?.layoutChanged && result?.newLayout) {
    //                         // const newLocalFormViewMetadata = {
    //                         //     ...localFormViewMetadata,
    //                         //     data: {
    //                         //         ...localFormViewMetadata.data,
    //                         //         solidView: {
    //                         //             ...localFormViewMetadata.data.solidView,
    //                         //             layout: result.newLayout
    //                         //         }
    //                         //     }
    //                         // };
    //                         // localFormViewMetadata = newLocalFormViewMetadata;
    //                         // customLayout = result.newLayout;
    //                         // setFormViewMetaData(localFormViewMetadata);
    //                         setFormViewLayout(result.newLayout);
    //                         // if (params.customLayout) {
    //                         //     setFormViewLayout(customLayout);
    //                         // } else {
    //                         //     setFormViewLayout(localFormViewMetadata.data.solidView.layout);
    //                         // }
    //                     }
    //                     if (result && result?.dataChanged && result?.newFormData) {
    //                         formViewData = result.newFormData;
    //                         setInitialEntityData(formViewData);
    //                     }
    //                 } catch (error) {
    //                     console.error(ERROR_MESSAGES.ON_FORM_LOAD, error);
    //                 }
    //             }
    //         }
    //     };
    //     handleOnFormLayoutLoadEvent();
    //     handleOnFormDataLoadEvent();
    //     handleOnFormLoadEvent();
    // }, [solidFormViewMetaData, solidFormViewData]);
    useEffect(function () {
        var runFormEvents = function () { return __awaiter(void 0, void 0, void 0, function () {
            var workingLayout, workingFormData, baseEvent, onFormLayoutLoadFn, fn, result, e_1, onFormDataLoadFn, fn, result, e_2, onFormLoadFn, fn, result, e_3;
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
            return __generator(this, function (_p) {
                switch (_p.label) {
                    case 0:
                        if (!solidFormViewMetaData)
                            return [2 /*return*/];
                        workingLayout = (_a = params.customLayout) !== null && _a !== void 0 ? _a : (_c = (_b = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _b === void 0 ? void 0 : _b.solidView) === null || _c === void 0 ? void 0 : _c.layout;
                        workingFormData = solidFormViewData === null || solidFormViewData === void 0 ? void 0 : solidFormViewData.data;
                        baseEvent = {
                            parentData: params === null || params === void 0 ? void 0 : params.parentData,
                            fieldsMetadata: solidFormViewMetaData,
                            viewMetadata: (_d = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _d === void 0 ? void 0 : _d.solidView,
                            queryParams: {
                                actionName: actionName,
                                actionType: actionType,
                                actionContext: actionContext,
                            },
                        };
                        onFormLayoutLoadFn = (_g = (_f = (_e = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _e === void 0 ? void 0 : _e.solidView) === null || _f === void 0 ? void 0 : _f.layout) === null || _g === void 0 ? void 0 : _g.onFormLayoutLoad;
                        if (!onFormLayoutLoadFn) return [3 /*break*/, 4];
                        fn = getExtensionFunction(onFormLayoutLoadFn);
                        if (!fn) return [3 /*break*/, 4];
                        _p.label = 1;
                    case 1:
                        _p.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, fn(__assign(__assign({}, baseEvent), { type: "onFormLayoutLoad", formData: workingFormData, formViewLayout: workingLayout }))];
                    case 2:
                        result = _p.sent();
                        if ((result === null || result === void 0 ? void 0 : result.layoutChanged) && (result === null || result === void 0 ? void 0 : result.newLayout)) {
                            workingLayout = result.newLayout;
                        }
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _p.sent();
                        console.error(ERROR_MESSAGES.DYNAMIC_FUNCTION_ERROR, e_1);
                        return [3 /*break*/, 4];
                    case 4:
                        onFormDataLoadFn = (_k = (_j = (_h = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _h === void 0 ? void 0 : _h.solidView) === null || _j === void 0 ? void 0 : _j.layout) === null || _k === void 0 ? void 0 : _k.onFormDataLoad;
                        if (!onFormDataLoadFn) return [3 /*break*/, 8];
                        fn = getExtensionFunction(onFormDataLoadFn);
                        if (!fn) return [3 /*break*/, 8];
                        _p.label = 5;
                    case 5:
                        _p.trys.push([5, 7, , 8]);
                        return [4 /*yield*/, fn(__assign(__assign({}, baseEvent), { type: "onFormDataLoad", formData: workingFormData, formViewLayout: workingLayout }))];
                    case 6:
                        result = _p.sent();
                        if ((result === null || result === void 0 ? void 0 : result.dataChanged) && (result === null || result === void 0 ? void 0 : result.newFormData)) {
                            workingFormData = result.newFormData;
                        }
                        return [3 /*break*/, 8];
                    case 7:
                        e_2 = _p.sent();
                        console.error(ERROR_MESSAGES.DYNAMIC_FUNCTION_ERROR, e_2);
                        return [3 /*break*/, 8];
                    case 8:
                        onFormLoadFn = (_o = (_m = (_l = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _l === void 0 ? void 0 : _l.solidView) === null || _m === void 0 ? void 0 : _m.layout) === null || _o === void 0 ? void 0 : _o.onFormLoad;
                        if (!onFormLoadFn) return [3 /*break*/, 12];
                        fn = getExtensionFunction(onFormLoadFn);
                        if (!fn) return [3 /*break*/, 12];
                        _p.label = 9;
                    case 9:
                        _p.trys.push([9, 11, , 12]);
                        return [4 /*yield*/, fn(__assign(__assign({}, baseEvent), { type: "onFormLoad", formData: workingFormData, formViewLayout: workingLayout }))];
                    case 10:
                        result = _p.sent();
                        if ((result === null || result === void 0 ? void 0 : result.layoutChanged) && (result === null || result === void 0 ? void 0 : result.newLayout)) {
                            workingLayout = result.newLayout;
                        }
                        if ((result === null || result === void 0 ? void 0 : result.dataChanged) && (result === null || result === void 0 ? void 0 : result.newFormData)) {
                            workingFormData = result.newFormData;
                        }
                        return [3 /*break*/, 12];
                    case 11:
                        e_3 = _p.sent();
                        console.error(ERROR_MESSAGES.ON_FORM_LOAD, e_3);
                        return [3 /*break*/, 12];
                    case 12:
                        /** ----------------------------
                         * 5. Commit once to React state
                         * ----------------------------- */
                        if (workingLayout) {
                            setFormViewLayout(workingLayout);
                        }
                        if (workingFormData) {
                            setInitialEntityData(workingFormData);
                        }
                        return [2 /*return*/];
                }
            });
        }); };
        runFormEvents();
    }, [solidFormViewMetaData, solidFormViewData]);
    useEffect(function () {
        if (solidFormViewData) {
            setInitialEntityData(solidFormViewData.data);
        }
    }, [solidFormViewData]);
    var formik;
    // If either the metadata or the data of this form is loading, then we simply render a loading screen...
    if (solidFormViewMetaDataIsLoading || solidFormViewDataIsLoading || !formViewLayout) {
        formik = useFormik({
            initialValues: {},
            validationSchema: Yup.object(),
            enableReinitialize: true,
            onSubmit: onFormikSubmit,
        });
        return _jsx(SolidFormViewShimmerLoading, {});
    }
    // At this point everything required to render the form is loaded, so we go ahead and start rendering things dynamically...
    else {
        // Initialize formik...
        var solidView_1 = solidFormViewMetaData.data.solidView;
        var solidFieldsMetadata_1 = solidFormViewMetaData.data.solidFieldsMetadata;
        var validationSchema = {};
        var initialValues = {};
        for (var i = 0; i < (layoutFields === null || layoutFields === void 0 ? void 0 : layoutFields.length); i++) {
            var formLayoutField = layoutFields[i];
            var fieldMetadata = solidFieldsMetadata_1[formLayoutField.attrs.name];
            var fieldContext = {
                fieldMetadata: fieldMetadata,
                field: formLayoutField,
                data: initialEntityData,
                solidFormViewMetaData: solidFormViewMetaData,
                modelName: params.modelName
            };
            if (params.parentData) {
                fieldContext.parentData = params.parentData;
            }
            var solidField = fieldFactory(fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.type, fieldContext);
            if (!(fieldMetadata === null || fieldMetadata === void 0 ? void 0 : fieldMetadata.type)) {
                var errorMessage = ((_d = formLayoutField === null || formLayoutField === void 0 ? void 0 : formLayoutField.attrs) === null || _d === void 0 ? void 0 : _d.label) ? (_e = formLayoutField === null || formLayoutField === void 0 ? void 0 : formLayoutField.attrs) === null || _e === void 0 ? void 0 : _e.label : formLayoutField.attrs.name;
                if (!errorFields.includes(errorMessage)) {
                    // errorFields.push(errorMessage);
                }
            }
            if (solidField) {
                // @ts-ignore
                validationSchema[formLayoutField.attrs.name] = solidField.validationSchema();
                // @ts-ignore
                initialValues[formLayoutField.attrs.name] = solidField.initialValue();
            }
        }
        formik = useFormik({
            initialValues: initialValues,
            validationSchema: Yup.object(validationSchema),
            enableReinitialize: true,
            onSubmit: onFormikSubmit,
        });
        var formFieldOnXXX_1 = function (event, eventType) { return __awaiter(void 0, void 0, void 0, function () {
            var _a, fieldName, value, type, checked, changeHandler, dynamicChangeHandler, event_1, updatedFormInfo;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        // Invoke the formik change 
                        if (eventType === 'onFieldChange') {
                            formik.handleChange(event);
                        }
                        _a = event.target, fieldName = _a.name, value = _a.value, type = _a.type, checked = _a.checked;
                        changeHandler = solidView_1.layout.attrs[eventType];
                        if (!changeHandler) {
                            changeHandler = solidView_1.layout[eventType];
                        }
                        if (!changeHandler) return [3 /*break*/, 3];
                        dynamicChangeHandler = getExtensionFunction(changeHandler);
                        if (!dynamicChangeHandler) return [3 /*break*/, 2];
                        event_1 = {
                            fieldsMetadata: solidFieldsMetadata_1,
                            formData: formik.values,
                            modifiedField: fieldName,
                            modifiedFieldValue: value,
                            // @ts-ignore
                            // TODO: HP & OR: This will be fixed once we figure out how to get types exported from core-ui
                            type: eventType,
                            viewMetadata: solidView_1,
                            formViewLayout: formViewLayout,
                            queryParams: {
                                actionName: actionName,
                                actionContext: actionContext,
                                actionType: actionType
                            }
                        };
                        return [4 /*yield*/, dynamicChangeHandler(event_1)];
                    case 1:
                        updatedFormInfo = _b.sent();
                        // If dataChanged is true, update Formik values
                        if ((updatedFormInfo === null || updatedFormInfo === void 0 ? void 0 : updatedFormInfo.dataChanged) && updatedFormInfo.newFormData) {
                            // This does one field at a time.
                            // TODO: does the below fire change events again?
                            Object.entries(updatedFormInfo.newFormData).forEach(function (_a) {
                                var key = _a[0], newValue = _a[1];
                                formik.setFieldValue(key, newValue);
                            });
                            // This does all at once.
                            // formik.setValues({
                            //     ...formik.values,
                            //     ...updatedFormInfo.newFormData
                            // });
                        }
                        // if layout has changed then we need to re-render.
                        if ((updatedFormInfo === null || updatedFormInfo === void 0 ? void 0 : updatedFormInfo.layoutChanged) && updatedFormInfo.newLayout) {
                            // setFormViewMetaData({ ...formViewMetaData, layout: updatedFormInfo.newLayout });
                            // TODO: this will trigger a useEffect dependent on formViewMetadata that invokes setFormViewLayout, 
                            // TODO: which means that this will not work if custom layout has been injected as a prop.
                            setFormViewLayout(updatedFormInfo.newLayout);
                            // setFormViewMetaData((prevMetaData: any) => {
                            //     const updatedFormViewMetadata = {
                            //         ...prevMetaData,
                            //         data: {
                            //             ...prevMetaData.data,
                            //             solidView: {
                            //                 ...prevMetaData.data.solidView,
                            //                 layout: updatedFormInfo.newLayout,
                            //             },
                            //         },
                            //     };
                            //     return updatedFormViewMetadata;
                            // });
                        }
                        return [3 /*break*/, 3];
                    case 2:
                        // TODO: Show an error popup and stop form rendering ideallly...
                        console.log(ERROR_MESSAGES.UNABLE_LOAD_DYNAMIC_MODULE, changeHandler);
                        _b.label = 3;
                    case 3: return [2 /*return*/];
                }
            });
        }); };
        // Now render the form dynamically...
        var renderFormElementDynamically_1 = function (element, recursiveFVMD) {
            var type = element.type, attrs = element.attrs, body = element.body, children = element.children;
            // const key = attrs?.name ?? generateRandomKey();
            var key = attrs === null || attrs === void 0 ? void 0 : attrs.label;
            var visible = attrs === null || attrs === void 0 ? void 0 : attrs.visible;
            if (visible === undefined || visible === null) {
                visible = true;
            }
            // console.log(`Resolved visibility of form element ${ key } to ${ visible } `);
            // console.log(`Form element ${ key }: `, attrs);
            var visibleToRole = (attrs === null || attrs === void 0 ? void 0 : attrs.roles) || [];
            if (visibleToRole.length > 0) {
                if (!hasAnyRole(user === null || user === void 0 ? void 0 : user.roles, visibleToRole)) {
                    return _jsx(_Fragment, {});
                }
            }
            switch (type) {
                case "form":
                    if (!children)
                        children = [];
                    return _jsx("div", { children: children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD, formik); }) }, key);
                case "div":
                    if (!children)
                        children = [];
                    return _jsx("div", __assign({}, attrs, { children: children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD, formik); }) }), key);
                case "span":
                    return _jsx("span", __assign({}, attrs, { children: body }), key);
                case "p":
                    return _jsx("p", __assign({ className: attrs === null || attrs === void 0 ? void 0 : attrs.className }, attrs, { children: body }), key);
                case "h1":
                    return _jsx("h1", __assign({}, attrs, { children: body }), key);
                case "h2":
                    return _jsx("h2", __assign({}, attrs, { children: body }), key);
                case "ul":
                    if (!children)
                        children = [];
                    return _jsx("ul", __assign({}, attrs, { children: children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD, formik); }) }), key);
                case "li":
                    return _jsx("li", __assign({}, attrs, { children: body }), key);
                case "sheet":
                    return _jsx(SolidSheet, { children: children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD, formik); }) }, key);
                case "group":
                    if (visible === true) {
                        return _jsx(SolidGroup, { attrs: attrs, children: children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD, formik); }) }, key);
                    }
                    break;
                case "row":
                    if (visible === true) {
                        return _jsx(SolidRow, { attrs: attrs, children: children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD, formik); }) }, key);
                    }
                    break;
                case "column":
                    if (visible === true) {
                        return _jsx(SolidColumn, { attrs: attrs, children: children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD, formik); }) }, key);
                    }
                    break;
                case "field":
                    if (visible === true) {
                        // const fieldMetadata = solidFieldsMetadata[attrs.name];
                        var fieldMetadata = recursiveFVMD.data.solidFieldsMetadata[attrs.name];
                        // Read only permission if there is no update permission on model and router doesnt contains new
                        var readOnlyPermission = !actionsAllowed.includes("".concat(permissionExpression(params.modelName, 'update'))) && params.id !== "new";
                        return _jsx(SolidField, { field: element, formik: formik, fieldMetadata: fieldMetadata, initialEntityData: solidFormViewData ? solidFormViewData.data : {}, solidFormViewMetaData: recursiveFVMD, modelName: params.modelName, readOnly: readOnlyPermission, viewMode: viewMode, onChange: formFieldOnXXX_1, onBlur: formFieldOnXXX_1, setLightboxUrls: setLightboxUrls, setOpenLightbox: setOpenLightbox, parentFieldName: params.parentFieldName, parentData: params.parentData, onEmbeddedFormSave: function () { return setRefreshChatterMessage(true); } }, attrs.name);
                    }
                    break;
                case "notebook":
                    if (visible === true) {
                        return _jsx(SolidNotebook, { activeTab: searchParams.get("activeTab") || "", embeded: params.embeded, children: children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD, formik); }) }, key);
                    }
                    break;
                case "page":
                    if (visible === true) {
                        var fields = children.flatMap(function (child) { return getLayoutFields(child); });
                        var pageChildren = children.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD); });
                        return SolidPage({ children: pageChildren, attrs: attrs, key: key, formik: formik, fields: fields });
                    }
                    break;
                case "custom":
                    if (visible === true) {
                        var widgetName = attrs === null || attrs === void 0 ? void 0 : attrs.widget;
                        var fieldMetadata = recursiveFVMD.data.solidFieldsMetadata[attrs.name];
                        if (widgetName) {
                            // widgetName, formik, field, fieldMetadata, solidFormViewMetaData
                            return _jsx(SolidDynamicWidget, { widgetName: widgetName, field: element, formik: formik, fieldMetadata: fieldMetadata, solidFormViewMetaData: recursiveFVMD, solidFormViewData: solidFormViewData }, key);
                        }
                    }
                    break;
                default:
                    return null;
            }
        };
        var renderFormDynamically = function (recursiveFVMD, formViewLayout) {
            if (!recursiveFVMD) {
                return;
            }
            if (!recursiveFVMD.data) {
                return;
            }
            var solidView = recursiveFVMD.data.solidView;
            var solidFieldsMetadata = recursiveFVMD.data.solidFieldsMetadata;
            if (!solidView || !solidFieldsMetadata) {
                return;
            }
            var updatedLayout = [formViewLayout];
            var dynamicForm = updatedLayout.map(function (element) { return renderFormElementDynamically_1(element, recursiveFVMD); });
            return dynamicForm;
        };
        var handleDeleteEntity_1 = function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                deleteEntity(solidFormViewData.data.id);
                setDeleteDialogVisible(false);
                if (params.embeded == true) {
                    setRedirectToList(false);
                }
                else {
                    setRedirectToList(true);
                }
                return [2 /*return*/];
            });
        }); };
        var onDeleteClose_1 = function () {
            setDeleteDialogVisible(false);
        };
        var dynamicHeader = (_f = solidView_1.layout) === null || _f === void 0 ? void 0 : _f.header;
        var DynamicHeaderComponent = null;
        if (dynamicHeader) {
            DynamicHeaderComponent = getExtensionComponent(dynamicHeader);
        }
        var customFormComponentEdit = solidView_1.layout.attrs.customFormComponentEdit;
        var customFormComponentNew = solidView_1.layout.attrs.customFormComponentNew;
        var DynamicFormComponentEdit = null;
        var DynamicFormComponentNew = null;
        if (customFormComponentEdit) {
            DynamicFormComponentEdit = getExtensionComponent(customFormComponentEdit);
        }
        if (customFormComponentNew) {
            DynamicFormComponentNew = getExtensionComponent(customFormComponentNew);
        }
        var handleChatterExpandClick_1 = function (option) {
            setShowChatter(true);
            if (option === 'info') {
                setDefaultTabViewOptionIndex(0);
            }
            else if (option === 'chatter') {
                setDefaultTabViewOptionIndex(1);
                setRefreshChatterMessage(true);
            }
            else {
                setDefaultTabViewOptionIndex(2);
            }
        };
        //en 4 null
        var handleLocaleChangeRedirect = function (locale, defaultEntityLocaleId, viewMode) {
            var _a, _b;
            var newViewMode = viewMode;
            var defaultApplicableLocales = ((_a = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _a === void 0 ? void 0 : _a.applicableLocales) || [];
            //fr 4
            var matchingLocale = defaultApplicableLocales.find(function (loc) {
                return loc.defaultEntityLocaleId &&
                    loc.entityId &&
                    loc.locale === locale;
            });
            // Extract the base path from the current pathname, removing query params if any
            var basePath = pathname.split('?')[0];
            // Determine entity part of the path (new or existing entityId)
            var entityPart = (_b = matchingLocale === null || matchingLocale === void 0 ? void 0 : matchingLocale.entityId) !== null && _b !== void 0 ? _b : 'new';
            if (entityPart === 'new' && viewMode === 'view') {
                newViewMode = 'edit';
            }
            else if (entityPart !== 'new' && viewMode === 'view') {
                newViewMode = 'view';
            }
            else {
                newViewMode = 'edit';
            }
            // Construct new pathname using existing basePath and replacing entity segment
            var updatedPath = basePath.replace(/\/form\/[^/]+/, "/form/".concat(entityPart));
            var queryParams = new URLSearchParams({
                viewMode: newViewMode,
                locale: locale,
                defaultEntityLocaleId: defaultEntityLocaleId,
            });
            router.push("".concat(updatedPath, "?").concat(queryParams.toString()));
        };
        var handleConfirmAccept = function () {
            var _a;
            (_a = confirmResolveRef.current) === null || _a === void 0 ? void 0 : _a.call(confirmResolveRef, true);
            setConfirmVisible(false);
        };
        var handleConfirmReject = function () {
            var _a;
            (_a = confirmResolveRef.current) === null || _a === void 0 ? void 0 : _a.call(confirmResolveRef, false);
            setConfirmVisible(false);
        };
        var handleDraftPublishWorkFlow = function (type) { return __awaiter(void 0, void 0, void 0, function () {
            var userChoice, result;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, confirmDialogWithPromise()];
                    case 1:
                        userChoice = _b.sent();
                        if (!userChoice)
                            return [2 /*return*/];
                        if (!(type === "publish")) return [3 /*break*/, 3];
                        return [4 /*yield*/, publishSolidEntity(params.id).unwrap()];
                    case 2:
                        result = _b.sent();
                        showToast(toast, "success", ERROR_MESSAGES.SAVED, ERROR_MESSAGES.MARK_PUBLISH);
                        return [3 /*break*/, 5];
                    case 3: return [4 /*yield*/, unpublishSolidEntity(params.id).unwrap()];
                    case 4:
                        result = _b.sent();
                        showToast(toast, "success", ERROR_MESSAGES.SAVED, ERROR_MESSAGES.MARK_UNPUBLISH);
                        _b.label = 5;
                    case 5:
                        console.log("publish/unpublish result", result);
                        // Set updated publish value from API response
                        setPublished((_a = result === null || result === void 0 ? void 0 : result.data) === null || _a === void 0 ? void 0 : _a.publishedAt);
                        return [2 /*return*/];
                }
            });
        }); };
        var isVideoOrAudio_1 = function (url) {
            var _a;
            // Remove query params if present
            var cleanUrl = url.split("?")[0];
            var ext = (_a = cleanUrl.split(".").pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
            // Combined list of supported media extensions
            var mediaExt = ["mp4", "webm", "ogg", "mov", "mp3", "wav", "m4a", "aac"];
            return ext ? mediaExt.includes(ext) : false;
        };
        var controlsList = ["nodownload", "nofullscreen", "noremoteplayback"];
        var slides = lightboxUrls.map(function (item) {
            var url = item.src || item.downloadUrl || "";
            if (isVideoOrAudio_1(url)) {
                return {
                    type: "video",
                    sources: [{ src: url, type: "video/mp4", }],
                };
            }
            return { src: url };
        });
        var hasMedia = slides.some(function (s) { return s.type === "video"; });
        return (_jsxs("div", { className: "solid-form-wrapper", children: [_jsx(Toast, { ref: toast }), _jsxs("div", { className: "solid-form-section", style: { borderRight: params.embeded !== true ? '1px solid var(--primary-light-color)' : '' }, children: [_jsxs("form", { style: { width: '100%' }, onSubmit: formik.handleSubmit, children: [_jsx(SolidFormActionHeader, { formik: formik, formData: solidFormViewData === null || solidFormViewData === void 0 ? void 0 : solidFormViewData.data, params: params, actionsAllowed: actionsAllowed, formViewLayout: formViewLayout, solidView: solidView_1, solidFormViewMetaData: solidFormViewMetaData, initialEntityData: solidFormViewData ? solidFormViewData.data : {}, setDeleteDialogVisible: setDeleteDialogVisible, setLayoutDialogVisible: setLayoutDialogVisible, setRedirectToList: setRedirectToList, viewMode: viewMode, setViewMode: setViewMode, solidWorkflowFieldValue: solidWorkflowFieldValue, setSolidWorkflowFieldValue: setSolidWorkflowFieldValue, draftEnabled: (_j = (_h = (_g = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _g === void 0 ? void 0 : _g.solidView) === null || _h === void 0 ? void 0 : _h.model) === null || _j === void 0 ? void 0 : _j.draftPublishWorkflow, publish: published, internationalisationEnabled: (_m = (_l = (_k = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _k === void 0 ? void 0 : _k.solidView) === null || _l === void 0 ? void 0 : _l.model) === null || _m === void 0 ? void 0 : _m.internationalisation, handleDraftPublishWorkFlow: handleDraftPublishWorkFlow, onStepperUpdate: function () { return setRefreshChatterMessage(true); }, isSubmitting: isSubmitting }), _jsxs("div", { className: "px-4 py-3 md:p-4 solid-form-content ".concat(params.embeded === true ? 'h-auto' : ''), style: { maxHeight: params.embeded === true ? '80vh' : '', overflowY: 'auto' }, children: [DynamicHeaderComponent && _jsx(DynamicHeaderComponent, {}), params.id === 'new' && DynamicFormComponentNew ? (_jsx(DynamicFormComponentNew, { params: params })) : params.id !== 'new' && DynamicFormComponentEdit ? (_jsx(DynamicFormComponentEdit, { params: params })) : (renderFormDynamically(formViewMetaData, formViewLayout))] })] }), _jsx(SolidFormFooter, { params: params })] }), params.embeded !== true &&
                    _jsxs("div", { className: "chatter-section ".concat(isShowChatter === false ? 'collapsed' : 'open'), style: { width: chatterLocaleWidth }, children: [isShowChatter && (_jsx("div", { style: {
                                    width: 5,
                                    cursor: 'col-resize',
                                    position: 'absolute',
                                    left: 0,
                                    top: 0,
                                    bottom: 0,
                                    zIndex: 9,
                                }, onMouseDown: function () { return setIsResizingChatterLocale(true); } })), isShowChatter === true &&
                                _jsx(Button, { icon: "pi pi-angle-double-right", size: "small", text: true, className: "chatter-collapse-btn", style: { width: 30, height: 30, aspectRatio: '1/1' }, onClick: function () { return setShowChatter(false); } }), isShowChatter === false ?
                                _jsxs("div", { className: "flex flex-column gap-2 justify-content-center p-2", children: [((_q = (_p = (_o = solidFormViewMetaData === null || solidFormViewMetaData === void 0 ? void 0 : solidFormViewMetaData.data) === null || _o === void 0 ? void 0 : _o.solidView) === null || _p === void 0 ? void 0 : _p.model) === null || _q === void 0 ? void 0 : _q.draftPublishWorkflow) &&
                                            _jsx("div", { className: "chatter-collapsed-content", onClick: function () { return handleChatterExpandClick_1('info'); }, children: "Info" }), _jsx("div", { className: "chatter-collapsed-content", onClick: function () { return handleChatterExpandClick_1('chatter'); }, children: "Audit Trail" }), mcpUrl &&
                                            (_jsx("div", { className: "chatter-collapsed-content", onClick: function () { return handleChatterExpandClick_1('solidx-ai'); }, children: _jsxs("div", { className: "flex gap-2", children: [" ", _jsx(SolidXAIIcon, {}), " SolidX AI "] }) })), _jsx(Button, { icon: "pi pi-chevron-left", size: "small", className: "px-0", style: { width: 30 }, onClick: function () { return handleChatterExpandClick_1('default'); } })] })
                                :
                                    _jsx(SolidChatterLocaleTabView, { createMode: createMode, setSelectedLocale: setSelectedLocale, selectedLocale: selectedLocale, solidFormViewMetaData: solidFormViewMetaData, id: params.id, refreshChatterMessage: refreshChatterMessage, setRefreshChatterMessage: setRefreshChatterMessage, activeTab: defaultTabViewOptionIndex, viewMode: viewMode, defaultEntityLocaleId: defaultEntityLocaleId, handleLocaleChangeRedirect: handleLocaleChangeRedirect, solidFormViewData: solidFormViewData, published: published, actionsAllowed: actionsAllowed, mcpUrl: mcpUrl })] }), _jsx(Dialog, { visible: isDeleteDialogVisible, header: "Confirm Delete", className: "solid-confirm-dialog", modal: true, footer: function () { return (_jsxs("div", { className: "flex justify-content-center", children: [_jsx(Button, { label: "Yes", icon: "pi pi-check", className: 'small-button', severity: "danger", autoFocus: true, onClick: function () { return handleDeleteEntity_1(); } }), _jsx(Button, { label: "No", icon: "pi pi-times", className: 'small-button', onClick: onDeleteClose_1 })] })); }, onHide: function () { return setDeleteDialogVisible(false); }, children: _jsx("p", { children: "Are you sure you want to delete?" }) }), _jsx(Dialog, { visible: isLayoutDialogVisible, header: "Change Form Layout", modal: true, onHide: function () { return setLayoutDialogVisible(false); }, style: { width: '50vw' }, breakpoints: {
                        '960px': '80vw',
                        '641px': '95vw'
                    }, contentClassName: "p-3 pt-0 lg:p-4", children: _jsx(SolidFormUserViewLayout, { solidFormViewMetaData: solidFormViewMetaData, setLayoutDialogVisible: setLayoutDialogVisible }) }), openLightbox &&
                    _jsx(Lightbox, __assign({ open: openLightbox, plugins: hasMedia
                            ? [Counter, Download, Video] // add Video plugin if needed
                            : [Counter, Download], close: function () { return setOpenLightbox(false); }, slides: __spreadArray([], slides, true) }, (hasMedia && {
                        video: {
                            controls: true,
                            playsInline: true,
                            autoPlay: false,
                            loop: false,
                            muted: false,
                            disablePictureInPicture: false,
                            disableRemotePlayback: false,
                            controlsList: controlsList.join(" "),
                            crossOrigin: "anonymous",
                            preload: "auto",
                        },
                    }))), _jsx(ConfirmDialog, { visible: confirmVisible, onHide: function () { return setConfirmVisible(false); }, header: "Confirmation", acceptLabel: "Yes, confrim", rejectLabel: "No, cancel", acceptClassName: "p-button-danger", rejectClassName: "p-button-text", position: "center", accept: handleConfirmAccept, reject: handleConfirmReject, message: _jsx("div", { className: "flex flex-col items-center justify-center text-center space-y-3", children: _jsxs("p", { className: "text-gray-800 text-base", children: ["Are you sure you want to ", published !== null ? 'unpublish' : 'publish', "?"] }) }) })] }));
    }
};
export default SolidFormView;
