// @solidxai/core-ui/dist/components/core/form/SolidFormFooter.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useLazyNavigationQuery } from "../../../redux/api/modelApi";
import { useEffect, useState } from "react";
import qs from "qs";
import { queryObjectToQueryStringByUrl, queryStringToQueryObjectByUrl } from "../list/SolidListView";
import { usePathname } from "../../../hooks/usePathname";
import { useSearchParams } from "../../../hooks/useSearchParams";
import { useRouter } from "../../../hooks/useRouter";
import { Button } from "primereact/button";
export var SolidFormFooter = function (_a) {
    var params = _a.params;
    var pathname = usePathname();
    var searchParams = useSearchParams();
    var router = useRouter();
    var _b = useState(null), prevNav = _b[0], setPrevNav = _b[1];
    var _c = useState(null), nextNav = _c[0], setNextNav = _c[1];
    var _d = useState(null), meta = _d[0], setMeta = _d[1];
    var _e = useLazyNavigationQuery(), triggerGetNavigation = _e[0], isLoading = _e[1].isLoading;
    // -----------------------------
    // Helper: update local storage
    // -----------------------------
    var updatePaginationInLocalStorage = function (offset, limit) {
        var listPath = window.location.pathname.replace(/\/form\/[^/]+/, "/list");
        var queryObject = queryStringToQueryObjectByUrl(listPath);
        var updatedQueryObj = __assign(__assign({}, queryObject), { offset: offset, limit: limit });
        queryObjectToQueryStringByUrl(listPath, updatedQueryObj);
    };
    var getNewUrl = function (recordId) {
        // Replace the id after /form/
        var newPathname = pathname.replace(/\/form\/[^/]+/, "/form/".concat(recordId));
        // Clone search params (immutable in Next.js)
        var params = new URLSearchParams(searchParams.toString());
        // Remove only userKeyField
        params.delete("userKeyField");
        var query = params.toString();
        return query ? "".concat(newPathname, "?").concat(query) : newPathname;
    };
    // -----------------------------
    // Click handlers
    // -----------------------------
    var handlePrev = function () {
        if (!prevNav)
            return;
        updatePaginationInLocalStorage(prevNav.offset, prevNav.limit);
        var newUrl = getNewUrl(prevNav.recordId);
        router.push(newUrl);
    };
    var handleNext = function () {
        if (!nextNav)
            return;
        updatePaginationInLocalStorage(nextNav.offset, nextNav.limit);
        var newUrl = getNewUrl(nextNav.recordId);
        router.push(newUrl);
    };
    // -----------------------------
    // Fetch navigation data
    // -----------------------------
    useEffect(function () {
        if (params.id !== "new") {
            var fetchNavigation = function () { return __awaiter(void 0, void 0, void 0, function () {
                var listPath, queryObject, queryData, queryString, response;
                var _a, _b, _c, _d, _e, _f;
                return __generator(this, function (_g) {
                    switch (_g.label) {
                        case 0:
                            listPath = window.location.pathname.replace(/\/form\/[^/]+/, "/list");
                            queryObject = queryStringToQueryObjectByUrl(listPath);
                            queryData = {
                                offset: queryObject.offset || 0,
                                limit: queryObject.limit || 25,
                                filters: queryObject.filters,
                                fields: ["id"],
                                modelName: params.modelName,
                                recordId: params.id,
                                sort: queryObject.sort
                            };
                            queryString = qs.stringify(queryData, {
                                encodeValuesOnly: true,
                            });
                            return [4 /*yield*/, triggerGetNavigation(queryString).unwrap()];
                        case 1:
                            response = _g.sent();
                            console.log("response nav", response);
                            if (response.statusCode == 200) {
                                setPrevNav((_b = (_a = response === null || response === void 0 ? void 0 : response.data) === null || _a === void 0 ? void 0 : _a.prev) !== null && _b !== void 0 ? _b : null);
                                setNextNav((_d = (_c = response === null || response === void 0 ? void 0 : response.data) === null || _c === void 0 ? void 0 : _c.next) !== null && _d !== void 0 ? _d : null);
                                setMeta((_f = (_e = response === null || response === void 0 ? void 0 : response.data) === null || _e === void 0 ? void 0 : _e.meta) !== null && _f !== void 0 ? _f : null);
                            }
                            return [2 /*return*/];
                    }
                });
            }); };
            fetchNavigation();
        }
    }, [params.id]);
    // -----------------------------
    // UI
    // -----------------------------
    return (_jsxs("div", { className: "flex justify-content-end align-items-center gap-2 p-1", style: { borderTop: "1px solid var(--surface-border)" }, children: [meta &&
                _jsx("span", { children: "".concat(meta.currentIndexGlobal, " of ").concat(meta.totalRecords) }), prevNav && (_jsx(Button, { icon: "pi pi-angle-left", className: "p-button-sm p-button-text", onClick: function () { return handlePrev(); }, disabled: isLoading, tooltip: "Previous", tooltipOptions: { position: "top" } })), nextNav && (_jsx(Button, { icon: "pi pi-angle-right", className: "p-button-sm p-button-text", onClick: function () { return handleNext(); }, disabled: isLoading, tooltip: "Next", tooltipOptions: { position: "top" } }))] }));
};
