// @solidxai/core-ui/dist/components/core/form/SolidFormViewContextMenuHeaderButton.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Button } from 'primereact/button';
import { hasAnyRole } from '../../../helpers/rolesHelper';
import { useSession } from "../../../hooks/useSession";
export function SolidFormViewContextMenuHeaderButton(_a) {
    var _b, _c, _d, _e, _f, _g, _h, _j;
    var button = _a.button, params = _a.params, formik = _a.formik, solidFormViewMetaData = _a.solidFormViewMetaData, handleCustomButtonClick = _a.handleCustomButtonClick, formData = _a.formData;
    var _k = useSession(), session = _k.data, status = _k.status;
    var user = session === null || session === void 0 ? void 0 : session.user;
    var hasRole = !((_b = button.attrs) === null || _b === void 0 ? void 0 : _b.roles) || ((_c = button === null || button === void 0 ? void 0 : button.attrs) === null || _c === void 0 ? void 0 : _c.roles.length) === 0 ? true : hasAnyRole(user === null || user === void 0 ? void 0 : user.roles, (_d = button === null || button === void 0 ? void 0 : button.attrs) === null || _d === void 0 ? void 0 : _d.roles);
    if (!hasRole)
        return null;
    if (((_e = button.attrs) === null || _e === void 0 ? void 0 : _e.visible) == false)
        return null;
    return (_jsx("div", { children: _jsx(Button, { text: true, type: "button", className: "w-full text-left gap-2 ".concat(((_f = button === null || button === void 0 ? void 0 : button.attrs) === null || _f === void 0 ? void 0 : _f.className) ? (_g = button === null || button === void 0 ? void 0 : button.attrs) === null || _g === void 0 ? void 0 : _g.className : ''), label: button.attrs.label, size: "small", iconPos: "left", icon: ((_h = button === null || button === void 0 ? void 0 : button.attrs) === null || _h === void 0 ? void 0 : _h.icon) ? (_j = button === null || button === void 0 ? void 0 : button.attrs) === null || _j === void 0 ? void 0 : _j.icon : "pi pi-pencil", onClick: function () {
                var event = {
                    action: button.attrs.action,
                    params: params,
                    formik: formik,
                    solidFormViewMetaData: solidFormViewMetaData.data,
                    formData: formData
                };
                handleCustomButtonClick(button.attrs, event);
            } }) }));
}
